{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import os, sys, itertools\n",
    "\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(148, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>630</td>\n",
       "      <td>66</td>\n",
       "      <td>T247M (ACG→ATG)</td>\n",
       "      <td>yafD</td>\n",
       "      <td>SNP</td>\n",
       "      <td>231861</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>0.72</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(231861, 231861)</td>\n",
       "      <td>{ECK120001594}</td>\n",
       "      <td>[{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...</td>\n",
       "      <td>{'ECK120001594': ['ECK120001594']}</td>\n",
       "      <td>[{'name': 'yafDE', 'RegulonDB ID': 'ECK1200291...</td>\n",
       "      <td>{'ECK120029169': ['ECK120001594']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'General function prediction only'}]</td>\n",
       "      <td>{'General function prediction only': ['ECK1200...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>631</td>\n",
       "      <td>126</td>\n",
       "      <td>R665S (CGT→AGT)</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3993145</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→A</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3993145, 3993145)</td>\n",
       "      <td>{ECK120000166}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>{'ECK120000166': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...</td>\n",
       "      <td>{'ECK120014835': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'CRP': ['cyaA']}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}]</td>\n",
       "      <td>{'Purine metabolism': ['ECK120014835']}</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism'}]</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>632</td>\n",
       "      <td>94</td>\n",
       "      <td>S59Y (TCC→TAC)</td>\n",
       "      <td>glpK</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4117047</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→T</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4117047, 4117047)</td>\n",
       "      <td>{ECK120000391}</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism'}]</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion'}]</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>633</td>\n",
       "      <td>72</td>\n",
       "      <td>T593P (ACC→CCC)</td>\n",
       "      <td>metL</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4131611</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>0.67</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4131611, 4131611)</td>\n",
       "      <td>{ECK120000583}</td>\n",
       "      <td>[{'name': 'metL', 'RegulonDB ID': 'ECK12000058...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'metL', 'RegulonDB ID': 'ECK12000058...</td>\n",
       "      <td>{'ECK120000583': ['ECK120000583']}</td>\n",
       "      <td>[{'name': 'metBL', 'RegulonDB ID': 'ECK1200150...</td>\n",
       "      <td>{'ECK120015057': ['ECK120000583']}</td>\n",
       "      <td>[{'name': 'MetJ', 'significantly associated co...</td>\n",
       "      <td>{'MetJ': ['metBL'], 'PhoP': ['metBL'], 'HypT':...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism'}]</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>634</td>\n",
       "      <td>107</td>\n",
       "      <td>E641K (GAA→AAA)</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4183165</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→A</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4183165, 4183165)</td>\n",
       "      <td>{ECK120000885}</td>\n",
       "      <td>[{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000885': ['ECK120000885']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000885']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "      <td>{'Purine metabolism': ['ECK120016992'], 'Pyrim...</td>\n",
       "      <td>[{'name': 'Transcription'}]</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index          Details mutation target annotation Mutation Type  Position  \\\n",
       "630    66  T247M (ACG→ATG)                       yafD           SNP    231861   \n",
       "631   126  R665S (CGT→AGT)                       cyaA           SNP   3993145   \n",
       "632    94   S59Y (TCC→TAC)                       glpK           SNP   4117047   \n",
       "633    72  T593P (ACC→CCC)                       metL           SNP   4131611   \n",
       "634   107  E641K (GAA→AAA)                       rpoB           SNP   4183165   \n",
       "\n",
       "    Reference Seq Sequence Change  ale  exp  flask  isolate  presence  \\\n",
       "630     NC_000913             C→T    1  GYD    280        0      0.72   \n",
       "631     NC_000913             C→A    1  GYD    280        0      1.00   \n",
       "632     NC_000913             G→T    1  GYD    280        0      1.00   \n",
       "633     NC_000913             A→C    1  GYD    280        0      0.67   \n",
       "634     NC_000913             G→A    1  GYD    280        0      1.00   \n",
       "\n",
       "     tech_rep  coding               range gene RegulonDB ID  \\\n",
       "630         1    True    (231861, 231861)    {ECK120001594}   \n",
       "631         1    True  (3993145, 3993145)    {ECK120000166}   \n",
       "632         1    True  (4117047, 4117047)    {ECK120000391}   \n",
       "633         1    True  (4131611, 4131611)    {ECK120000583}   \n",
       "634         1    True  (4183165, 4183165)    {ECK120000885}   \n",
       "\n",
       "                                      genetic features   oriC  pseudogene  \\\n",
       "630  [{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...  False       False   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...  False       False   \n",
       "632  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...  False       False   \n",
       "633  [{'name': 'metL', 'RegulonDB ID': 'ECK12000058...  False       False   \n",
       "634  [{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "\n",
       "    TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "630   {}       {}  {}                    {}         {}     True   \n",
       "631   {}       {}  {}                    {}         {}     True   \n",
       "632   {}       {}  {}                    {}         {}     True   \n",
       "633   {}       {}  {}                    {}         {}     True   \n",
       "634   {}       {}  {}                    {}         {}     True   \n",
       "\n",
       "                                      genomic features  \\\n",
       "630  [{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...   \n",
       "632  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "633  [{'name': 'metL', 'RegulonDB ID': 'ECK12000058...   \n",
       "634  [{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...   \n",
       "\n",
       "                  genetic feature links  \\\n",
       "630  {'ECK120001594': ['ECK120001594']}   \n",
       "631  {'ECK120000166': ['ECK120000166']}   \n",
       "632  {'ECK120000391': ['ECK120000391']}   \n",
       "633  {'ECK120000583': ['ECK120000583']}   \n",
       "634  {'ECK120000885': ['ECK120000885']}   \n",
       "\n",
       "                                               operons  \\\n",
       "630  [{'name': 'yafDE', 'RegulonDB ID': 'ECK1200291...   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...   \n",
       "632  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "633  [{'name': 'metBL', 'RegulonDB ID': 'ECK1200150...   \n",
       "634  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "\n",
       "                           operon links  \\\n",
       "630  {'ECK120029169': ['ECK120001594']}   \n",
       "631  {'ECK120014835': ['ECK120000166']}   \n",
       "632  {'ECK120014723': ['ECK120000391']}   \n",
       "633  {'ECK120015057': ['ECK120000583']}   \n",
       "634  {'ECK120016992': ['ECK120000885']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "630                                                 []   \n",
       "631  [{'name': 'CRP', 'significantly associated con...   \n",
       "632  [{'name': 'GlpR', 'significantly associated co...   \n",
       "633  [{'name': 'MetJ', 'significantly associated co...   \n",
       "634                                                 []   \n",
       "\n",
       "                                       regulator links  \\\n",
       "630                                                 {}   \n",
       "631                                  {'CRP': ['cyaA']}   \n",
       "632            {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "633  {'MetJ': ['metBL'], 'PhoP': ['metBL'], 'HypT':...   \n",
       "634                                                 {}   \n",
       "\n",
       "                                              pathways  \\\n",
       "630                                                 []   \n",
       "631                    [{'name': 'Purine metabolism'}]   \n",
       "632              [{'name': 'Glycerolipid metabolism'}]   \n",
       "633  [{'name': 'Glycine, serine and threonine metab...   \n",
       "634  [{'name': 'Purine metabolism'}, {'name': 'Pyri...   \n",
       "\n",
       "                                         pathway links  \\\n",
       "630                                                 {}   \n",
       "631            {'Purine metabolism': ['ECK120014835']}   \n",
       "632      {'Glycerolipid metabolism': ['ECK120014723']}   \n",
       "633  {'Glycine, serine and threonine metabolism': [...   \n",
       "634  {'Purine metabolism': ['ECK120016992'], 'Pyrim...   \n",
       "\n",
       "                                                  COGs  \\\n",
       "630     [{'name': 'General function prediction only'}]   \n",
       "631  [{'name': 'Nucleotide transport and metabolism'}]   \n",
       "632     [{'name': 'Energy production and conversion'}]   \n",
       "633  [{'name': 'Amino acid transport and metabolism'}]   \n",
       "634                        [{'name': 'Transcription'}]   \n",
       "\n",
       "                                             COG links temperature  \\\n",
       "630  {'General function prediction only': ['ECK1200...  37 celsius   \n",
       "631  {'Nucleotide transport and metabolism': ['ECK1...  37 celsius   \n",
       "632  {'Energy production and conversion': ['ECK1200...  37 celsius   \n",
       "633  {'Amino acid transport and metabolism': ['ECK1...  37 celsius   \n",
       "634                {'Transcription': ['ECK120016992']}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "630  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "631  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "632  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "633  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "634  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "    taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "630      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "631      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "632      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "633      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "634      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "    sulfur-source calcium-source  \n",
       "630   MgSO4(0.24)     CaCl2(0.1)  \n",
       "631   MgSO4(0.24)     CaCl2(0.1)  \n",
       "632   MgSO4(0.24)     CaCl2(0.1)  \n",
       "633   MgSO4(0.24)     CaCl2(0.1)  \n",
       "634   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# exp_muts_df = pd.read_pickle(\"./data/9_df.pkl\")  # Currently skipping imodulons for the sake of the AVA MS\n",
    "exp_muts_df = pd.read_pickle(\"./data/8_df.pkl\")\n",
    "display(exp_muts_df.shape, exp_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Posttranslational modification, protein turnover, chaperones</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Secondary metabolites biosynthesis, transport and catabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lipid transport and metabolism</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell motility</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Mobilome: prophages, transposons</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count\n",
       "General function prediction only                                          5\n",
       "Nucleotide transport and metabolism                                      23\n",
       "Energy production and conversion                                         39\n",
       "Amino acid transport and metabolism                                      76\n",
       "Transcription                                                            27\n",
       "Carbohydrate transport and metabolism                                    31\n",
       "Replication, recombination and repair                                     3\n",
       "Posttranslational modification, protein turnove...                        1\n",
       "Inorganic ion transport and metabolism                                    1\n",
       "Coenzyme transport and metabolism                                         3\n",
       "Translation, ribosomal structure and biogenesis                          43\n",
       "Secondary metabolites biosynthesis, transport a...                        1\n",
       "Lipid transport and metabolism                                            3\n",
       "Cell motility                                                             2\n",
       "Signal transduction mechanisms                                            7\n",
       "Defense mechanisms                                                        1\n",
       "Cell wall/membrane/envelope biogenesis                                   15\n",
       "Mobilome: prophages, transposons                                          1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a bit behind annotations currently used in AVA MS.\n",
    "# Still uses same approach to counted, just not all data-types have been consolidated into the COGs feat col\n",
    "\n",
    "\n",
    "comp_mut_df = pd.DataFrame(columns=[\"observed mutation count\"])\n",
    "for i, r in exp_muts_df.iterrows():\n",
    "    for f, links in r[\"COG links\"].items():\n",
    "        if \"unknown\" not in f:\n",
    "            if f not in comp_mut_df.index:\n",
    "                df = pd.DataFrame({\"observed mutation count\": len(links)}, index=[f])\n",
    "                comp_mut_df = comp_mut_df.append(df)\n",
    "            else:\n",
    "                comp_mut_df.loc[f, \"observed mutation count\"] += len(links)\n",
    "            \n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "282"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(comp_mut_df[\"observed mutation count\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in exp_muts_df.iterrows():\n",
    "    for feat_d in r[\"COGs\"]:\n",
    "        if \"unknown\" not in feat_d[\"name\"]:\n",
    "            feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"name\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG description</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>660007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "      <td>305849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>679809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>212728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>589566</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        COG length\n",
       "COG description                                   \n",
       "Amino acid transport and metabolism         660007\n",
       "Inorganic ion transport and metabolism      305849\n",
       "Carbohydrate transport and metabolism       679809\n",
       "Coenzyme transport and metabolism           212728\n",
       "Energy production and conversion            589566"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COG_len_df = pd.read_pickle(\"./data/COG_len_df.pkl\")\n",
    "COG_len_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>5</td>\n",
       "      <td>254764</td>\n",
       "      <td>0.044958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>23</td>\n",
       "      <td>160339</td>\n",
       "      <td>0.028295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>39</td>\n",
       "      <td>589566</td>\n",
       "      <td>0.104041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>76</td>\n",
       "      <td>660007</td>\n",
       "      <td>0.116472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>27</td>\n",
       "      <td>424603</td>\n",
       "      <td>0.074930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>31</td>\n",
       "      <td>679809</td>\n",
       "      <td>0.119966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>3</td>\n",
       "      <td>309957</td>\n",
       "      <td>0.054698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Posttranslational modification, protein turnover, chaperones</td>\n",
       "      <td>1</td>\n",
       "      <td>270912</td>\n",
       "      <td>0.047808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>305849</td>\n",
       "      <td>0.053973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>3</td>\n",
       "      <td>212728</td>\n",
       "      <td>0.037540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>43</td>\n",
       "      <td>393020</td>\n",
       "      <td>0.069357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Secondary metabolites biosynthesis, transport and catabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>52093</td>\n",
       "      <td>0.009193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lipid transport and metabolism</td>\n",
       "      <td>3</td>\n",
       "      <td>183775</td>\n",
       "      <td>0.032431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell motility</td>\n",
       "      <td>2</td>\n",
       "      <td>144533</td>\n",
       "      <td>0.025506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>7</td>\n",
       "      <td>339255</td>\n",
       "      <td>0.059869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>1</td>\n",
       "      <td>135950</td>\n",
       "      <td>0.023991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>15</td>\n",
       "      <td>520583</td>\n",
       "      <td>0.091868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Mobilome: prophages, transposons</td>\n",
       "      <td>1</td>\n",
       "      <td>28915</td>\n",
       "      <td>0.005103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count  \\\n",
       "General function prediction only                                          5   \n",
       "Nucleotide transport and metabolism                                      23   \n",
       "Energy production and conversion                                         39   \n",
       "Amino acid transport and metabolism                                      76   \n",
       "Transcription                                                            27   \n",
       "Carbohydrate transport and metabolism                                    31   \n",
       "Replication, recombination and repair                                     3   \n",
       "Posttranslational modification, protein turnove...                        1   \n",
       "Inorganic ion transport and metabolism                                    1   \n",
       "Coenzyme transport and metabolism                                         3   \n",
       "Translation, ribosomal structure and biogenesis                          43   \n",
       "Secondary metabolites biosynthesis, transport a...                        1   \n",
       "Lipid transport and metabolism                                            3   \n",
       "Cell motility                                                             2   \n",
       "Signal transduction mechanisms                                            7   \n",
       "Defense mechanisms                                                        1   \n",
       "Cell wall/membrane/envelope biogenesis                                   15   \n",
       "Mobilome: prophages, transposons                                          1   \n",
       "\n",
       "                                                    length  proportion  \n",
       "General function prediction only                    254764    0.044958  \n",
       "Nucleotide transport and metabolism                 160339    0.028295  \n",
       "Energy production and conversion                    589566    0.104041  \n",
       "Amino acid transport and metabolism                 660007    0.116472  \n",
       "Transcription                                       424603    0.074930  \n",
       "Carbohydrate transport and metabolism               679809    0.119966  \n",
       "Replication, recombination and repair               309957    0.054698  \n",
       "Posttranslational modification, protein turnove...  270912    0.047808  \n",
       "Inorganic ion transport and metabolism              305849    0.053973  \n",
       "Coenzyme transport and metabolism                   212728    0.037540  \n",
       "Translation, ribosomal structure and biogenesis     393020    0.069357  \n",
       "Secondary metabolites biosynthesis, transport a...   52093    0.009193  \n",
       "Lipid transport and metabolism                      183775    0.032431  \n",
       "Cell motility                                       144533    0.025506  \n",
       "Signal transduction mechanisms                      339255    0.059869  \n",
       "Defense mechanisms                                  135950    0.023991  \n",
       "Cell wall/membrane/envelope biogenesis              520583    0.091868  \n",
       "Mobilome: prophages, transposons                     28915    0.005103  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "comp_mut_df[\"length\"] = comp_mut_df.apply(lambda r: COG_len_df.loc[r.index, \"COG length\"])\n",
    "total_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>5</td>\n",
       "      <td>254764</td>\n",
       "      <td>0.044958</td>\n",
       "      <td>9968</td>\n",
       "      <td>0.9968</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>23</td>\n",
       "      <td>160339</td>\n",
       "      <td>0.028295</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>39</td>\n",
       "      <td>589566</td>\n",
       "      <td>0.104041</td>\n",
       "      <td>392</td>\n",
       "      <td>0.0392</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>76</td>\n",
       "      <td>660007</td>\n",
       "      <td>0.116472</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>27</td>\n",
       "      <td>424603</td>\n",
       "      <td>0.074930</td>\n",
       "      <td>1170</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     observed mutation count  length  \\\n",
       "General function prediction only                           5  254764   \n",
       "Nucleotide transport and metabolism                       23  160339   \n",
       "Energy production and conversion                          39  589566   \n",
       "Amino acid transport and metabolism                       76  660007   \n",
       "Transcription                                             27  424603   \n",
       "\n",
       "                                     proportion  equal or larger count  \\\n",
       "General function prediction only       0.044958                   9968   \n",
       "Nucleotide transport and metabolism    0.028295                      0   \n",
       "Energy production and conversion       0.104041                    392   \n",
       "Amino acid transport and metabolism    0.116472                      0   \n",
       "Transcription                          0.074930                   1170   \n",
       "\n",
       "                                     p value  significant  \n",
       "General function prediction only      0.9968        False  \n",
       "Nucleotide transport and metabolism   0.0000         True  \n",
       "Energy production and conversion      0.0392         True  \n",
       "Amino acid transport and metabolism   0.0000         True  \n",
       "Transcription                         0.1170        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "import numpy as np\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=int(num_muts),\n",
    "        p=proportion_l,\n",
    "        replace=True,\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>General function prediction only</td>\n",
       "      <td>5</td>\n",
       "      <td>254764</td>\n",
       "      <td>0.044958</td>\n",
       "      <td>9968</td>\n",
       "      <td>0.9968</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>23</td>\n",
       "      <td>160339</td>\n",
       "      <td>0.028295</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>39</td>\n",
       "      <td>589566</td>\n",
       "      <td>0.104041</td>\n",
       "      <td>392</td>\n",
       "      <td>0.0392</td>\n",
       "      <td>True</td>\n",
       "      <td>0.7056</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>76</td>\n",
       "      <td>660007</td>\n",
       "      <td>0.116472</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>27</td>\n",
       "      <td>424603</td>\n",
       "      <td>0.074930</td>\n",
       "      <td>1170</td>\n",
       "      <td>0.1170</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     observed mutation count  length  \\\n",
       "General function prediction only                           5  254764   \n",
       "Nucleotide transport and metabolism                       23  160339   \n",
       "Energy production and conversion                          39  589566   \n",
       "Amino acid transport and metabolism                       76  660007   \n",
       "Transcription                                             27  424603   \n",
       "\n",
       "                                     proportion  equal or larger count  \\\n",
       "General function prediction only       0.044958                   9968   \n",
       "Nucleotide transport and metabolism    0.028295                      0   \n",
       "Energy production and conversion       0.104041                    392   \n",
       "Amino acid transport and metabolism    0.116472                      0   \n",
       "Transcription                          0.074930                   1170   \n",
       "\n",
       "                                     p value  significant  corrected p value  \\\n",
       "General function prediction only      0.9968        False             1.0000   \n",
       "Nucleotide transport and metabolism   0.0000         True             0.0000   \n",
       "Energy production and conversion      0.0392         True             0.7056   \n",
       "Amino acid transport and metabolism   0.0000         True             0.0000   \n",
       "Transcription                         0.1170        False             1.0000   \n",
       "\n",
       "                                     corrected significance  \n",
       "General function prediction only                      False  \n",
       "Nucleotide transport and metabolism                    True  \n",
       "Energy production and conversion                      False  \n",
       "Amino acid transport and metabolism                    True  \n",
       "Transcription                                         False  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, r in exp_muts_df.iterrows():\n",
    "    for d in r[\"COGs\"]:\n",
    "        if d[\"name\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"name\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(exp_muts_df.exp.unique())==1)\n",
    "exp = exp_muts_df.exp.unique()[0]\n",
    "comp_mut_df.to_csv(\"./data/supp/permutation_test_results_per_experiment/\"+exp+\"_mut_COG_signif.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_muts_df.to_pickle(\"./data/10_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     link_d = r[\"COG links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"COGs\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += len(hits)\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.2, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 mutated COGs\")\n",
    "# ax.get_legend().remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     link_d = r[\"COG links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"COGs\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#             if t not in targ_cnt_d.keys():\n",
    "#                 targ_cnt_d[t] = 0\n",
    "#             targ_cnt_d[t] += len(hits)\n",
    "# if len(targ_cnt_d) > 0:\n",
    "#     df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "#     df = df.sort_values(\"mutated feature count\")\n",
    "#     df = df[-20:]\n",
    "#     ax = df.plot.barh(y='mutated feature count',\n",
    "#     #                   figsize=(10,25),\n",
    "#     #                   figsize=(5,10),\n",
    "#                       color=\"#4C72B0\",\n",
    "#                       width=1\n",
    "#                      )\n",
    "#     for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#         ax.text(v, i - 0.2, str(v))\n",
    "#     sns.despine(ax=ax, top=True, right=True)\n",
    "#     ax.set_title(\"Signifcant mutated COGs\")\n",
    "#     ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
