{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import os, sys, itertools\n",
    "\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(114, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2527</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "      <td>[crl]</td>\n",
       "      <td>DEL</td>\n",
       "      <td>257908</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ776 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(257908, 258683)</td>\n",
       "      <td>{ECK120001080, ECK125229290, ECK125229291}</td>\n",
       "      <td>[{'name': 'insA9', 'RegulonDB ID': 'ECK1252292...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'insB9', 'RegulonDB ID': 'ECK1252292...</td>\n",
       "      <td>{'ECK125229291': ['ECK125229291'], 'ECK1252292...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2528</td>\n",
       "      <td>3</td>\n",
       "      <td>coding (163/210 nt)</td>\n",
       "      <td>cspC</td>\n",
       "      <td>INS</td>\n",
       "      <td>1907273</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>(CGTCCTG)1→2</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1907273, 1907273)</td>\n",
       "      <td>{ECK120002092}</td>\n",
       "      <td>[{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...</td>\n",
       "      <td>{'ECK120002092': ['ECK120002092']}</td>\n",
       "      <td>[{'name': 'yobF-cspC', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028784': ['ECK120002092']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Transcription'}]</td>\n",
       "      <td>{'Transcription': ['ECK120028784']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2529</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>insB1, insA</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1978503</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ776 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1978503, 1979278)</td>\n",
       "      <td>{ECK120003616, ECK120023915, ECK120003617}</td>\n",
       "      <td>[{'name': 'insB-5', 'RegulonDB ID': 'ECK120003...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120017150, ECK120013472}</td>\n",
       "      <td>{ECK125095446, ECK120034252}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'flhDC TFBS', 'RegulonDB ID': 'ECK12...</td>\n",
       "      <td>{'ECK120003616': ['ECK120013472', 'ECK12000361...</td>\n",
       "      <td>[{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...</td>\n",
       "      <td>{'ECK120014934': ['ECK120003616', 'ECK12000361...</td>\n",
       "      <td>[{'name': 'YjjQ', 'significantly associated co...</td>\n",
       "      <td>{'YjjQ': ['flhDC', 'flhDC'], 'Fur': ['flhDC', ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Mobilome: prophages, transposons'},...</td>\n",
       "      <td>{'Mobilome: prophages, transposons': ['ECK1250...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2530</td>\n",
       "      <td>32</td>\n",
       "      <td>intergenic (‑1/+1)</td>\n",
       "      <td>gatC/gatC</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2173363</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(2173363, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>{'ECK120002276': ['ECK120002276']}</td>\n",
       "      <td>[{'name': 'gatYZABCD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120015849': ['ECK120002276']}</td>\n",
       "      <td>[{'name': 'GatR', 'significantly associated co...</td>\n",
       "      <td>{'GatR': ['gatYZABCD'], 'ArcA': ['gatYZABCD'],...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2531</td>\n",
       "      <td>4</td>\n",
       "      <td>N34K (AAT→AAG)</td>\n",
       "      <td>mreB</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3400986</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3400986, 3400986)</td>\n",
       "      <td>{ECK120000601}</td>\n",
       "      <td>[{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>{'ECK120000601': ['ECK120000601']}</td>\n",
       "      <td>[{'name': 'mreBCD', 'RegulonDB ID': 'ECK120030...</td>\n",
       "      <td>{'ECK120030764': ['ECK120000601']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Cell cycle control, cell division, ...</td>\n",
       "      <td>{'Cell cycle control, cell division, chromosom...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index              Details mutation target annotation Mutation Type  \\\n",
       "2527    28                 None                      [crl]           DEL   \n",
       "2528     3  coding (163/210 nt)                       cspC           INS   \n",
       "2529    29                 None                insB1, insA           DEL   \n",
       "2530    32   intergenic (‑1/+1)                  gatC/gatC           DEL   \n",
       "2531     4       N34K (AAT→AAG)                       mreB           SNP   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale             exp  flask  \\\n",
       "2527    257908     NC_000913         Δ776 bp    1  TOL_putrescine     50   \n",
       "2528   1907273     NC_000913    (CGTCCTG)1→2    1  TOL_putrescine     50   \n",
       "2529   1978503     NC_000913         Δ776 bp    1  TOL_putrescine     50   \n",
       "2530   2173363     NC_000913           Δ2 bp    1  TOL_putrescine     50   \n",
       "2531   3400986     NC_000913             A→C    1  TOL_putrescine     50   \n",
       "\n",
       "      isolate  presence  tech_rep  coding               range  \\\n",
       "2527        1       1.0         1    True    (257908, 258683)   \n",
       "2528        1       1.0         1    True  (1907273, 1907273)   \n",
       "2529        1       1.0         1    True  (1978503, 1979278)   \n",
       "2530        1       1.0         1   False  (2173363, 2173364)   \n",
       "2531        1       1.0         1    True  (3400986, 3400986)   \n",
       "\n",
       "                               gene RegulonDB ID  \\\n",
       "2527  {ECK120001080, ECK125229290, ECK125229291}   \n",
       "2528                              {ECK120002092}   \n",
       "2529  {ECK120003616, ECK120023915, ECK120003617}   \n",
       "2530                              {ECK120002276}   \n",
       "2531                              {ECK120000601}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "2527  [{'name': 'insA9', 'RegulonDB ID': 'ECK1252292...  False       False   \n",
       "2528  [{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...  False       False   \n",
       "2529  [{'name': 'insB-5', 'RegulonDB ID': 'ECK120003...  False       False   \n",
       "2530  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...  False       False   \n",
       "2531  [{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...  False       False   \n",
       "\n",
       "                              TFBS                      promoter RBS  \\\n",
       "2527                            {}                            {}  {}   \n",
       "2528                            {}                            {}  {}   \n",
       "2529  {ECK120017150, ECK120013472}  {ECK125095446, ECK120034252}  {}   \n",
       "2530                            {}                            {}  {}   \n",
       "2531                            {}                            {}  {}   \n",
       "\n",
       "     attenuator terminator terminator  genetic  \\\n",
       "2527                    {}         {}     True   \n",
       "2528                    {}         {}     True   \n",
       "2529                    {}         {}     True   \n",
       "2530                    {}         {}    False   \n",
       "2531                    {}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "2527  [{'name': 'insB9', 'RegulonDB ID': 'ECK1252292...   \n",
       "2528  [{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...   \n",
       "2529  [{'name': 'flhDC TFBS', 'RegulonDB ID': 'ECK12...   \n",
       "2530  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...   \n",
       "2531  [{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "2527  {'ECK125229291': ['ECK125229291'], 'ECK1252292...   \n",
       "2528                 {'ECK120002092': ['ECK120002092']}   \n",
       "2529  {'ECK120003616': ['ECK120013472', 'ECK12000361...   \n",
       "2530                 {'ECK120002276': ['ECK120002276']}   \n",
       "2531                 {'ECK120000601': ['ECK120000601']}   \n",
       "\n",
       "                                                operons  \\\n",
       "2527                                                 []   \n",
       "2528  [{'name': 'yobF-cspC', 'RegulonDB ID': 'ECK120...   \n",
       "2529  [{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...   \n",
       "2530  [{'name': 'gatYZABCD', 'RegulonDB ID': 'ECK120...   \n",
       "2531  [{'name': 'mreBCD', 'RegulonDB ID': 'ECK120030...   \n",
       "\n",
       "                                           operon links  \\\n",
       "2527                                                 {}   \n",
       "2528                 {'ECK120028784': ['ECK120002092']}   \n",
       "2529  {'ECK120014934': ['ECK120003616', 'ECK12000361...   \n",
       "2530                 {'ECK120015849': ['ECK120002276']}   \n",
       "2531                 {'ECK120030764': ['ECK120000601']}   \n",
       "\n",
       "                                             regulators  \\\n",
       "2527                                                 []   \n",
       "2528                                                 []   \n",
       "2529  [{'name': 'YjjQ', 'significantly associated co...   \n",
       "2530  [{'name': 'GatR', 'significantly associated co...   \n",
       "2531                                                 []   \n",
       "\n",
       "                                        regulator links pathways  \\\n",
       "2527                                                 {}       []   \n",
       "2528                                                 {}       []   \n",
       "2529  {'YjjQ': ['flhDC', 'flhDC'], 'Fur': ['flhDC', ...       []   \n",
       "2530  {'GatR': ['gatYZABCD'], 'ArcA': ['gatYZABCD'],...       []   \n",
       "2531                                                 {}       []   \n",
       "\n",
       "     pathway links                                               COGs  \\\n",
       "2527            {}                                                 []   \n",
       "2528            {}                        [{'name': 'Transcription'}]   \n",
       "2529            {}  [{'name': 'Mobilome: prophages, transposons'},...   \n",
       "2530            {}                                                 []   \n",
       "2531            {}  [{'name': 'Cell cycle control, cell division, ...   \n",
       "\n",
       "                                              COG links temperature  \\\n",
       "2527                                                 {}  37 celsius   \n",
       "2528                {'Transcription': ['ECK120028784']}  37 celsius   \n",
       "2529  {'Mobilome: prophages, transposons': ['ECK1250...  37 celsius   \n",
       "2530                                                 {}  37 celsius   \n",
       "2531  {'Cell cycle control, cell division, chromosom...  37 celsius   \n",
       "\n",
       "     carbon-source  supplement strain-description taxonomy-id base-media  \\\n",
       "2527    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2528    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2529    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2530    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2531    glucose(2)  putrescine                 WT      511145         M9   \n",
       "\n",
       "     nitrogen-source      phosphorous-source sulfur-source calcium-source  \n",
       "2527        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2528        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2529        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2530        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2531        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# exp_muts_df = pd.read_pickle(\"./data/9_df.pkl\")  # Currently skipping imodulons for the sake of the AVA MS\n",
    "exp_muts_df = pd.read_pickle(\"./data/8_df.pkl\")\n",
    "display(exp_muts_df.shape, exp_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Mobilome: prophages, transposons</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell cycle control, cell division, chromosome partitioning</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell motility</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Posttranslational modification, protein turnover, chaperones</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count\n",
       "Transcription                                                            22\n",
       "Mobilome: prophages, transposons                                         28\n",
       "Cell cycle control, cell division, chromosome p...                        7\n",
       "Carbohydrate transport and metabolism                                     6\n",
       "Amino acid transport and metabolism                                      15\n",
       "Cell motility                                                             3\n",
       "Inorganic ion transport and metabolism                                   10\n",
       "Nucleotide transport and metabolism                                       3\n",
       "Signal transduction mechanisms                                            4\n",
       "Translation, ribosomal structure and biogenesis                           8\n",
       "Replication, recombination and repair                                     2\n",
       "Cell wall/membrane/envelope biogenesis                                    4\n",
       "Posttranslational modification, protein turnove...                        2\n",
       "Defense mechanisms                                                        2\n",
       "Energy production and conversion                                          2"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a bit behind annotations currently used in AVA MS.\n",
    "# Still uses same approach to counted, just not all data-types have been consolidated into the COGs feat col\n",
    "\n",
    "\n",
    "comp_mut_df = pd.DataFrame(columns=[\"observed mutation count\"])\n",
    "for i, r in exp_muts_df.iterrows():\n",
    "    for f, links in r[\"COG links\"].items():\n",
    "        if \"unknown\" not in f:\n",
    "            if f not in comp_mut_df.index:\n",
    "                df = pd.DataFrame({\"observed mutation count\": len(links)}, index=[f])\n",
    "                comp_mut_df = comp_mut_df.append(df)\n",
    "            else:\n",
    "                comp_mut_df.loc[f, \"observed mutation count\"] += len(links)\n",
    "            \n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "118"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(comp_mut_df[\"observed mutation count\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in exp_muts_df.iterrows():\n",
    "    for feat_d in r[\"COGs\"]:\n",
    "        if \"unknown\" not in feat_d[\"name\"]:\n",
    "            feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"name\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COG description</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>660007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "      <td>305849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>679809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Coenzyme transport and metabolism</td>\n",
       "      <td>212728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>589566</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        COG length\n",
       "COG description                                   \n",
       "Amino acid transport and metabolism         660007\n",
       "Inorganic ion transport and metabolism      305849\n",
       "Carbohydrate transport and metabolism       679809\n",
       "Coenzyme transport and metabolism           212728\n",
       "Energy production and conversion            589566"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COG_len_df = pd.read_pickle(\"./data/COG_len_df.pkl\")\n",
    "COG_len_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>22</td>\n",
       "      <td>424603</td>\n",
       "      <td>0.083610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Mobilome: prophages, transposons</td>\n",
       "      <td>28</td>\n",
       "      <td>28915</td>\n",
       "      <td>0.005694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell cycle control, cell division, chromosome partitioning</td>\n",
       "      <td>7</td>\n",
       "      <td>115057</td>\n",
       "      <td>0.022656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>6</td>\n",
       "      <td>679809</td>\n",
       "      <td>0.133864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>15</td>\n",
       "      <td>660007</td>\n",
       "      <td>0.129965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell motility</td>\n",
       "      <td>3</td>\n",
       "      <td>144533</td>\n",
       "      <td>0.028461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "      <td>10</td>\n",
       "      <td>305849</td>\n",
       "      <td>0.060226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>3</td>\n",
       "      <td>160339</td>\n",
       "      <td>0.031573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>4</td>\n",
       "      <td>339255</td>\n",
       "      <td>0.066804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Translation, ribosomal structure and biogenesis</td>\n",
       "      <td>8</td>\n",
       "      <td>393020</td>\n",
       "      <td>0.077391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>2</td>\n",
       "      <td>309957</td>\n",
       "      <td>0.061035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell wall/membrane/envelope biogenesis</td>\n",
       "      <td>4</td>\n",
       "      <td>520583</td>\n",
       "      <td>0.102510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Posttranslational modification, protein turnover, chaperones</td>\n",
       "      <td>2</td>\n",
       "      <td>270912</td>\n",
       "      <td>0.053346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Defense mechanisms</td>\n",
       "      <td>2</td>\n",
       "      <td>135950</td>\n",
       "      <td>0.026770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>2</td>\n",
       "      <td>589566</td>\n",
       "      <td>0.116094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count  \\\n",
       "Transcription                                                            22   \n",
       "Mobilome: prophages, transposons                                         28   \n",
       "Cell cycle control, cell division, chromosome p...                        7   \n",
       "Carbohydrate transport and metabolism                                     6   \n",
       "Amino acid transport and metabolism                                      15   \n",
       "Cell motility                                                             3   \n",
       "Inorganic ion transport and metabolism                                   10   \n",
       "Nucleotide transport and metabolism                                       3   \n",
       "Signal transduction mechanisms                                            4   \n",
       "Translation, ribosomal structure and biogenesis                           8   \n",
       "Replication, recombination and repair                                     2   \n",
       "Cell wall/membrane/envelope biogenesis                                    4   \n",
       "Posttranslational modification, protein turnove...                        2   \n",
       "Defense mechanisms                                                        2   \n",
       "Energy production and conversion                                          2   \n",
       "\n",
       "                                                    length  proportion  \n",
       "Transcription                                       424603    0.083610  \n",
       "Mobilome: prophages, transposons                     28915    0.005694  \n",
       "Cell cycle control, cell division, chromosome p...  115057    0.022656  \n",
       "Carbohydrate transport and metabolism               679809    0.133864  \n",
       "Amino acid transport and metabolism                 660007    0.129965  \n",
       "Cell motility                                       144533    0.028461  \n",
       "Inorganic ion transport and metabolism              305849    0.060226  \n",
       "Nucleotide transport and metabolism                 160339    0.031573  \n",
       "Signal transduction mechanisms                      339255    0.066804  \n",
       "Translation, ribosomal structure and biogenesis     393020    0.077391  \n",
       "Replication, recombination and repair               309957    0.061035  \n",
       "Cell wall/membrane/envelope biogenesis              520583    0.102510  \n",
       "Posttranslational modification, protein turnove...  270912    0.053346  \n",
       "Defense mechanisms                                  135950    0.026770  \n",
       "Energy production and conversion                    589566    0.116094  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "comp_mut_df[\"length\"] = comp_mut_df.apply(lambda r: COG_len_df.loc[r.index, \"COG length\"])\n",
    "total_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>22</td>\n",
       "      <td>424603</td>\n",
       "      <td>0.083610</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Mobilome: prophages, transposons</td>\n",
       "      <td>28</td>\n",
       "      <td>28915</td>\n",
       "      <td>0.005694</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell cycle control, cell division, chromosome partitioning</td>\n",
       "      <td>7</td>\n",
       "      <td>115057</td>\n",
       "      <td>0.022656</td>\n",
       "      <td>201</td>\n",
       "      <td>0.0201</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>6</td>\n",
       "      <td>679809</td>\n",
       "      <td>0.133864</td>\n",
       "      <td>9987</td>\n",
       "      <td>0.9987</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>15</td>\n",
       "      <td>660007</td>\n",
       "      <td>0.129965</td>\n",
       "      <td>5820</td>\n",
       "      <td>0.5820</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count  \\\n",
       "Transcription                                                            22   \n",
       "Mobilome: prophages, transposons                                         28   \n",
       "Cell cycle control, cell division, chromosome p...                        7   \n",
       "Carbohydrate transport and metabolism                                     6   \n",
       "Amino acid transport and metabolism                                      15   \n",
       "\n",
       "                                                    length  proportion  \\\n",
       "Transcription                                       424603    0.083610   \n",
       "Mobilome: prophages, transposons                     28915    0.005694   \n",
       "Cell cycle control, cell division, chromosome p...  115057    0.022656   \n",
       "Carbohydrate transport and metabolism               679809    0.133864   \n",
       "Amino acid transport and metabolism                 660007    0.129965   \n",
       "\n",
       "                                                    equal or larger count  \\\n",
       "Transcription                                                           1   \n",
       "Mobilome: prophages, transposons                                        0   \n",
       "Cell cycle control, cell division, chromosome p...                    201   \n",
       "Carbohydrate transport and metabolism                                9987   \n",
       "Amino acid transport and metabolism                                  5820   \n",
       "\n",
       "                                                    p value  significant  \n",
       "Transcription                                        0.0001         True  \n",
       "Mobilome: prophages, transposons                     0.0000         True  \n",
       "Cell cycle control, cell division, chromosome p...   0.0201         True  \n",
       "Carbohydrate transport and metabolism                0.9987        False  \n",
       "Amino acid transport and metabolism                  0.5820        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "import numpy as np\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=int(num_muts),\n",
    "        p=proportion_l,\n",
    "        replace=True,\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Transcription</td>\n",
       "      <td>22</td>\n",
       "      <td>424603</td>\n",
       "      <td>0.083610</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Mobilome: prophages, transposons</td>\n",
       "      <td>28</td>\n",
       "      <td>28915</td>\n",
       "      <td>0.005694</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cell cycle control, cell division, chromosome partitioning</td>\n",
       "      <td>7</td>\n",
       "      <td>115057</td>\n",
       "      <td>0.022656</td>\n",
       "      <td>201</td>\n",
       "      <td>0.0201</td>\n",
       "      <td>True</td>\n",
       "      <td>0.3015</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>6</td>\n",
       "      <td>679809</td>\n",
       "      <td>0.133864</td>\n",
       "      <td>9987</td>\n",
       "      <td>0.9987</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>15</td>\n",
       "      <td>660007</td>\n",
       "      <td>0.129965</td>\n",
       "      <td>5820</td>\n",
       "      <td>0.5820</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count  \\\n",
       "Transcription                                                            22   \n",
       "Mobilome: prophages, transposons                                         28   \n",
       "Cell cycle control, cell division, chromosome p...                        7   \n",
       "Carbohydrate transport and metabolism                                     6   \n",
       "Amino acid transport and metabolism                                      15   \n",
       "\n",
       "                                                    length  proportion  \\\n",
       "Transcription                                       424603    0.083610   \n",
       "Mobilome: prophages, transposons                     28915    0.005694   \n",
       "Cell cycle control, cell division, chromosome p...  115057    0.022656   \n",
       "Carbohydrate transport and metabolism               679809    0.133864   \n",
       "Amino acid transport and metabolism                 660007    0.129965   \n",
       "\n",
       "                                                    equal or larger count  \\\n",
       "Transcription                                                           1   \n",
       "Mobilome: prophages, transposons                                        0   \n",
       "Cell cycle control, cell division, chromosome p...                    201   \n",
       "Carbohydrate transport and metabolism                                9987   \n",
       "Amino acid transport and metabolism                                  5820   \n",
       "\n",
       "                                                    p value  significant  \\\n",
       "Transcription                                        0.0001         True   \n",
       "Mobilome: prophages, transposons                     0.0000         True   \n",
       "Cell cycle control, cell division, chromosome p...   0.0201         True   \n",
       "Carbohydrate transport and metabolism                0.9987        False   \n",
       "Amino acid transport and metabolism                  0.5820        False   \n",
       "\n",
       "                                                    corrected p value  \\\n",
       "Transcription                                                  0.0015   \n",
       "Mobilome: prophages, transposons                               0.0000   \n",
       "Cell cycle control, cell division, chromosome p...             0.3015   \n",
       "Carbohydrate transport and metabolism                          1.0000   \n",
       "Amino acid transport and metabolism                            1.0000   \n",
       "\n",
       "                                                    corrected significance  \n",
       "Transcription                                                         True  \n",
       "Mobilome: prophages, transposons                                      True  \n",
       "Cell cycle control, cell division, chromosome p...                   False  \n",
       "Carbohydrate transport and metabolism                                False  \n",
       "Amino acid transport and metabolism                                  False  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, r in exp_muts_df.iterrows():\n",
    "    for d in r[\"COGs\"]:\n",
    "        if d[\"name\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"name\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(exp_muts_df.exp.unique())==1)\n",
    "exp = exp_muts_df.exp.unique()[0]\n",
    "comp_mut_df.to_csv(\"./data/supp/permutation_test_results_per_experiment/\"+exp+\"_mut_COG_signif.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_muts_df.to_pickle(\"./data/10_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     link_d = r[\"COG links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"COGs\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += len(hits)\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.2, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 mutated COGs\")\n",
    "# ax.get_legend().remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     link_d = r[\"COG links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"COGs\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#             if t not in targ_cnt_d.keys():\n",
    "#                 targ_cnt_d[t] = 0\n",
    "#             targ_cnt_d[t] += len(hits)\n",
    "# if len(targ_cnt_d) > 0:\n",
    "#     df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "#     df = df.sort_values(\"mutated feature count\")\n",
    "#     df = df[-20:]\n",
    "#     ax = df.plot.barh(y='mutated feature count',\n",
    "#     #                   figsize=(10,25),\n",
    "#     #                   figsize=(5,10),\n",
    "#                       color=\"#4C72B0\",\n",
    "#                       width=1\n",
    "#                      )\n",
    "#     for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#         ax.text(v, i - 0.2, str(v))\n",
    "#     sns.despine(ax=ax, top=True, right=True)\n",
    "#     ax.set_title(\"Signifcant mutated COGs\")\n",
    "#     ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
