{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.feature import get_feat_d_from_ID\n",
    "from util.genome import is_overlap\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46, 43)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1232</td>\n",
       "      <td>4</td>\n",
       "      <td>R121C (CGC→TGC)</td>\n",
       "      <td>xylR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3735339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3735339, 3735339)</td>\n",
       "      <td>{ECK120002449}</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>{'ECK120002449': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014482': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120014482']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1233</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1234</td>\n",
       "      <td>6</td>\n",
       "      <td>L770R (CTC→CGC)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187658, 4187658)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1235</td>\n",
       "      <td>9</td>\n",
       "      <td>G164A (GGC→GCC)</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3211597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3211597, 3211597)</td>\n",
       "      <td>{ECK120000233}</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>{'ECK120000233': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120014515': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'LexA', 'significantly associated co...</td>\n",
       "      <td>{'LexA': ['rpsU-dnaG-rpoD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Replication, recombination and repa...</td>\n",
       "      <td>{'Replication, recombination and repair': ['EC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1236</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index          Details mutation target annotation Mutation Type  \\\n",
       "1232     4  R121C (CGC→TGC)                       xylR           SNP   \n",
       "1233     5             None                        rph           DEL   \n",
       "1234     6  L770R (CTC→CGC)                       rpoC           SNP   \n",
       "1235     9  G164A (GGC→GCC)                       dnaG           SNP   \n",
       "1236     5             None                        rph           DEL   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale      exp  flask  isolate  \\\n",
       "1232   3735339           NaN             C→T    1  SSW_XYL    116        1   \n",
       "1233   3815859           NaN          Δ82 bp    1  SSW_XYL    116        1   \n",
       "1234   4187658           NaN             T→G    1  SSW_XYL    116        1   \n",
       "1235   3211597           NaN             G→C    2  SSW_XYL    109        1   \n",
       "1236   3815859           NaN          Δ82 bp    2  SSW_XYL    109        1   \n",
       "\n",
       "      presence  tech_rep  coding               range gene RegulonDB ID  \\\n",
       "1232       1.0         1    True  (3735339, 3735339)    {ECK120002449}   \n",
       "1233       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "1234       1.0         1    True  (4187658, 4187658)    {ECK120000886}   \n",
       "1235       1.0         1    True  (3211597, 3211597)    {ECK120000233}   \n",
       "1236       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...  False       False   \n",
       "1233  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1236  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "\n",
       "     TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "1232   {}       {}  {}                    {}         {}     True   \n",
       "1233   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "1234   {}       {}  {}                    {}         {}     True   \n",
       "1235   {}       {}  {}                    {}         {}     True   \n",
       "1236   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...   \n",
       "1233  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...   \n",
       "1236  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1232                 {'ECK120002449': ['ECK120002449']}   \n",
       "1233  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "1234                 {'ECK120000886': ['ECK120000886']}   \n",
       "1235                 {'ECK120000233': ['ECK120000233']}   \n",
       "1236  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "\n",
       "                                                operons  \\\n",
       "1232  [{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...   \n",
       "1233  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "1234  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "1235  [{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...   \n",
       "1236  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "\n",
       "                                           operon links  \\\n",
       "1232                 {'ECK120014482': ['ECK120002449']}   \n",
       "1233  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "1234                 {'ECK120016992': ['ECK120000886']}   \n",
       "1235                 {'ECK120014515': ['ECK120000233']}   \n",
       "1236  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "\n",
       "                                             regulators  \\\n",
       "1232  [{'name': 'Fis', 'significantly associated con...   \n",
       "1233                                                 []   \n",
       "1234                                                 []   \n",
       "1235  [{'name': 'LexA', 'significantly associated co...   \n",
       "1236                                                 []   \n",
       "\n",
       "                                        regulator links temperature  \\\n",
       "1232  {'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...  37 celsius   \n",
       "1233                                                 {}  37 celsius   \n",
       "1234                                                 {}  37 celsius   \n",
       "1235                       {'LexA': ['rpsU-dnaG-rpoD']}  37 celsius   \n",
       "1236                                                 {}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "1232     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1233     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1234     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1235     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1236     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "     taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "1232      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1233      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1234      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1235      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1236      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "     sulfur-source calcium-source  \\\n",
       "1232   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1233   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1234   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1235   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1236   MgSO4(0.24)     CaCl2(0.1)   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1232  [{'name': 'Transcription', 'mutation set count...   \n",
       "1233  [{'name': 'Nucleotide transport and metabolism...   \n",
       "1234  [{'name': 'Transcription', 'mutation set count...   \n",
       "1235  [{'name': 'Replication, recombination and repa...   \n",
       "1236  [{'name': 'Nucleotide transport and metabolism...   \n",
       "\n",
       "                                              COG links  \n",
       "1232                {'Transcription': ['ECK120014482']}  \n",
       "1233  {'Nucleotide transport and metabolism': ['ECK1...  \n",
       "1234                {'Transcription': ['ECK120016992']}  \n",
       "1235  {'Replication, recombination and repair': ['EC...  \n",
       "1236  {'Nucleotide transport and metabolism': ['ECK1...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "exp_muts_df = pd.read_pickle(\"./data/10_df.pkl\")\n",
    "display(exp_muts_df.shape, exp_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ArcA</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Fis</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CRP</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>XylR</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LexA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H-NS</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lrp</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IHF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FadR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>QseB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>OmpR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Fur</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MatA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>HdfR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AcrR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LrhA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>YjjQ</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RcsAB</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FliZ</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FNR</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SlyA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LldR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TreR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       observed mutation count\n",
       "ArcA                         4\n",
       "Fis                          2\n",
       "CRP                          6\n",
       "XylR                         2\n",
       "LexA                         1\n",
       "H-NS                         3\n",
       "Lrp                          2\n",
       "IHF                          2\n",
       "FadR                         1\n",
       "QseB                         1\n",
       "OmpR                         1\n",
       "Fur                          3\n",
       "MatA                         1\n",
       "HdfR                         1\n",
       "AcrR                         1\n",
       "LrhA                         1\n",
       "YjjQ                         1\n",
       "RcsAB                        1\n",
       "FliZ                         1\n",
       "FNR                          2\n",
       "SlyA                         1\n",
       "LldR                         1\n",
       "TreR                         1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a bit behind annotations currently used in AVA MS.\n",
    "# Still uses same approach to counted, just not all data-types have been consolidated into the regulators feat col\n",
    "\n",
    "\n",
    "comp_mut_df = pd.DataFrame(columns=[\"observed mutation count\"])\n",
    "for i, r in exp_muts_df.iterrows():\n",
    "    for f, links in r[\"regulator links\"].items():\n",
    "        if f not in comp_mut_df.index:\n",
    "            df = pd.DataFrame({\"observed mutation count\": len(links)}, index=[f])\n",
    "            comp_mut_df = comp_mut_df.append(df)\n",
    "        else:\n",
    "            comp_mut_df.loc[f, \"observed mutation count\"] += len(links)\n",
    "            \n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in exp_muts_df.iterrows():\n",
    "    for feat_d in r[\"regulators\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"name\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulon length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcription factor</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>RcdA</td>\n",
       "      <td>6808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RutR</td>\n",
       "      <td>23573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>YpdB</td>\n",
       "      <td>1324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GatR</td>\n",
       "      <td>5453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>UhpA</td>\n",
       "      <td>1537</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      regulon length\n",
       "transcription factor                \n",
       "RcdA                            6808\n",
       "RutR                           23573\n",
       "YpdB                            1324\n",
       "GatR                            5453\n",
       "UhpA                            1537"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reg_len_df = pd.read_pickle(\"./data/regulon_len_df.pkl\")\n",
    "reg_len_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reg_len_df[reg_len_df.index.str.contains(\"unknown\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ArcA</td>\n",
       "      <td>4</td>\n",
       "      <td>247929</td>\n",
       "      <td>0.088824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Fis</td>\n",
       "      <td>2</td>\n",
       "      <td>271077</td>\n",
       "      <td>0.097117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CRP</td>\n",
       "      <td>6</td>\n",
       "      <td>718479</td>\n",
       "      <td>0.257406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>XylR</td>\n",
       "      <td>2</td>\n",
       "      <td>9420</td>\n",
       "      <td>0.003375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LexA</td>\n",
       "      <td>1</td>\n",
       "      <td>82592</td>\n",
       "      <td>0.029590</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      observed mutation count  length  proportion\n",
       "ArcA                        4  247929    0.088824\n",
       "Fis                         2  271077    0.097117\n",
       "CRP                         6  718479    0.257406\n",
       "XylR                        2    9420    0.003375\n",
       "LexA                        1   82592    0.029590"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "comp_mut_df[\"length\"] = comp_mut_df.apply(lambda r: reg_len_df.loc[r.index, \"regulon length\"])\n",
    "total_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ArcA</td>\n",
       "      <td>4</td>\n",
       "      <td>247929</td>\n",
       "      <td>0.088824</td>\n",
       "      <td>4791</td>\n",
       "      <td>0.4791</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Fis</td>\n",
       "      <td>2</td>\n",
       "      <td>271077</td>\n",
       "      <td>0.097117</td>\n",
       "      <td>9189</td>\n",
       "      <td>0.9189</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CRP</td>\n",
       "      <td>6</td>\n",
       "      <td>718479</td>\n",
       "      <td>0.257406</td>\n",
       "      <td>9644</td>\n",
       "      <td>0.9644</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>XylR</td>\n",
       "      <td>2</td>\n",
       "      <td>9420</td>\n",
       "      <td>0.003375</td>\n",
       "      <td>84</td>\n",
       "      <td>0.0084</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LexA</td>\n",
       "      <td>1</td>\n",
       "      <td>82592</td>\n",
       "      <td>0.029590</td>\n",
       "      <td>7018</td>\n",
       "      <td>0.7018</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      observed mutation count  length  proportion  equal or larger count  \\\n",
       "ArcA                        4  247929    0.088824                   4791   \n",
       "Fis                         2  271077    0.097117                   9189   \n",
       "CRP                         6  718479    0.257406                   9644   \n",
       "XylR                        2    9420    0.003375                     84   \n",
       "LexA                        1   82592    0.029590                   7018   \n",
       "\n",
       "      p value  significant  \n",
       "ArcA   0.4791        False  \n",
       "Fis    0.9189        False  \n",
       "CRP    0.9644        False  \n",
       "XylR   0.0084         True  \n",
       "LexA   0.7018        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=int(num_muts),\n",
    "        p=proportion_l,\n",
    "        replace=True,\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ArcA</td>\n",
       "      <td>4</td>\n",
       "      <td>247929</td>\n",
       "      <td>0.088824</td>\n",
       "      <td>4791</td>\n",
       "      <td>0.4791</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Fis</td>\n",
       "      <td>2</td>\n",
       "      <td>271077</td>\n",
       "      <td>0.097117</td>\n",
       "      <td>9189</td>\n",
       "      <td>0.9189</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CRP</td>\n",
       "      <td>6</td>\n",
       "      <td>718479</td>\n",
       "      <td>0.257406</td>\n",
       "      <td>9644</td>\n",
       "      <td>0.9644</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>XylR</td>\n",
       "      <td>2</td>\n",
       "      <td>9420</td>\n",
       "      <td>0.003375</td>\n",
       "      <td>84</td>\n",
       "      <td>0.0084</td>\n",
       "      <td>True</td>\n",
       "      <td>0.1932</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>LexA</td>\n",
       "      <td>1</td>\n",
       "      <td>82592</td>\n",
       "      <td>0.029590</td>\n",
       "      <td>7018</td>\n",
       "      <td>0.7018</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      observed mutation count  length  proportion  equal or larger count  \\\n",
       "ArcA                        4  247929    0.088824                   4791   \n",
       "Fis                         2  271077    0.097117                   9189   \n",
       "CRP                         6  718479    0.257406                   9644   \n",
       "XylR                        2    9420    0.003375                     84   \n",
       "LexA                        1   82592    0.029590                   7018   \n",
       "\n",
       "      p value  significant  corrected p value  corrected significance  \n",
       "ArcA   0.4791        False             1.0000                   False  \n",
       "Fis    0.9189        False             1.0000                   False  \n",
       "CRP    0.9644        False             1.0000                   False  \n",
       "XylR   0.0084         True             0.1932                   False  \n",
       "LexA   0.7018        False             1.0000                   False  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "# comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, r in exp_muts_df.iterrows():\n",
    "    for d in r[\"regulators\"]:\n",
    "        if d[\"name\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"name\"]][\"corrected significance\"]\n",
    "#             d[\"significant\"] = comp_mut_df.loc[d[\"name\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "exp_muts_df.to_pickle(\"./data/11_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _is_signif(name, regulons):\n",
    "#     is_signif = False\n",
    "#     for d in regulons:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# reg_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     reg_link_d = r[\"regulator links\"]\n",
    "#     for reg, hits in reg_link_d.items():\n",
    "#         t = reg        \n",
    "#         if _is_signif(reg, r[\"regulators\"]):\n",
    "#             t = '*' + t\n",
    "#         if t not in reg_cnt_d.keys(): reg_cnt_d[t] = 0\n",
    "#         reg_cnt_d[t] += len(hits)\n",
    "# df = pd.DataFrame.from_dict(reg_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.2, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 Regulon mutated feature count\")\n",
    "# ax.get_legend().remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reg_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     reg_link_d = r[\"regulator links\"]\n",
    "#     for reg, hits in reg_link_d.items():\n",
    "#         t = reg\n",
    "#         if _is_signif(reg, r[\"regulators\"]):\n",
    "#             t = '*' + t\n",
    "#             if t not in reg_cnt_d.keys(): reg_cnt_d[t] = 0\n",
    "#             reg_cnt_d[t] += len(hits)\n",
    "\n",
    "# if len(reg_cnt_d) > 0:\n",
    "#     df = pd.DataFrame.from_dict(reg_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "#     df = df.sort_values(\"mutated feature count\")\n",
    "#     ax = df.plot.barh(y='mutated feature count',\n",
    "#     #                   figsize=(10,25),\n",
    "#                       figsize=(5,10),\n",
    "#                       color=\"#4C72B0\",\n",
    "#                       width=1\n",
    "#                      )\n",
    "#     for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#         ax.text(v, i - 0.2, str(v))\n",
    "#     sns.despine(ax=ax, top=True, right=True)\n",
    "#     ax.set_title(\"Significant regulon mutated feature count\")\n",
    "#     ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
