{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100\n",
    "\n",
    "import sys, os\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(148, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>630</td>\n",
       "      <td>66</td>\n",
       "      <td>T247M (ACG→ATG)</td>\n",
       "      <td>yafD</td>\n",
       "      <td>SNP</td>\n",
       "      <td>231861</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>0.72</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(231861, 231861)</td>\n",
       "      <td>{ECK120001594}</td>\n",
       "      <td>[{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...</td>\n",
       "      <td>{'ECK120001594': ['ECK120001594']}</td>\n",
       "      <td>[{'name': 'yafDE', 'RegulonDB ID': 'ECK1200291...</td>\n",
       "      <td>{'ECK120029169': ['ECK120001594']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'General function prediction only', ...</td>\n",
       "      <td>{'General function prediction only': ['ECK1200...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>631</td>\n",
       "      <td>126</td>\n",
       "      <td>R665S (CGT→AGT)</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3993145</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→A</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3993145, 3993145)</td>\n",
       "      <td>{ECK120000166}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>{'ECK120000166': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...</td>\n",
       "      <td>{'ECK120014835': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'CRP': ['cyaA']}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}]</td>\n",
       "      <td>{'Purine metabolism': ['ECK120014835']}</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>632</td>\n",
       "      <td>94</td>\n",
       "      <td>S59Y (TCC→TAC)</td>\n",
       "      <td>glpK</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4117047</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→T</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4117047, 4117047)</td>\n",
       "      <td>{ECK120000391}</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism'}]</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>633</td>\n",
       "      <td>72</td>\n",
       "      <td>T593P (ACC→CCC)</td>\n",
       "      <td>metL</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4131611</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>0.67</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4131611, 4131611)</td>\n",
       "      <td>{ECK120000583}</td>\n",
       "      <td>[{'name': 'metL', 'RegulonDB ID': 'ECK12000058...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'metL', 'RegulonDB ID': 'ECK12000058...</td>\n",
       "      <td>{'ECK120000583': ['ECK120000583']}</td>\n",
       "      <td>[{'name': 'metBL', 'RegulonDB ID': 'ECK1200150...</td>\n",
       "      <td>{'ECK120015057': ['ECK120000583']}</td>\n",
       "      <td>[{'name': 'MetJ', 'significantly associated co...</td>\n",
       "      <td>{'MetJ': ['metBL'], 'PhoP': ['metBL'], 'HypT':...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>634</td>\n",
       "      <td>107</td>\n",
       "      <td>E641K (GAA→AAA)</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4183165</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→A</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4183165, 4183165)</td>\n",
       "      <td>{ECK120000885}</td>\n",
       "      <td>[{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000885': ['ECK120000885']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000885']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "      <td>{'Purine metabolism': ['ECK120016992'], 'Pyrim...</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index          Details mutation target annotation Mutation Type  Position  \\\n",
       "630    66  T247M (ACG→ATG)                       yafD           SNP    231861   \n",
       "631   126  R665S (CGT→AGT)                       cyaA           SNP   3993145   \n",
       "632    94   S59Y (TCC→TAC)                       glpK           SNP   4117047   \n",
       "633    72  T593P (ACC→CCC)                       metL           SNP   4131611   \n",
       "634   107  E641K (GAA→AAA)                       rpoB           SNP   4183165   \n",
       "\n",
       "    Reference Seq Sequence Change  ale  exp  flask  isolate  presence  \\\n",
       "630     NC_000913             C→T    1  GYD    280        0      0.72   \n",
       "631     NC_000913             C→A    1  GYD    280        0      1.00   \n",
       "632     NC_000913             G→T    1  GYD    280        0      1.00   \n",
       "633     NC_000913             A→C    1  GYD    280        0      0.67   \n",
       "634     NC_000913             G→A    1  GYD    280        0      1.00   \n",
       "\n",
       "     tech_rep  coding               range gene RegulonDB ID  \\\n",
       "630         1    True    (231861, 231861)    {ECK120001594}   \n",
       "631         1    True  (3993145, 3993145)    {ECK120000166}   \n",
       "632         1    True  (4117047, 4117047)    {ECK120000391}   \n",
       "633         1    True  (4131611, 4131611)    {ECK120000583}   \n",
       "634         1    True  (4183165, 4183165)    {ECK120000885}   \n",
       "\n",
       "                                      genetic features   oriC  pseudogene  \\\n",
       "630  [{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...  False       False   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...  False       False   \n",
       "632  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...  False       False   \n",
       "633  [{'name': 'metL', 'RegulonDB ID': 'ECK12000058...  False       False   \n",
       "634  [{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "\n",
       "    TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "630   {}       {}  {}                    {}         {}     True   \n",
       "631   {}       {}  {}                    {}         {}     True   \n",
       "632   {}       {}  {}                    {}         {}     True   \n",
       "633   {}       {}  {}                    {}         {}     True   \n",
       "634   {}       {}  {}                    {}         {}     True   \n",
       "\n",
       "                                      genomic features  \\\n",
       "630  [{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...   \n",
       "632  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "633  [{'name': 'metL', 'RegulonDB ID': 'ECK12000058...   \n",
       "634  [{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...   \n",
       "\n",
       "                  genetic feature links  \\\n",
       "630  {'ECK120001594': ['ECK120001594']}   \n",
       "631  {'ECK120000166': ['ECK120000166']}   \n",
       "632  {'ECK120000391': ['ECK120000391']}   \n",
       "633  {'ECK120000583': ['ECK120000583']}   \n",
       "634  {'ECK120000885': ['ECK120000885']}   \n",
       "\n",
       "                                               operons  \\\n",
       "630  [{'name': 'yafDE', 'RegulonDB ID': 'ECK1200291...   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...   \n",
       "632  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "633  [{'name': 'metBL', 'RegulonDB ID': 'ECK1200150...   \n",
       "634  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "\n",
       "                           operon links  \\\n",
       "630  {'ECK120029169': ['ECK120001594']}   \n",
       "631  {'ECK120014835': ['ECK120000166']}   \n",
       "632  {'ECK120014723': ['ECK120000391']}   \n",
       "633  {'ECK120015057': ['ECK120000583']}   \n",
       "634  {'ECK120016992': ['ECK120000885']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "630                                                 []   \n",
       "631  [{'name': 'CRP', 'significantly associated con...   \n",
       "632  [{'name': 'GlpR', 'significantly associated co...   \n",
       "633  [{'name': 'MetJ', 'significantly associated co...   \n",
       "634                                                 []   \n",
       "\n",
       "                                       regulator links  \\\n",
       "630                                                 {}   \n",
       "631                                  {'CRP': ['cyaA']}   \n",
       "632            {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "633  {'MetJ': ['metBL'], 'PhoP': ['metBL'], 'HypT':...   \n",
       "634                                                 {}   \n",
       "\n",
       "                                              pathways  \\\n",
       "630                                                 []   \n",
       "631                    [{'name': 'Purine metabolism'}]   \n",
       "632              [{'name': 'Glycerolipid metabolism'}]   \n",
       "633  [{'name': 'Glycine, serine and threonine metab...   \n",
       "634  [{'name': 'Purine metabolism'}, {'name': 'Pyri...   \n",
       "\n",
       "                                         pathway links  \\\n",
       "630                                                 {}   \n",
       "631            {'Purine metabolism': ['ECK120014835']}   \n",
       "632      {'Glycerolipid metabolism': ['ECK120014723']}   \n",
       "633  {'Glycine, serine and threonine metabolism': [...   \n",
       "634  {'Purine metabolism': ['ECK120016992'], 'Pyrim...   \n",
       "\n",
       "                                                  COGs  \\\n",
       "630  [{'name': 'General function prediction only', ...   \n",
       "631  [{'name': 'Nucleotide transport and metabolism...   \n",
       "632  [{'name': 'Energy production and conversion', ...   \n",
       "633  [{'name': 'Amino acid transport and metabolism...   \n",
       "634  [{'name': 'Transcription', 'mutation set count...   \n",
       "\n",
       "                                             COG links temperature  \\\n",
       "630  {'General function prediction only': ['ECK1200...  37 celsius   \n",
       "631  {'Nucleotide transport and metabolism': ['ECK1...  37 celsius   \n",
       "632  {'Energy production and conversion': ['ECK1200...  37 celsius   \n",
       "633  {'Amino acid transport and metabolism': ['ECK1...  37 celsius   \n",
       "634                {'Transcription': ['ECK120016992']}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "630  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "631  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "632  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "633  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "634  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "    taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "630      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "631      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "632      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "633      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "634      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "    sulfur-source calcium-source  \n",
       "630   MgSO4(0.24)     CaCl2(0.1)  \n",
       "631   MgSO4(0.24)     CaCl2(0.1)  \n",
       "632   MgSO4(0.24)     CaCl2(0.1)  \n",
       "633   MgSO4(0.24)     CaCl2(0.1)  \n",
       "634   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assoc_muts_df = pd.read_pickle(\"./data/11_df.pkl\")\n",
    "display(assoc_muts_df.shape, assoc_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycerolipid metabolism</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Fructose and mannose metabolism</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Nitrogen metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Ubiquinone and other terpenoid-quinone biosynthesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Biosynthesis of siderophore group nonribosomal peptides</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Photosynthesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Methane metabolism</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Vitamin B6 metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbon fixation in photosynthetic organisms</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Valine, leucine and isoleucine biosynthesis</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pyruvate metabolism</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Phenylalanine, tyrosine and tryptophan biosynthesis</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Novobiocin biosynthesis</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tropane, piperidine and pyridine alkaloid biosynthesis</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glyoxylate and dicarboxylate metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count\n",
       "Purine metabolism                                                        36\n",
       "Glycerolipid metabolism                                                  31\n",
       "Glycine, serine and threonine metabolism                                  2\n",
       "Cysteine and methionine metabolism                                        2\n",
       "Lysine biosynthesis                                                       2\n",
       "Pyrimidine metabolism                                                    21\n",
       "Fructose and mannose metabolism                                           4\n",
       "Drug metabolism - other enzymes                                           2\n",
       "Nitrogen metabolism                                                       1\n",
       "Ubiquinone and other terpenoid-quinone biosynth...                        1\n",
       "Biosynthesis of siderophore group nonribosomal ...                        1\n",
       "Oxidative phosphorylation                                                 1\n",
       "Photosynthesis                                                            1\n",
       "Methane metabolism                                                        4\n",
       "Vitamin B6 metabolism                                                     1\n",
       "Glycolysis / Gluconeogenesis                                              3\n",
       "Pentose phosphate pathway                                                 4\n",
       "Carbon fixation in photosynthetic organisms                               3\n",
       "Valine, leucine and isoleucine biosynthesis                              24\n",
       "Pyruvate metabolism                                                       6\n",
       "Alanine, aspartate and glutamate metabolism                               4\n",
       "Histidine metabolism                                                     36\n",
       "Tyrosine metabolism                                                       6\n",
       "Phenylalanine metabolism                                                  6\n",
       "Phenylalanine, tyrosine and tryptophan biosynth...                        6\n",
       "Novobiocin biosynthesis                                                   6\n",
       "Tropane, piperidine and pyridine alkaloid biosy...                        6\n",
       "Glyoxylate and dicarboxylate metabolism                                   1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_mut_df = pd.DataFrame(columns=[\"observed mutation count\"])\n",
    "for i, r in assoc_muts_df.iterrows():\n",
    "    for f, links in r[\"pathway links\"].items():\n",
    "        if f not in comp_mut_df.index:\n",
    "            df = pd.DataFrame({\"observed mutation count\": len(links)}, index=[f])\n",
    "            comp_mut_df = comp_mut_df.append(df)\n",
    "        else:\n",
    "            comp_mut_df.loc[f, \"observed mutation count\"] += len(links)\n",
    "\n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in assoc_muts_df.iterrows():\n",
    "    for feat_d in r[\"pathways\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"name\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>54804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>36129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "      <td>18189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Vitamin B6 metabolism</td>\n",
       "      <td>10138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>51305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          pathway length\n",
       "pathway_name                                            \n",
       "Glycine, serine and threonine metabolism           54804\n",
       "Cysteine and methionine metabolism                 36129\n",
       "Lysine biosynthesis                                18189\n",
       "Vitamin B6 metabolism                              10138\n",
       "Pentose phosphate pathway                          51305"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_len_df = pd.read_pickle(\"./data/pathway_len_df.pkl\")\n",
    "pathway_len_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "if len(comp_mut_df):\n",
    "    comp_mut_df[\"length\"] = comp_mut_df.apply(lambda r: pathway_len_df.loc[r.index, \"pathway length\"])\n",
    "    total_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "    comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "    comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>36</td>\n",
       "      <td>154034</td>\n",
       "      <td>0.121141</td>\n",
       "      <td>399</td>\n",
       "      <td>0.0399</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycerolipid metabolism</td>\n",
       "      <td>31</td>\n",
       "      <td>16744</td>\n",
       "      <td>0.013168</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>2</td>\n",
       "      <td>54804</td>\n",
       "      <td>0.043101</td>\n",
       "      <td>9996</td>\n",
       "      <td>0.9996</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>2</td>\n",
       "      <td>36129</td>\n",
       "      <td>0.028414</td>\n",
       "      <td>9865</td>\n",
       "      <td>0.9865</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "      <td>2</td>\n",
       "      <td>18189</td>\n",
       "      <td>0.014305</td>\n",
       "      <td>8263</td>\n",
       "      <td>0.8263</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          observed mutation count  length  \\\n",
       "Purine metabolism                                              36  154034   \n",
       "Glycerolipid metabolism                                        31   16744   \n",
       "Glycine, serine and threonine metabolism                        2   54804   \n",
       "Cysteine and methionine metabolism                              2   36129   \n",
       "Lysine biosynthesis                                             2   18189   \n",
       "\n",
       "                                          proportion  equal or larger count  \\\n",
       "Purine metabolism                           0.121141                    399   \n",
       "Glycerolipid metabolism                     0.013168                      0   \n",
       "Glycine, serine and threonine metabolism    0.043101                   9996   \n",
       "Cysteine and methionine metabolism          0.028414                   9865   \n",
       "Lysine biosynthesis                         0.014305                   8263   \n",
       "\n",
       "                                          p value  significant  \n",
       "Purine metabolism                          0.0399         True  \n",
       "Glycerolipid metabolism                    0.0000         True  \n",
       "Glycine, serine and threonine metabolism   0.9996        False  \n",
       "Cysteine and methionine metabolism         0.9865        False  \n",
       "Lysine biosynthesis                        0.8263        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "if len(comp_mut_df):\n",
    "    comp_mut_df['equal or larger count'] = 0\n",
    "    component_l = comp_mut_df.index.tolist()\n",
    "    proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "    num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "    # slides from UW speak about the relationship of permutations and precision.\n",
    "    # should read through them to get a sense of what is the least iterations that should execute.\n",
    "    num_itr = 10000\n",
    "    for i in range(0, num_itr):\n",
    "        random_mut_comp_a = np.random.choice(\n",
    "            a=component_l,\n",
    "            size=int(num_muts),\n",
    "            p=proportion_l,\n",
    "            replace=True,\n",
    "        )\n",
    "        random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "        for component in random_mut_comp_count_d.keys():\n",
    "            if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "                comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "    comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "        lambda count: count/num_itr)\n",
    "\n",
    "    comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "    display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats import multitest\n",
    "\n",
    "\n",
    "if len(comp_mut_df):\n",
    "    pvals = comp_mut_df[\"p value\"]\n",
    "    rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "        pvals=pvals,\n",
    "        alpha=ENRICH_ALPHA,\n",
    "        method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "    pd.options.display.max_rows = 500\n",
    "    comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "    comp_mut_df[\"corrected significance\"] = rejects\n",
    "    comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "if len(comp_mut_df):\n",
    "    comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(comp_mut_df):\n",
    "    for i, r in assoc_muts_df.iterrows():\n",
    "        for d in r[\"pathways\"]:\n",
    "            if d[\"name\"] in comp_mut_df.index:\n",
    "                d[\"significantly enriched\"] = comp_mut_df.loc[d[\"name\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(assoc_muts_df.exp.unique())==1)\n",
    "exp = assoc_muts_df.exp.unique()[0]\n",
    "comp_mut_df.to_csv(\"./data/supp/permutation_test_results_per_experiment/\"+exp+\"_mut_pwy_signif.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_muts_df.to_pickle(\"./data/12_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     link_d = r[\"pathway links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"pathways\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += len(hits)\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.2, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 mutated pathways\")\n",
    "# ax.get_legend().remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     link_d = r[\"pathway links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"pathways\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#             if t not in targ_cnt_d.keys():\n",
    "#                 targ_cnt_d[t] = 0\n",
    "#             targ_cnt_d[t] += len(hits)\n",
    "            \n",
    "# if len(targ_cnt_d) > 0:\n",
    "#     df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "#     df = df.sort_values(\"mutated feature count\")\n",
    "#     df = df[-20:]\n",
    "#     display(df)\n",
    "#     ax = df.plot.barh(y='mutated feature count',\n",
    "#     #                   figsize=(10,25),\n",
    "#                       figsize=(5,10),\n",
    "#                       color=\"#4C72B0\",\n",
    "#                       width=1\n",
    "#                      )\n",
    "#     for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#         ax.text(v, i - 0.2, str(v))\n",
    "#     sns.despine(ax=ax, top=True, right=True)\n",
    "#     ax.set_title(\"Significant mutated pathways\")\n",
    "#     ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
