{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100\n",
    "\n",
    "import sys, os\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(114, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2527</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "      <td>[crl]</td>\n",
       "      <td>DEL</td>\n",
       "      <td>257908</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ776 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(257908, 258683)</td>\n",
       "      <td>{ECK120001080, ECK125229291, ECK125229290}</td>\n",
       "      <td>[{'name': 'insA9', 'RegulonDB ID': 'ECK1252292...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'insB9', 'RegulonDB ID': 'ECK1252292...</td>\n",
       "      <td>{'ECK125229291': ['ECK125229291'], 'ECK1252292...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2528</td>\n",
       "      <td>3</td>\n",
       "      <td>coding (163/210 nt)</td>\n",
       "      <td>cspC</td>\n",
       "      <td>INS</td>\n",
       "      <td>1907273</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>(CGTCCTG)1→2</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1907273, 1907273)</td>\n",
       "      <td>{ECK120002092}</td>\n",
       "      <td>[{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...</td>\n",
       "      <td>{'ECK120002092': ['ECK120002092']}</td>\n",
       "      <td>[{'name': 'yobF-cspC', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028784': ['ECK120002092']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120028784']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2529</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>insB1, insA</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1978503</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ776 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1978503, 1979278)</td>\n",
       "      <td>{ECK120023915, ECK120003617, ECK120003616}</td>\n",
       "      <td>[{'name': 'insB-5', 'RegulonDB ID': 'ECK120003...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120017150, ECK120013472}</td>\n",
       "      <td>{ECK125095446, ECK120034252}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'flhDC TFBS', 'RegulonDB ID': 'ECK12...</td>\n",
       "      <td>{'ECK120003616': ['ECK120013472', 'ECK12000361...</td>\n",
       "      <td>[{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...</td>\n",
       "      <td>{'ECK120014934': ['ECK120003616', 'ECK12000361...</td>\n",
       "      <td>[{'name': 'YjjQ', 'significantly associated co...</td>\n",
       "      <td>{'YjjQ': ['flhDC', 'flhDC'], 'Fur': ['flhDC', ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Mobilome: prophages, transposons', ...</td>\n",
       "      <td>{'Mobilome: prophages, transposons': ['ECK1250...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2530</td>\n",
       "      <td>32</td>\n",
       "      <td>intergenic (‑1/+1)</td>\n",
       "      <td>gatC/gatC</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2173363</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(2173363, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>{'ECK120002276': ['ECK120002276']}</td>\n",
       "      <td>[{'name': 'gatYZABCD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120015849': ['ECK120002276']}</td>\n",
       "      <td>[{'name': 'GatR', 'significantly associated co...</td>\n",
       "      <td>{'GatR': ['gatYZABCD'], 'ArcA': ['gatYZABCD'],...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2531</td>\n",
       "      <td>4</td>\n",
       "      <td>N34K (AAT→AAG)</td>\n",
       "      <td>mreB</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3400986</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3400986, 3400986)</td>\n",
       "      <td>{ECK120000601}</td>\n",
       "      <td>[{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>{'ECK120000601': ['ECK120000601']}</td>\n",
       "      <td>[{'name': 'mreBCD', 'RegulonDB ID': 'ECK120030...</td>\n",
       "      <td>{'ECK120030764': ['ECK120000601']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Cell cycle control, cell division, ...</td>\n",
       "      <td>{'Cell cycle control, cell division, chromosom...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index              Details mutation target annotation Mutation Type  \\\n",
       "2527    28                 None                      [crl]           DEL   \n",
       "2528     3  coding (163/210 nt)                       cspC           INS   \n",
       "2529    29                 None                insB1, insA           DEL   \n",
       "2530    32   intergenic (‑1/+1)                  gatC/gatC           DEL   \n",
       "2531     4       N34K (AAT→AAG)                       mreB           SNP   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale             exp  flask  \\\n",
       "2527    257908     NC_000913         Δ776 bp    1  TOL_putrescine     50   \n",
       "2528   1907273     NC_000913    (CGTCCTG)1→2    1  TOL_putrescine     50   \n",
       "2529   1978503     NC_000913         Δ776 bp    1  TOL_putrescine     50   \n",
       "2530   2173363     NC_000913           Δ2 bp    1  TOL_putrescine     50   \n",
       "2531   3400986     NC_000913             A→C    1  TOL_putrescine     50   \n",
       "\n",
       "      isolate  presence  tech_rep  coding               range  \\\n",
       "2527        1       1.0         1    True    (257908, 258683)   \n",
       "2528        1       1.0         1    True  (1907273, 1907273)   \n",
       "2529        1       1.0         1    True  (1978503, 1979278)   \n",
       "2530        1       1.0         1   False  (2173363, 2173364)   \n",
       "2531        1       1.0         1    True  (3400986, 3400986)   \n",
       "\n",
       "                               gene RegulonDB ID  \\\n",
       "2527  {ECK120001080, ECK125229291, ECK125229290}   \n",
       "2528                              {ECK120002092}   \n",
       "2529  {ECK120023915, ECK120003617, ECK120003616}   \n",
       "2530                              {ECK120002276}   \n",
       "2531                              {ECK120000601}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "2527  [{'name': 'insA9', 'RegulonDB ID': 'ECK1252292...  False       False   \n",
       "2528  [{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...  False       False   \n",
       "2529  [{'name': 'insB-5', 'RegulonDB ID': 'ECK120003...  False       False   \n",
       "2530  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...  False       False   \n",
       "2531  [{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...  False       False   \n",
       "\n",
       "                              TFBS                      promoter RBS  \\\n",
       "2527                            {}                            {}  {}   \n",
       "2528                            {}                            {}  {}   \n",
       "2529  {ECK120017150, ECK120013472}  {ECK125095446, ECK120034252}  {}   \n",
       "2530                            {}                            {}  {}   \n",
       "2531                            {}                            {}  {}   \n",
       "\n",
       "     attenuator terminator terminator  genetic  \\\n",
       "2527                    {}         {}     True   \n",
       "2528                    {}         {}     True   \n",
       "2529                    {}         {}     True   \n",
       "2530                    {}         {}    False   \n",
       "2531                    {}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "2527  [{'name': 'insB9', 'RegulonDB ID': 'ECK1252292...   \n",
       "2528  [{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...   \n",
       "2529  [{'name': 'flhDC TFBS', 'RegulonDB ID': 'ECK12...   \n",
       "2530  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...   \n",
       "2531  [{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "2527  {'ECK125229291': ['ECK125229291'], 'ECK1252292...   \n",
       "2528                 {'ECK120002092': ['ECK120002092']}   \n",
       "2529  {'ECK120003616': ['ECK120013472', 'ECK12000361...   \n",
       "2530                 {'ECK120002276': ['ECK120002276']}   \n",
       "2531                 {'ECK120000601': ['ECK120000601']}   \n",
       "\n",
       "                                                operons  \\\n",
       "2527                                                 []   \n",
       "2528  [{'name': 'yobF-cspC', 'RegulonDB ID': 'ECK120...   \n",
       "2529  [{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...   \n",
       "2530  [{'name': 'gatYZABCD', 'RegulonDB ID': 'ECK120...   \n",
       "2531  [{'name': 'mreBCD', 'RegulonDB ID': 'ECK120030...   \n",
       "\n",
       "                                           operon links  \\\n",
       "2527                                                 {}   \n",
       "2528                 {'ECK120028784': ['ECK120002092']}   \n",
       "2529  {'ECK120014934': ['ECK120003616', 'ECK12000361...   \n",
       "2530                 {'ECK120015849': ['ECK120002276']}   \n",
       "2531                 {'ECK120030764': ['ECK120000601']}   \n",
       "\n",
       "                                             regulators  \\\n",
       "2527                                                 []   \n",
       "2528                                                 []   \n",
       "2529  [{'name': 'YjjQ', 'significantly associated co...   \n",
       "2530  [{'name': 'GatR', 'significantly associated co...   \n",
       "2531                                                 []   \n",
       "\n",
       "                                        regulator links pathways  \\\n",
       "2527                                                 {}       []   \n",
       "2528                                                 {}       []   \n",
       "2529  {'YjjQ': ['flhDC', 'flhDC'], 'Fur': ['flhDC', ...       []   \n",
       "2530  {'GatR': ['gatYZABCD'], 'ArcA': ['gatYZABCD'],...       []   \n",
       "2531                                                 {}       []   \n",
       "\n",
       "     pathway links                                               COGs  \\\n",
       "2527            {}                                                 []   \n",
       "2528            {}  [{'name': 'Transcription', 'mutation set count...   \n",
       "2529            {}  [{'name': 'Mobilome: prophages, transposons', ...   \n",
       "2530            {}                                                 []   \n",
       "2531            {}  [{'name': 'Cell cycle control, cell division, ...   \n",
       "\n",
       "                                              COG links temperature  \\\n",
       "2527                                                 {}  37 celsius   \n",
       "2528                {'Transcription': ['ECK120028784']}  37 celsius   \n",
       "2529  {'Mobilome: prophages, transposons': ['ECK1250...  37 celsius   \n",
       "2530                                                 {}  37 celsius   \n",
       "2531  {'Cell cycle control, cell division, chromosom...  37 celsius   \n",
       "\n",
       "     carbon-source  supplement strain-description taxonomy-id base-media  \\\n",
       "2527    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2528    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2529    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2530    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2531    glucose(2)  putrescine                 WT      511145         M9   \n",
       "\n",
       "     nitrogen-source      phosphorous-source sulfur-source calcium-source  \n",
       "2527        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2528        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2529        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2530        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2531        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assoc_muts_df = pd.read_pickle(\"./data/11_df.pkl\")\n",
    "display(assoc_muts_df.shape, assoc_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose and glucuronate interconversions</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glyoxylate and dicarboxylate metabolism</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Phenylalanine, tyrosine and tryptophan biosynthesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Novobiocin biosynthesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Isoquinoline alkaloid biosynthesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tropane, piperidine and pyridine alkaloid biosynthesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pyruvate metabolism</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Carbon fixation in photosynthetic organisms</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Amino sugar and nucleotide sugar metabolism</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Peptidoglycan biosynthesis</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Valine, leucine and isoleucine biosynthesis</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    observed mutation count\n",
       "Purine metabolism                                                        12\n",
       "Pyrimidine metabolism                                                    10\n",
       "Pentose phosphate pathway                                                 4\n",
       "Pentose and glucuronate interconversions                                  2\n",
       "Arginine and proline metabolism                                           3\n",
       "Glyoxylate and dicarboxylate metabolism                                   2\n",
       "Drug metabolism - other enzymes                                           3\n",
       "Cysteine and methionine metabolism                                        1\n",
       "Lysine biosynthesis                                                       1\n",
       "Tyrosine metabolism                                                       1\n",
       "Phenylalanine metabolism                                                  1\n",
       "Phenylalanine, tyrosine and tryptophan biosynth...                        1\n",
       "Novobiocin biosynthesis                                                   1\n",
       "Isoquinoline alkaloid biosynthesis                                        1\n",
       "Tropane, piperidine and pyridine alkaloid biosy...                        1\n",
       "Glycolysis / Gluconeogenesis                                              1\n",
       "Pyruvate metabolism                                                       3\n",
       "Carbon fixation in photosynthetic organisms                               1\n",
       "Amino sugar and nucleotide sugar metabolism                               2\n",
       "Peptidoglycan biosynthesis                                                2\n",
       "Valine, leucine and isoleucine biosynthesis                               8\n",
       "Alanine, aspartate and glutamate metabolism                               1"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_mut_df = pd.DataFrame(columns=[\"observed mutation count\"])\n",
    "for i, r in assoc_muts_df.iterrows():\n",
    "    for f, links in r[\"pathway links\"].items():\n",
    "        if f not in comp_mut_df.index:\n",
    "            df = pd.DataFrame({\"observed mutation count\": len(links)}, index=[f])\n",
    "            comp_mut_df = comp_mut_df.append(df)\n",
    "        else:\n",
    "            comp_mut_df.loc[f, \"observed mutation count\"] += len(links)\n",
    "\n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in assoc_muts_df.iterrows():\n",
    "    for feat_d in r[\"pathways\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"name\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>54804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>36129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "      <td>18189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Vitamin B6 metabolism</td>\n",
       "      <td>10138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>51305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          pathway length\n",
       "pathway_name                                            \n",
       "Glycine, serine and threonine metabolism           54804\n",
       "Cysteine and methionine metabolism                 36129\n",
       "Lysine biosynthesis                                18189\n",
       "Vitamin B6 metabolism                              10138\n",
       "Pentose phosphate pathway                          51305"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_len_df = pd.read_pickle(\"./data/pathway_len_df.pkl\")\n",
    "pathway_len_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "if len(comp_mut_df):\n",
    "    comp_mut_df[\"length\"] = comp_mut_df.apply(lambda r: pathway_len_df.loc[r.index, \"pathway length\"])\n",
    "    total_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "    comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "    comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>12</td>\n",
       "      <td>154034</td>\n",
       "      <td>0.146006</td>\n",
       "      <td>1835</td>\n",
       "      <td>0.1835</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>10</td>\n",
       "      <td>89371</td>\n",
       "      <td>0.084713</td>\n",
       "      <td>374</td>\n",
       "      <td>0.0374</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>4</td>\n",
       "      <td>51305</td>\n",
       "      <td>0.048631</td>\n",
       "      <td>3497</td>\n",
       "      <td>0.3497</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose and glucuronate interconversions</td>\n",
       "      <td>2</td>\n",
       "      <td>32217</td>\n",
       "      <td>0.030538</td>\n",
       "      <td>5671</td>\n",
       "      <td>0.5671</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>3</td>\n",
       "      <td>71864</td>\n",
       "      <td>0.068119</td>\n",
       "      <td>8013</td>\n",
       "      <td>0.8013</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          observed mutation count  length  \\\n",
       "Purine metabolism                                              12  154034   \n",
       "Pyrimidine metabolism                                          10   89371   \n",
       "Pentose phosphate pathway                                       4   51305   \n",
       "Pentose and glucuronate interconversions                        2   32217   \n",
       "Arginine and proline metabolism                                 3   71864   \n",
       "\n",
       "                                          proportion  equal or larger count  \\\n",
       "Purine metabolism                           0.146006                   1835   \n",
       "Pyrimidine metabolism                       0.084713                    374   \n",
       "Pentose phosphate pathway                   0.048631                   3497   \n",
       "Pentose and glucuronate interconversions    0.030538                   5671   \n",
       "Arginine and proline metabolism             0.068119                   8013   \n",
       "\n",
       "                                          p value  significant  \n",
       "Purine metabolism                          0.1835        False  \n",
       "Pyrimidine metabolism                      0.0374         True  \n",
       "Pentose phosphate pathway                  0.3497        False  \n",
       "Pentose and glucuronate interconversions   0.5671        False  \n",
       "Arginine and proline metabolism            0.8013        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "if len(comp_mut_df):\n",
    "    comp_mut_df['equal or larger count'] = 0\n",
    "    component_l = comp_mut_df.index.tolist()\n",
    "    proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "    num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "    # slides from UW speak about the relationship of permutations and precision.\n",
    "    # should read through them to get a sense of what is the least iterations that should execute.\n",
    "    num_itr = 10000\n",
    "    for i in range(0, num_itr):\n",
    "        random_mut_comp_a = np.random.choice(\n",
    "            a=component_l,\n",
    "            size=int(num_muts),\n",
    "            p=proportion_l,\n",
    "            replace=True,\n",
    "        )\n",
    "        random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "        for component in random_mut_comp_count_d.keys():\n",
    "            if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "                comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "    comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "        lambda count: count/num_itr)\n",
    "\n",
    "    comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "    display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats import multitest\n",
    "\n",
    "\n",
    "if len(comp_mut_df):\n",
    "    pvals = comp_mut_df[\"p value\"]\n",
    "    rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "        pvals=pvals,\n",
    "        alpha=ENRICH_ALPHA,\n",
    "        method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "    pd.options.display.max_rows = 500\n",
    "    comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "    comp_mut_df[\"corrected significance\"] = rejects\n",
    "    comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "if len(comp_mut_df):\n",
    "    comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(comp_mut_df):\n",
    "    for i, r in assoc_muts_df.iterrows():\n",
    "        for d in r[\"pathways\"]:\n",
    "            if d[\"name\"] in comp_mut_df.index:\n",
    "                d[\"significantly enriched\"] = comp_mut_df.loc[d[\"name\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(assoc_muts_df.exp.unique())==1)\n",
    "exp = assoc_muts_df.exp.unique()[0]\n",
    "comp_mut_df.to_csv(\"./data/supp/permutation_test_results_per_experiment/\"+exp+\"_mut_pwy_signif.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_muts_df.to_pickle(\"./data/12_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     link_d = r[\"pathway links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"pathways\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += len(hits)\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.2, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 mutated pathways\")\n",
    "# ax.get_legend().remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     link_d = r[\"pathway links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"pathways\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#             if t not in targ_cnt_d.keys():\n",
    "#                 targ_cnt_d[t] = 0\n",
    "#             targ_cnt_d[t] += len(hits)\n",
    "            \n",
    "# if len(targ_cnt_d) > 0:\n",
    "#     df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "#     df = df.sort_values(\"mutated feature count\")\n",
    "#     df = df[-20:]\n",
    "#     display(df)\n",
    "#     ax = df.plot.barh(y='mutated feature count',\n",
    "#     #                   figsize=(10,25),\n",
    "#                       figsize=(5,10),\n",
    "#                       color=\"#4C72B0\",\n",
    "#                       width=1\n",
    "#                      )\n",
    "#     for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#         ax.text(v, i - 0.2, str(v))\n",
    "#     sns.despine(ax=ax, top=True, right=True)\n",
    "#     ax.set_title(\"Significant mutated pathways\")\n",
    "#     ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
