{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from util.gene import get_gene_bnum\n",
    "pd.options.display.max_columns = 100\n",
    "\n",
    "import sys, os\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.operon import get_operon_ID_set\n",
    "from util.genome import NON_K12_EXP_L\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46, 43)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1232</td>\n",
       "      <td>4</td>\n",
       "      <td>R121C (CGC→TGC)</td>\n",
       "      <td>xylR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3735339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3735339, 3735339)</td>\n",
       "      <td>{ECK120002449}</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>{'ECK120002449': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014482': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120014482']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1233</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1234</td>\n",
       "      <td>6</td>\n",
       "      <td>L770R (CTC→CGC)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187658, 4187658)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1235</td>\n",
       "      <td>9</td>\n",
       "      <td>G164A (GGC→GCC)</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3211597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3211597, 3211597)</td>\n",
       "      <td>{ECK120000233}</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>{'ECK120000233': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120014515': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'LexA', 'significantly associated co...</td>\n",
       "      <td>{'LexA': ['rpsU-dnaG-rpoD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Replication, recombination and repa...</td>\n",
       "      <td>{'Replication, recombination and repair': ['EC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1236</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index          Details mutation target annotation Mutation Type  \\\n",
       "1232     4  R121C (CGC→TGC)                       xylR           SNP   \n",
       "1233     5             None                        rph           DEL   \n",
       "1234     6  L770R (CTC→CGC)                       rpoC           SNP   \n",
       "1235     9  G164A (GGC→GCC)                       dnaG           SNP   \n",
       "1236     5             None                        rph           DEL   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale      exp  flask  isolate  \\\n",
       "1232   3735339           NaN             C→T    1  SSW_XYL    116        1   \n",
       "1233   3815859           NaN          Δ82 bp    1  SSW_XYL    116        1   \n",
       "1234   4187658           NaN             T→G    1  SSW_XYL    116        1   \n",
       "1235   3211597           NaN             G→C    2  SSW_XYL    109        1   \n",
       "1236   3815859           NaN          Δ82 bp    2  SSW_XYL    109        1   \n",
       "\n",
       "      presence  tech_rep  coding               range gene RegulonDB ID  \\\n",
       "1232       1.0         1    True  (3735339, 3735339)    {ECK120002449}   \n",
       "1233       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "1234       1.0         1    True  (4187658, 4187658)    {ECK120000886}   \n",
       "1235       1.0         1    True  (3211597, 3211597)    {ECK120000233}   \n",
       "1236       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...  False       False   \n",
       "1233  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1236  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "\n",
       "     TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "1232   {}       {}  {}                    {}         {}     True   \n",
       "1233   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "1234   {}       {}  {}                    {}         {}     True   \n",
       "1235   {}       {}  {}                    {}         {}     True   \n",
       "1236   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...   \n",
       "1233  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...   \n",
       "1236  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1232                 {'ECK120002449': ['ECK120002449']}   \n",
       "1233  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "1234                 {'ECK120000886': ['ECK120000886']}   \n",
       "1235                 {'ECK120000233': ['ECK120000233']}   \n",
       "1236  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "\n",
       "                                                operons  \\\n",
       "1232  [{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...   \n",
       "1233  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "1234  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "1235  [{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...   \n",
       "1236  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "\n",
       "                                           operon links  \\\n",
       "1232                 {'ECK120014482': ['ECK120002449']}   \n",
       "1233  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "1234                 {'ECK120016992': ['ECK120000886']}   \n",
       "1235                 {'ECK120014515': ['ECK120000233']}   \n",
       "1236  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "\n",
       "                                             regulators  \\\n",
       "1232  [{'name': 'Fis', 'significantly associated con...   \n",
       "1233                                                 []   \n",
       "1234                                                 []   \n",
       "1235  [{'name': 'LexA', 'significantly associated co...   \n",
       "1236                                                 []   \n",
       "\n",
       "                                        regulator links temperature  \\\n",
       "1232  {'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...  37 celsius   \n",
       "1233                                                 {}  37 celsius   \n",
       "1234                                                 {}  37 celsius   \n",
       "1235                       {'LexA': ['rpsU-dnaG-rpoD']}  37 celsius   \n",
       "1236                                                 {}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "1232     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1233     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1234     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1235     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1236     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "     taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "1232      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1233      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1234      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1235      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1236      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "     sulfur-source calcium-source  \\\n",
       "1232   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1233   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1234   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1235   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1236   MgSO4(0.24)     CaCl2(0.1)   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1232  [{'name': 'Transcription', 'mutation set count...   \n",
       "1233  [{'name': 'Nucleotide transport and metabolism...   \n",
       "1234  [{'name': 'Transcription', 'mutation set count...   \n",
       "1235  [{'name': 'Replication, recombination and repa...   \n",
       "1236  [{'name': 'Nucleotide transport and metabolism...   \n",
       "\n",
       "                                              COG links  \n",
       "1232                {'Transcription': ['ECK120014482']}  \n",
       "1233  {'Nucleotide transport and metabolism': ['ECK1...  \n",
       "1234                {'Transcription': ['ECK120016992']}  \n",
       "1235  {'Replication, recombination and repair': ['EC...  \n",
       "1236  {'Nucleotide transport and metabolism': ['ECK1...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assoc_muts_df = pd.read_pickle(\"./data/11_df.pkl\")\n",
    "display(assoc_muts_df.shape, assoc_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "TU_objects_df = pd.read_csv(\"./data/RegulonDB10/tu_objects_tmp.txt\",\n",
    "                            sep=\"\\t\",\n",
    "                            comment='#',\n",
    "                            header=None,\n",
    "                            )\n",
    "\n",
    "TU_objects_df.columns = [\n",
    "    \"TRANSCRIPTION_UNIT_ID\",\n",
    "    \"NUMTU\",\n",
    "    \"TU_POSLEFT\",\n",
    "    \"TU_POSRIGHT\",\n",
    "    \"TU_TYPE\",\n",
    "    \"TU_OBJECT_CLASS\",\n",
    "    \"TU_OBJECT_ID\",\n",
    "    \"TU_OBJECT_NAME\",\n",
    "    \"TU_OBJECT_POSLEFT\",\n",
    "    \"TU_OBJECT_POSRIGHT\",\n",
    "    \"TU_OBJECT_STRAND\",\n",
    "    \"TU_OBJECT_COLORCLASS\",\n",
    "    \"TU_OBJECT_DESCRIPTION\",\n",
    "    \"TU_OBJECT_SIGMA\",\n",
    "    \"TU_OBJECT_EVIDENCE\",\n",
    "    \"TU_OBJECT_RI_TYPE\",\n",
    "    \"TU_OBJECT_TYPE\",\n",
    "    \"EVIDENCE\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBJECT_ID</th>\n",
       "      <th>OBJECT_SYNONYM_NAME</th>\n",
       "      <th>OS_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>EG10001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>ECK4045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>b4053</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>b0764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      OBJECT_ID OBJECT_SYNONYM_NAME  OS_INTERNAL_COMMENT KEY_ID_ORG\n",
       "0  ECK120000001             EG10001                  NaN      ECK12\n",
       "1  ECK120000001             ECK4045                  NaN      ECK12\n",
       "2  ECK120000001               b4053                  NaN      ECK12\n",
       "3  ECK120000001                alr5                  NaN      ECK12\n",
       "4  ECK120000002               b0764                  NaN      ECK12"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_synonym_df = pd.read_csv(\n",
    "    \"./data/RegulonDB10/object_synonym.txt\",\n",
    "    sep=\"\\t\",\n",
    "    comment='#',\n",
    "    header=None,\n",
    "    quoting=3\n",
    ")\n",
    "gene_synonym_df.columns = [\"OBJECT_ID\", \"OBJECT_SYNONYM_NAME\", \"OS_INTERNAL_COMMENT\", \"KEY_ID_ORG\"]\n",
    "gene_synonym_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq_locus_tag</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>b0002</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>b0002</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>b0002</td>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>b0003</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>b0004</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  refseq_locus_tag                              pathway_name\n",
       "0            b0002  Glycine, serine and threonine metabolism\n",
       "1            b0002        Cysteine and methionine metabolism\n",
       "2            b0002                       Lysine biosynthesis\n",
       "6            b0003  Glycine, serine and threonine metabolism\n",
       "7            b0004  Glycine, serine and threonine metabolism"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_df = pd.read_csv(\n",
    "    \"./data/pathways/511145.12.PATRIC.pathway.tab\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "# Some of the entries have different entries for columsn in ec_description and ec_number and therefore can make duplicate entries when associating pathways to genes\n",
    "pathway_df = pathway_df.drop(columns=['genome_id',\n",
    " 'genome_name',\n",
    " 'patric_id',\n",
    " 'alt_locus_tag',\n",
    " 'gene',\n",
    " 'product',\n",
    " 'ec_number',\n",
    " 'ec_description',\n",
    " 'pathway_id',])\n",
    "pathway_df = pathway_df.drop_duplicates()\n",
    "\n",
    "pathway_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1232</td>\n",
       "      <td>4</td>\n",
       "      <td>R121C (CGC→TGC)</td>\n",
       "      <td>xylR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3735339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3735339, 3735339)</td>\n",
       "      <td>{ECK120002449}</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>{'ECK120002449': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014482': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120014482']}</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1233</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism'}, {'name': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1234</td>\n",
       "      <td>6</td>\n",
       "      <td>L770R (CTC→CGC)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187658, 4187658)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1235</td>\n",
       "      <td>9</td>\n",
       "      <td>G164A (GGC→GCC)</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3211597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3211597, 3211597)</td>\n",
       "      <td>{ECK120000233}</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>{'ECK120000233': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120014515': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'LexA', 'significantly associated co...</td>\n",
       "      <td>{'LexA': ['rpsU-dnaG-rpoD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Replication, recombination and repa...</td>\n",
       "      <td>{'Replication, recombination and repair': ['EC...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1236</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism'}, {'name': '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index          Details mutation target annotation Mutation Type  \\\n",
       "1232     4  R121C (CGC→TGC)                       xylR           SNP   \n",
       "1233     5             None                        rph           DEL   \n",
       "1234     6  L770R (CTC→CGC)                       rpoC           SNP   \n",
       "1235     9  G164A (GGC→GCC)                       dnaG           SNP   \n",
       "1236     5             None                        rph           DEL   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale      exp  flask  isolate  \\\n",
       "1232   3735339           NaN             C→T    1  SSW_XYL    116        1   \n",
       "1233   3815859           NaN          Δ82 bp    1  SSW_XYL    116        1   \n",
       "1234   4187658           NaN             T→G    1  SSW_XYL    116        1   \n",
       "1235   3211597           NaN             G→C    2  SSW_XYL    109        1   \n",
       "1236   3815859           NaN          Δ82 bp    2  SSW_XYL    109        1   \n",
       "\n",
       "      presence  tech_rep  coding               range gene RegulonDB ID  \\\n",
       "1232       1.0         1    True  (3735339, 3735339)    {ECK120002449}   \n",
       "1233       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "1234       1.0         1    True  (4187658, 4187658)    {ECK120000886}   \n",
       "1235       1.0         1    True  (3211597, 3211597)    {ECK120000233}   \n",
       "1236       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...  False       False   \n",
       "1233  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1236  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "\n",
       "     TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "1232   {}       {}  {}                    {}         {}     True   \n",
       "1233   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "1234   {}       {}  {}                    {}         {}     True   \n",
       "1235   {}       {}  {}                    {}         {}     True   \n",
       "1236   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...   \n",
       "1233  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...   \n",
       "1236  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1232                 {'ECK120002449': ['ECK120002449']}   \n",
       "1233  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "1234                 {'ECK120000886': ['ECK120000886']}   \n",
       "1235                 {'ECK120000233': ['ECK120000233']}   \n",
       "1236  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "\n",
       "                                                operons  \\\n",
       "1232  [{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...   \n",
       "1233  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "1234  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "1235  [{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...   \n",
       "1236  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "\n",
       "                                           operon links  \\\n",
       "1232                 {'ECK120014482': ['ECK120002449']}   \n",
       "1233  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "1234                 {'ECK120016992': ['ECK120000886']}   \n",
       "1235                 {'ECK120014515': ['ECK120000233']}   \n",
       "1236  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "\n",
       "                                             regulators  \\\n",
       "1232  [{'name': 'Fis', 'significantly associated con...   \n",
       "1233                                                 []   \n",
       "1234                                                 []   \n",
       "1235  [{'name': 'LexA', 'significantly associated co...   \n",
       "1236                                                 []   \n",
       "\n",
       "                                        regulator links temperature  \\\n",
       "1232  {'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...  37 celsius   \n",
       "1233                                                 {}  37 celsius   \n",
       "1234                                                 {}  37 celsius   \n",
       "1235                       {'LexA': ['rpsU-dnaG-rpoD']}  37 celsius   \n",
       "1236                                                 {}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "1232     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1233     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1234     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1235     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1236     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "     taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "1232      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1233      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1234      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1235      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1236      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "     sulfur-source calcium-source  \\\n",
       "1232   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1233   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1234   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1235   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1236   MgSO4(0.24)     CaCl2(0.1)   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1232  [{'name': 'Transcription', 'mutation set count...   \n",
       "1233  [{'name': 'Nucleotide transport and metabolism...   \n",
       "1234  [{'name': 'Transcription', 'mutation set count...   \n",
       "1235  [{'name': 'Replication, recombination and repa...   \n",
       "1236  [{'name': 'Nucleotide transport and metabolism...   \n",
       "\n",
       "                                              COG links  \\\n",
       "1232                {'Transcription': ['ECK120014482']}   \n",
       "1233  {'Nucleotide transport and metabolism': ['ECK1...   \n",
       "1234                {'Transcription': ['ECK120016992']}   \n",
       "1235  {'Replication, recombination and repair': ['EC...   \n",
       "1236  {'Nucleotide transport and metabolism': ['ECK1...   \n",
       "\n",
       "                                               pathways  \n",
       "1232                                                 []  \n",
       "1233  [{'name': 'Pyrimidine metabolism'}, {'name': '...  \n",
       "1234  [{'name': 'Purine metabolism'}, {'name': 'Pyri...  \n",
       "1235                                                 []  \n",
       "1236  [{'name': 'Pyrimidine metabolism'}, {'name': '...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_pathways(gene_id):\n",
    "    annots = []\n",
    "    bnum = get_gene_bnum(gene_id, gene_synonym_df)\n",
    "    df = pathway_df[pathway_df[\"refseq_locus_tag\"]==bnum]\n",
    "    for pathway in df[\"pathway_name\"]:\n",
    "        d = {\"name\": pathway}\n",
    "        annots.append(d)\n",
    "    return annots\n",
    "\n",
    "\n",
    "def get_pathways(mut_row):\n",
    "    annots = []\n",
    "    if mut_row.exp not in NON_K12_EXP_L:\n",
    "        for feat_d in mut_row[\"genomic features\"]:\n",
    "            # !!! This part only handles mutations that land within a genomic feature.\n",
    "            # Need to also implement logic for handling mutations that land within intergenic regions.\n",
    "            if feat_d[\"genetic\"]:\n",
    "                annots += _get_pathways(feat_d[\"RegulonDB ID\"])\n",
    "            # This will catch whether no features exist in the feat_ID_set\n",
    "            # or if those features are just integenic regions.\n",
    "            # Get operon genetic feature link dict through overlapping mutation and operon ranges.\n",
    "            # All genetic features are annotated within both the \"genetic features\" and \"genetic feature links\" column.\n",
    "            # If no integenic genomic feature is hit by a mutation, use the genetic feature range in an overlap analysis with operons.\n",
    "            else:\n",
    "                TU_IDs = set(TU_objects_df[TU_objects_df[\"TU_OBJECT_ID\"]==feat_d[\"RegulonDB ID\"]][\"TRANSCRIPTION_UNIT_ID\"])\n",
    "                for TU_ID in TU_IDs:\n",
    "                    subset_TU_genes_df = TU_objects_df[(TU_objects_df[\"TRANSCRIPTION_UNIT_ID\"]==TU_ID) & (TU_objects_df[\"TU_OBJECT_CLASS\"]=='GN')]\n",
    "                    for gene_id in subset_TU_genes_df[\"TU_OBJECT_ID\"].drop_duplicates():\n",
    "                        if str(gene_id) != 'nan' and gene_id != '':  # RegulonDB can return 'nan' or '' gene names\n",
    "                            annots += _get_pathways(gene_id)\n",
    "    return annots\n",
    "\n",
    "\n",
    "assoc_muts_df[\"pathways\"] = assoc_muts_df.apply(get_pathways, axis=1)\n",
    "assoc_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "TU_df = pd.read_csv(\"./data/RegulonDB10/transcription_unit.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "TU_df.columns = [\n",
    "    \"TRANSCRIPTION_UNIT_ID\",\n",
    "    \"PROMOTER_ID\",\n",
    "    \"TRANSCRIPTION_UNIT_NAME\",\n",
    "    \"OPERON_ID\",\n",
    "    \"TRANSCRIPTION_UNIT_NOTE\",\n",
    "    \"TU_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "\n",
    "\n",
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1232</td>\n",
       "      <td>4</td>\n",
       "      <td>R121C (CGC→TGC)</td>\n",
       "      <td>xylR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3735339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3735339, 3735339)</td>\n",
       "      <td>{ECK120002449}</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>{'ECK120002449': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014482': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120014482']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1233</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism'}, {'name': '...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014627'], 'D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1234</td>\n",
       "      <td>6</td>\n",
       "      <td>L770R (CTC→CGC)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187658, 4187658)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "      <td>{'Purine metabolism': ['ECK120016992'], 'Pyrim...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1235</td>\n",
       "      <td>9</td>\n",
       "      <td>G164A (GGC→GCC)</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3211597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3211597, 3211597)</td>\n",
       "      <td>{ECK120000233}</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>{'ECK120000233': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120014515': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'LexA', 'significantly associated co...</td>\n",
       "      <td>{'LexA': ['rpsU-dnaG-rpoD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Replication, recombination and repa...</td>\n",
       "      <td>{'Replication, recombination and repair': ['EC...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1236</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism'}, {'name': '...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014627'], 'D...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index          Details mutation target annotation Mutation Type  \\\n",
       "1232     4  R121C (CGC→TGC)                       xylR           SNP   \n",
       "1233     5             None                        rph           DEL   \n",
       "1234     6  L770R (CTC→CGC)                       rpoC           SNP   \n",
       "1235     9  G164A (GGC→GCC)                       dnaG           SNP   \n",
       "1236     5             None                        rph           DEL   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale      exp  flask  isolate  \\\n",
       "1232   3735339           NaN             C→T    1  SSW_XYL    116        1   \n",
       "1233   3815859           NaN          Δ82 bp    1  SSW_XYL    116        1   \n",
       "1234   4187658           NaN             T→G    1  SSW_XYL    116        1   \n",
       "1235   3211597           NaN             G→C    2  SSW_XYL    109        1   \n",
       "1236   3815859           NaN          Δ82 bp    2  SSW_XYL    109        1   \n",
       "\n",
       "      presence  tech_rep  coding               range gene RegulonDB ID  \\\n",
       "1232       1.0         1    True  (3735339, 3735339)    {ECK120002449}   \n",
       "1233       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "1234       1.0         1    True  (4187658, 4187658)    {ECK120000886}   \n",
       "1235       1.0         1    True  (3211597, 3211597)    {ECK120000233}   \n",
       "1236       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...  False       False   \n",
       "1233  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1236  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "\n",
       "     TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "1232   {}       {}  {}                    {}         {}     True   \n",
       "1233   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "1234   {}       {}  {}                    {}         {}     True   \n",
       "1235   {}       {}  {}                    {}         {}     True   \n",
       "1236   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...   \n",
       "1233  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...   \n",
       "1236  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1232                 {'ECK120002449': ['ECK120002449']}   \n",
       "1233  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "1234                 {'ECK120000886': ['ECK120000886']}   \n",
       "1235                 {'ECK120000233': ['ECK120000233']}   \n",
       "1236  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "\n",
       "                                                operons  \\\n",
       "1232  [{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...   \n",
       "1233  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "1234  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "1235  [{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...   \n",
       "1236  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "\n",
       "                                           operon links  \\\n",
       "1232                 {'ECK120014482': ['ECK120002449']}   \n",
       "1233  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "1234                 {'ECK120016992': ['ECK120000886']}   \n",
       "1235                 {'ECK120014515': ['ECK120000233']}   \n",
       "1236  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "\n",
       "                                             regulators  \\\n",
       "1232  [{'name': 'Fis', 'significantly associated con...   \n",
       "1233                                                 []   \n",
       "1234                                                 []   \n",
       "1235  [{'name': 'LexA', 'significantly associated co...   \n",
       "1236                                                 []   \n",
       "\n",
       "                                        regulator links temperature  \\\n",
       "1232  {'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...  37 celsius   \n",
       "1233                                                 {}  37 celsius   \n",
       "1234                                                 {}  37 celsius   \n",
       "1235                       {'LexA': ['rpsU-dnaG-rpoD']}  37 celsius   \n",
       "1236                                                 {}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "1232     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1233     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1234     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1235     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1236     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "     taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "1232      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1233      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1234      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1235      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1236      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "     sulfur-source calcium-source  \\\n",
       "1232   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1233   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1234   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1235   MgSO4(0.24)     CaCl2(0.1)   \n",
       "1236   MgSO4(0.24)     CaCl2(0.1)   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1232  [{'name': 'Transcription', 'mutation set count...   \n",
       "1233  [{'name': 'Nucleotide transport and metabolism...   \n",
       "1234  [{'name': 'Transcription', 'mutation set count...   \n",
       "1235  [{'name': 'Replication, recombination and repa...   \n",
       "1236  [{'name': 'Nucleotide transport and metabolism...   \n",
       "\n",
       "                                              COG links  \\\n",
       "1232                {'Transcription': ['ECK120014482']}   \n",
       "1233  {'Nucleotide transport and metabolism': ['ECK1...   \n",
       "1234                {'Transcription': ['ECK120016992']}   \n",
       "1235  {'Replication, recombination and repair': ['EC...   \n",
       "1236  {'Nucleotide transport and metabolism': ['ECK1...   \n",
       "\n",
       "                                               pathways  \\\n",
       "1232                                                 []   \n",
       "1233  [{'name': 'Pyrimidine metabolism'}, {'name': '...   \n",
       "1234  [{'name': 'Purine metabolism'}, {'name': 'Pyri...   \n",
       "1235                                                 []   \n",
       "1236  [{'name': 'Pyrimidine metabolism'}, {'name': '...   \n",
       "\n",
       "                                          pathway links  \n",
       "1232                                                 {}  \n",
       "1233  {'Pyrimidine metabolism': ['ECK120014627'], 'D...  \n",
       "1234  {'Purine metabolism': ['ECK120016992'], 'Pyrim...  \n",
       "1235                                                 {}  \n",
       "1236  {'Pyrimidine metabolism': ['ECK120014627'], 'D...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _add_links_to_d(bnum, d, op_ID_set):\n",
    "    for pathway in pathway_df[pathway_df[\"refseq_locus_tag\"]==bnum][\"pathway_name\"]:\n",
    "        if pathway not in d.keys(): d[pathway] = []\n",
    "        d[pathway] += list(op_ID_set)\n",
    "\n",
    "\n",
    "def get_pathway_operon_links_d(mut_row):\n",
    "    pathway_operon_d = dict()\n",
    "    if mut_row.exp not in NON_K12_EXP_L:\n",
    "        for feat_d in mut_row[\"genomic features\"]:\n",
    "            if feat_d[\"genetic\"]:\n",
    "                op_ID_set = get_operon_ID_set(feat_d[\"RegulonDB ID\"], TU_objects_df, TU_df, operon_df)\n",
    "                if str(feat_d[\"RegulonDB ID\"]) != 'nan' and feat_d[\"RegulonDB ID\"] != '':\n",
    "                    bnum = get_gene_bnum(feat_d[\"RegulonDB ID\"], gene_synonym_df)\n",
    "                    _add_links_to_d(bnum, pathway_operon_d, op_ID_set)\n",
    "            else:\n",
    "                TU_IDs = set(TU_objects_df[TU_objects_df[\"TU_OBJECT_ID\"]==feat_d[\"RegulonDB ID\"]][\"TRANSCRIPTION_UNIT_ID\"])\n",
    "                for TU_ID in TU_IDs:\n",
    "                    subset_TU_genes_df = TU_objects_df[(TU_objects_df[\"TRANSCRIPTION_UNIT_ID\"]==TU_ID) & (TU_objects_df[\"TU_OBJECT_CLASS\"]=='GN')]\n",
    "\n",
    "                    op_ID_set = set()\n",
    "                    for _, TU_gene in subset_TU_genes_df.iterrows():\n",
    "                         op_ID_set = op_ID_set | get_operon_ID_set(TU_gene[\"TU_OBJECT_ID\"], TU_objects_df, TU_df, operon_df)\n",
    "\n",
    "                    for gene_id in subset_TU_genes_df[\"TU_OBJECT_ID\"].drop_duplicates():\n",
    "                        if str(gene_id) != 'nan' and gene_id != '':  # RegulonDB can return 'nan' or '' gene names\n",
    "                            bnum = get_gene_bnum(gene_id, gene_synonym_df)\n",
    "                            _add_links_to_d(bnum, pathway_operon_d, op_ID_set)\n",
    "    return pathway_operon_d\n",
    "\n",
    "                    \n",
    "assoc_muts_df[\"pathway links\"] = assoc_muts_df.apply(get_pathway_operon_links_d, axis=1)\n",
    "assoc_muts_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Starch and sucrose metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycerolipid metabolism</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 observed mutation count\n",
       "Pyrimidine metabolism                                  9\n",
       "Drug metabolism - other enzymes                        4\n",
       "Purine metabolism                                      4\n",
       "Starch and sucrose metabolism                          1\n",
       "Glycerolipid metabolism                                1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a bit behind annotations currently used in AVA MS.\n",
    "# Still uses same approach to counted, just not all data-types have been consolidated into the regulators feat col\n",
    "\n",
    "\n",
    "comp_mut_df = pd.DataFrame(columns=[\"observed mutation count\"])\n",
    "for i, r in assoc_muts_df.iterrows():\n",
    "    for f, links in r[\"pathway links\"].items():\n",
    "        if f not in comp_mut_df.index:\n",
    "            df = pd.DataFrame({\"observed mutation count\": len(links)}, index=[f])\n",
    "            comp_mut_df = comp_mut_df.append(df)\n",
    "        else:\n",
    "            comp_mut_df.loc[f, \"observed mutation count\"] += len(links)\n",
    "\n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in assoc_muts_df.iterrows():\n",
    "    for feat_d in r[\"pathways\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"name\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pathway_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>54804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>36129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lysine biosynthesis</td>\n",
       "      <td>18189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Vitamin B6 metabolism</td>\n",
       "      <td>10138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>51305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          pathway length\n",
       "pathway_name                                            \n",
       "Glycine, serine and threonine metabolism           54804\n",
       "Cysteine and methionine metabolism                 36129\n",
       "Lysine biosynthesis                                18189\n",
       "Vitamin B6 metabolism                              10138\n",
       "Pentose phosphate pathway                          51305"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_len_df = pd.read_pickle(\"./data/pathway_len_df.pkl\")\n",
    "pathway_len_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>9</td>\n",
       "      <td>89371</td>\n",
       "      <td>0.269703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>4</td>\n",
       "      <td>7286</td>\n",
       "      <td>0.021988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>4</td>\n",
       "      <td>154034</td>\n",
       "      <td>0.464843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Starch and sucrose metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>63933</td>\n",
       "      <td>0.192937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycerolipid metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>16744</td>\n",
       "      <td>0.050530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 observed mutation count  length  proportion\n",
       "Pyrimidine metabolism                                  9   89371    0.269703\n",
       "Drug metabolism - other enzymes                        4    7286    0.021988\n",
       "Purine metabolism                                      4  154034    0.464843\n",
       "Starch and sucrose metabolism                          1   63933    0.192937\n",
       "Glycerolipid metabolism                                1   16744    0.050530"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "comp_mut_df[\"length\"] = comp_mut_df.apply(lambda r: pathway_len_df.loc[r.index, \"pathway length\"])\n",
    "total_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>9</td>\n",
       "      <td>89371</td>\n",
       "      <td>0.269703</td>\n",
       "      <td>451</td>\n",
       "      <td>0.0451</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>4</td>\n",
       "      <td>7286</td>\n",
       "      <td>0.021988</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>4</td>\n",
       "      <td>154034</td>\n",
       "      <td>0.464843</td>\n",
       "      <td>9951</td>\n",
       "      <td>0.9951</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Starch and sucrose metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>63933</td>\n",
       "      <td>0.192937</td>\n",
       "      <td>9839</td>\n",
       "      <td>0.9839</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycerolipid metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>16744</td>\n",
       "      <td>0.050530</td>\n",
       "      <td>6108</td>\n",
       "      <td>0.6108</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 observed mutation count  length  proportion  \\\n",
       "Pyrimidine metabolism                                  9   89371    0.269703   \n",
       "Drug metabolism - other enzymes                        4    7286    0.021988   \n",
       "Purine metabolism                                      4  154034    0.464843   \n",
       "Starch and sucrose metabolism                          1   63933    0.192937   \n",
       "Glycerolipid metabolism                                1   16744    0.050530   \n",
       "\n",
       "                                 equal or larger count  p value  significant  \n",
       "Pyrimidine metabolism                              451   0.0451         True  \n",
       "Drug metabolism - other enzymes                      8   0.0008         True  \n",
       "Purine metabolism                                 9951   0.9951        False  \n",
       "Starch and sucrose metabolism                     9839   0.9839        False  \n",
       "Glycerolipid metabolism                           6108   0.6108        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=int(num_muts),\n",
    "        p=proportion_l,\n",
    "        replace=True,\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>9</td>\n",
       "      <td>89371</td>\n",
       "      <td>0.269703</td>\n",
       "      <td>451</td>\n",
       "      <td>0.0451</td>\n",
       "      <td>True</td>\n",
       "      <td>0.2255</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>4</td>\n",
       "      <td>7286</td>\n",
       "      <td>0.021988</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0040</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>4</td>\n",
       "      <td>154034</td>\n",
       "      <td>0.464843</td>\n",
       "      <td>9951</td>\n",
       "      <td>0.9951</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Starch and sucrose metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>63933</td>\n",
       "      <td>0.192937</td>\n",
       "      <td>9839</td>\n",
       "      <td>0.9839</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Glycerolipid metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td>16744</td>\n",
       "      <td>0.050530</td>\n",
       "      <td>6108</td>\n",
       "      <td>0.6108</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 observed mutation count  length  proportion  \\\n",
       "Pyrimidine metabolism                                  9   89371    0.269703   \n",
       "Drug metabolism - other enzymes                        4    7286    0.021988   \n",
       "Purine metabolism                                      4  154034    0.464843   \n",
       "Starch and sucrose metabolism                          1   63933    0.192937   \n",
       "Glycerolipid metabolism                                1   16744    0.050530   \n",
       "\n",
       "                                 equal or larger count  p value  significant  \\\n",
       "Pyrimidine metabolism                              451   0.0451         True   \n",
       "Drug metabolism - other enzymes                      8   0.0008         True   \n",
       "Purine metabolism                                 9951   0.9951        False   \n",
       "Starch and sucrose metabolism                     9839   0.9839        False   \n",
       "Glycerolipid metabolism                           6108   0.6108        False   \n",
       "\n",
       "                                 corrected p value  corrected significance  \n",
       "Pyrimidine metabolism                       0.2255                   False  \n",
       "Drug metabolism - other enzymes             0.0040                    True  \n",
       "Purine metabolism                           1.0000                   False  \n",
       "Starch and sucrose metabolism               1.0000                   False  \n",
       "Glycerolipid metabolism                     1.0000                   False  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "# comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, r in assoc_muts_df.iterrows():\n",
    "    for d in r[\"pathways\"]:\n",
    "        if d[\"name\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"name\"]][\"corrected significance\"]\n",
    "#             d[\"significant\"] = comp_mut_df.loc[d[\"name\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_muts_df.to_pickle(\"./data/12_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     link_d = r[\"pathway links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"pathways\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += len(hits)\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.2, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 mutated pathways\")\n",
    "# ax.get_legend().remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_name(name, json):\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"name\"] == name:\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return is_signif\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     link_d = r[\"pathway links\"]\n",
    "#     for targ_name, hits in link_d.items():\n",
    "#         is_signif = _get_name_signif_from_name(targ_name, r[\"pathways\"])\n",
    "#         t = str(targ_name)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#             if t not in targ_cnt_d.keys():\n",
    "#                 targ_cnt_d[t] = 0\n",
    "#             targ_cnt_d[t] += len(hits)\n",
    "            \n",
    "# if len(targ_cnt_d) > 0:\n",
    "#     df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "#     df = df.sort_values(\"mutated feature count\")\n",
    "#     df = df[-20:]\n",
    "#     display(df)\n",
    "#     ax = df.plot.barh(y='mutated feature count',\n",
    "#     #                   figsize=(10,25),\n",
    "#                       figsize=(5,10),\n",
    "#                       color=\"#4C72B0\",\n",
    "#                       width=1\n",
    "#                      )\n",
    "#     for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#         ax.text(v, i - 0.2, str(v))\n",
    "#     sns.despine(ax=ax, top=True, right=True)\n",
    "#     ax.set_title(\"Significant mutated pathways\")\n",
    "#     ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
