{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.alemutdf import get_all_sample_mut_df, get_gene_mut_count_mat, get_multi_exp_max_freq_mut_df, get_mut_type_avg_frac_across_class_df\n",
    "from util.mut import is_coding_mut, get_original_nuc_mut_range\n",
    "from util.metadata import get_condition_val_dict, get_condition_field_val_set\n",
    "from util.genome import get_feature_hit_set, is_overlap, get_promoter_range_from_RegulonDB_df_row, NON_K12_EXP_L "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1278, 13)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/1_df.pkl\")\n",
    "display(all_muts_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Renaming the \"Gene\" column to \"mutation target annotation\" because this is more accurate.\n",
    "all_muts_df = all_muts_df.rename(index=str, columns={\"Gene\": \"mutation target annotation\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build genetic feature annotation\n",
    "Need to build genetic feature annotations before genomic features annotations since the genetic annotations are used to fill gaps with unknown genomic targets."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## K12 specific genetic annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>GENE_POSLEFT</th>\n",
       "      <th>GENE_POSRIGHT</th>\n",
       "      <th>GENE_STRAND</th>\n",
       "      <th>GENE_SEQUENCE</th>\n",
       "      <th>GC_CONTENT</th>\n",
       "      <th>CRI_SCORE</th>\n",
       "      <th>GENE_NOTE</th>\n",
       "      <th>GENE_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>GENE_TYPE</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr</td>\n",
       "      <td>4265782.0</td>\n",
       "      <td>4266861.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...</td>\n",
       "      <td>55.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4265782, 4266861)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>modB</td>\n",
       "      <td>795862.0</td>\n",
       "      <td>796551.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...</td>\n",
       "      <td>54.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(795862, 796551)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ECK120000003</td>\n",
       "      <td>cysZ</td>\n",
       "      <td>2531463.0</td>\n",
       "      <td>2532224.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...</td>\n",
       "      <td>50.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2531463, 2532224)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ECK120000004</td>\n",
       "      <td>dfp</td>\n",
       "      <td>3812731.0</td>\n",
       "      <td>3813951.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...</td>\n",
       "      <td>53.64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3812731, 3813951)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ECK120000005</td>\n",
       "      <td>dcuB</td>\n",
       "      <td>4347404.0</td>\n",
       "      <td>4348744.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...</td>\n",
       "      <td>52.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4347404, 4348744)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GENE_ID GENE_NAME  GENE_POSLEFT  GENE_POSRIGHT GENE_STRAND  \\\n",
       "0  ECK120000001       alr     4265782.0      4266861.0     forward   \n",
       "1  ECK120000002      modB      795862.0       796551.0     forward   \n",
       "2  ECK120000003      cysZ     2531463.0      2532224.0     forward   \n",
       "3  ECK120000004       dfp     3812731.0      3813951.0     forward   \n",
       "4  ECK120000005      dcuB     4347404.0      4348744.0     reverse   \n",
       "\n",
       "                                       GENE_SEQUENCE  GC_CONTENT  CRI_SCORE  \\\n",
       "0  ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...       55.93        NaN   \n",
       "1  ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...       54.06        NaN   \n",
       "2  ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...       50.13        NaN   \n",
       "3  ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...       53.64        NaN   \n",
       "4  ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...       52.27        NaN   \n",
       "\n",
       "  GENE_NOTE  GENE_INTERNAL_COMMENT KEY_ID_ORG GENE_TYPE               range  \n",
       "0       NaN                    NaN      ECK12       NaN  (4265782, 4266861)  \n",
       "1       NaN                    NaN      ECK12       NaN    (795862, 796551)  \n",
       "2       NaN                    NaN      ECK12       NaN  (2531463, 2532224)  \n",
       "3       NaN                    NaN      ECK12       NaN  (3812731, 3813951)  \n",
       "4       NaN                    NaN      ECK12       NaN  (4347404, 4348744)  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = pd.read_csv(\n",
    "    \"./data/RegulonDB10/gene.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "gene_df.columns = [\n",
    "    \"GENE_ID\",\n",
    "    \"GENE_NAME\",\n",
    "    \"GENE_POSLEFT\",\n",
    "    \"GENE_POSRIGHT\",\n",
    "    \"GENE_STRAND\",\n",
    "    \"GENE_SEQUENCE\",\n",
    "    \"GC_CONTENT\",\n",
    "    \"CRI_SCORE\",\n",
    "    \"GENE_NOTE\",\n",
    "    \"GENE_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\",\n",
    "    \"GENE_TYPE\"\n",
    "]\n",
    "\n",
    "def get_gene_range(row):\n",
    "    r = ()\n",
    "    if not pd.isna(row[\"GENE_POSLEFT\"]) and not pd.isna(row[\"GENE_POSRIGHT\"]):\n",
    "        r = (int(row[\"GENE_POSLEFT\"]), int(row[\"GENE_POSRIGHT\"])) \n",
    "    return r\n",
    "\n",
    "gene_df[\"range\"] = gene_df.apply(lambda r: get_gene_range(r), axis=1)\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1308318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187550, 4187550)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4400313</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4400313, 4400313)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>DEL</td>\n",
       "      <td>702352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Details mutation target annotation Mutation Type  \\\n",
       "2           R110G (CGT→GGT)                       clsA           SNP   \n",
       "6                                                  rph           DEL   \n",
       "7           A734V (GCG→GTG)                       rpoC           SNP   \n",
       "8             D9A (GAT→GCT)                        hfq           SNP   \n",
       "0  coding (380‑400/1149 nt)                       nagA           DEL   \n",
       "\n",
       "   Position Reference Seq Sequence Change  ale  exp  flask  isolate  presence  \\\n",
       "2   1308318           NaN             G→C    1  42C    124        1       1.0   \n",
       "6   3815859           NaN          Δ82 bp    1  42C    124        1       1.0   \n",
       "7   4187550           NaN             C→T    1  42C    124        1       1.0   \n",
       "8   4400313           NaN             A→C    1  42C    124        1       1.0   \n",
       "0    702352           NaN          Δ21 bp    1  42C    124        1       1.0   \n",
       "\n",
       "   tech_rep  coding               range  \n",
       "2         1    True  (1308318, 1308318)  \n",
       "6         1    True  (3815859, 3815940)  \n",
       "7         1    True  (4187550, 4187550)  \n",
       "8         1    True  (4400313, 4400313)  \n",
       "0         1    True    (702352, 702372)  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_muts_df[\"range\"] = all_muts_df.apply(get_original_nuc_mut_range, axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1308318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187550, 4187550)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4400313</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4400313, 4400313)</td>\n",
       "      <td>{ECK120000431}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>DEL</td>\n",
       "      <td>702352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Details mutation target annotation Mutation Type  \\\n",
       "2           R110G (CGT→GGT)                       clsA           SNP   \n",
       "6                                                  rph           DEL   \n",
       "7           A734V (GCG→GTG)                       rpoC           SNP   \n",
       "8             D9A (GAT→GCT)                        hfq           SNP   \n",
       "0  coding (380‑400/1149 nt)                       nagA           DEL   \n",
       "\n",
       "   Position Reference Seq Sequence Change  ale  exp  flask  isolate  presence  \\\n",
       "2   1308318           NaN             G→C    1  42C    124        1       1.0   \n",
       "6   3815859           NaN          Δ82 bp    1  42C    124        1       1.0   \n",
       "7   4187550           NaN             C→T    1  42C    124        1       1.0   \n",
       "8   4400313           NaN             A→C    1  42C    124        1       1.0   \n",
       "0    702352           NaN          Δ21 bp    1  42C    124        1       1.0   \n",
       "\n",
       "   tech_rep  coding               range gene RegulonDB ID  \n",
       "2         1    True  (1308318, 1308318)    {ECK120001556}  \n",
       "6         1    True  (3815859, 3815940)    {ECK120000854}  \n",
       "7         1    True  (4187550, 4187550)    {ECK120000886}  \n",
       "8         1    True  (4400313, 4400313)    {ECK120000431}  \n",
       "0         1    True    (702352, 702372)    {ECK120000625}  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_muts_df[\"gene RegulonDB ID\"] = all_muts_df.apply(\n",
    "    lambda r: get_feature_hit_set(\n",
    "        r[\"range\"], gene_df, \"range\", \"GENE_ID\") if r.exp not in NON_K12_EXP_L else set(), axis=1\n",
    ")  # Maybe instead just have a list of K12 experiments\n",
    "\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>RegulonDB ID</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>thrL/thrA</td>\n",
       "      <td>ECK120001251/ECK120000987</td>\n",
       "      <td>(256, 336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>thrA/thrB</td>\n",
       "      <td>ECK120000987/ECK120000988</td>\n",
       "      <td>(2800, 2800)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>thrC/yaaX</td>\n",
       "      <td>ECK120000989/ECK120002701</td>\n",
       "      <td>(5021, 5233)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>yaaX/yaaA</td>\n",
       "      <td>ECK120002701/ECK120000009</td>\n",
       "      <td>(5531, 5682)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>yaaA/yaaJ</td>\n",
       "      <td>ECK120000009/ECK120001508</td>\n",
       "      <td>(6460, 6528)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        name               RegulonDB ID         range\n",
       "0  thrL/thrA  ECK120001251/ECK120000987    (256, 336)\n",
       "1  thrA/thrB  ECK120000987/ECK120000988  (2800, 2800)\n",
       "2  thrC/yaaX  ECK120000989/ECK120002701  (5021, 5233)\n",
       "3  yaaX/yaaA  ECK120002701/ECK120000009  (5531, 5682)\n",
       "4  yaaA/yaaJ  ECK120000009/ECK120001508  (6460, 6528)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "temp_gene_df = gene_df.copy()\n",
    "temp_gene_df = temp_gene_df.sort_values(by=[\"GENE_POSLEFT\"])  # Getting the in order according to position\n",
    "temp_gene_df = temp_gene_df.reset_index()\n",
    "\n",
    "intergenic_df = pd.DataFrame(columns=[\"name\", \"RegulonDB ID\", \"range\"])\n",
    "# Don't want to iterate to the last one since the gene will be considered in i-1\n",
    "for i in range(0, len(temp_gene_df) - 1):\n",
    "    # check if there even exists an intergenic region between genes.\n",
    "    if (temp_gene_df.loc[i + 1][\"GENE_POSLEFT\"] - temp_gene_df.loc[i][\"GENE_POSRIGHT\"]) > 1:\n",
    "        intergenic_df = intergenic_df.append(\n",
    "            {\"name\": str(temp_gene_df.loc[i][\"GENE_NAME\"]) + '/' + str(temp_gene_df.loc[i + 1][\"GENE_NAME\"]),\n",
    "            \"RegulonDB ID\": temp_gene_df.loc[i][\"GENE_ID\"] + '/' + temp_gene_df.loc[i + 1][\"GENE_ID\"],\n",
    "            \"range\": (int(temp_gene_df.loc[i][\"GENE_POSRIGHT\"] + 1), int(temp_gene_df.loc[i + 1][\"GENE_POSLEFT\"] - 1))},\n",
    "            ignore_index=True\n",
    "    )\n",
    "display(intergenic_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1308318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187550, 4187550)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4400313</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4400313, 4400313)</td>\n",
       "      <td>{ECK120000431}</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>DEL</td>\n",
       "      <td>702352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Details mutation target annotation Mutation Type  \\\n",
       "2           R110G (CGT→GGT)                       clsA           SNP   \n",
       "6                                                  rph           DEL   \n",
       "7           A734V (GCG→GTG)                       rpoC           SNP   \n",
       "8             D9A (GAT→GCT)                        hfq           SNP   \n",
       "0  coding (380‑400/1149 nt)                       nagA           DEL   \n",
       "\n",
       "   Position Reference Seq Sequence Change  ale  exp  flask  isolate  presence  \\\n",
       "2   1308318           NaN             G→C    1  42C    124        1       1.0   \n",
       "6   3815859           NaN          Δ82 bp    1  42C    124        1       1.0   \n",
       "7   4187550           NaN             C→T    1  42C    124        1       1.0   \n",
       "8   4400313           NaN             A→C    1  42C    124        1       1.0   \n",
       "0    702352           NaN          Δ21 bp    1  42C    124        1       1.0   \n",
       "\n",
       "   tech_rep  coding               range gene RegulonDB ID  \\\n",
       "2         1    True  (1308318, 1308318)    {ECK120001556}   \n",
       "6         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "7         1    True  (4187550, 4187550)    {ECK120000886}   \n",
       "8         1    True  (4400313, 4400313)    {ECK120000431}   \n",
       "0         1    True    (702352, 702372)    {ECK120000625}   \n",
       "\n",
       "                                    genetic features  \n",
       "2  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  \n",
       "6  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  \n",
       "7  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  \n",
       "8  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...  \n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def get_feature_range(feat_id, feat_type):\n",
    "    r = ()\n",
    "    feat_id_col = 0\n",
    "    range_col = \"range\"\n",
    "    feat_id_col = \"GENE_ID\"\n",
    "    feat_type_df = gene_df\n",
    "\n",
    "    df = feat_type_df[feat_type_df[feat_id_col] == feat_id]\n",
    "    if not df.empty:\n",
    "        r = df.iloc[0][range_col]  # Assuming all rows describe the same range.\n",
    "    l = [int(i) for i in r]\n",
    "    return tuple(l)\n",
    "\n",
    "\n",
    "# This will include all implicit intergenic regions, where get_genomic_features(...) does not.\n",
    "# TODO: this may not handle wrapping around the genome from the last gene to the first gene. Write test for this.\n",
    "def get_ordered_genetic_targets(genetic_targets):\n",
    "    ordered_genetic_targets = sorted(\n",
    "        genetic_targets, key=lambda k: k[\"range\"][0])\n",
    "    return ordered_genetic_targets\n",
    "\n",
    "\n",
    "test_genetic_targets = [\n",
    "    {'RegulonDB ID': \"ECK120000003\",\n",
    "     \"range\": (2531463, 2532224)},\n",
    "    {'RegulonDB ID': \"ECK120000001\",\n",
    "     \"range\": (4265782, 4266861)},\n",
    "    {'RegulonDB ID': \"ECK120000002\",\n",
    "     \"range\": (795862, 796551)}]\n",
    "assert(get_ordered_genetic_targets(test_genetic_targets) ==\n",
    "       [{'RegulonDB ID': 'ECK120000002', 'range': (795862, 796551)},\n",
    "        {'RegulonDB ID': 'ECK120000003', 'range': (2531463, 2532224)},\n",
    "        {'RegulonDB ID': 'ECK120000001', 'range': (4265782, 4266861)}])\n",
    "\n",
    "\n",
    "#Simply returning the first gene name in the second column that has the RegulonDB ID in the first column.\n",
    "def get_gene_name(RegulonDB_ID, debug=False):\n",
    "    if debug:\n",
    "        display(RegulonDB_ID, gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID])\n",
    "    return gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID].iloc[0][\"GENE_NAME\"]\n",
    "\n",
    "\n",
    "def get_genetic_feat_JSON(mut_row):\n",
    "    l = []\n",
    "    for gene_id in mut_row[\"gene RegulonDB ID\"]:\n",
    "        l.append({\n",
    "            # IDs are unique; assuming only be one entry per DF.\n",
    "            \"name\": get_gene_name(gene_id),\n",
    "            \"RegulonDB ID\": gene_id,\n",
    "            \"range\": get_feature_range(gene_id, \"gene RegulonDB ID\"),\n",
    "            \"feature type\": \"gene\"\n",
    "        })\n",
    "\n",
    "    intergenic_regions = get_feature_hit_set(mut_row[\"range\"], intergenic_df, \"range\", \"RegulonDB ID\")\n",
    "    for intergenic_region in intergenic_regions:\n",
    "        l.append({\n",
    "            \"RegulonDB ID\": intergenic_region,\n",
    "            \"name\": str(get_gene_name(intergenic_region.split('/')[0])) + '/' + str(get_gene_name(intergenic_region.split('/')[1])),\n",
    "            \"range\": intergenic_df[intergenic_df[\"RegulonDB ID\"] == intergenic_region].iloc[0][\"range\"],\n",
    "            \"feature type\": \"unknown\"\n",
    "        })\n",
    "\n",
    "    return l\n",
    "\n",
    "\n",
    "all_muts_df[\"genetic features\"] = all_muts_df.apply(lambda r: get_genetic_feat_JSON(r) if r.exp not in NON_K12_EXP_L else [], axis=1)\n",
    "display(all_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## E. coli B (REL606) genetic annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OBJECT_ID</th>\n",
       "      <th>OBJECT_SYNONYM_NAME</th>\n",
       "      <th>OS_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>EG10001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>ECK4045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>b4053</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>b0764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      OBJECT_ID OBJECT_SYNONYM_NAME  OS_INTERNAL_COMMENT KEY_ID_ORG\n",
       "0  ECK120000001             EG10001                  NaN      ECK12\n",
       "1  ECK120000001             ECK4045                  NaN      ECK12\n",
       "2  ECK120000001               b4053                  NaN      ECK12\n",
       "3  ECK120000001                alr5                  NaN      ECK12\n",
       "4  ECK120000002               b0764                  NaN      ECK12"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_synonym_df = pd.read_csv(\n",
    "    \"./data/RegulonDB10/object_synonym.txt\",\n",
    "    sep=\"\\t\",\n",
    "    comment='#',\n",
    "    header=None,\n",
    "    quoting=3\n",
    ")\n",
    "gene_synonym_df.columns = [\"OBJECT_ID\", \"OBJECT_SYNONYM_NAME\", \"OS_INTERNAL_COMMENT\", \"KEY_ID_ORG\"]\n",
    "gene_synonym_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1308318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187550, 4187550)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4400313</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4400313, 4400313)</td>\n",
       "      <td>{ECK120000431}</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>DEL</td>\n",
       "      <td>702352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>42C</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Details mutation target annotation Mutation Type  \\\n",
       "2           R110G (CGT→GGT)                       clsA           SNP   \n",
       "6                                                  rph           DEL   \n",
       "7           A734V (GCG→GTG)                       rpoC           SNP   \n",
       "8             D9A (GAT→GCT)                        hfq           SNP   \n",
       "0  coding (380‑400/1149 nt)                       nagA           DEL   \n",
       "\n",
       "   Position Reference Seq Sequence Change  ale  exp  flask  isolate  presence  \\\n",
       "2   1308318           NaN             G→C    1  42C    124        1       1.0   \n",
       "6   3815859           NaN          Δ82 bp    1  42C    124        1       1.0   \n",
       "7   4187550           NaN             C→T    1  42C    124        1       1.0   \n",
       "8   4400313           NaN             A→C    1  42C    124        1       1.0   \n",
       "0    702352           NaN          Δ21 bp    1  42C    124        1       1.0   \n",
       "\n",
       "   tech_rep  coding               range gene RegulonDB ID  \\\n",
       "2         1    True  (1308318, 1308318)    {ECK120001556}   \n",
       "6         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "7         1    True  (4187550, 4187550)    {ECK120000886}   \n",
       "8         1    True  (4400313, 4400313)    {ECK120000431}   \n",
       "0         1    True    (702352, 702372)    {ECK120000625}   \n",
       "\n",
       "                                    genetic features  \n",
       "2  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  \n",
       "6  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  \n",
       "7  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  \n",
       "8  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...  \n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO: replace into util.mut\n",
    "def get_clean_mut_gene_list(gene_list_str):\n",
    "    gene_list_str = gene_list_str.replace(\" > \", \" \")\n",
    "    gene_list_str = gene_list_str.replace(\"]\", \"\")\n",
    "    gene_list_str = gene_list_str.replace(\"[\", \"\")\n",
    "    gene_list_str = gene_list_str.replace(\"<i><b>\", '')\n",
    "    gene_list_str = gene_list_str.replace(\"</b><BR>\", '')\n",
    "    gene_list_str = gene_list_str.replace(\"</i>\", '')\n",
    "    if \"genes\" in gene_list_str:\n",
    "        start_idx = gene_list_str.rfind(\"genes\")+len(\"genes\")\n",
    "        gene_list_str = gene_list_str[start_idx:]\n",
    "    mut_gene_list = gene_list_str.split(\", \")\n",
    "    clean_mut_gene_list = [gene for gene in mut_gene_list]\n",
    "    return clean_mut_gene_list\n",
    "\n",
    "\n",
    "test_s = '<i><b>3 genes</b><BR>[manB], manC, [cpsG]</i>'\n",
    "assert(get_clean_mut_gene_list(test_s) == [\"manB\", \"manC\", \"cpsG\"])\n",
    "\n",
    "\n",
    "# TODO: put this within util.gene\n",
    "def get_gene_RegDB_ID(gene_name):\n",
    "    regDB_ID = \"\"\n",
    "    df = gene_df[gene_df[\"GENE_NAME\"] == gene_name]\n",
    "    id_col_name = \"GENE_ID\"\n",
    "    if len(df) == 0:  # Gene name used might be a synonym\n",
    "        df = gene_synonym_df[gene_synonym_df[\"OBJECT_SYNONYM_NAME\"] == gene_name]\n",
    "        id_col_name = \"OBJECT_ID\"\n",
    "    if len(df) > 0:\n",
    "        regDB_ID = df.iloc[0][id_col_name]\n",
    "    return regDB_ID\n",
    "\n",
    "\n",
    "def get_genetic_feat_JSON_LTEE(mut_row):\n",
    "    l = []\n",
    "    mut_targ_annot = mut_row[\"mutation target annotation\"]\n",
    "\n",
    "    # Since trying to match genes to RegDB IDs, if genetic target is an \"ECB\" gene, will automatically be ignored.\n",
    "    if '/' in mut_targ_annot and \",\" not in mut_targ_annot:\n",
    "        regDB_ID_l = []\n",
    "        interg_l = mut_targ_annot.split('/')\n",
    "        for g in interg_l:\n",
    "            regDB_ID = get_gene_RegDB_ID(g)\n",
    "            if regDB_ID != \"\":\n",
    "                regDB_ID_l.append(regDB_ID)\n",
    "        if len(regDB_ID_l) == 2:  # only 2 gene annotations ever involved in intergenic region annotations.\n",
    "            if len(gene_df[gene_df[\"GENE_ID\"] == regDB_ID_l[0]]) and len(gene_df[gene_df[\"GENE_ID\"] == regDB_ID_l[1]]):\n",
    "                l.append({\n",
    "                    \"RegulonDB ID\": regDB_ID_l[0] + '/' + regDB_ID_l[1],\n",
    "                    \"name\": get_gene_name(regDB_ID_l[0]) + '/' + get_gene_name(regDB_ID_l[1]),\n",
    "                    \"feature type\": \"unknown\"\n",
    "                })\n",
    "    # TODO: for refactoring, could combine this condition and the final condition into one\n",
    "    elif ',' in mut_targ_annot:  # list of genes\n",
    "        g_l = get_clean_mut_gene_list(mut_targ_annot)\n",
    "        regDB_ID_s = set()  # making this a set so that no genes are duplicated.\n",
    "        for g in g_l:\n",
    "            regDB_ID = get_gene_RegDB_ID(g)\n",
    "            if regDB_ID != \"\":\n",
    "                regDB_ID_s.add(regDB_ID)\n",
    "        if len(regDB_ID_s) > 0:\n",
    "            for regDB_ID in regDB_ID_s:\n",
    "                if len(gene_df[gene_df[\"GENE_ID\"] == regDB_ID]) > 0:\n",
    "                    l.append({\n",
    "                        \"RegulonDB ID\": regDB_ID,\n",
    "                        \"name\": get_gene_name(regDB_ID),\n",
    "                        \"feature type\": \"gene\"\n",
    "                    })\n",
    "    else:  # just a single gene\n",
    "        regDB_ID = get_gene_RegDB_ID(mut_targ_annot)\n",
    "        if regDB_ID != \"\" and len(gene_df[gene_df[\"GENE_ID\"] == regDB_ID]) > 0:\n",
    "            l.append({\n",
    "                \"RegulonDB ID\": regDB_ID,\n",
    "                \"name\": get_gene_name(regDB_ID),\n",
    "                \"feature type\": \"gene\"\n",
    "            })\n",
    "\n",
    "    return l\n",
    "\n",
    "\n",
    "all_muts_df[\"genetic features\"] = all_muts_df.apply(\n",
    "    # will default to what is already in r[\"genetic features\"] which is previously annotated above.\n",
    "    lambda r: get_genetic_feat_JSON_LTEE(\n",
    "        r) if r.exp in NON_K12_EXP_L else r[\"genetic features\"],\n",
    "    axis=1\n",
    ")\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_muts_df.to_pickle(\"./data/1_5_df.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
