{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.alemutdf import get_all_sample_mut_df, get_multi_exp_max_freq_mut_df\n",
    "from util.mut import is_coding_mut\n",
    "from util.metadata import get_condition_val_dict, get_condition_field_val_set\n",
    "from util.genome import get_K12_pos_from_BOP27"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:00<00:00, 240.03it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 306.15it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 264.58it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 264.34it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 235.25it/s]\n",
      "  0%|          | 0/20 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_untreated_mut.csv\n",
      "./data/muts/pub_growth//42C.csv\n",
      "./data/muts/pub_growth//Iron-ALE_mut.csv\n",
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2_mut.csv\n",
      "./data/muts/pub_growth//gallium-nitrate_mut.csv\n",
      "./data/muts/pub_growth//TOL adipic acid_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 252.94it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 244.81it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 242.16it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 252.43it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 245.35it/s]\n",
      "  0%|          | 0/67 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-blaTEM1_untreated_mut.csv\n",
      "./data/muts/pub_growth//tpiKO_mut.csv\n",
      "./data/muts/pub_growth//TOL n-butanol_mut.csv\n",
      "./data/muts/pub_growth//pvdJ_KO_P_aeruginosa_spent_medium_mut.csv\n",
      "./data/muts/pub_growth//tpiPAE_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 67/67 [00:00<00:00, 235.60it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 304.39it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 256.62it/s]\n",
      "  0%|          | 0/24 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//PGI.csv\n",
      "./data/muts/pub_growth//TOL octanoic acid_mut.csv\n",
      "./data/muts/pub_growth//TOL glutaric acid_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24/24 [00:00<00:00, 260.95it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 248.06it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 250.36it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 248.79it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 176.84it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 265.61it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2_mut.csv\n",
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2_mut.csv\n",
      "./data/muts/pub_growth//TOL putrescine_mut.csv\n",
      "./data/muts/pub_growth//WT_P_aeruginosa_spent_medium_mut.csv\n",
      "./data/muts/pub_growth//EG_pCA24N-fucO_mut.csv\n",
      "./data/muts/pub_growth//TOL hexamethylenediamine_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 256.78it/s]\n",
      "100%|██████████| 29/29 [00:00<00:00, 302.00it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 300.09it/s]\n",
      "  0%|          | 0/35 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//SSW GLU XYL.csv\n",
      "./data/muts/pub_growth//SER.csv\n",
      "./data/muts/pub_growth//SSW GLU AC.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 35/35 [00:00<00:00, 302.21it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 239.21it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 239.45it/s]\n",
      "100%|██████████| 2/2 [00:00<00:00, 228.06it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 257.46it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 293.29it/s]\n",
      "  0%|          | 0/20 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_doxycycline_mut.csv\n",
      "./data/muts/pub_growth//evo_mut.csv\n",
      "./data/muts/pub_growth//JCL260__isobutanol_production_host-M9-MG1655gbk_mut.csv\n",
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2_mut.csv\n",
      "./data/muts/pub_growth//SSW AC.csv\n",
      "./data/muts/pub_growth//TOL propanediol_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 251.63it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 259.10it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 243.08it/s]\n",
      "100%|██████████| 11/11 [00:00<00:00, 244.22it/s]\n",
      "  0%|          | 0/48 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//TOL 2,3-butanediol_mut.csv\n",
      "./data/muts/pub_growth//sdh_mut.csv\n",
      "./data/muts/pub_growth//tpiVCH_mut.csv\n",
      "./data/muts/pub_growth//GLU.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 48/48 [00:00<00:00, 232.42it/s]\n",
      "100%|██████████| 24/24 [00:00<00:00, 242.99it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 242.45it/s]\n",
      "  0%|          | 0/56 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//pgiBME_mut.csv\n",
      "./data/muts/pub_growth//MG1655_trimethoprim_mut.csv\n",
      "./data/muts/pub_growth//tpiHSA_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 56/56 [00:00<00:00, 234.09it/s]\n",
      "100%|██████████| 13/13 [00:00<00:00, 302.77it/s]\n",
      " 67%|██████▋   | 24/36 [00:00<00:00, 239.39it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//SSW GLU GLY.csv\n",
      "./data/muts/pub_growth//MG1655-pBGT_ceftazidime_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 36/36 [00:00<00:00, 235.71it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 261.62it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 236.30it/s]\n",
      "  0%|          | 0/31 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//TOL hexanoic acid_mut.csv\n",
      "./data/muts/pub_growth//wt_mut.csv\n",
      "./data/muts/pub_growth//application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 31/31 [00:00<00:00, 227.36it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 245.70it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 245.78it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 254.94it/s]\n",
      "100%|██████████| 13/13 [00:00<00:00, 243.66it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 235.90it/s]\n",
      "  0%|          | 0/13 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//pyocyanin_supplemented_medium_mut.csv\n",
      "./data/muts/pub_growth//MG1655_ceftazidime_mut.csv\n",
      "./data/muts/pub_growth//OxidizeME PQ3_mut.csv\n",
      "./data/muts/pub_growth//pgiHSA_mut.csv\n",
      "./data/muts/pub_growth//pgiVCH_mut.csv\n",
      "./data/muts/pub_growth//tpi_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 13/13 [00:00<00:00, 236.64it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 237.73it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 261.36it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 261.47it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_chloramphenicol_mut.csv\n",
      "./data/muts/pub_growth//TOL coumaric acid_mut.csv\n",
      "./data/muts/pub_growth//TOL isobutyric acid_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 4/4 [00:00<00:00, 295.39it/s]\n",
      "100%|██████████| 25/25 [00:00<00:00, 239.33it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 239.14it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 306.35it/s]\n",
      "  0%|          | 0/13 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//SSW GLY.csv\n",
      "./data/muts/pub_growth//ILT_MG1655_mut.csv\n",
      "./data/muts/pub_growth//untreated_gallium-nitrate.csv\n",
      "./data/muts/pub_growth//SSW XYL.csv\n",
      "./data/muts/pub_growth//pts_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 13/13 [00:00<00:00, 234.48it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 265.69it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 256.38it/s]\n",
      "100%|██████████| 26/26 [00:00<00:00, 235.10it/s]\n",
      "  0%|          | 0/67 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2_mut.csv\n",
      "./data/muts/pub_growth//untreated_pCA24N-fucO_mut.csv\n",
      "./data/muts/pub_growth//pgiPAE_mut.csv\n",
      "./data/muts/pub_growth//GYD.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 67/67 [00:00<00:00, 217.65it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 236.86it/s]\n",
      "100%|██████████| 13/13 [00:00<00:00, 243.45it/s]\n",
      " 36%|███▋      | 24/66 [00:00<00:00, 239.80it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_anchestor_mut.csv\n",
      "./data/muts/pub_growth//pgiKO_mut.csv\n",
      "./data/muts/pub_growth//pgi_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 66/66 [00:00<00:00, 235.58it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 234.41it/s]\n",
      "100%|██████████| 12/12 [00:00<00:00, 302.20it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 242.34it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 238.94it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-pBGT_untreated_mut.csv\n",
      "./data/muts/pub_growth//C13.csv\n",
      "./data/muts/pub_growth//MG1655-blaTEM1_ceftazidime_mut.csv\n",
      "./data/muts/pub_growth//gnd_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'42C',\n",
       " 'C13',\n",
       " 'EG_pCA24N-fucO',\n",
       " 'GLU',\n",
       " 'GYD',\n",
       " 'ILT_MG1655',\n",
       " 'Iron-ALE',\n",
       " 'JCL260__isobutanol_production_host-M9-MG1655gbk',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2',\n",
       " 'MG1655-blaTEM1_ceftazidime',\n",
       " 'MG1655-blaTEM1_untreated',\n",
       " 'MG1655-pBGT_ceftazidime',\n",
       " 'MG1655-pBGT_untreated',\n",
       " 'MG1655_anchestor',\n",
       " 'MG1655_ceftazidime',\n",
       " 'MG1655_chloramphenicol',\n",
       " 'MG1655_doxycycline',\n",
       " 'MG1655_trimethoprim',\n",
       " 'MG1655_untreated',\n",
       " 'OxidizeME_PQ3',\n",
       " 'PGI',\n",
       " 'SER',\n",
       " 'SSW_AC',\n",
       " 'SSW_GLU_AC',\n",
       " 'SSW_GLU_GLY',\n",
       " 'SSW_GLU_XYL',\n",
       " 'SSW_GLY',\n",
       " 'SSW_XYL',\n",
       " 'TOL_2,3-butanediol',\n",
       " 'TOL_adipic_acid',\n",
       " 'TOL_coumaric_acid',\n",
       " 'TOL_glutaric_acid',\n",
       " 'TOL_hexamethylenediamine',\n",
       " 'TOL_hexanoic_acid',\n",
       " 'TOL_isobutyric_acid',\n",
       " 'TOL_n-butanol',\n",
       " 'TOL_octanoic_acid',\n",
       " 'TOL_propanediol',\n",
       " 'TOL_putrescine',\n",
       " 'WT_P_aeruginosa_spent_medium',\n",
       " 'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk',\n",
       " 'evo',\n",
       " 'gallium-nitrate',\n",
       " 'gnd',\n",
       " 'pgi',\n",
       " 'pgiBME',\n",
       " 'pgiHSA',\n",
       " 'pgiKO',\n",
       " 'pgiPAE',\n",
       " 'pgiVCH',\n",
       " 'pts',\n",
       " 'pvdJ_KO_P_aeruginosa_spent_medium',\n",
       " 'pyocyanin_supplemented_medium',\n",
       " 'sdh',\n",
       " 'tpi',\n",
       " 'tpiHSA',\n",
       " 'tpiKO',\n",
       " 'tpiPAE',\n",
       " 'tpiVCH',\n",
       " 'untreated_pCA24N-fucO',\n",
       " 'wt'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = get_all_sample_mut_df(\"./data/muts/pub_growth/\")\n",
    "# unpublished_muts_df = get_all_sample_mut_df(\"./data/muts/unpublished/\")  # I've currently placed GYD into the public mutation set.\n",
    "# muts_df = muts_df.append(unpublished_muts_df)\n",
    "display(set(muts_df[\"exp\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## remove problem and starting strain mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Different experiments have different strings for position (some with commas, some without), therefore going ahead and changing them all to integers\n",
    "muts_df.Position = muts_df.Position.apply(lambda x: int(str(x).replace(\",\",\"\")))\n",
    "muts_df.Position = muts_df.Position.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17509"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1397381</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ13,756 bp</td>\n",
       "      <td></td>\n",
       "      <td>[ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608368.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.61</td>\n",
       "      <td>4542237</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (201‑209/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1397381</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ13,756 bp</td>\n",
       "      <td></td>\n",
       "      <td>[ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608368.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>4542456</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (420‑428/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608472.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.21</td>\n",
       "      <td>4542831</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS5 (+) +4 bp</td>\n",
       "      <td>intergenic (+198/‑281)</td>\n",
       "      <td>fimE, fimA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608473.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "6   MG1655_untreated    1     16        0         1      1.00   1397381   \n",
       "14  MG1655_untreated    1     16        0         1      0.61   4542237   \n",
       "6   MG1655_untreated    2     16        0         1      1.00   1397381   \n",
       "11  MG1655_untreated    2     16        0         1      0.22   4542456   \n",
       "12  MG1655_untreated    2     16        0         1      0.21   4542831   \n",
       "\n",
       "   Mutation Type Sequence Change                  Details  \\\n",
       "6            DEL      Δ13,756 bp                            \n",
       "14           MOB   IS1 (+) +9 bp  coding (201‑209/597 nt)   \n",
       "6            DEL      Δ13,756 bp                            \n",
       "11           MOB   IS1 (+) +9 bp  coding (420‑428/597 nt)   \n",
       "12           MOB   IS5 (+) +4 bp   intergenic (+198/‑281)   \n",
       "\n",
       "                                                 Gene Reference Seq     Mut ID  \n",
       "6   [ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...     NC_000913  1608368.0  \n",
       "14                                               fimE     NC_000913  1608475.0  \n",
       "6   [ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...     NC_000913  1608368.0  \n",
       "11                                               fimE     NC_000913  1608472.0  \n",
       "12                                         fimE, fimA     NC_000913  1608473.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = muts_df[~(\n",
    "# For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "#     ((muts_df.Gene == \"gatC\")\n",
    "#      | (muts_df.Gene == \"gatC, gatC\"))\n",
    "    ((muts_df.Position == 2173361)\n",
    "       | (muts_df.Position == 2173363)\n",
    "       | (muts_df.Position == 2173364))\n",
    "    & ((muts_df[\"Sequence Change\"] == \"Δ2 bp\")\n",
    "       | (muts_df[\"Sequence Change\"] == \"Δ1 bp\"))\n",
    ")]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"[crl]\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 257908)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ776 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"insB1, insA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1978503)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ776 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"ychE, oppA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1299499)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ1,199 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "# For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "#     ((muts_df.Gene == \"glpR\")\n",
    "#      | (muts_df.Gene == \"glpR, glpR\")\n",
    "#      | (muts_df.Gene == \"glpR/glpR\"))\n",
    "    (muts_df.Position == 3560455)\n",
    "    & (muts_df[\"Sequence Change\"] == \"+G\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"gltP, yjcO\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 4296381)\n",
    "    & (muts_df[\"Sequence Change\"] == \"+GC\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"gltP, yjcO\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 4296380)\n",
    "    & (muts_df[\"Sequence Change\"] == \"+CG\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"insB1,insA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1978503)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ776 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     ((muts_df.Gene == \"insA/uspC\") | (muts_df.Gene == \"insA, uspC\"))  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1979486)\n",
    "    & (muts_df[\"Sequence Change\"] == \"IS5 (+) +4 bp\"))]\n",
    "\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"corA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.exp != \"GLU\")  # This is the ALE that originated the corA mutation that is found in all CCK samples, therefore should track\n",
    "    & (muts_df.Position == 4001645)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ5 bp\"))]\n",
    "\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"wcaA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.exp != \"GLU\")  # This is the ALE that originated the corA mutation that is found in all CCK samples, therefore should track\n",
    "    & (muts_df.Position == 2132787)\n",
    "    & (muts_df[\"Details\"] == \"I204S (ATC→AGC)\"))]\n",
    "\n",
    "\n",
    "p = {4293212,\n",
    " 4293403,\n",
    " 4293430,\n",
    " 4296060,\n",
    " 4296154,\n",
    " 4296189,\n",
    " 4296190,\n",
    " 4296191,\n",
    " 4296267,\n",
    " 4296268,\n",
    " 4296286,\n",
    " 4296380,\n",
    " 4296382}\n",
    "ch = {'+C', '+G', 'A→C', 'A→G', 'A→T', 'C→A', 'C→T', 'T→C'}\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"gltP, yjcO\")\n",
    "    (muts_df[\"Position\"].isin(p))\n",
    "    & (muts_df[\"Sequence Change\"].isin(ch))\n",
    "       )]\n",
    "\n",
    "display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "reduced_ecoli_ale = 'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk'\n",
    "MG1655_pBGT_untreated = \"MG1655-pBGT_untreated\"\n",
    "MG1655_untreated = \"MG1655_untreated\"\n",
    "MG1655_blaTEM1_untreated = \"MG1655-blaTEM1_untreated\"\n",
    "muts_to_remove = [\n",
    "    {\"experiment\":MG1655_blaTEM1_untreated,\"position\":1397381,\"sequence change\":\"Δ13,756 bp\"},\n",
    "    {\"experiment\":MG1655_untreated,\"position\":1397381,\"sequence change\":\"Δ13,756 bp\"},\n",
    "    {\"experiment\":MG1655_pBGT_untreated,\"position\":1397381,\"sequence change\":\"Δ13,756 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":389898,\"sequence change\":\"Δ9,908 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1388889,\"sequence change\":\"Δ9,734 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2466545,\"sequence change\":\"Δ9,634 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":729365,\"sequence change\":\"Δ9,598 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3815859,\"sequence change\":\"Δ82 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1400325,\"sequence change\":\"Δ81,931 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":380112,\"sequence change\":\"Δ8,537 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2509631,\"sequence change\":\"Δ8,319 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3761993,\"sequence change\":\"Δ8,250 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":257908,\"sequence change\":\"Δ776 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1978503,\"sequence change\":\"Δ776 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1040502,\"sequence change\":\"Δ73,763 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3090347,\"sequence change\":\"Δ708 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":675552,\"sequence change\":\"Δ7,819 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":4400289,\"sequence change\":\"Δ7 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":819776,\"sequence change\":\"Δ62,977 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":263514,\"sequence change\":\"Δ61,897 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3184782,\"sequence change\":\"Δ6,916 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":997868,\"sequence change\":\"Δ6,818 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2558699,\"sequence change\":\"Δ6,790 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3618992,\"sequence change\":\"Δ6,688 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":4496676,\"sequence change\":\"Δ53,036 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":167484,\"sequence change\":\"Δ5,964 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":603467,\"sequence change\":\"Δ5,885 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3362162,\"sequence change\":\"Δ5,481 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":765208,\"sequence change\":\"Δ5,405 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":15389,\"sequence change\":\"Δ5,175 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1509562,\"sequence change\":\"Δ46,201 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":332366,\"sequence change\":\"Δ44,951 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":4555491,\"sequence change\":\"Δ41,523 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":508873,\"sequence change\":\"Δ40,744 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1223090,\"sequence change\":\"Δ4 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2101397,\"sequence change\":\"Δ36,321 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2756160,\"sequence change\":\"Δ35,091 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2204350,\"sequence change\":\"Δ32,368 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2994992,\"sequence change\":\"Δ3,877 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2286401,\"sequence change\":\"Δ3,780 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1997112,\"sequence change\":\"Δ26,567 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1197137,\"sequence change\":\"Δ25,940 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":4299628,\"sequence change\":\"Δ25,538 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1627518,\"sequence change\":\"Δ25,244 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":565055,\"sequence change\":\"Δ21,054 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2846301,\"sequence change\":\"Δ20,968 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3651293,\"sequence change\":\"Δ2,422 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2173363,\"sequence change\":\"Δ2 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3888432,\"sequence change\":\"Δ18,167 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1962566,\"sequence change\":\"Δ16,764 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2320144,\"sequence change\":\"Δ16,489 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3453928,\"sequence change\":\"Δ15,926 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2601201,\"sequence change\":\"Δ13,595 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":2165151,\"sequence change\":\"Δ12,060 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1129397,\"sequence change\":\"Δ11,591 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3593299,\"sequence change\":\"Δ10 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3720634,\"sequence change\":\"Δ1,442 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1872030,\"sequence change\":\"Δ1,437 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1299499,\"sequence change\":\"Δ1,199 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":687860,\"sequence change\":\"Δ1,185 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3581139,\"sequence change\":\"3915 bp→6 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":1197137,\"sequence change\":\"25956 bp→12 bp\"},\n",
    "    {\"experiment\":reduced_ecoli_ale,\"position\":3110680,\"sequence change\":\"25702 bp→GA\"},\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17184"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.61</td>\n",
       "      <td>4542237</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (201‑209/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>4542456</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (420‑428/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608472.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.21</td>\n",
       "      <td>4542831</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS5 (+) +4 bp</td>\n",
       "      <td>intergenic (+198/‑281)</td>\n",
       "      <td>fimE, fimA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608473.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.49</td>\n",
       "      <td>4542191</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D52A (GAC→GCC)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608466.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4542298</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +8 bp</td>\n",
       "      <td>coding (262‑269/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608467.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "14  MG1655_untreated    1     16        0         1      0.61   4542237   \n",
       "11  MG1655_untreated    2     16        0         1      0.22   4542456   \n",
       "12  MG1655_untreated    2     16        0         1      0.21   4542831   \n",
       "8   MG1655_untreated    3     16        0         1      0.49   4542191   \n",
       "9   MG1655_untreated    3     16        0         1      0.33   4542298   \n",
       "\n",
       "   Mutation Type Sequence Change                  Details        Gene  \\\n",
       "14           MOB   IS1 (+) +9 bp  coding (201‑209/597 nt)        fimE   \n",
       "11           MOB   IS1 (+) +9 bp  coding (420‑428/597 nt)        fimE   \n",
       "12           MOB   IS5 (+) +4 bp   intergenic (+198/‑281)  fimE, fimA   \n",
       "8            SNP             A→C           D52A (GAC→GCC)        fimE   \n",
       "9            MOB   IS1 (+) +8 bp  coding (262‑269/597 nt)        fimE   \n",
       "\n",
       "   Reference Seq     Mut ID  \n",
       "14     NC_000913  1608475.0  \n",
       "11     NC_000913  1608472.0  \n",
       "12     NC_000913  1608473.0  \n",
       "8      NC_000913  1608466.0  \n",
       "9      NC_000913  1608467.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for mut in muts_to_remove:\n",
    "    muts_df = muts_df[~(\n",
    "        (muts_df[\"exp\"] != mut[\"experiment\"])\n",
    "        & (muts_df[\"Position\"] == mut[\"position\"])\n",
    "        & (muts_df[\"Sequence Change\"] == mut[\"sequence change\"])\n",
    "    )]\n",
    "display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove experiment specific starting strain mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the following mutations out if they come from specific experiments.\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"cyoB\") & (muts_df[\"Details\"]==\"W190* (TGG→TAG)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"eno\") & (muts_df[\"Details\"]==\"L193Q (CTG→CAG)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"chbF\") & (muts_df[\"Details\"]==\"V133E (GTA→GAA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"fumC\") & (muts_df[\"Details\"]==\"G354G (GGT→GGC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"yegE\") & (muts_df[\"Details\"]==\"S683Y (TCC→TAC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"rcsC\") & (muts_df[\"Details\"]==\"L594F (TTG→TTC)\")))]\n",
    "\n",
    "\n",
    "muts_df = muts_df[(~((muts_df.exp==\"GYD\") & (muts_df[\"Gene\"]==\"fumC\") & (muts_df[\"Details\"]==\"G354G (GGT→GGC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"GYD\") & (muts_df[\"Gene\"]==\"yegE\") & (muts_df[\"Details\"]==\"S683Y (TCC→TAC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"GYD\") & (muts_df[\"Gene\"]==\"rcsC\") & (muts_df[\"Details\"]==\"L594F (TTG→TTC)\")))]\n",
    "\n",
    "\n",
    "cck_exp_list = [\n",
    "    \"Central_carbon_knockout_tpiA\",\n",
    "    \"Central_carbon_knockout_PTS\",\n",
    "    \"Central_carbon_knockout_pgi\",\n",
    "    \"Central_carbon_knockout_sdh\",\n",
    "    \"Central_carbon_knockout_gnd\",\n",
    "    \"Central_carbon_knockout_glucose_evolution\"\n",
    "]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "    (muts_df.exp.isin(cck_exp_list))\n",
    "    & (\n",
    "        (muts_df[\"Gene\"]==\"wcaA\")\n",
    "        | (muts_df[\"Gene\"]==\"corA\")\n",
    "    )\n",
    ")]  # The details of the mutations to filter out should probably include the sequence change."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Removing MG1655_anchestor mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to use with filtering out mutations \n",
    "muts_df = muts_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16116</th>\n",
       "      <td>MG1655_anchestor</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4390754</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>A252A (GCC→GCA)</td>\n",
       "      <td>rsgA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608306.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16117</th>\n",
       "      <td>MG1655_anchestor</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>803662</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>L54I (CTT→ATT)</td>\n",
       "      <td>ybhJ</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16118</th>\n",
       "      <td>MG1655_anchestor</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1905761</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→A</td>\n",
       "      <td>G25D (GGT→GAT)</td>\n",
       "      <td>mntP</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608303.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "16116  MG1655_anchestor    0      0        1         1       1.0   4390754   \n",
       "16117  MG1655_anchestor    0      0        1         1       1.0    803662   \n",
       "16118  MG1655_anchestor    0      0        1         1       1.0   1905761   \n",
       "\n",
       "      Mutation Type Sequence Change          Details  Gene Reference Seq  \\\n",
       "16116           SNP             G→T  A252A (GCC→GCA)  rsgA     NC_000913   \n",
       "16117           SNP             C→A   L54I (CTT→ATT)  ybhJ     NC_000913   \n",
       "16118           SNP             G→A   G25D (GGT→GAT)  mntP     NC_000913   \n",
       "\n",
       "          Mut ID  \n",
       "16116  1608306.0  \n",
       "16117  1608301.0  \n",
       "16118  1608303.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Not many. Going to manual remove (see cell below)\n",
    "# This is for the Ecoli_resistance_caz\n",
    "# Sebastian often uses the name \"MG1655_anchestor\" \n",
    "MG1655_anchestor = muts_df[muts_df.exp==\"MG1655_anchestor\"].copy()\n",
    "MG1655_anchestor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16980\n",
      "16977\n"
     ]
    }
   ],
   "source": [
    "print(len(muts_df))\n",
    "muts_df = muts_df[~(muts_df.index.isin(list(MG1655_anchestor.index)))]\n",
    "print(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16920\n"
     ]
    }
   ],
   "source": [
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_chloramphenicol\") & (muts_df[\"Position\"]==4390754) & (muts_df[\"Details\"]==\"A252A (GCC→GCA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_chloramphenicol\") & (muts_df[\"Position\"]==803662) & (muts_df[\"Details\"]==\"L54I (CTT→ATT)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_chloramphenicol\") & (muts_df[\"Position\"]==1905761) & (muts_df[\"Details\"]==\"G25D (GGT→GAT)\")))]\n",
    "\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_doxycycline\") & (muts_df[\"Position\"]==4390754) & (muts_df[\"Details\"]==\"A252A (GCC→GCA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_doxycycline\") & (muts_df[\"Position\"]==803662) & (muts_df[\"Details\"]==\"L54I (CTT→ATT)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_doxycycline\") & (muts_df[\"Position\"]==1905761) & (muts_df[\"Details\"]==\"G25D (GGT→GAT)\")))]\n",
    "\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_trimethoprim\") & (muts_df[\"Position\"]==4390754) & (muts_df[\"Details\"]==\"A252A (GCC→GCA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_trimethoprim\") & (muts_df[\"Position\"]==803662) & (muts_df[\"Details\"]==\"L54I (CTT→ATT)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_trimethoprim\") & (muts_df[\"Position\"]==1905761) & (muts_df[\"Details\"]==\"G25D (GGT→GAT)\")))]\n",
    "print(len(muts_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Though the other Ecoli_resistance_caz experiments have \"anchestor\" data on the aledb.ucsd.edu server, their mutations aren't uploaded, therefore not going to filter anything from them at this point."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Starting strain mutations according to samples within experiments with starting strain AFIRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to use with filtering out mutations \n",
    "muts_df = muts_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.61</td>\n",
       "      <td>4542237</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (201‑209/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608475.0</td>\n",
       "      <td>1 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>4542456</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (420‑428/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608472.0</td>\n",
       "      <td>2 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.21</td>\n",
       "      <td>4542831</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS5 (+) +4 bp</td>\n",
       "      <td>intergenic (+198/‑281)</td>\n",
       "      <td>fimE, fimA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608473.0</td>\n",
       "      <td>2 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.49</td>\n",
       "      <td>4542191</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D52A (GAC→GCC)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608466.0</td>\n",
       "      <td>3 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.33</td>\n",
       "      <td>4542298</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +8 bp</td>\n",
       "      <td>coding (262‑269/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608467.0</td>\n",
       "      <td>3 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16915</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1697375</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→C</td>\n",
       "      <td>Q222Q (CAA→CAG)</td>\n",
       "      <td>hdhA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>799561.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16916</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3951735</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ326 bp</td>\n",
       "      <td>pseudogene (194‑519/663 nt)</td>\n",
       "      <td>ilvG</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1114004.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16917</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3700725</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>I55F (ATC→TTC)</td>\n",
       "      <td>yhjV</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1566613.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16918</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3968459</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>Q182* (CAA→TAA)</td>\n",
       "      <td>wecA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1566614.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16919</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>4184543</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>P1100Q (CCG→CAG)</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1112480.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16920 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "0      MG1655_untreated    1     16        0         1      0.61   4542237   \n",
       "1      MG1655_untreated    2     16        0         1      0.22   4542456   \n",
       "2      MG1655_untreated    2     16        0         1      0.21   4542831   \n",
       "3      MG1655_untreated    3     16        0         1      0.49   4542191   \n",
       "4      MG1655_untreated    3     16        0         1      0.33   4542298   \n",
       "...                 ...  ...    ...      ...       ...       ...       ...   \n",
       "16915               gnd   19    171        2         1      1.00   1697375   \n",
       "16916               gnd   19    171        2         1      1.00   3951735   \n",
       "16917               gnd   19    171        2         1      1.00   3700725   \n",
       "16918               gnd   19    171        2         1      1.00   3968459   \n",
       "16919               gnd   19    171        2         1      1.00   4184543   \n",
       "\n",
       "      Mutation Type Sequence Change                      Details        Gene  \\\n",
       "0               MOB   IS1 (+) +9 bp      coding (201‑209/597 nt)        fimE   \n",
       "1               MOB   IS1 (+) +9 bp      coding (420‑428/597 nt)        fimE   \n",
       "2               MOB   IS5 (+) +4 bp       intergenic (+198/‑281)  fimE, fimA   \n",
       "3               SNP             A→C               D52A (GAC→GCC)        fimE   \n",
       "4               MOB   IS1 (+) +8 bp      coding (262‑269/597 nt)        fimE   \n",
       "...             ...             ...                          ...         ...   \n",
       "16915           SNP             T→C              Q222Q (CAA→CAG)        hdhA   \n",
       "16916           DEL         Δ326 bp  pseudogene (194‑519/663 nt)        ilvG   \n",
       "16917           SNP             A→T               I55F (ATC→TTC)        yhjV   \n",
       "16918           SNP             C→T              Q182* (CAA→TAA)        wecA   \n",
       "16919           SNP             C→A             P1100Q (CCG→CAG)        rpoB   \n",
       "\n",
       "      Reference Seq     Mut ID      sample  \n",
       "0         NC_000913  1608475.0    1 16 0 1  \n",
       "1         NC_000913  1608472.0    2 16 0 1  \n",
       "2         NC_000913  1608473.0    2 16 0 1  \n",
       "3         NC_000913  1608466.0    3 16 0 1  \n",
       "4         NC_000913  1608467.0    3 16 0 1  \n",
       "...             ...        ...         ...  \n",
       "16915     NC_000913   799561.0  19 171 2 1  \n",
       "16916     NC_000913  1114004.0  19 171 2 1  \n",
       "16917     NC_000913  1566613.0  19 171 2 1  \n",
       "16918     NC_000913  1566614.0  19 171 2 1  \n",
       "16919     NC_000913  1112480.0  19 171 2 1  \n",
       "\n",
       "[16920 rows x 14 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_sample_name(ale, flask, isolate, tech_rep):\n",
    "    sample_name = str(int(ale)) + \" \" + str(int(flask)) + \" \" + str(int(isolate)) + \" \" + str(int(tech_rep))\n",
    "    return sample_name\n",
    "\n",
    "muts_df[\"sample\"] = muts_df.apply(lambda r: get_sample_name(r.ale, r.flask, r.isolate, r.tech_rep), axis=1)\n",
    "muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove starting strain mutations from all experiments according to each experiment's starting strain\n",
    "muts_df = muts_df.copy()\n",
    "starting_strains = muts_df[muts_df[\"sample\"]==\"0 0 1 1\"]\n",
    "muts_df = muts_df[muts_df[\"sample\"]!=\"0 0 1 1\"]  # remove starting strain samples\n",
    "for e, exp_starting_strain_muts in starting_strains.groupby(\"exp\"):\n",
    "    rows_to_filter = (\n",
    "        muts_df[\"exp\"].isin(exp_starting_strain_muts[\"exp\"])\n",
    "        & muts_df[\"Position\"].isin(exp_starting_strain_muts[\"Position\"])\n",
    "        & muts_df[\"Sequence Change\"].isin(exp_starting_strain_muts[\"Sequence Change\"])\n",
    "        & muts_df[\"Details\"].isin(exp_starting_strain_muts[\"Details\"])\n",
    "    )\n",
    "    muts_df.drop(muts_df[rows_to_filter].index, inplace=True)  # have to ensure that muts_df indeces are unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Check that no weird mutations references are included in dataset, since they won't match to RegulonDB's data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP009273 ['TOL_isobutyric_acid']\n",
      "NC_000913 ['MG1655_untreated' 'Iron-ALE'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2'\n",
      " 'gallium-nitrate' 'TOL_adipic_acid' 'MG1655-blaTEM1_untreated' 'tpiKO'\n",
      " 'TOL_n-butanol' 'pvdJ_KO_P_aeruginosa_spent_medium' 'TOL_octanoic_acid'\n",
      " 'TOL_glutaric_acid'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2' 'TOL_putrescine'\n",
      " 'WT_P_aeruginosa_spent_medium' 'EG_pCA24N-fucO'\n",
      " 'TOL_hexamethylenediamine' 'MG1655_doxycycline' 'evo'\n",
      " 'JCL260__isobutanol_production_host-M9-MG1655gbk'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2' 'TOL_propanediol'\n",
      " 'TOL_2,3-butanediol' 'sdh' 'GLU' 'MG1655_trimethoprim'\n",
      " 'MG1655-pBGT_ceftazidime' 'TOL_hexanoic_acid' 'wt'\n",
      " 'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk'\n",
      " 'pyocyanin_supplemented_medium' 'MG1655_ceftazidime' 'OxidizeME_PQ3'\n",
      " 'tpi' 'MG1655_chloramphenicol' 'TOL_coumaric_acid' 'TOL_isobutyric_acid'\n",
      " 'ILT_MG1655' 'pts' 'MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2'\n",
      " 'GYD' 'pgi' 'MG1655-pBGT_untreated' 'MG1655-blaTEM1_ceftazidime' 'gnd']\n",
      "NC_000913_3_KO_pgi ['pgiKO']\n",
      "NC_000913_3_bme_pgi ['pgiBME']\n",
      "NC_000913_3_hsa_pgi ['pgiHSA']\n",
      "NC_000913_3_hsa_tpiA ['tpiHSA']\n",
      "NC_000913_3_pae_pgi ['pgiPAE']\n",
      "NC_000913_3_pae_tpiA ['tpiPAE']\n",
      "NC_000913_3_vch_pgi ['pgiVCH']\n",
      "NC_000913_3_vch_tpiA ['tpiVCH']\n"
     ]
    }
   ],
   "source": [
    "for ref_seq, df in muts_df.groupby(\"Reference Seq\"):\n",
    "    print(ref_seq, df.exp.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "According the the prevous results, TOL_isobutyric_acid has one weird mutation. Going to remove it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NC_000913', nan, 'NC_000913_3_pae_tpiA', 'NC_000913_3_vch_tpiA',\n",
       "       'NC_000913_3_bme_pgi', 'NC_000913_3_hsa_tpiA',\n",
       "       'NC_000913_3_hsa_pgi', 'NC_000913_3_vch_pgi',\n",
       "       'NC_000913_3_pae_pgi', 'NC_000913_3_KO_pgi'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df = muts_df[muts_df[\"Reference Seq\"] != \"CP009273\"]\n",
    "muts_df[\"Reference Seq\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "388\n",
      "375\n"
     ]
    }
   ],
   "source": [
    "# Amount of ALEs before removal\n",
    "print(len(muts_df.apply(lambda r: r.exp + \" \" + str(r.ale), axis=1).unique()))\n",
    "\n",
    "# removing hypermutators since they will affect associations\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"GLU\")\n",
    "        & ((muts_df.ale == 5) | (muts_df.ale == 7))\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_2,3-butanediol\")\n",
    "        & (muts_df.ale == 3)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_adipic_acid\")\n",
    "        & (muts_df.ale == 5)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_hexamethylenediamine\")\n",
    "        & (muts_df.ale == 6)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_octanoic_acid\")\n",
    "        & (muts_df.ale == 5)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_propanediol\")\n",
    "        & ((muts_df.ale == 1) | (muts_df.ale == 2) | (muts_df.ale == 3) | (muts_df.ale == 4) | (muts_df.ale == 5) | (muts_df.ale == 7) | (muts_df.ale == 8))\n",
    "    )\n",
    "]\n",
    "\n",
    "print(len(muts_df.apply(lambda r: r.exp + \" \" + str(r.ale), axis=1).unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "373\n"
     ]
    }
   ],
   "source": [
    "# removing samples not incuded in GYD MS for AVA MS\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"GYD\")\n",
    "        & ((muts_df.ale == 35) | (muts_df.ale == 34))\n",
    "    )\n",
    "]\n",
    "print(len(muts_df.apply(lambda r: r.exp + \" \" + str(r.ale), axis=1).unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10008\n",
      "9514\n"
     ]
    }
   ],
   "source": [
    "print(len(muts_df))\n",
    "muts_df = muts_df[muts_df.presence >= 0.5]\n",
    "print(len(muts_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Just get ALE-unique mutations \n",
    "Since the majority of our comparisons have to do with endpoints, just look at endpoints.  \n",
    "If not limiting to endpoint, the mutations used in associated mutations to conditions will need filtering.  \n",
    "This occurs in a later NB of the pipeline.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>42C</td>\n",
       "      <td>2</td>\n",
       "      <td>163</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>149214</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A130T (GCC→ACC)</td>\n",
       "      <td>panB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2 163 1 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "9   42C    1    124        1         1       1.0    702352           DEL   \n",
       "10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "11  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "12  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "21  42C    2    163        1         1       1.0    149214           SNP   \n",
       "\n",
       "   Sequence Change                   Details  Gene Reference Seq  Mut ID  \\\n",
       "9           Δ21 bp  coding (380‑400/1149 nt)  nagA           NaN     NaN   \n",
       "10             G→C           R110G (CGT→GGT)  clsA           NaN     NaN   \n",
       "11             C→T           A734V (GCG→GTG)  rpoC           NaN     NaN   \n",
       "12             A→C             D9A (GAT→GCT)   hfq           NaN     NaN   \n",
       "21             C→T           A130T (GCC→ACC)  panB           NaN     NaN   \n",
       "\n",
       "       sample  \n",
       "9   1 124 1 1  \n",
       "10  1 124 1 1  \n",
       "11  1 124 1 1  \n",
       "12  1 124 1 1  \n",
       "21  2 163 1 1  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df = get_multi_exp_max_freq_mut_df(muts_df, endpoint_flask_only=False)\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensure all mutations are using NC_000913 positions.\n",
    "genome_pos_transform_exp_l = [\"ERS\",\n",
    "                              'Central_carbon_knockout_PTS',\n",
    "                              'Central_carbon_knockout_glucose_evolution',\n",
    "                              'Central_carbon_knockout_gnd',\n",
    "                              'Central_carbon_knockout_pgi',\n",
    "                              'Central_carbon_knockout_sdh',\n",
    "                              'Central_carbon_knockout_tpiA',]\n",
    "muts_df.Position = muts_df.apply(lambda row: get_K12_pos_from_BOP27(row.Position) if row.exp in genome_pos_transform_exp_l else row.Position, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>42C</td>\n",
       "      <td>2</td>\n",
       "      <td>163</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>149214</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A130T (GCC→ACC)</td>\n",
       "      <td>panB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2 163 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "9   42C    1    124        1         1       1.0    702352           DEL   \n",
       "10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "11  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "12  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "21  42C    2    163        1         1       1.0    149214           SNP   \n",
       "\n",
       "   Sequence Change                   Details  Gene Reference Seq  Mut ID  \\\n",
       "9           Δ21 bp  coding (380‑400/1149 nt)  nagA           NaN     NaN   \n",
       "10             G→C           R110G (CGT→GGT)  clsA           NaN     NaN   \n",
       "11             C→T           A734V (GCG→GTG)  rpoC           NaN     NaN   \n",
       "12             A→C             D9A (GAT→GCT)   hfq           NaN     NaN   \n",
       "21             C→T           A130T (GCC→ACC)  panB           NaN     NaN   \n",
       "\n",
       "       sample  coding  \n",
       "9   1 124 1 1    True  \n",
       "10  1 124 1 1    True  \n",
       "11  1 124 1 1    True  \n",
       "12  1 124 1 1    True  \n",
       "21  2 163 1 1    True  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df[\"coding\"] = muts_df.Details.apply(is_coding_mut)\n",
    "muts_df[\"Gene\"] = muts_df.apply(lambda row: row[\"Gene\"].replace(\", \",\"/\") if not row[\"coding\"] else row[\"Gene\"], axis=1)\n",
    "\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>42C</td>\n",
       "      <td>2</td>\n",
       "      <td>163</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>149214</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A130T (GCC→ACC)</td>\n",
       "      <td>panB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2 163 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "9   42C    1    124        1         1       1.0    702352           DEL   \n",
       "10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "11  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "12  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "21  42C    2    163        1         1       1.0    149214           SNP   \n",
       "\n",
       "   Sequence Change                   Details  Gene Reference Seq  Mut ID  \\\n",
       "9           Δ21 bp  coding (380‑400/1149 nt)  nagA           NaN     NaN   \n",
       "10             G→C           R110G (CGT→GGT)  clsA           NaN     NaN   \n",
       "11             C→T           A734V (GCG→GTG)  rpoC           NaN     NaN   \n",
       "12             A→C             D9A (GAT→GCT)   hfq           NaN     NaN   \n",
       "21             C→T           A130T (GCC→ACC)  panB           NaN     NaN   \n",
       "\n",
       "       sample  coding  \n",
       "9   1 124 1 1    True  \n",
       "10  1 124 1 1    True  \n",
       "11  1 124 1 1    True  \n",
       "12  1 124 1 1    True  \n",
       "21  2 163 1 1    True  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This work is also currently duplicated NB4 \n",
    "muts_df[\"Gene\"] = muts_df[\"Gene\"].apply(lambda a: \"rph\" if a == \"[rph], [rph]\" else a)\n",
    "muts_df[\"Gene\"] = muts_df[\"Gene\"].apply(lambda a: \"rph\" if a == \"[rph],[rph]\" else a)\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# removing unnecessary column\n",
    "muts_df = muts_df.drop(columns=[\"Mut ID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "muts_df.to_pickle(\"./data/1_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6940"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(muts_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
