{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.alemutdf import get_all_sample_mut_df, get_multi_exp_max_freq_mut_df\n",
    "from util.mut import is_coding_mut\n",
    "from util.metadata import get_condition_val_dict, get_condition_field_val_set\n",
    "from util.genome import get_K12_pos_from_BOP27"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:00<00:00, 240.42it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 302.93it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 256.28it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 258.60it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 240.55it/s]\n",
      "  0%|          | 0/20 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_untreated_mut.csv\n",
      "./data/muts/pub_growth//42C.csv\n",
      "./data/muts/pub_growth//Iron-ALE_mut.csv\n",
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2_mut.csv\n",
      "./data/muts/pub_growth//gallium-nitrate_mut.csv\n",
      "./data/muts/pub_growth//TOL adipic acid_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 253.71it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 244.62it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 236.57it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 253.89it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 243.22it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-blaTEM1_untreated_mut.csv\n",
      "./data/muts/pub_growth//tpiKO_mut.csv\n",
      "./data/muts/pub_growth//TOL n-butanol_mut.csv\n",
      "./data/muts/pub_growth//pvdJ_KO_P_aeruginosa_spent_medium_mut.csv\n",
      "./data/muts/pub_growth//tpiPAE_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 67/67 [00:00<00:00, 237.87it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 306.58it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 257.72it/s]\n",
      "  0%|          | 0/24 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//PGI.csv\n",
      "./data/muts/pub_growth//TOL octanoic acid_mut.csv\n",
      "./data/muts/pub_growth//TOL glutaric acid_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24/24 [00:00<00:00, 254.96it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 258.30it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 249.08it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 253.46it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 233.10it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 248.93it/s]\n",
      "  0%|          | 0/20 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2_mut.csv\n",
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2_mut.csv\n",
      "./data/muts/pub_growth//TOL putrescine_mut.csv\n",
      "./data/muts/pub_growth//WT_P_aeruginosa_spent_medium_mut.csv\n",
      "./data/muts/pub_growth//EG_pCA24N-fucO_mut.csv\n",
      "./data/muts/pub_growth//TOL hexamethylenediamine_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 254.14it/s]\n",
      "100%|██████████| 29/29 [00:00<00:00, 305.09it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 298.83it/s]\n",
      "  0%|          | 0/35 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//SSW GLU XYL.csv\n",
      "./data/muts/pub_growth//SER.csv\n",
      "./data/muts/pub_growth//SSW GLU AC.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 35/35 [00:00<00:00, 302.33it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 243.73it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 242.88it/s]\n",
      "100%|██████████| 2/2 [00:00<00:00, 237.74it/s]\n",
      "100%|██████████| 8/8 [00:00<00:00, 246.76it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 304.18it/s]\n",
      "  0%|          | 0/20 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_doxycycline_mut.csv\n",
      "./data/muts/pub_growth//evo_mut.csv\n",
      "./data/muts/pub_growth//JCL260__isobutanol_production_host-M9-MG1655gbk_mut.csv\n",
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2_mut.csv\n",
      "./data/muts/pub_growth//SSW AC.csv\n",
      "./data/muts/pub_growth//TOL propanediol_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 251.63it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 258.43it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 242.43it/s]\n",
      "100%|██████████| 11/11 [00:00<00:00, 242.76it/s]\n",
      "  0%|          | 0/48 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//TOL 2,3-butanediol_mut.csv\n",
      "./data/muts/pub_growth//sdh_mut.csv\n",
      "./data/muts/pub_growth//tpiVCH_mut.csv\n",
      "./data/muts/pub_growth//GLU.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 48/48 [00:00<00:00, 234.30it/s]\n",
      "100%|██████████| 24/24 [00:00<00:00, 242.24it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 243.90it/s]\n",
      "  0%|          | 0/56 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//pgiBME_mut.csv\n",
      "./data/muts/pub_growth//MG1655_trimethoprim_mut.csv\n",
      "./data/muts/pub_growth//tpiHSA_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 56/56 [00:00<00:00, 236.44it/s]\n",
      "100%|██████████| 13/13 [00:00<00:00, 306.27it/s]\n",
      " 69%|██████▉   | 25/36 [00:00<00:00, 241.93it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//SSW GLU GLY.csv\n",
      "./data/muts/pub_growth//MG1655-pBGT_ceftazidime_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 36/36 [00:00<00:00, 235.48it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 258.29it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 238.80it/s]\n",
      "  0%|          | 0/31 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//TOL hexanoic acid_mut.csv\n",
      "./data/muts/pub_growth//wt_mut.csv\n",
      "./data/muts/pub_growth//application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 31/31 [00:00<00:00, 230.60it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 238.08it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 242.67it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 258.45it/s]\n",
      "100%|██████████| 13/13 [00:00<00:00, 240.49it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 245.08it/s]\n",
      "  0%|          | 0/13 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//pyocyanin_supplemented_medium_mut.csv\n",
      "./data/muts/pub_growth//MG1655_ceftazidime_mut.csv\n",
      "./data/muts/pub_growth//OxidizeME PQ3_mut.csv\n",
      "./data/muts/pub_growth//pgiHSA_mut.csv\n",
      "./data/muts/pub_growth//pgiVCH_mut.csv\n",
      "./data/muts/pub_growth//tpi_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 13/13 [00:00<00:00, 239.10it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 242.07it/s]\n",
      "100%|██████████| 20/20 [00:00<00:00, 253.94it/s]\n",
      "  0%|          | 0/20 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_chloramphenicol_mut.csv\n",
      "./data/muts/pub_growth//TOL coumaric acid_mut.csv\n",
      "./data/muts/pub_growth//TOL isobutyric acid_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 254.09it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 310.80it/s]\n",
      "100%|██████████| 25/25 [00:00<00:00, 237.96it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 240.53it/s]\n",
      "100%|██████████| 4/4 [00:00<00:00, 304.46it/s]\n",
      "  0%|          | 0/13 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//SSW GLY.csv\n",
      "./data/muts/pub_growth//ILT_MG1655_mut.csv\n",
      "./data/muts/pub_growth//untreated_gallium-nitrate.csv\n",
      "./data/muts/pub_growth//SSW XYL.csv\n",
      "./data/muts/pub_growth//pts_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 13/13 [00:00<00:00, 239.60it/s]\n",
      "100%|██████████| 7/7 [00:00<00:00, 256.67it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 258.78it/s]\n",
      "100%|██████████| 26/26 [00:00<00:00, 239.71it/s]\n",
      "  0%|          | 0/67 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2_mut.csv\n",
      "./data/muts/pub_growth//untreated_pCA24N-fucO_mut.csv\n",
      "./data/muts/pub_growth//pgiPAE_mut.csv\n",
      "./data/muts/pub_growth//GYD.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 67/67 [00:00<00:00, 238.89it/s]\n",
      "100%|██████████| 1/1 [00:00<00:00, 264.27it/s]\n",
      "100%|██████████| 13/13 [00:00<00:00, 241.01it/s]\n",
      " 38%|███▊      | 25/66 [00:00<00:00, 240.68it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655_anchestor_mut.csv\n",
      "./data/muts/pub_growth//pgiKO_mut.csv\n",
      "./data/muts/pub_growth//pgi_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 66/66 [00:00<00:00, 236.22it/s]\n",
      "100%|██████████| 6/6 [00:00<00:00, 241.73it/s]\n",
      "100%|██████████| 12/12 [00:00<00:00, 307.19it/s]\n",
      "100%|██████████| 5/5 [00:00<00:00, 244.74it/s]\n",
      "100%|██████████| 10/10 [00:00<00:00, 243.78it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/muts/pub_growth//MG1655-pBGT_untreated_mut.csv\n",
      "./data/muts/pub_growth//C13.csv\n",
      "./data/muts/pub_growth//MG1655-blaTEM1_ceftazidime_mut.csv\n",
      "./data/muts/pub_growth//gnd_mut.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'42C',\n",
       " 'C13',\n",
       " 'EG_pCA24N-fucO',\n",
       " 'GLU',\n",
       " 'GYD',\n",
       " 'ILT_MG1655',\n",
       " 'Iron-ALE',\n",
       " 'JCL260__isobutanol_production_host-M9-MG1655gbk',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2',\n",
       " 'MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2',\n",
       " 'MG1655-blaTEM1_ceftazidime',\n",
       " 'MG1655-blaTEM1_untreated',\n",
       " 'MG1655-pBGT_ceftazidime',\n",
       " 'MG1655-pBGT_untreated',\n",
       " 'MG1655_anchestor',\n",
       " 'MG1655_ceftazidime',\n",
       " 'MG1655_chloramphenicol',\n",
       " 'MG1655_doxycycline',\n",
       " 'MG1655_trimethoprim',\n",
       " 'MG1655_untreated',\n",
       " 'OxidizeME_PQ3',\n",
       " 'PGI',\n",
       " 'SER',\n",
       " 'SSW_AC',\n",
       " 'SSW_GLU_AC',\n",
       " 'SSW_GLU_GLY',\n",
       " 'SSW_GLU_XYL',\n",
       " 'SSW_GLY',\n",
       " 'SSW_XYL',\n",
       " 'TOL_2,3-butanediol',\n",
       " 'TOL_adipic_acid',\n",
       " 'TOL_coumaric_acid',\n",
       " 'TOL_glutaric_acid',\n",
       " 'TOL_hexamethylenediamine',\n",
       " 'TOL_hexanoic_acid',\n",
       " 'TOL_isobutyric_acid',\n",
       " 'TOL_n-butanol',\n",
       " 'TOL_octanoic_acid',\n",
       " 'TOL_propanediol',\n",
       " 'TOL_putrescine',\n",
       " 'WT_P_aeruginosa_spent_medium',\n",
       " 'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk',\n",
       " 'evo',\n",
       " 'gallium-nitrate',\n",
       " 'gnd',\n",
       " 'pgi',\n",
       " 'pgiBME',\n",
       " 'pgiHSA',\n",
       " 'pgiKO',\n",
       " 'pgiPAE',\n",
       " 'pgiVCH',\n",
       " 'pts',\n",
       " 'pvdJ_KO_P_aeruginosa_spent_medium',\n",
       " 'pyocyanin_supplemented_medium',\n",
       " 'sdh',\n",
       " 'tpi',\n",
       " 'tpiHSA',\n",
       " 'tpiKO',\n",
       " 'tpiPAE',\n",
       " 'tpiVCH',\n",
       " 'untreated_pCA24N-fucO',\n",
       " 'wt'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = get_all_sample_mut_df(\"./data/muts/pub_growth/\")\n",
    "# unpublished_muts_df = get_all_sample_mut_df(\"./data/muts/unpublished/\")  # I've currently placed GYD into the public mutation set.\n",
    "# muts_df = muts_df.append(unpublished_muts_df)\n",
    "display(set(muts_df[\"exp\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## remove problem and starting strain mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Different experiments have different strings for position (some with commas, some without), therefore going ahead and changing them all to integers\n",
    "muts_df.Position = muts_df.Position.apply(lambda x: int(str(x).replace(\",\",\"\")))\n",
    "muts_df.Position = muts_df.Position.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17509"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1397381</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ13,756 bp</td>\n",
       "      <td></td>\n",
       "      <td>[ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608368.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.61</td>\n",
       "      <td>4542237</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (201‑209/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608475.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1397381</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ13,756 bp</td>\n",
       "      <td></td>\n",
       "      <td>[ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608368.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>4542456</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (420‑428/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608472.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.21</td>\n",
       "      <td>4542831</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS5 (+) +4 bp</td>\n",
       "      <td>intergenic (+198/‑281)</td>\n",
       "      <td>fimE, fimA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608473.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "6   MG1655_untreated    1     16        0         1      1.00   1397381   \n",
       "14  MG1655_untreated    1     16        0         1      0.61   4542237   \n",
       "6   MG1655_untreated    2     16        0         1      1.00   1397381   \n",
       "11  MG1655_untreated    2     16        0         1      0.22   4542456   \n",
       "12  MG1655_untreated    2     16        0         1      0.21   4542831   \n",
       "\n",
       "   Mutation Type Sequence Change                  Details  \\\n",
       "6            DEL      Δ13,756 bp                            \n",
       "14           MOB   IS1 (+) +9 bp  coding (201‑209/597 nt)   \n",
       "6            DEL      Δ13,756 bp                            \n",
       "11           MOB   IS1 (+) +9 bp  coding (420‑428/597 nt)   \n",
       "12           MOB   IS5 (+) +4 bp   intergenic (+198/‑281)   \n",
       "\n",
       "                                                 Gene Reference Seq     Mut ID  \n",
       "6   [ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...     NC_000913  1608368.0  \n",
       "14                                               fimE     NC_000913  1608475.0  \n",
       "6   [ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...     NC_000913  1608368.0  \n",
       "11                                               fimE     NC_000913  1608472.0  \n",
       "12                                         fimE, fimA     NC_000913  1608473.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = muts_df[~(\n",
    "# For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "#     ((muts_df.Gene == \"gatC\")\n",
    "#      | (muts_df.Gene == \"gatC, gatC\"))\n",
    "    ((muts_df.Position == 2173361)\n",
    "       | (muts_df.Position == 2173363)\n",
    "       | (muts_df.Position == 2173364))\n",
    "    & ((muts_df[\"Sequence Change\"] == \"Δ2 bp\")\n",
    "       | (muts_df[\"Sequence Change\"] == \"Δ1 bp\"))\n",
    ")]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"[crl]\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 257908)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ776 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"insB1, insA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1978503)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ776 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"ychE, oppA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1299499)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ1,199 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "# For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "#     ((muts_df.Gene == \"glpR\")\n",
    "#      | (muts_df.Gene == \"glpR, glpR\")\n",
    "#      | (muts_df.Gene == \"glpR/glpR\"))\n",
    "    (muts_df.Position == 3560455)\n",
    "    & (muts_df[\"Sequence Change\"] == \"+G\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"gltP, yjcO\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 4296381)\n",
    "    & (muts_df[\"Sequence Change\"] == \"+GC\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"gltP, yjcO\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 4296380)\n",
    "    & (muts_df[\"Sequence Change\"] == \"+CG\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"insB1,insA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1978503)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ776 bp\"))]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     ((muts_df.Gene == \"insA/uspC\") | (muts_df.Gene == \"insA, uspC\"))  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.Position == 1979486)\n",
    "    & (muts_df[\"Sequence Change\"] == \"IS5 (+) +4 bp\"))]\n",
    "\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"corA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.exp != \"GLU\")  # This is the ALE that originated the corA mutation that is found in all CCK samples, therefore should track\n",
    "    & (muts_df.Position == 4001645)\n",
    "    & (muts_df[\"Sequence Change\"] == \"Δ5 bp\"))]\n",
    "\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"wcaA\")  # For whatever reason, the gene annotations can change between version of breseq or runs\n",
    "    (muts_df.exp != \"GLU\")  # This is the ALE that originated the corA mutation that is found in all CCK samples, therefore should track\n",
    "    & (muts_df.Position == 2132787)\n",
    "    & (muts_df[\"Details\"] == \"I204S (ATC→AGC)\"))]\n",
    "\n",
    "\n",
    "p = {4293212,\n",
    " 4293403,\n",
    " 4293430,\n",
    " 4296060,\n",
    " 4296154,\n",
    " 4296189,\n",
    " 4296190,\n",
    " 4296191,\n",
    " 4296267,\n",
    " 4296268,\n",
    " 4296286,\n",
    " 4296380,\n",
    " 4296382}\n",
    "ch = {'+C', '+G', 'A→C', 'A→G', 'A→T', 'C→A', 'C→T', 'T→C'}\n",
    "muts_df = muts_df[~(\n",
    "#     (muts_df.Gene == \"gltP, yjcO\")\n",
    "    (muts_df[\"Position\"].isin(p))\n",
    "    & (muts_df[\"Sequence Change\"].isin(ch))\n",
    "       )]\n",
    "\n",
    "display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove experiment specific starting strain mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the following mutations out if they come from specific experiments.\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"cyoB\") & (muts_df[\"Details\"]==\"W190* (TGG→TAG)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"eno\") & (muts_df[\"Details\"]==\"L193Q (CTG→CAG)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"chbF\") & (muts_df[\"Details\"]==\"V133E (GTA→GAA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"fumC\") & (muts_df[\"Details\"]==\"G354G (GGT→GGC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"yegE\") & (muts_df[\"Details\"]==\"S683Y (TCC→TAC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"SER\") & (muts_df[\"Gene\"]==\"rcsC\") & (muts_df[\"Details\"]==\"L594F (TTG→TTC)\")))]\n",
    "\n",
    "\n",
    "muts_df = muts_df[(~((muts_df.exp==\"GYD\") & (muts_df[\"Gene\"]==\"fumC\") & (muts_df[\"Details\"]==\"G354G (GGT→GGC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"GYD\") & (muts_df[\"Gene\"]==\"yegE\") & (muts_df[\"Details\"]==\"S683Y (TCC→TAC)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"GYD\") & (muts_df[\"Gene\"]==\"rcsC\") & (muts_df[\"Details\"]==\"L594F (TTG→TTC)\")))]\n",
    "\n",
    "\n",
    "cck_exp_list = [\n",
    "    \"Central_carbon_knockout_tpiA\",\n",
    "    \"Central_carbon_knockout_PTS\",\n",
    "    \"Central_carbon_knockout_pgi\",\n",
    "    \"Central_carbon_knockout_sdh\",\n",
    "    \"Central_carbon_knockout_gnd\",\n",
    "    \"Central_carbon_knockout_glucose_evolution\"\n",
    "]\n",
    "\n",
    "muts_df = muts_df[~(\n",
    "    (muts_df.exp.isin(cck_exp_list))\n",
    "    & (\n",
    "        (muts_df[\"Gene\"]==\"wcaA\")\n",
    "        | (muts_df[\"Gene\"]==\"corA\")\n",
    "    )\n",
    ")]  # The details of the mutations to filter out should probably include the sequence change."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Removing MG1655_anchestor mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to use with filtering out mutations \n",
    "muts_df = muts_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16349</th>\n",
       "      <td>MG1655_anchestor</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4390754</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>A252A (GCC→GCA)</td>\n",
       "      <td>rsgA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608306.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16350</th>\n",
       "      <td>MG1655_anchestor</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>803662</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>L54I (CTT→ATT)</td>\n",
       "      <td>ybhJ</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608301.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16351</th>\n",
       "      <td>MG1655_anchestor</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1905761</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→A</td>\n",
       "      <td>G25D (GGT→GAT)</td>\n",
       "      <td>mntP</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608303.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "16349  MG1655_anchestor    0      0        1         1       1.0   4390754   \n",
       "16350  MG1655_anchestor    0      0        1         1       1.0    803662   \n",
       "16351  MG1655_anchestor    0      0        1         1       1.0   1905761   \n",
       "\n",
       "      Mutation Type Sequence Change          Details  Gene Reference Seq  \\\n",
       "16349           SNP             G→T  A252A (GCC→GCA)  rsgA     NC_000913   \n",
       "16350           SNP             C→A   L54I (CTT→ATT)  ybhJ     NC_000913   \n",
       "16351           SNP             G→A   G25D (GGT→GAT)  mntP     NC_000913   \n",
       "\n",
       "          Mut ID  \n",
       "16349  1608306.0  \n",
       "16350  1608301.0  \n",
       "16351  1608303.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Not many. Going to manual remove (see cell below)\n",
    "# This is for the Ecoli_resistance_caz\n",
    "# Sebastian often uses the name \"MG1655_anchestor\" \n",
    "MG1655_anchestor = muts_df[muts_df.exp==\"MG1655_anchestor\"].copy()\n",
    "MG1655_anchestor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17305\n",
      "17302\n"
     ]
    }
   ],
   "source": [
    "print(len(muts_df))\n",
    "muts_df = muts_df[~(muts_df.index.isin(list(MG1655_anchestor.index)))]\n",
    "print(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17245\n"
     ]
    }
   ],
   "source": [
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_chloramphenicol\") & (muts_df[\"Position\"]==4390754) & (muts_df[\"Details\"]==\"A252A (GCC→GCA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_chloramphenicol\") & (muts_df[\"Position\"]==803662) & (muts_df[\"Details\"]==\"L54I (CTT→ATT)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_chloramphenicol\") & (muts_df[\"Position\"]==1905761) & (muts_df[\"Details\"]==\"G25D (GGT→GAT)\")))]\n",
    "\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_doxycycline\") & (muts_df[\"Position\"]==4390754) & (muts_df[\"Details\"]==\"A252A (GCC→GCA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_doxycycline\") & (muts_df[\"Position\"]==803662) & (muts_df[\"Details\"]==\"L54I (CTT→ATT)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_doxycycline\") & (muts_df[\"Position\"]==1905761) & (muts_df[\"Details\"]==\"G25D (GGT→GAT)\")))]\n",
    "\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_trimethoprim\") & (muts_df[\"Position\"]==4390754) & (muts_df[\"Details\"]==\"A252A (GCC→GCA)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_trimethoprim\") & (muts_df[\"Position\"]==803662) & (muts_df[\"Details\"]==\"L54I (CTT→ATT)\")))]\n",
    "muts_df = muts_df[(~((muts_df.exp==\"MG1655_trimethoprim\") & (muts_df[\"Position\"]==1905761) & (muts_df[\"Details\"]==\"G25D (GGT→GAT)\")))]\n",
    "print(len(muts_df))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Though the other Ecoli_resistance_caz experiments have \"anchestor\" data on the aledb.ucsd.edu server, their mutations aren't uploaded, therefore not going to filter anything from them at this point."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Starting strain mutations according to samples within experiments with starting strain AFIRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# to use with filtering out mutations \n",
    "muts_df = muts_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1397381</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ13,756 bp</td>\n",
       "      <td></td>\n",
       "      <td>[ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608368.0</td>\n",
       "      <td>1 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.61</td>\n",
       "      <td>4542237</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (201‑209/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608475.0</td>\n",
       "      <td>1 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1397381</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ13,756 bp</td>\n",
       "      <td></td>\n",
       "      <td>[ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608368.0</td>\n",
       "      <td>2 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>4542456</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS1 (+) +9 bp</td>\n",
       "      <td>coding (420‑428/597 nt)</td>\n",
       "      <td>fimE</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608472.0</td>\n",
       "      <td>2 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MG1655_untreated</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.21</td>\n",
       "      <td>4542831</td>\n",
       "      <td>MOB</td>\n",
       "      <td>IS5 (+) +4 bp</td>\n",
       "      <td>intergenic (+198/‑281)</td>\n",
       "      <td>fimE, fimA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1608473.0</td>\n",
       "      <td>2 16 0 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17240</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3951735</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ326 bp</td>\n",
       "      <td>pseudogene (194‑519/663 nt)</td>\n",
       "      <td>ilvG</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1114004.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17241</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3700725</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→T</td>\n",
       "      <td>I55F (ATC→TTC)</td>\n",
       "      <td>yhjV</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1566613.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17242</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3968459</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>Q182* (CAA→TAA)</td>\n",
       "      <td>wecA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1566614.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17243</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>4184543</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>P1100Q (CCG→CAG)</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1112480.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17244</th>\n",
       "      <td>gnd</td>\n",
       "      <td>19</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td></td>\n",
       "      <td>[rph],[rph]</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>1102506.0</td>\n",
       "      <td>19 171 2 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17245 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "0      MG1655_untreated    1     16        0         1      1.00   1397381   \n",
       "1      MG1655_untreated    1     16        0         1      0.61   4542237   \n",
       "2      MG1655_untreated    2     16        0         1      1.00   1397381   \n",
       "3      MG1655_untreated    2     16        0         1      0.22   4542456   \n",
       "4      MG1655_untreated    2     16        0         1      0.21   4542831   \n",
       "...                 ...  ...    ...      ...       ...       ...       ...   \n",
       "17240               gnd   19    171        2         1      1.00   3951735   \n",
       "17241               gnd   19    171        2         1      1.00   3700725   \n",
       "17242               gnd   19    171        2         1      1.00   3968459   \n",
       "17243               gnd   19    171        2         1      1.00   4184543   \n",
       "17244               gnd   19    171        2         1      1.00   3815859   \n",
       "\n",
       "      Mutation Type Sequence Change                      Details  \\\n",
       "0               DEL      Δ13,756 bp                                \n",
       "1               MOB   IS1 (+) +9 bp      coding (201‑209/597 nt)   \n",
       "2               DEL      Δ13,756 bp                                \n",
       "3               MOB   IS1 (+) +9 bp      coding (420‑428/597 nt)   \n",
       "4               MOB   IS5 (+) +4 bp       intergenic (+198/‑281)   \n",
       "...             ...             ...                          ...   \n",
       "17240           DEL         Δ326 bp  pseudogene (194‑519/663 nt)   \n",
       "17241           SNP             A→T               I55F (ATC→TTC)   \n",
       "17242           SNP             C→T              Q182* (CAA→TAA)   \n",
       "17243           SNP             C→A             P1100Q (CCG→CAG)   \n",
       "17244           DEL          Δ82 bp                                \n",
       "\n",
       "                                                    Gene Reference Seq  \\\n",
       "0      [ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...     NC_000913   \n",
       "1                                                   fimE     NC_000913   \n",
       "2      [ynaJ],uspE,fnr,ogt,abgT,abgB,abgA,abgR,mcaS,s...     NC_000913   \n",
       "3                                                   fimE     NC_000913   \n",
       "4                                             fimE, fimA     NC_000913   \n",
       "...                                                  ...           ...   \n",
       "17240                                               ilvG     NC_000913   \n",
       "17241                                               yhjV     NC_000913   \n",
       "17242                                               wecA     NC_000913   \n",
       "17243                                               rpoB     NC_000913   \n",
       "17244                                        [rph],[rph]     NC_000913   \n",
       "\n",
       "          Mut ID      sample  \n",
       "0      1608368.0    1 16 0 1  \n",
       "1      1608475.0    1 16 0 1  \n",
       "2      1608368.0    2 16 0 1  \n",
       "3      1608472.0    2 16 0 1  \n",
       "4      1608473.0    2 16 0 1  \n",
       "...          ...         ...  \n",
       "17240  1114004.0  19 171 2 1  \n",
       "17241  1566613.0  19 171 2 1  \n",
       "17242  1566614.0  19 171 2 1  \n",
       "17243  1112480.0  19 171 2 1  \n",
       "17244  1102506.0  19 171 2 1  \n",
       "\n",
       "[17245 rows x 14 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_sample_name(ale, flask, isolate, tech_rep):\n",
    "    sample_name = str(int(ale)) + \" \" + str(int(flask)) + \" \" + str(int(isolate)) + \" \" + str(int(tech_rep))\n",
    "    return sample_name\n",
    "\n",
    "muts_df[\"sample\"] = muts_df.apply(lambda r: get_sample_name(r.ale, r.flask, r.isolate, r.tech_rep), axis=1)\n",
    "muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove starting strain mutations from all experiments according to each experiment's starting strain\n",
    "muts_df = muts_df.copy()\n",
    "starting_strains = muts_df[muts_df[\"sample\"]==\"0 0 1 1\"]\n",
    "muts_df = muts_df[muts_df[\"sample\"]!=\"0 0 1 1\"]  # remove starting strain samples\n",
    "for e, exp_starting_strain_muts in starting_strains.groupby(\"exp\"):\n",
    "    rows_to_filter = (\n",
    "        muts_df[\"exp\"].isin(exp_starting_strain_muts[\"exp\"])\n",
    "        & muts_df[\"Position\"].isin(exp_starting_strain_muts[\"Position\"])\n",
    "        & muts_df[\"Sequence Change\"].isin(exp_starting_strain_muts[\"Sequence Change\"])\n",
    "        & muts_df[\"Details\"].isin(exp_starting_strain_muts[\"Details\"])\n",
    "    )\n",
    "    muts_df.drop(muts_df[rows_to_filter].index, inplace=True)  # have to ensure that muts_df indeces are unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Check that no weird mutations references are included in dataset, since they won't match to RegulonDB's data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CP009273 ['TOL_isobutyric_acid']\n",
      "NC_000913 ['MG1655_untreated' 'Iron-ALE'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2'\n",
      " 'gallium-nitrate' 'TOL_adipic_acid' 'MG1655-blaTEM1_untreated' 'tpiKO'\n",
      " 'TOL_n-butanol' 'pvdJ_KO_P_aeruginosa_spent_medium' 'TOL_octanoic_acid'\n",
      " 'TOL_glutaric_acid'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2' 'TOL_putrescine'\n",
      " 'WT_P_aeruginosa_spent_medium' 'EG_pCA24N-fucO'\n",
      " 'TOL_hexamethylenediamine' 'MG1655_doxycycline' 'evo'\n",
      " 'JCL260__isobutanol_production_host-M9-MG1655gbk'\n",
      " 'MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2' 'TOL_propanediol'\n",
      " 'TOL_2,3-butanediol' 'sdh' 'GLU' 'MG1655_trimethoprim'\n",
      " 'MG1655-pBGT_ceftazidime' 'TOL_hexanoic_acid' 'wt'\n",
      " 'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk'\n",
      " 'pyocyanin_supplemented_medium' 'MG1655_ceftazidime' 'OxidizeME_PQ3'\n",
      " 'tpi' 'MG1655_chloramphenicol' 'TOL_coumaric_acid' 'TOL_isobutyric_acid'\n",
      " 'ILT_MG1655' 'pts' 'MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2'\n",
      " 'GYD' 'pgi' 'MG1655-pBGT_untreated' 'MG1655-blaTEM1_ceftazidime' 'gnd']\n",
      "NC_000913_3_KO_pgi ['pgiKO']\n",
      "NC_000913_3_bme_pgi ['pgiBME']\n",
      "NC_000913_3_hsa_pgi ['pgiHSA']\n",
      "NC_000913_3_hsa_tpiA ['tpiHSA']\n",
      "NC_000913_3_pae_pgi ['pgiPAE']\n",
      "NC_000913_3_pae_tpiA ['tpiPAE']\n",
      "NC_000913_3_vch_pgi ['pgiVCH']\n",
      "NC_000913_3_vch_tpiA ['tpiVCH']\n"
     ]
    }
   ],
   "source": [
    "for ref_seq, df in muts_df.groupby(\"Reference Seq\"):\n",
    "    print(ref_seq, df.exp.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "According the the prevous results, TOL_isobutyric_acid has one weird mutation. Going to remove it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NC_000913', nan, 'NC_000913_3_pae_tpiA', 'NC_000913_3_vch_tpiA',\n",
       "       'NC_000913_3_bme_pgi', 'NC_000913_3_hsa_tpiA',\n",
       "       'NC_000913_3_hsa_pgi', 'NC_000913_3_vch_pgi',\n",
       "       'NC_000913_3_pae_pgi', 'NC_000913_3_KO_pgi'], dtype=object)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df = muts_df[muts_df[\"Reference Seq\"] != \"CP009273\"]\n",
    "muts_df[\"Reference Seq\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Just get ALE-unique mutations \n",
    "Since the majority of our comparisons have to do with endpoints, just look at endpoints.  \n",
    "If not limiting to endpoint, the mutations used in associated mutations to conditions will need filtering.  \n",
    "This occurs in a later NB of the pipeline.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td></td>\n",
       "      <td>[rph], [rph]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "15  42C    1    124        1         1       1.0    702352           DEL   \n",
       "16  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "17  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "18  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "19  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "   Sequence Change                   Details          Gene Reference Seq  \\\n",
       "15          Δ21 bp  coding (380‑400/1149 nt)          nagA           NaN   \n",
       "16             G→C           R110G (CGT→GGT)          clsA           NaN   \n",
       "17          Δ82 bp                            [rph], [rph]           NaN   \n",
       "18             C→T           A734V (GCG→GTG)          rpoC           NaN   \n",
       "19             A→C             D9A (GAT→GCT)           hfq           NaN   \n",
       "\n",
       "    Mut ID     sample  \n",
       "15     NaN  1 124 1 1  \n",
       "16     NaN  1 124 1 1  \n",
       "17     NaN  1 124 1 1  \n",
       "18     NaN  1 124 1 1  \n",
       "19     NaN  1 124 1 1  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df = get_multi_exp_max_freq_mut_df(muts_df, endpoint_flask_only=False)\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensure all mutations are using NC_000913 positions.\n",
    "genome_pos_transform_exp_l = [\"ERS\",\n",
    "                              'Central_carbon_knockout_PTS',\n",
    "                              'Central_carbon_knockout_glucose_evolution',\n",
    "                              'Central_carbon_knockout_gnd',\n",
    "                              'Central_carbon_knockout_pgi',\n",
    "                              'Central_carbon_knockout_sdh',\n",
    "                              'Central_carbon_knockout_tpiA',]\n",
    "muts_df.Position = muts_df.apply(lambda row: get_K12_pos_from_BOP27(row.Position) if row.exp in genome_pos_transform_exp_l else row.Position, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td></td>\n",
       "      <td>[rph], [rph]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "15  42C    1    124        1         1       1.0    702352           DEL   \n",
       "16  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "17  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "18  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "19  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "   Sequence Change                   Details          Gene Reference Seq  \\\n",
       "15          Δ21 bp  coding (380‑400/1149 nt)          nagA           NaN   \n",
       "16             G→C           R110G (CGT→GGT)          clsA           NaN   \n",
       "17          Δ82 bp                            [rph], [rph]           NaN   \n",
       "18             C→T           A734V (GCG→GTG)          rpoC           NaN   \n",
       "19             A→C             D9A (GAT→GCT)           hfq           NaN   \n",
       "\n",
       "    Mut ID     sample  coding  \n",
       "15     NaN  1 124 1 1    True  \n",
       "16     NaN  1 124 1 1    True  \n",
       "17     NaN  1 124 1 1    True  \n",
       "18     NaN  1 124 1 1    True  \n",
       "19     NaN  1 124 1 1    True  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df[\"coding\"] = muts_df.Details.apply(is_coding_mut)\n",
    "muts_df[\"Gene\"] = muts_df.apply(lambda row: row[\"Gene\"].replace(\", \",\"/\") if not row[\"coding\"] else row[\"Gene\"], axis=1)\n",
    "\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10871\n",
      "10497\n"
     ]
    }
   ],
   "source": [
    "print(len(muts_df))\n",
    "muts_df = muts_df[muts_df.presence >= 0.5]\n",
    "print(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td></td>\n",
       "      <td>rph</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "15  42C    1    124        1         1       1.0    702352           DEL   \n",
       "16  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "17  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "18  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "19  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "   Sequence Change                   Details  Gene Reference Seq  Mut ID  \\\n",
       "15          Δ21 bp  coding (380‑400/1149 nt)  nagA           NaN     NaN   \n",
       "16             G→C           R110G (CGT→GGT)  clsA           NaN     NaN   \n",
       "17          Δ82 bp                             rph           NaN     NaN   \n",
       "18             C→T           A734V (GCG→GTG)  rpoC           NaN     NaN   \n",
       "19             A→C             D9A (GAT→GCT)   hfq           NaN     NaN   \n",
       "\n",
       "       sample  coding  \n",
       "15  1 124 1 1    True  \n",
       "16  1 124 1 1    True  \n",
       "17  1 124 1 1    True  \n",
       "18  1 124 1 1    True  \n",
       "19  1 124 1 1    True  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This work is also currently duplicated NB4 \n",
    "muts_df[\"Gene\"] = muts_df[\"Gene\"].apply(lambda a: \"rph\" if a == \"[rph], [rph]\" else a)\n",
    "muts_df[\"Gene\"] = muts_df[\"Gene\"].apply(lambda a: \"rph\" if a == \"[rph],[rph]\" else a)\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# removing unnecessary column\n",
    "muts_df = muts_df.drop(columns=[\"Mut ID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "388\n",
      "375\n"
     ]
    }
   ],
   "source": [
    "# Amount of ALEs before removal\n",
    "print(len(muts_df.apply(lambda r: r.exp + \" \" + str(r.ale), axis=1).unique()))\n",
    "\n",
    "# removing hypermutators since they will affect associations\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"GLU\")\n",
    "        & ((muts_df.ale == 5) | (muts_df.ale == 7))\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_2,3-butanediol\")\n",
    "        & (muts_df.ale == 3)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_adipic_acid\")\n",
    "        & (muts_df.ale == 5)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_hexamethylenediamine\")\n",
    "        & (muts_df.ale == 6)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_octanoic_acid\")\n",
    "        & (muts_df.ale == 5)\n",
    "    )\n",
    "]\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"TOL_propanediol\")\n",
    "        & ((muts_df.ale == 1) | (muts_df.ale == 2) | (muts_df.ale == 3) | (muts_df.ale == 4) | (muts_df.ale == 5) | (muts_df.ale == 7) | (muts_df.ale == 8))\n",
    "    )\n",
    "]\n",
    "\n",
    "print(len(muts_df.apply(lambda r: r.exp + \" \" + str(r.ale), axis=1).unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "373\n"
     ]
    }
   ],
   "source": [
    "# removing samples not incuded in GYD MS for AVA MS\n",
    "\n",
    "muts_df = muts_df[\n",
    "    ~(\n",
    "        (muts_df.exp == \"GYD\")\n",
    "        & ((muts_df.ale == 35) | (muts_df.ale == 34))\n",
    "    )\n",
    "]\n",
    "print(len(muts_df.apply(lambda r: r.exp + \" \" + str(r.ale), axis=1).unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "muts_df.to_pickle(\"./data/1_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "63"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(len(muts_df.exp.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42C\n",
      "C13\n",
      "EG_pCA24N-fucO\n",
      "GLU\n",
      "GYD\n",
      "ILT_MG1655\n",
      "Iron-ALE\n",
      "JCL260__isobutanol_production_host-M9-MG1655gbk\n",
      "MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2\n",
      "MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2\n",
      "MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2\n",
      "MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2\n",
      "MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2\n",
      "MG1655-blaTEM1_ceftazidime\n",
      "MG1655-blaTEM1_untreated\n",
      "MG1655-pBGT_ceftazidime\n",
      "MG1655-pBGT_untreated\n",
      "MG1655_ceftazidime\n",
      "MG1655_chloramphenicol\n",
      "MG1655_doxycycline\n",
      "MG1655_trimethoprim\n",
      "MG1655_untreated\n",
      "OxidizeME_PQ3\n",
      "PGI\n",
      "SER\n",
      "SSW_AC\n",
      "SSW_GLU_AC\n",
      "SSW_GLU_GLY\n",
      "SSW_GLU_XYL\n",
      "SSW_GLY\n",
      "SSW_XYL\n",
      "TOL_2,3-butanediol\n",
      "TOL_adipic_acid\n",
      "TOL_coumaric_acid\n",
      "TOL_glutaric_acid\n",
      "TOL_hexamethylenediamine\n",
      "TOL_hexanoic_acid\n",
      "TOL_isobutyric_acid\n",
      "TOL_n-butanol\n",
      "TOL_octanoic_acid\n",
      "TOL_propanediol\n",
      "TOL_putrescine\n",
      "WT_P_aeruginosa_spent_medium\n",
      "application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk\n",
      "evo\n",
      "gallium-nitrate\n",
      "gnd\n",
      "pgi\n",
      "pgiBME\n",
      "pgiHSA\n",
      "pgiKO\n",
      "pgiPAE\n",
      "pgiVCH\n",
      "pts\n",
      "pvdJ_KO_P_aeruginosa_spent_medium\n",
      "pyocyanin_supplemented_medium\n",
      "sdh\n",
      "tpi\n",
      "tpiHSA\n",
      "tpiKO\n",
      "tpiPAE\n",
      "tpiVCH\n",
      "wt\n"
     ]
    }
   ],
   "source": [
    "for e in muts_df.exp.unique():\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "373"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "array(['42C 1', '42C 2', '42C 3', '42C 4', '42C 5', '42C 6', '42C 7',\n",
       "       '42C 8', '42C 9', '42C 10', 'C13 1', 'C13 2', 'C13 3', 'C13 4',\n",
       "       'C13 5', 'C13 6', 'EG_pCA24N-fucO 1', 'GLU 3', 'GLU 4', 'GLU 6',\n",
       "       'GLU 8', 'GLU 9', 'GLU 10', 'GYD 1', 'GYD 2', 'GYD 3', 'GYD 4',\n",
       "       'GYD 5', 'GYD 6', 'GYD 7', 'GYD 8', 'GYD 9', 'GYD 10', 'GYD 11',\n",
       "       'GYD 12', 'GYD 13', 'GYD 14', 'GYD 15', 'GYD 16', 'GYD 17',\n",
       "       'GYD 18', 'GYD 19', 'GYD 20', 'GYD 21', 'GYD 22', 'GYD 23',\n",
       "       'GYD 24', 'GYD 25', 'GYD 26', 'GYD 27', 'GYD 28', 'GYD 29',\n",
       "       'GYD 30', 'ILT_MG1655 1', 'ILT_MG1655 2', 'ILT_MG1655 3',\n",
       "       'ILT_MG1655 4', 'ILT_MG1655 9', 'ILT_MG1655 10', 'ILT_MG1655 11',\n",
       "       'ILT_MG1655 12', 'Iron-ALE 17', 'Iron-ALE 18', 'Iron-ALE 19',\n",
       "       'Iron-ALE 20', 'JCL260__isobutanol_production_host-M9-MG1655gbk 1',\n",
       "       'JCL260__isobutanol_production_host-M9-MG1655gbk 2',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-D-2-deoxyribose2 1',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2 1',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-D-arabinose2 2',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2 1',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-D-lyxose2 2',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2 1',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-m-tartrate2 2',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2 1',\n",
       "       'MG1655-M9-NC_000913_3gb-stationary-37-monomethyl_succinate2 2',\n",
       "       'MG1655-blaTEM1_ceftazidime 1', 'MG1655-blaTEM1_ceftazidime 2',\n",
       "       'MG1655-blaTEM1_ceftazidime 3', 'MG1655-blaTEM1_ceftazidime 4',\n",
       "       'MG1655-blaTEM1_ceftazidime 5', 'MG1655-blaTEM1_untreated 1',\n",
       "       'MG1655-blaTEM1_untreated 2', 'MG1655-blaTEM1_untreated 3',\n",
       "       'MG1655-blaTEM1_untreated 4', 'MG1655-blaTEM1_untreated 5',\n",
       "       'MG1655-blaTEM1_untreated 6', 'MG1655-pBGT_ceftazidime 1',\n",
       "       'MG1655-pBGT_ceftazidime 2', 'MG1655-pBGT_ceftazidime 3',\n",
       "       'MG1655-pBGT_ceftazidime 4', 'MG1655-pBGT_ceftazidime 5',\n",
       "       'MG1655-pBGT_ceftazidime 6', 'MG1655-pBGT_ceftazidime 7',\n",
       "       'MG1655-pBGT_ceftazidime 8', 'MG1655-pBGT_ceftazidime 9',\n",
       "       'MG1655-pBGT_ceftazidime 10', 'MG1655-pBGT_ceftazidime 11',\n",
       "       'MG1655-pBGT_untreated 1', 'MG1655-pBGT_untreated 2',\n",
       "       'MG1655-pBGT_untreated 3', 'MG1655-pBGT_untreated 4',\n",
       "       'MG1655-pBGT_untreated 5', 'MG1655-pBGT_untreated 6',\n",
       "       'MG1655_ceftazidime 1', 'MG1655_ceftazidime 2',\n",
       "       'MG1655_ceftazidime 3', 'MG1655_ceftazidime 4',\n",
       "       'MG1655_ceftazidime 5', 'MG1655_chloramphenicol 1',\n",
       "       'MG1655_chloramphenicol 2', 'MG1655_chloramphenicol 3',\n",
       "       'MG1655_chloramphenicol 4', 'MG1655_chloramphenicol 5',\n",
       "       'MG1655_doxycycline 1', 'MG1655_doxycycline 2',\n",
       "       'MG1655_doxycycline 3', 'MG1655_doxycycline 4',\n",
       "       'MG1655_doxycycline 5', 'MG1655_trimethoprim 1',\n",
       "       'MG1655_trimethoprim 2', 'MG1655_trimethoprim 3',\n",
       "       'MG1655_trimethoprim 4', 'MG1655_trimethoprim 5',\n",
       "       'MG1655_untreated 1', 'MG1655_untreated 2', 'MG1655_untreated 3',\n",
       "       'MG1655_untreated 4', 'MG1655_untreated 5', 'MG1655_untreated 6',\n",
       "       'OxidizeME_PQ3 18', 'PGI 1', 'PGI 2', 'PGI 3', 'PGI 4', 'PGI 5',\n",
       "       'PGI 6', 'PGI 7', 'PGI 8', 'PGI 9', 'PGI 10', 'SER 3', 'SER 4',\n",
       "       'SER 5', 'SSW_AC 1', 'SSW_AC 2', 'SSW_AC 3', 'SSW_AC 4',\n",
       "       'SSW_GLU_AC 7', 'SSW_GLU_AC 8', 'SSW_GLU_AC 9', 'SSW_GLU_GLY 4',\n",
       "       'SSW_GLU_GLY 5', 'SSW_GLU_GLY 6', 'SSW_GLU_XYL 1', 'SSW_GLU_XYL 2',\n",
       "       'SSW_GLU_XYL 3', 'SSW_GLY 1', 'SSW_GLY 2', 'SSW_XYL 1',\n",
       "       'SSW_XYL 2', 'SSW_XYL 3', 'SSW_XYL 4', 'TOL_2,3-butanediol 1',\n",
       "       'TOL_2,3-butanediol 2', 'TOL_2,3-butanediol 4',\n",
       "       'TOL_2,3-butanediol 5', 'TOL_2,3-butanediol 6',\n",
       "       'TOL_2,3-butanediol 7', 'TOL_2,3-butanediol 8',\n",
       "       'TOL_adipic_acid 1', 'TOL_adipic_acid 2', 'TOL_adipic_acid 3',\n",
       "       'TOL_adipic_acid 4', 'TOL_adipic_acid 6', 'TOL_adipic_acid 7',\n",
       "       'TOL_adipic_acid 8', 'TOL_coumaric_acid 1', 'TOL_coumaric_acid 2',\n",
       "       'TOL_coumaric_acid 3', 'TOL_coumaric_acid 4',\n",
       "       'TOL_coumaric_acid 5', 'TOL_coumaric_acid 6',\n",
       "       'TOL_coumaric_acid 7', 'TOL_coumaric_acid 8',\n",
       "       'TOL_glutaric_acid 1', 'TOL_glutaric_acid 2',\n",
       "       'TOL_glutaric_acid 3', 'TOL_glutaric_acid 4',\n",
       "       'TOL_glutaric_acid 5', 'TOL_glutaric_acid 6',\n",
       "       'TOL_glutaric_acid 7', 'TOL_glutaric_acid 8',\n",
       "       'TOL_hexamethylenediamine 1', 'TOL_hexamethylenediamine 2',\n",
       "       'TOL_hexamethylenediamine 3', 'TOL_hexamethylenediamine 4',\n",
       "       'TOL_hexamethylenediamine 5', 'TOL_hexamethylenediamine 7',\n",
       "       'TOL_hexamethylenediamine 8', 'TOL_hexanoic_acid 1',\n",
       "       'TOL_hexanoic_acid 2', 'TOL_hexanoic_acid 3',\n",
       "       'TOL_hexanoic_acid 4', 'TOL_hexanoic_acid 5',\n",
       "       'TOL_hexanoic_acid 6', 'TOL_hexanoic_acid 7',\n",
       "       'TOL_isobutyric_acid 1', 'TOL_isobutyric_acid 2',\n",
       "       'TOL_isobutyric_acid 3', 'TOL_isobutyric_acid 4',\n",
       "       'TOL_isobutyric_acid 5', 'TOL_isobutyric_acid 6',\n",
       "       'TOL_isobutyric_acid 7', 'TOL_isobutyric_acid 8',\n",
       "       'TOL_n-butanol 1', 'TOL_n-butanol 2', 'TOL_n-butanol 3',\n",
       "       'TOL_n-butanol 4', 'TOL_n-butanol 5', 'TOL_n-butanol 6',\n",
       "       'TOL_n-butanol 7', 'TOL_n-butanol 8', 'TOL_octanoic_acid 1',\n",
       "       'TOL_octanoic_acid 2', 'TOL_octanoic_acid 3',\n",
       "       'TOL_octanoic_acid 4', 'TOL_octanoic_acid 6',\n",
       "       'TOL_octanoic_acid 7', 'TOL_propanediol 6', 'TOL_putrescine 1',\n",
       "       'TOL_putrescine 2', 'TOL_putrescine 3', 'TOL_putrescine 4',\n",
       "       'TOL_putrescine 5', 'TOL_putrescine 6', 'TOL_putrescine 7',\n",
       "       'WT_P_aeruginosa_spent_medium 2', 'WT_P_aeruginosa_spent_medium 3',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 1',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 2',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 3',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 4',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 5',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 6',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 7',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 8',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 9',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 10',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 11',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 12',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 13',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 14',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 15',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 16',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 17',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 18',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 19',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 20',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 21',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 22',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 23',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 24',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 25',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 26',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 27',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 28',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 29',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 30',\n",
       "       'application_of_ALE_to_re-optimize_growth_performance_of_genome-reduced_strain_MS56-M9-MG1655gbk 31',\n",
       "       'evo 23', 'evo 24', 'gallium-nitrate 1', 'gallium-nitrate 2',\n",
       "       'gallium-nitrate 3', 'gallium-nitrate 4', 'gallium-nitrate 5',\n",
       "       'gnd 17', 'gnd 18', 'gnd 19', 'pgi 1', 'pgi 2', 'pgi 3', 'pgi 4',\n",
       "       'pgi 5', 'pgi 6', 'pgi 7', 'pgi 8', 'pgi 9', 'pgi 10', 'pgi 11',\n",
       "       'pgi 12', 'pgi 13', 'pgi 14', 'pgi 15', 'pgi 16', 'pgiBME 13',\n",
       "       'pgiBME 14', 'pgiBME 16', 'pgiBME 18', 'pgiBME 19', 'pgiBME 20',\n",
       "       'pgiBME 21', 'pgiBME 22', 'pgiBME 24', 'pgiHSA 12', 'pgiHSA 14',\n",
       "       'pgiHSA 16', 'pgiHSA 17', 'pgiHSA 19', 'pgiHSA 20', 'pgiKO 3',\n",
       "       'pgiKO 4', 'pgiKO 19', 'pgiKO 20', 'pgiPAE 9', 'pgiPAE 10',\n",
       "       'pgiPAE 11', 'pgiPAE 12', 'pgiPAE 24', 'pgiPAE 25', 'pgiPAE 29',\n",
       "       'pgiVCH 30', 'pgiVCH 31', 'pgiVCH 32', 'pts 9', 'pts 10', 'pts 11',\n",
       "       'pts 12', 'pvdJ_KO_P_aeruginosa_spent_medium 1',\n",
       "       'pvdJ_KO_P_aeruginosa_spent_medium 2',\n",
       "       'pvdJ_KO_P_aeruginosa_spent_medium 3',\n",
       "       'pyocyanin_supplemented_medium 1',\n",
       "       'pyocyanin_supplemented_medium 2',\n",
       "       'pyocyanin_supplemented_medium 3', 'sdh 20', 'sdh 22', 'sdh 23',\n",
       "       'tpi 13', 'tpi 14', 'tpi 15', 'tpi 16', 'tpiHSA 1', 'tpiHSA 2',\n",
       "       'tpiHSA 3', 'tpiHSA 4', 'tpiHSA 5', 'tpiHSA 6', 'tpiHSA 7',\n",
       "       'tpiHSA 8', 'tpiHSA 9', 'tpiHSA 10', 'tpiKO 1', 'tpiKO 3',\n",
       "       'tpiKO 4', 'tpiKO 7', 'tpiKO 9', 'tpiPAE 13', 'tpiPAE 14',\n",
       "       'tpiPAE 15', 'tpiPAE 16', 'tpiPAE 17', 'tpiPAE 18', 'tpiPAE 21',\n",
       "       'tpiPAE 22', 'tpiPAE 23', 'tpiPAE 24', 'tpiVCH 25', 'tpiVCH 26',\n",
       "       'tpiVCH 27', 'tpiVCH 28', 'wt 1', 'wt 2'], dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df[\"exp ALE\"] = muts_df.apply(lambda r: r.exp + ' ' + str(r.ale), axis=1)\n",
    "display(len(muts_df[\"exp ALE\"].unique()), muts_df[\"exp ALE\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7042"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(muts_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
