{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting the mapping between genetic and genomic features is done separately from annotating the genetic features since the annotation of the genetic features has to happen before genomic features, though we need the genomic features to get the mapping between genomic and genetic. This work therefore comes after getting the genomic and genetic features. In summary, the following describes the necessary order:  \n",
    "1) genetic features  \n",
    "2) genomic features  \n",
    "3) mapping between genomic and genetic features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:21:41.675909Z",
     "iopub.status.busy": "2021-01-01T04:21:41.673380Z",
     "iopub.status.idle": "2021-01-01T04:21:42.197236Z",
     "shell.execute_reply": "2021-01-01T04:21:42.197533Z"
    }
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import copy\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.gene import get_coding_genetic_target_len_d, get_intergenic_len_d\n",
    "from util.genome import get_feature_hit_set\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:21:42.200461Z",
     "iopub.status.busy": "2021-01-01T04:21:42.199894Z",
     "iopub.status.idle": "2021-01-01T04:21:42.299923Z",
     "shell.execute_reply": "2021-01-01T04:21:42.299548Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5676, 27)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td></td>\n",
       "      <td>rph</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187550, 4187550)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4400313, 4400313)</td>\n",
       "      <td>{ECK120000431}</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change                   Details mutation target annotation  \\\n",
       "0          Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "1             G→C           R110G (CGT→GGT)                       clsA   \n",
       "2          Δ82 bp                                                  rph   \n",
       "3             C→T           A734V (GCG→GTG)                       rpoC   \n",
       "4             A→C             D9A (GAT→GCT)                        hfq   \n",
       "\n",
       "  Reference Seq     sample  coding               range gene RegulonDB ID  \\\n",
       "0           NaN  1 124 1 1    True    (702352, 702372)    {ECK120000625}   \n",
       "1           NaN  1 124 1 1    True  (1308318, 1308318)    {ECK120001556}   \n",
       "2           NaN  1 124 1 1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "3           NaN  1 124 1 1    True  (4187550, 4187550)    {ECK120000886}   \n",
       "4           NaN  1 124 1 1    True  (4400313, 4400313)    {ECK120000431}   \n",
       "\n",
       "                                    genetic features   oriC  pseudogene TFBS  \\\n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  False       False   {}   \n",
       "1  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  False       False   {}   \n",
       "2  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   {}   \n",
       "3  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   {}   \n",
       "4  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...  False       False   {}   \n",
       "\n",
       "  promoter RBS attenuator terminator terminator  genetic  \\\n",
       "0       {}  {}                    {}         {}     True   \n",
       "1       {}  {}                    {}         {}     True   \n",
       "2       {}  {}        {ECK125144791}         {}     True   \n",
       "3       {}  {}                    {}         {}     True   \n",
       "4       {}  {}                    {}         {}     True   \n",
       "\n",
       "                                    genomic features  \n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  \n",
       "1  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  \n",
       "2  [{'name': 'rph-pyrE attenuator terminator', 'R...  \n",
       "3  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  \n",
       "4  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/2_2_df.pkl\")\n",
    "display(all_muts_df.shape, all_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:21:42.302040Z",
     "iopub.status.busy": "2021-01-01T04:21:42.301689Z",
     "iopub.status.idle": "2021-01-01T04:21:42.303546Z",
     "shell.execute_reply": "2021-01-01T04:21:42.303163Z"
    }
   },
   "outputs": [],
   "source": [
    "# # # DEBUG\n",
    "# all_muts_df = all_muts_df[all_muts_df.exp==\"SSW_GLU_XYL\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:21:42.307126Z",
     "iopub.status.busy": "2021-01-01T04:21:42.306576Z",
     "iopub.status.idle": "2021-01-01T04:22:04.218419Z",
     "shell.execute_reply": "2021-01-01T04:22:04.218053Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>{'ECK120000625': ['ECK120000625']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>{'ECK120001556': ['ECK120001556']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td></td>\n",
       "      <td>rph</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>A734V (GCG→GTG)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187550, 4187550)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>D9A (GAT→GCT)</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4400313, 4400313)</td>\n",
       "      <td>{ECK120000431}</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "      <td>{'ECK120000431': ['ECK120000431']}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change                   Details mutation target annotation  \\\n",
       "0          Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "1             G→C           R110G (CGT→GGT)                       clsA   \n",
       "2          Δ82 bp                                                  rph   \n",
       "3             C→T           A734V (GCG→GTG)                       rpoC   \n",
       "4             A→C             D9A (GAT→GCT)                        hfq   \n",
       "\n",
       "  Reference Seq     sample  coding               range gene RegulonDB ID  \\\n",
       "0           NaN  1 124 1 1    True    (702352, 702372)    {ECK120000625}   \n",
       "1           NaN  1 124 1 1    True  (1308318, 1308318)    {ECK120001556}   \n",
       "2           NaN  1 124 1 1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "3           NaN  1 124 1 1    True  (4187550, 4187550)    {ECK120000886}   \n",
       "4           NaN  1 124 1 1    True  (4400313, 4400313)    {ECK120000431}   \n",
       "\n",
       "                                    genetic features   oriC  pseudogene TFBS  \\\n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  False       False   {}   \n",
       "1  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  False       False   {}   \n",
       "2  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   {}   \n",
       "3  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   {}   \n",
       "4  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...  False       False   {}   \n",
       "\n",
       "  promoter RBS attenuator terminator terminator  genetic  \\\n",
       "0       {}  {}                    {}         {}     True   \n",
       "1       {}  {}                    {}         {}     True   \n",
       "2       {}  {}        {ECK125144791}         {}     True   \n",
       "3       {}  {}                    {}         {}     True   \n",
       "4       {}  {}                    {}         {}     True   \n",
       "\n",
       "                                    genomic features  \\\n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...   \n",
       "1  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...   \n",
       "2  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "3  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "4  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...   \n",
       "\n",
       "                               genetic feature links  \n",
       "0                 {'ECK120000625': ['ECK120000625']}  \n",
       "1                 {'ECK120001556': ['ECK120001556']}  \n",
       "2  {'ECK120000854': ['ECK125144791', 'ECK12000085...  \n",
       "3                 {'ECK120000886': ['ECK120000886']}  \n",
       "4                 {'ECK120000431': ['ECK120000431']}  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_genetic_links_d(mut_row):\n",
    "    gen_links_d = dict()\n",
    "    # generates a DF of all genomic features for this mutation (row) to use as input for mapping genomic features\n",
    "    # to genetic features according to range.\n",
    "    # This mapping is done by ranges only, unlike other mapping since there is no explicit mapping between\n",
    "    # genomic and genetic features. \n",
    "    geno_feat_df = pd.DataFrame(mut_row[\"genomic features\"])\n",
    "    for gen_feat_d in mut_row[\"genetic features\"]:\n",
    "        o = get_feature_hit_set(gen_feat_d[\"range\"], geno_feat_df, \"range\", \"RegulonDB ID\")\n",
    "        gen_links_d[gen_feat_d[\"RegulonDB ID\"]] = list(o)\n",
    "    return gen_links_d\n",
    "\n",
    "\n",
    "all_muts_df[\"genetic feature links\"] = all_muts_df.apply(get_genetic_links_d, axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:22:04.222265Z",
     "iopub.status.busy": "2021-01-01T04:22:04.221914Z",
     "iopub.status.idle": "2021-01-01T04:22:04.223782Z",
     "shell.execute_reply": "2021-01-01T04:22:04.223409Z"
    }
   },
   "outputs": [],
   "source": [
    "# I think I'm only going to have to implement the remove of a genetic feature function once.\n",
    "def filter_feats(feat_json, feats_to_remove):\n",
    "    output = feat_json.copy()\n",
    "    for f in feats_to_remove:\n",
    "        for d in feat_json:\n",
    "            if d[\"RegulonDB ID\"] == f:\n",
    "                output.remove(d)\n",
    "    return output\n",
    "\n",
    "\n",
    "i = [{\n",
    "    'name': 'g1',\n",
    "    'RegulonDB ID': '1',\n",
    "    'range': (1, 2),\n",
    "    'feature type': 'gene'},\n",
    "    {'name': 'g2',\n",
    "     'RegulonDB ID': '2',\n",
    "     'range': (2, 3),\n",
    "     'feature type': 'gene'}]\n",
    "e = [{'name': 'g1', 'RegulonDB ID': '1', 'range': (1, 2), 'feature type': 'gene'}]\n",
    "o = filter_feats(i, {'2'})\n",
    "assert(o == e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:22:04.227446Z",
     "iopub.status.busy": "2021-01-01T04:22:04.226838Z",
     "iopub.status.idle": "2021-01-01T04:22:04.241494Z",
     "shell.execute_reply": "2021-01-01T04:22:04.241840Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>GENE_POSLEFT</th>\n",
       "      <th>GENE_POSRIGHT</th>\n",
       "      <th>GENE_STRAND</th>\n",
       "      <th>GENE_SEQUENCE</th>\n",
       "      <th>GC_CONTENT</th>\n",
       "      <th>CRI_SCORE</th>\n",
       "      <th>GENE_NOTE</th>\n",
       "      <th>GENE_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>GENE_TYPE</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr</td>\n",
       "      <td>4265782.0</td>\n",
       "      <td>4266861.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...</td>\n",
       "      <td>55.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4265782, 4266861)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>modB</td>\n",
       "      <td>795862.0</td>\n",
       "      <td>796551.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...</td>\n",
       "      <td>54.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(795862, 796551)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120000003</td>\n",
       "      <td>cysZ</td>\n",
       "      <td>2531463.0</td>\n",
       "      <td>2532224.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...</td>\n",
       "      <td>50.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2531463, 2532224)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120000004</td>\n",
       "      <td>dfp</td>\n",
       "      <td>3812731.0</td>\n",
       "      <td>3813951.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...</td>\n",
       "      <td>53.64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3812731, 3813951)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120000005</td>\n",
       "      <td>dcuB</td>\n",
       "      <td>4347404.0</td>\n",
       "      <td>4348744.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...</td>\n",
       "      <td>52.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4347404, 4348744)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GENE_ID GENE_NAME  GENE_POSLEFT  GENE_POSRIGHT GENE_STRAND  \\\n",
       "0  ECK120000001       alr     4265782.0      4266861.0     forward   \n",
       "1  ECK120000002      modB      795862.0       796551.0     forward   \n",
       "2  ECK120000003      cysZ     2531463.0      2532224.0     forward   \n",
       "3  ECK120000004       dfp     3812731.0      3813951.0     forward   \n",
       "4  ECK120000005      dcuB     4347404.0      4348744.0     reverse   \n",
       "\n",
       "                                       GENE_SEQUENCE  GC_CONTENT  CRI_SCORE  \\\n",
       "0  ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...       55.93        NaN   \n",
       "1  ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...       54.06        NaN   \n",
       "2  ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...       50.13        NaN   \n",
       "3  ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...       53.64        NaN   \n",
       "4  ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...       52.27        NaN   \n",
       "\n",
       "  GENE_NOTE  GENE_INTERNAL_COMMENT KEY_ID_ORG GENE_TYPE               range  \n",
       "0       NaN                    NaN      ECK12       NaN  (4265782, 4266861)  \n",
       "1       NaN                    NaN      ECK12       NaN    (795862, 796551)  \n",
       "2       NaN                    NaN      ECK12       NaN  (2531463, 2532224)  \n",
       "3       NaN                    NaN      ECK12       NaN  (3812731, 3813951)  \n",
       "4       NaN                    NaN      ECK12       NaN  (4347404, 4348744)  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = pd.read_pickle(\"./data/gene_df.pkl\")\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:22:04.250615Z",
     "iopub.status.busy": "2021-01-01T04:22:04.250230Z",
     "iopub.status.idle": "2021-01-01T04:22:04.252126Z",
     "shell.execute_reply": "2021-01-01T04:22:04.251742Z"
    }
   },
   "outputs": [],
   "source": [
    "# fix for ASW-1628\n",
    "# remove links to coding regions for non-coding features that overlap in both coding and non-coding regions\n",
    "# assuming that all overlapping features are non-coding.\n",
    "\n",
    "\n",
    "# want to filter out non-coding features from coding links.\n",
    "# The overlap between of a cis-reg feature with an intergenic and coding region is real\n",
    "# though its not the relationship we want to model.\n",
    "def _filter_coding_noncoding_overlapping_links(gen_link_d):\n",
    "    ret_d = copy.deepcopy(gen_link_d)\n",
    "    gen_feats_to_remove = set()\n",
    "\n",
    "    # using these for iteration\n",
    "    coding_links = dict()\n",
    "    noncoding_links = dict()\n",
    "    for k, v in gen_link_d.items():\n",
    "        if '/' in k:\n",
    "            noncoding_links[k] = v\n",
    "        else:\n",
    "            coding_links[k] = v\n",
    "\n",
    "    # changing the given gen_link_d to return\n",
    "    for ncl_geno_feats in noncoding_links.values():\n",
    "        for ncl_geno_feat in ncl_geno_feats:\n",
    "            # There's the potential to filter for overlapping genes here as well\n",
    "            for cl_gen_feat, cl_geno_feats in coding_links.items():\n",
    "                # Check if non-coding feature linked with intergenic region (ncl_geno_feats) overlaps (linked) with coding region.\n",
    "                if ncl_geno_feat not in set(gene_df[\"GENE_ID\"]):  # Ensuring feat to filter is cis-reg, otherwise don't filter\n",
    "                    if ncl_geno_feat in cl_geno_feats:                    \n",
    "                        if len(cl_geno_feats) == 1:\n",
    "                            gen_feats_to_remove.add(cl_gen_feat)\n",
    "                            del ret_d[cl_gen_feat]\n",
    "                        else:\n",
    "                            # remove geno_feat link to the coding gen_feat (ncl_geno_feat == to feat in cl_geno_feats that want to remove)\n",
    "                            if ncl_geno_feat in ret_d[cl_gen_feat]:\n",
    "                                ret_d[cl_gen_feat].remove(ncl_geno_feat)\n",
    "\n",
    "    return ret_d, gen_feats_to_remove\n",
    "\n",
    "\n",
    "i = {'ECK120001080': ['ECK125229290', 'ECK125229291', 'ECK120001080'],\n",
    "     'ECK125229290': ['ECK125229290', 'ECK125229291', 'ECK120001080'],\n",
    "     'ECK125229291': ['ECK125229290', 'ECK125229291', 'ECK120001080'],\n",
    "     'ECK125229291/ECK125256962': ['ECK120001080']}  # This should be removed since ECK120001080 == crl gene, though will be done in a filtering step below\n",
    "e = {'ECK120001080': ['ECK125229290', 'ECK125229291', 'ECK120001080'],\n",
    "     'ECK125229290': ['ECK125229290', 'ECK125229291', 'ECK120001080'],\n",
    "     'ECK125229291': ['ECK125229290', 'ECK125229291', 'ECK120001080'],\n",
    "     'ECK125229291/ECK125256962': ['ECK120001080']}\n",
    "o, _ = _filter_coding_noncoding_overlapping_links(i)\n",
    "assert(o == e)\n",
    "\n",
    "i = {'ECK120000854': ['ECK120000854', 'ECK125144791'],\n",
    "     'ECK120000799/ECK120000854': ['ECK125144791']}\n",
    "e = {'ECK120000854': ['ECK120000854'],\n",
    "     'ECK120000799/ECK120000854': ['ECK125144791']}\n",
    "o, _ = _filter_coding_noncoding_overlapping_links(i)\n",
    "assert(o == e)\n",
    "\n",
    "# Remove the coding region entry if the only item linking is removed\n",
    "i = {'ECK120000854': ['ECK125144791'],\n",
    "     'ECK120000799/ECK120000854': ['ECK125144791']}\n",
    "e = {'ECK120000799/ECK120000854': ['ECK125144791']}\n",
    "o, _ = _filter_coding_noncoding_overlapping_links(i)\n",
    "assert(o == e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:22:04.276832Z",
     "iopub.status.busy": "2021-01-01T04:22:04.257934Z",
     "iopub.status.idle": "2021-01-01T04:22:53.846324Z",
     "shell.execute_reply": "2021-01-01T04:22:53.845929Z"
    }
   },
   "outputs": [],
   "source": [
    "def _f(r):\n",
    "    link_d, gen_feats_to_remove = _filter_coding_noncoding_overlapping_links(r[\"genetic feature links\"])\n",
    "    gen_feats = filter_feats(r[\"genetic features\"], gen_feats_to_remove)\n",
    "    return pd.Series([gen_feats, link_d])\n",
    "\n",
    "\n",
    "# replicates the necessary parts of a dataframe row\n",
    "# have to use RegulonDB IDs in testing for the gene_df condition in _filter_overlapping_genes_link above to work.\n",
    "i = {\n",
    "    \"genetic features\": [{'name': 'g1',\n",
    "                          'RegulonDB ID': '1'},\n",
    "                         {'name': 'g1/g2',\n",
    "                          'RegulonDB ID': '1/2'},  # Intergenic region\n",
    "                         {'name': 'f3',\n",
    "                          'RegulonDB ID': '3'},\n",
    "                         ],\n",
    "    \"genetic feature links\": {'1': ['3'],\n",
    "                              '1/2': ['3']}\n",
    "}\n",
    "e = pd.Series([\n",
    "    [{'name': 'g1/g2', 'RegulonDB ID': '1/2'}, {'name': 'f3', 'RegulonDB ID': '3'}],\n",
    "    {'1/2': ['3']}\n",
    "])\n",
    "o = _f(i)\n",
    "assert(o.equals(e))\n",
    "\n",
    "\n",
    "all_muts_df[[\"genetic features\", \"genetic feature links\"]] = all_muts_df.apply(_f, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:22:53.855086Z",
     "iopub.status.busy": "2021-01-01T04:22:53.854700Z",
     "iopub.status.idle": "2021-01-01T04:23:04.419325Z",
     "shell.execute_reply": "2021-01-01T04:23:04.419653Z"
    }
   },
   "outputs": [],
   "source": [
    "# fix for ASW-1628\n",
    "\n",
    "# 1) Check if genomic feature is a gene\n",
    "# 2) If genetic feature it links to isn't the same gene, remove the link\n",
    "# 2.1) If the length of links is one, just remove the linked genetic feature altogether like I had with overlapping non-coding features.\n",
    "\n",
    "\n",
    "# filters out links due to overlapping genes\n",
    "# and when genes for some reason get linked with integenic regions\n",
    "def _filter_gene_link_only_to_itself(link_d):\n",
    "    ret_d = copy.deepcopy(link_d)\n",
    "    gen_feats_to_remove = set()\n",
    "    coding_links = dict()\n",
    "    noncoding_links = dict()\n",
    "    for k, v in link_d.items():\n",
    "        if '/' in k:\n",
    "            noncoding_links[k] = v\n",
    "        else:\n",
    "            coding_links[k] = v\n",
    "\n",
    "    # filter out links due to overlapping genes\n",
    "    for cl_gen_feat, cl_geno_feats in coding_links.items():\n",
    "        for cl_geno_feat in cl_geno_feats:            \n",
    "            # if both gen and geno feat are genes.\n",
    "            if (cl_gen_feat in gene_df.GENE_ID.unique()) & (cl_geno_feat in gene_df.GENE_ID.unique()):\n",
    "                # geno feat isn't gen feat, remove link.\n",
    "                if cl_gen_feat != cl_geno_feat:\n",
    "                    if len(cl_geno_feats) == 1:\n",
    "                        gen_feats_to_remove.add(cl_gen_feat)\n",
    "                        del ret_d[cl_gen_feat]\n",
    "                    else:\n",
    "                        ret_d[cl_gen_feat].remove(cl_geno_feat)\n",
    "                        \n",
    "    # filters for when genes for some reason get linked with integenic regions\n",
    "    # the simplest thing to do is just remove any link of a gene to an intergenic region.\n",
    "    for ncl_gen_feat, ncl_geno_feats in noncoding_links.items():\n",
    "        for ncl_geno_feat in ncl_geno_feats:\n",
    "            if (ncl_gen_feat not in gene_df.GENE_ID.unique()) & (ncl_geno_feat in gene_df.GENE_ID.unique()):\n",
    "                if len(ncl_geno_feats) == 1:\n",
    "                    gen_feats_to_remove.add(ncl_gen_feat)\n",
    "                    del ret_d[ncl_gen_feat]\n",
    "                else:\n",
    "                    ret_d[ncl_gen_feat].remove(ncl_geno_feat)\n",
    "    \n",
    "    return ret_d, gen_feats_to_remove\n",
    "\n",
    "\n",
    "i = {'ECK120001080': ['ECK120001080', 'ECK125229291', 'ECK125229290'],\n",
    "     'ECK125229290': ['ECK120001080', 'ECK125229291', 'ECK125229290'],\n",
    "     'ECK125229291': ['ECK120001080', 'ECK125229291', 'ECK125229290'],}\n",
    "e = {'ECK120001080': ['ECK120001080'],\n",
    "     'ECK125229290': ['ECK125229290'],\n",
    "     'ECK125229291': ['ECK125229291']}\n",
    "o, _ = _filter_gene_link_only_to_itself(i)\n",
    "assert(o == e)\n",
    "\n",
    "\n",
    "i = {'ECK120002224': ['ECK120010816', 'ECK120002224', 'ECK120000393'],\n",
    "     'ECK120000393': ['ECK120002224', 'ECK120000393']}\n",
    "e = {'ECK120002224': ['ECK120010816', 'ECK120002224'],\n",
    "     'ECK120000393': ['ECK120000393']}\n",
    "o, _ = _filter_gene_link_only_to_itself(i)\n",
    "assert(o == e)\n",
    "\n",
    "\n",
    "i = {'ECK120003617': ['ECK120023915', 'ECK120003617'],\n",
    "     'ECK120023915': ['ECK120003616', 'ECK120023915'],\n",
    "     'ECK120003616': ['ECK120003616']}\n",
    "e = {'ECK120003617': ['ECK120003617'],\n",
    "     'ECK120023915': ['ECK120023915'],\n",
    "     'ECK120003616': ['ECK120003616']}\n",
    "o, _ = _filter_gene_link_only_to_itself(i)\n",
    "assert(o == e)\n",
    "\n",
    "\n",
    "i = {'ECK120001080': ['ECK120001080'],\n",
    "     'ECK125229291/ECK125256962': ['ECK120001080']}\n",
    "e = {'ECK120001080': ['ECK120001080']}\n",
    "es = {'ECK125229291/ECK125256962'}\n",
    "o, os = _filter_gene_link_only_to_itself(i)\n",
    "assert(o == e)\n",
    "assert(os == es)\n",
    "\n",
    "\n",
    "def _f(r):\n",
    "    link_d, gen_feats_to_remove = _filter_gene_link_only_to_itself(r[\"genetic feature links\"])\n",
    "    gen_feats = filter_feats(r[\"genetic features\"], gen_feats_to_remove)\n",
    "    return pd.Series([gen_feats, link_d])\n",
    "\n",
    "\n",
    "all_muts_df[[\"genetic features\", \"genetic feature links\"]] = all_muts_df.apply(_f, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:23:04.425842Z",
     "iopub.status.busy": "2021-01-01T04:23:04.425467Z",
     "iopub.status.idle": "2021-01-01T04:23:08.712249Z",
     "shell.execute_reply": "2021-01-01T04:23:08.711833Z"
    }
   },
   "outputs": [],
   "source": [
    "# fix for ASW-1628\n",
    "# remove links between cis-regulatory features and overlapping genes\n",
    "\n",
    "\n",
    "def _filter_cis_reg_overlapping_genes(link_d):\n",
    "    ret_d = copy.deepcopy(link_d)\n",
    "    gen_feats_to_remove = set()\n",
    "    # This issue likely is only happening with genes and not intergenic regions.\n",
    "    coding_links = dict()\n",
    "    for k, v in link_d.items():\n",
    "        if '/' not in k:\n",
    "            coding_links[k] = v\n",
    "\n",
    "    assigned_geno_feats = set()\n",
    "    for gen_feat, geno_feats in coding_links.items():\n",
    "        for geno_feat in geno_feats:\n",
    "            # ensuring that it's a cis-regulatory feature\n",
    "            if (geno_feat not in gene_df.GENE_ID.unique()) & (\"/\" not in geno_feat):\n",
    "                if geno_feat in assigned_geno_feats:  # If the geno feat has already been assigned, remove it\n",
    "                    if len(geno_feats) == 1:\n",
    "                        gen_feats_to_remove.add(gen_feats)\n",
    "                        del ret_d[gen_feats]\n",
    "                    else:\n",
    "                        ret_d[gen_feat].remove(geno_feat)\n",
    "                else:\n",
    "                    assigned_geno_feats.add(geno_feat)\n",
    "\n",
    "    return ret_d, gen_feats_to_remove\n",
    "\n",
    "\n",
    "i = {'ECK120003617': ['ECK120003617'],\n",
    "     'ECK120023915': ['ECK120013472', 'ECK120023915', 'ECK125095446'],\n",
    "     'ECK120003616': ['ECK120013472', 'ECK120003616', 'ECK125095446'],\n",
    "     'ECK120000314/ECK120023915': ['ECK120017150'],\n",
    "     'ECK120003617/ECK120003618': ['ECK120034252']}\n",
    "\n",
    "e = {'ECK120003617': ['ECK120003617'],\n",
    "     'ECK120023915': ['ECK120013472', 'ECK120023915', 'ECK125095446'],\n",
    "     'ECK120003616': ['ECK120003616'],\n",
    "     'ECK120000314/ECK120023915': ['ECK120017150'],\n",
    "     'ECK120003617/ECK120003618': ['ECK120034252']}\n",
    "o, _ = _filter_cis_reg_overlapping_genes(i)\n",
    "assert(o == e)\n",
    "\n",
    "\n",
    "def _f(r):\n",
    "    link_d, gen_feats_to_remove = _filter_cis_reg_overlapping_genes(\n",
    "        r[\"genetic feature links\"])\n",
    "    gen_feats = filter_feats(r[\"genetic features\"], gen_feats_to_remove)\n",
    "    return pd.Series([gen_feats, link_d])\n",
    "\n",
    "\n",
    "all_muts_df[[\"genetic features\", \"genetic feature links\"]] = all_muts_df.apply(_f, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-01-01T04:23:08.742461Z",
     "iopub.status.busy": "2021-01-01T04:23:08.714378Z",
     "iopub.status.idle": "2021-01-01T04:23:08.745978Z",
     "shell.execute_reply": "2021-01-01T04:23:08.745487Z"
    }
   },
   "outputs": [],
   "source": [
    "all_muts_df.to_pickle(\"./data/2_2_1_df.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
