{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-03T03:00:30.259194Z",
     "iopub.status.busy": "2021-05-03T03:00:30.258841Z",
     "iopub.status.idle": "2021-05-03T03:00:30.441510Z",
     "shell.execute_reply": "2021-05-03T03:00:30.440934Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-03T03:00:30.445216Z",
     "iopub.status.busy": "2021-05-03T03:00:30.444758Z",
     "iopub.status.idle": "2021-05-03T03:00:30.606002Z",
     "shell.execute_reply": "2021-05-03T03:00:30.605631Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5762, 34)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>...</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>{'ECK120000625': ['ECK120000625']}</td>\n",
       "      <td>[{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...</td>\n",
       "      <td>{'ECK125162049': ['ECK120000625']}</td>\n",
       "      <td>[{'name': 'PhoP'}, {'name': 'CRP'}, {'name': '...</td>\n",
       "      <td>{'CRP': ['ECK125162049'], 'NagC': ['ECK1251620...</td>\n",
       "      <td>[{'name': 'Galactose metabolism'}, {'name': 'A...</td>\n",
       "      <td>{'Galactose metabolism': ['ECK125162049'], 'Am...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>...</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>{'ECK120001556': ['ECK120001556']}</td>\n",
       "      <td>[{'name': 'clsA-yciU', 'RegulonDB ID': 'ECK125...</td>\n",
       "      <td>{'ECK125147415': ['ECK120001556']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK120000854'], 'ECK1200007...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000079...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism'}, {'name': '...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014627'], 'D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>...</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "      <td>{'Purine metabolism': ['ECK120016992'], 'Pyrim...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>...</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "      <td>{'ECK120000431': ['ECK120000431']}</td>\n",
       "      <td>[{'name': 'yjeF-tsaE-amiB-mutL-miaA-hfq-hflXKC...</td>\n",
       "      <td>{'ECK125147410': ['ECK120000431']}</td>\n",
       "      <td>[{'name': 'CRP'}]</td>\n",
       "      <td>{'CRP': ['ECK125147410']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change  ... terminator genetic  \\\n",
       "0          Δ21 bp  ...         {}    True   \n",
       "1             G→C  ...         {}    True   \n",
       "2          Δ82 bp  ...         {}    True   \n",
       "3             C→T  ...         {}    True   \n",
       "4             A→C  ...         {}    True   \n",
       "\n",
       "                                    genomic features  \\\n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...   \n",
       "1  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...   \n",
       "2  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "3  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "4  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...   \n",
       "\n",
       "                               genetic feature links  \\\n",
       "0                 {'ECK120000625': ['ECK120000625']}   \n",
       "1                 {'ECK120001556': ['ECK120001556']}   \n",
       "2  {'ECK120000854': ['ECK120000854'], 'ECK1200007...   \n",
       "3                 {'ECK120000886': ['ECK120000886']}   \n",
       "4                 {'ECK120000431': ['ECK120000431']}   \n",
       "\n",
       "                                             operons  \\\n",
       "0  [{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...   \n",
       "1  [{'name': 'clsA-yciU', 'RegulonDB ID': 'ECK125...   \n",
       "2  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "3  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "4  [{'name': 'yjeF-tsaE-amiB-mutL-miaA-hfq-hflXKC...   \n",
       "\n",
       "                                        operon links  \\\n",
       "0                 {'ECK125162049': ['ECK120000625']}   \n",
       "1                 {'ECK125147415': ['ECK120001556']}   \n",
       "2  {'ECK120014627': ['ECK120000854', 'ECK12000079...   \n",
       "3                 {'ECK120016992': ['ECK120000886']}   \n",
       "4                 {'ECK125147410': ['ECK120000431']}   \n",
       "\n",
       "                                          regulators  \\\n",
       "0  [{'name': 'PhoP'}, {'name': 'CRP'}, {'name': '...   \n",
       "1                                                 []   \n",
       "2                                                 []   \n",
       "3                                                 []   \n",
       "4                                  [{'name': 'CRP'}]   \n",
       "\n",
       "                                     regulator links  \\\n",
       "0  {'CRP': ['ECK125162049'], 'NagC': ['ECK1251620...   \n",
       "1                                                 {}   \n",
       "2                                                 {}   \n",
       "3                                                 {}   \n",
       "4                          {'CRP': ['ECK125147410']}   \n",
       "\n",
       "                                            pathways  \\\n",
       "0  [{'name': 'Galactose metabolism'}, {'name': 'A...   \n",
       "1                                                 []   \n",
       "2  [{'name': 'Pyrimidine metabolism'}, {'name': '...   \n",
       "3  [{'name': 'Purine metabolism'}, {'name': 'Pyri...   \n",
       "4                                                 []   \n",
       "\n",
       "                                       pathway links  \n",
       "0  {'Galactose metabolism': ['ECK125162049'], 'Am...  \n",
       "1                                                 {}  \n",
       "2  {'Pyrimidine metabolism': ['ECK120014627'], 'D...  \n",
       "3  {'Purine metabolism': ['ECK120016992'], 'Pyrim...  \n",
       "4                                                 {}  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/2_5_df.pkl\")\n",
    "display(all_muts_df.shape, all_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-03T03:00:30.609510Z",
     "iopub.status.busy": "2021-05-03T03:00:30.608362Z",
     "iopub.status.idle": "2021-05-03T03:00:30.633312Z",
     "shell.execute_reply": "2021-05-03T03:00:30.633619Z"
    }
   },
   "outputs": [],
   "source": [
    "cog_df = pd.read_pickle(\"./data/COG_df.pkl\")\n",
    "TU_objects_df = pd.read_pickle(\"./data/TU_objects_df.pkl\")\n",
    "operon_df = pd.read_pickle(\"./data/operon_df.pkl\")\n",
    "TU_df = pd.read_pickle(\"./data/TU_df.pkl\")\n",
    "gene_synonym_df = pd.read_pickle(\"./data/gene_synonym_df.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!!! Currently using lists instead of sets for COG hits because I want to know the count of items per COG affected with each mutation, where a mutation can affect multiple unique genomic targets that belong to the same COG. If I used a set, would lose the magnitude of hits to a COG per mutations.  \n",
    "THIS SHOULD GET FIXED WITH COG LINK DICTS, WHERE THE NUMBER OF TIMES AN OPERON SHOWS UP CAN IN THE LINKING LIST CAN BE THE NUMBER OF GENES INVOVLED IN MUTATING THE COG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-03T03:00:30.640232Z",
     "iopub.status.busy": "2021-05-03T03:00:30.639839Z",
     "iopub.status.idle": "2021-05-03T03:01:09.159663Z",
     "shell.execute_reply": "2021-05-03T03:01:09.159302Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>...</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>{'ECK120000625': ['ECK120000625']}</td>\n",
       "      <td>[{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...</td>\n",
       "      <td>{'ECK125162049': ['ECK120000625']}</td>\n",
       "      <td>[{'name': 'PhoP'}, {'name': 'CRP'}, {'name': '...</td>\n",
       "      <td>{'CRP': ['ECK125162049'], 'NagC': ['ECK1251620...</td>\n",
       "      <td>[{'name': 'Galactose metabolism'}, {'name': 'A...</td>\n",
       "      <td>{'Galactose metabolism': ['ECK125162049'], 'Am...</td>\n",
       "      <td>[{'name': 'Carbohydrate transport and metaboli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>{'ECK120001556': ['ECK120001556']}</td>\n",
       "      <td>[{'name': 'clsA-yciU', 'RegulonDB ID': 'ECK125...</td>\n",
       "      <td>{'ECK125147415': ['ECK120001556']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Lipid transport and metabolism'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK120000854'], 'ECK1200007...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000079...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism'}, {'name': '...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014627'], 'D...</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "      <td>{'Purine metabolism': ['ECK120016992'], 'Pyrim...</td>\n",
       "      <td>[{'name': 'Transcription'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "      <td>{'ECK120000431': ['ECK120000431']}</td>\n",
       "      <td>[{'name': 'yjeF-tsaE-amiB-mutL-miaA-hfq-hflXKC...</td>\n",
       "      <td>{'ECK125147410': ['ECK120000431']}</td>\n",
       "      <td>[{'name': 'CRP'}]</td>\n",
       "      <td>{'CRP': ['ECK125147410']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Signal transduction mechanisms'}]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change  ... genetic  \\\n",
       "0          Δ21 bp  ...    True   \n",
       "1             G→C  ...    True   \n",
       "2          Δ82 bp  ...    True   \n",
       "3             C→T  ...    True   \n",
       "4             A→C  ...    True   \n",
       "\n",
       "                                    genomic features  \\\n",
       "0  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...   \n",
       "1  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...   \n",
       "2  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "3  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "4  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...   \n",
       "\n",
       "                               genetic feature links  \\\n",
       "0                 {'ECK120000625': ['ECK120000625']}   \n",
       "1                 {'ECK120001556': ['ECK120001556']}   \n",
       "2  {'ECK120000854': ['ECK120000854'], 'ECK1200007...   \n",
       "3                 {'ECK120000886': ['ECK120000886']}   \n",
       "4                 {'ECK120000431': ['ECK120000431']}   \n",
       "\n",
       "                                             operons  \\\n",
       "0  [{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...   \n",
       "1  [{'name': 'clsA-yciU', 'RegulonDB ID': 'ECK125...   \n",
       "2  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "3  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "4  [{'name': 'yjeF-tsaE-amiB-mutL-miaA-hfq-hflXKC...   \n",
       "\n",
       "                                        operon links  \\\n",
       "0                 {'ECK125162049': ['ECK120000625']}   \n",
       "1                 {'ECK125147415': ['ECK120001556']}   \n",
       "2  {'ECK120014627': ['ECK120000854', 'ECK12000079...   \n",
       "3                 {'ECK120016992': ['ECK120000886']}   \n",
       "4                 {'ECK125147410': ['ECK120000431']}   \n",
       "\n",
       "                                          regulators  \\\n",
       "0  [{'name': 'PhoP'}, {'name': 'CRP'}, {'name': '...   \n",
       "1                                                 []   \n",
       "2                                                 []   \n",
       "3                                                 []   \n",
       "4                                  [{'name': 'CRP'}]   \n",
       "\n",
       "                                     regulator links  \\\n",
       "0  {'CRP': ['ECK125162049'], 'NagC': ['ECK1251620...   \n",
       "1                                                 {}   \n",
       "2                                                 {}   \n",
       "3                                                 {}   \n",
       "4                          {'CRP': ['ECK125147410']}   \n",
       "\n",
       "                                            pathways  \\\n",
       "0  [{'name': 'Galactose metabolism'}, {'name': 'A...   \n",
       "1                                                 []   \n",
       "2  [{'name': 'Pyrimidine metabolism'}, {'name': '...   \n",
       "3  [{'name': 'Purine metabolism'}, {'name': 'Pyri...   \n",
       "4                                                 []   \n",
       "\n",
       "                                       pathway links  \\\n",
       "0  {'Galactose metabolism': ['ECK125162049'], 'Am...   \n",
       "1                                                 {}   \n",
       "2  {'Pyrimidine metabolism': ['ECK120014627'], 'D...   \n",
       "3  {'Purine metabolism': ['ECK120016992'], 'Pyrim...   \n",
       "4                                                 {}   \n",
       "\n",
       "                                                COGs  \n",
       "0  [{'name': 'Carbohydrate transport and metaboli...  \n",
       "1       [{'name': 'Lipid transport and metabolism'}]  \n",
       "2  [{'name': 'Nucleotide transport and metabolism'}]  \n",
       "3                        [{'name': 'Transcription'}]  \n",
       "4       [{'name': 'Signal transduction mechanisms'}]  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from util.gene import get_gene_bnum\n",
    "\n",
    "\n",
    "# This is currently not attributing gene without COG to \"Function unknown\"\n",
    "# Last time checking this, it wasn't necessary to add the addition \"Function unknown\" entries here\n",
    "# though these extra entries were required with the link dictionaries\n",
    "def _get_COGs(gene_id):\n",
    "    annots = []\n",
    "    bnum = get_gene_bnum(gene_id, gene_synonym_df)\n",
    "    gene_cog_df = cog_df[cog_df[\"locus\"] == bnum]\n",
    "    if len(gene_cog_df) > 0:\n",
    "        for COG in gene_cog_df[\"COG description\"]:\n",
    "            d = {\"name\": COG}\n",
    "            annots.append(d)\n",
    "    return annots\n",
    "\n",
    "\n",
    "def get_COGs(mut_row):\n",
    "    annots = []\n",
    "    for feat_d in mut_row[\"genomic features\"]:\n",
    "        if feat_d[\"genetic\"]:\n",
    "            annots += _get_COGs(feat_d[\"RegulonDB ID\"])\n",
    "        # This will catch whether no features exist in the feat_ID_set\n",
    "        # or if those features are just integenic regions.\n",
    "        # Get operon genetic feature link dict through overlapping mutation and operon ranges.\n",
    "        # All genetic features are annotated within both the \"genetic features\" and \"genetic feature links\" column.\n",
    "        # If no integenic genomic feature is hit by a mutation, use the genetic feature range in an overlap analysis with operons.\n",
    "        else:\n",
    "            TU_IDs = set(TU_objects_df[TU_objects_df[\"TU_OBJECT_ID\"] == feat_d[\"RegulonDB ID\"]][\"TRANSCRIPTION_UNIT_ID\"])\n",
    "            for TU_ID in TU_IDs:\n",
    "                subset_TU_genes_df = TU_objects_df[(TU_objects_df[\"TRANSCRIPTION_UNIT_ID\"] == TU_ID) & (\n",
    "                    TU_objects_df[\"TU_OBJECT_CLASS\"] == 'GN')]\n",
    "                for gene_id in subset_TU_genes_df[\"TU_OBJECT_ID\"].drop_duplicates():\n",
    "                    # RegulonDB can return 'nan' or '' gene names\n",
    "                    if str(gene_id) != 'nan' and gene_id != '':\n",
    "                        annots += _get_COGs(gene_id)\n",
    "    return annots\n",
    "\n",
    "\n",
    "test_mut_row_d = {\n",
    "    \"exp\": \"test exp\",\n",
    "    \"genomic features\":\n",
    "    [\n",
    "        {'name': 'rph-pyrE attenuator terminator',\n",
    "         'RegulonDB ID': 'ECK125144791',\n",
    "         'range': (3815799, 3815828),\n",
    "         'genetic': False,\n",
    "         'feature type': 'attenuator terminator',\n",
    "         'operon': 'rph-pyrE',\n",
    "         'mutation set count': 5,\n",
    "         'significant': True},\n",
    "        {'name': 'rph',\n",
    "         'RegulonDB ID': 'ECK120000854',\n",
    "         'range': (3815863, 3816549),\n",
    "         'genetic': True,\n",
    "         'feature type': 'gene',\n",
    "         'operon': 'rph-pyrE',\n",
    "         'mutation set count': 5,\n",
    "         'significant': True}\n",
    "    ]}\n",
    "assert(get_COGs(test_mut_row_d) == [{'name': 'Nucleotide transport and metabolism'}])\n",
    "\n",
    "\n",
    "all_muts_df[\"COGs\"] = all_muts_df.apply(lambda r: get_COGs(r), axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-03T03:01:09.169655Z",
     "iopub.status.busy": "2021-05-03T03:01:09.166933Z",
     "iopub.status.idle": "2021-05-03T03:02:37.026615Z",
     "shell.execute_reply": "2021-05-03T03:02:37.026252Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>...</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>{'ECK120000625': ['ECK120000625']}</td>\n",
       "      <td>[{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...</td>\n",
       "      <td>{'ECK125162049': ['ECK120000625']}</td>\n",
       "      <td>[{'name': 'PhoP'}, {'name': 'CRP'}, {'name': '...</td>\n",
       "      <td>{'CRP': ['ECK125162049'], 'NagC': ['ECK1251620...</td>\n",
       "      <td>[{'name': 'Galactose metabolism'}, {'name': 'A...</td>\n",
       "      <td>{'Galactose metabolism': ['ECK125162049'], 'Am...</td>\n",
       "      <td>[{'name': 'Carbohydrate transport and metaboli...</td>\n",
       "      <td>{'Carbohydrate transport and metabolism': ['EC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>{'ECK120001556': ['ECK120001556']}</td>\n",
       "      <td>[{'name': 'clsA-yciU', 'RegulonDB ID': 'ECK125...</td>\n",
       "      <td>{'ECK125147415': ['ECK120001556']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Lipid transport and metabolism'}]</td>\n",
       "      <td>{'Lipid transport and metabolism': ['ECK125147...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815859</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK120000854'], 'ECK1200007...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000079...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism'}, {'name': '...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014627'], 'D...</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism'}]</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4187550</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "      <td>{'Purine metabolism': ['ECK120016992'], 'Pyrim...</td>\n",
       "      <td>[{'name': 'Transcription'}]</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4400313</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...</td>\n",
       "      <td>{'ECK120000431': ['ECK120000431']}</td>\n",
       "      <td>[{'name': 'yjeF-tsaE-amiB-mutL-miaA-hfq-hflXKC...</td>\n",
       "      <td>{'ECK125147410': ['ECK120000431']}</td>\n",
       "      <td>[{'name': 'CRP'}]</td>\n",
       "      <td>{'CRP': ['ECK125147410']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Signal transduction mechanisms'}]</td>\n",
       "      <td>{'Signal transduction mechanisms': ['ECK125147...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  index  exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0     9  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1    10  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "2    11  42C    1    124        1         1       1.0   3815859           DEL   \n",
       "3    12  42C    1    124        1         1       1.0   4187550           SNP   \n",
       "4    13  42C    1    124        1         1       1.0   4400313           SNP   \n",
       "\n",
       "  Sequence Change  ...                                   genomic features  \\\n",
       "0          Δ21 bp  ...  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...   \n",
       "1             G→C  ...  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...   \n",
       "2          Δ82 bp  ...  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "3             C→T  ...  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "4             A→C  ...  [{'name': 'hfq', 'RegulonDB ID': 'ECK120000431...   \n",
       "\n",
       "                               genetic feature links  \\\n",
       "0                 {'ECK120000625': ['ECK120000625']}   \n",
       "1                 {'ECK120001556': ['ECK120001556']}   \n",
       "2  {'ECK120000854': ['ECK120000854'], 'ECK1200007...   \n",
       "3                 {'ECK120000886': ['ECK120000886']}   \n",
       "4                 {'ECK120000431': ['ECK120000431']}   \n",
       "\n",
       "                                             operons  \\\n",
       "0  [{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...   \n",
       "1  [{'name': 'clsA-yciU', 'RegulonDB ID': 'ECK125...   \n",
       "2  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "3  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "4  [{'name': 'yjeF-tsaE-amiB-mutL-miaA-hfq-hflXKC...   \n",
       "\n",
       "                                        operon links  \\\n",
       "0                 {'ECK125162049': ['ECK120000625']}   \n",
       "1                 {'ECK125147415': ['ECK120001556']}   \n",
       "2  {'ECK120014627': ['ECK120000854', 'ECK12000079...   \n",
       "3                 {'ECK120016992': ['ECK120000886']}   \n",
       "4                 {'ECK125147410': ['ECK120000431']}   \n",
       "\n",
       "                                          regulators  \\\n",
       "0  [{'name': 'PhoP'}, {'name': 'CRP'}, {'name': '...   \n",
       "1                                                 []   \n",
       "2                                                 []   \n",
       "3                                                 []   \n",
       "4                                  [{'name': 'CRP'}]   \n",
       "\n",
       "                                     regulator links  \\\n",
       "0  {'CRP': ['ECK125162049'], 'NagC': ['ECK1251620...   \n",
       "1                                                 {}   \n",
       "2                                                 {}   \n",
       "3                                                 {}   \n",
       "4                          {'CRP': ['ECK125147410']}   \n",
       "\n",
       "                                            pathways  \\\n",
       "0  [{'name': 'Galactose metabolism'}, {'name': 'A...   \n",
       "1                                                 []   \n",
       "2  [{'name': 'Pyrimidine metabolism'}, {'name': '...   \n",
       "3  [{'name': 'Purine metabolism'}, {'name': 'Pyri...   \n",
       "4                                                 []   \n",
       "\n",
       "                                       pathway links  \\\n",
       "0  {'Galactose metabolism': ['ECK125162049'], 'Am...   \n",
       "1                                                 {}   \n",
       "2  {'Pyrimidine metabolism': ['ECK120014627'], 'D...   \n",
       "3  {'Purine metabolism': ['ECK120016992'], 'Pyrim...   \n",
       "4                                                 {}   \n",
       "\n",
       "                                                COGs  \\\n",
       "0  [{'name': 'Carbohydrate transport and metaboli...   \n",
       "1       [{'name': 'Lipid transport and metabolism'}]   \n",
       "2  [{'name': 'Nucleotide transport and metabolism'}]   \n",
       "3                        [{'name': 'Transcription'}]   \n",
       "4       [{'name': 'Signal transduction mechanisms'}]   \n",
       "\n",
       "                                           COG links  \n",
       "0  {'Carbohydrate transport and metabolism': ['EC...  \n",
       "1  {'Lipid transport and metabolism': ['ECK125147...  \n",
       "2  {'Nucleotide transport and metabolism': ['ECK1...  \n",
       "3                {'Transcription': ['ECK120016992']}  \n",
       "4  {'Signal transduction mechanisms': ['ECK125147...  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from util.operon import get_operon_name_set, get_operon_ID_set\n",
    "\n",
    "\n",
    "def _add_links_to_d(bnum, d, op_ID_set):\n",
    "    gene_cog_df = cog_df[cog_df[\"locus\"]==bnum]\n",
    "    if len(gene_cog_df) > 0:\n",
    "        for cog in gene_cog_df[\"COG description\"]:\n",
    "            if cog not in d.keys(): d[cog] = []\n",
    "            d[cog] += list(op_ID_set)\n",
    "#     else:\n",
    "#         if \"Function unknown\" not in d.keys(): d[\"Function unknown\"] = []\n",
    "#         d[\"Function unknown\"] = list(op_ID_set)\n",
    "\n",
    "\n",
    "def get_COG_operon_links(mut_row):\n",
    "    COG_op_link_d = dict()\n",
    "    for feat_d in mut_row[\"genomic features\"]:\n",
    "\n",
    "#             link_exists = False\n",
    "\n",
    "        if feat_d[\"genetic\"]:\n",
    "            op_ID_set = get_operon_ID_set(feat_d[\"RegulonDB ID\"], TU_objects_df, TU_df, operon_df)\n",
    "            if str(feat_d[\"RegulonDB ID\"]) != 'nan' and feat_d[\"RegulonDB ID\"] != '':\n",
    "                bnum = get_gene_bnum(feat_d[\"RegulonDB ID\"], gene_synonym_df)\n",
    "                _add_links_to_d(bnum, COG_op_link_d, op_ID_set)\n",
    "#                     link_exists = True\n",
    "        else:\n",
    "            TU_IDs = set(TU_objects_df[TU_objects_df[\"TU_OBJECT_ID\"]==feat_d[\"RegulonDB ID\"]][\"TRANSCRIPTION_UNIT_ID\"])\n",
    "            for TU_ID in TU_IDs:\n",
    "                subset_TU_genes_df = TU_objects_df[(TU_objects_df[\"TRANSCRIPTION_UNIT_ID\"]==TU_ID) & (TU_objects_df[\"TU_OBJECT_CLASS\"]=='GN')]\n",
    "\n",
    "                op_ID_set = set()\n",
    "                for _, TU_gene in subset_TU_genes_df.iterrows():\n",
    "                     op_ID_set = op_ID_set | get_operon_ID_set(TU_gene[\"TU_OBJECT_ID\"], TU_objects_df, TU_df, operon_df)\n",
    "\n",
    "                for gene_id in subset_TU_genes_df[\"TU_OBJECT_ID\"].drop_duplicates():\n",
    "                    if str(gene_id) != 'nan' and gene_id != '':  # RegulonDB can return 'nan' or '' gene names\n",
    "                        bnum = get_gene_bnum(gene_id, gene_synonym_df)\n",
    "                        _add_links_to_d(bnum, COG_op_link_d, op_ID_set)\n",
    "#                             link_exists = True\n",
    "\n",
    "        # Per mutation feature\n",
    "        # Need to be looking at previously populated level of annotations depdending on how the current level is being populated\n",
    "        # for regulon, it just means that if there are no operons, go immediately to genomic features.\n",
    "#             if link_exists == False:\n",
    "#                 if \"Function unknown\" not in COG_op_link_d.keys(): COG_op_link_d[\"Function unknown\"] = []\n",
    "#                 if len(operon_name_set) > 0: # need to always check the nearest annotation type for defined features\n",
    "#                     COG_op_link_d[\"Function unknown\"] += list(operon_name_set)\n",
    "#                 else:\n",
    "#                     COG_op_link_d[\"Function unknown\"].append(feat_d[\"RegulonDB ID\"])\n",
    "\n",
    "    return COG_op_link_d\n",
    "\n",
    "\n",
    "all_muts_df[\"COG links\"] = all_muts_df.apply(get_COG_operon_links, axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-03T03:02:37.077790Z",
     "iopub.status.busy": "2021-05-03T03:02:37.073111Z",
     "iopub.status.idle": "2021-05-03T03:02:37.081872Z",
     "shell.execute_reply": "2021-05-03T03:02:37.081409Z"
    }
   },
   "outputs": [],
   "source": [
    "all_muts_df.to_pickle(\"./data/2_6_df.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
