{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.metadata import get_condition_val_dict, get_condition_field_val_set\n",
    "from util.gene import get_coding_genetic_target_len_d, get_intergenic_len_d\n",
    "from util.genome import get_feature_hit_set, is_overlap, get_promoter_range_from_RegulonDB_df_row\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46, 41)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1232</td>\n",
       "      <td>4</td>\n",
       "      <td>R121C (CGC→TGC)</td>\n",
       "      <td>xylR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3735339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3735339, 3735339)</td>\n",
       "      <td>{ECK120002449}</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>{'ECK120002449': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014482': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1233</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1234</td>\n",
       "      <td>6</td>\n",
       "      <td>L770R (CTC→CGC)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187658, 4187658)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1235</td>\n",
       "      <td>9</td>\n",
       "      <td>G164A (GGC→GCC)</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3211597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3211597, 3211597)</td>\n",
       "      <td>{ECK120000233}</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>{'ECK120000233': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120014515': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'LexA', 'significantly associated co...</td>\n",
       "      <td>{'LexA': ['rpsU-dnaG-rpoD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1236</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index          Details mutation target annotation Mutation Type  \\\n",
       "1232     4  R121C (CGC→TGC)                       xylR           SNP   \n",
       "1233     5             None                        rph           DEL   \n",
       "1234     6  L770R (CTC→CGC)                       rpoC           SNP   \n",
       "1235     9  G164A (GGC→GCC)                       dnaG           SNP   \n",
       "1236     5             None                        rph           DEL   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale      exp  flask  isolate  \\\n",
       "1232   3735339           NaN             C→T    1  SSW_XYL    116        1   \n",
       "1233   3815859           NaN          Δ82 bp    1  SSW_XYL    116        1   \n",
       "1234   4187658           NaN             T→G    1  SSW_XYL    116        1   \n",
       "1235   3211597           NaN             G→C    2  SSW_XYL    109        1   \n",
       "1236   3815859           NaN          Δ82 bp    2  SSW_XYL    109        1   \n",
       "\n",
       "      presence  tech_rep  coding               range gene RegulonDB ID  \\\n",
       "1232       1.0         1    True  (3735339, 3735339)    {ECK120002449}   \n",
       "1233       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "1234       1.0         1    True  (4187658, 4187658)    {ECK120000886}   \n",
       "1235       1.0         1    True  (3211597, 3211597)    {ECK120000233}   \n",
       "1236       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...  False       False   \n",
       "1233  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1236  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "\n",
       "     TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "1232   {}       {}  {}                    {}         {}     True   \n",
       "1233   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "1234   {}       {}  {}                    {}         {}     True   \n",
       "1235   {}       {}  {}                    {}         {}     True   \n",
       "1236   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...   \n",
       "1233  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...   \n",
       "1236  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1232                 {'ECK120002449': ['ECK120002449']}   \n",
       "1233  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "1234                 {'ECK120000886': ['ECK120000886']}   \n",
       "1235                 {'ECK120000233': ['ECK120000233']}   \n",
       "1236  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "\n",
       "                                                operons  \\\n",
       "1232  [{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...   \n",
       "1233  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "1234  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "1235  [{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...   \n",
       "1236  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "\n",
       "                                           operon links  \\\n",
       "1232                 {'ECK120014482': ['ECK120002449']}   \n",
       "1233  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "1234                 {'ECK120016992': ['ECK120000886']}   \n",
       "1235                 {'ECK120014515': ['ECK120000233']}   \n",
       "1236  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "\n",
       "                                             regulators  \\\n",
       "1232  [{'name': 'Fis', 'significantly associated con...   \n",
       "1233                                                 []   \n",
       "1234                                                 []   \n",
       "1235  [{'name': 'LexA', 'significantly associated co...   \n",
       "1236                                                 []   \n",
       "\n",
       "                                        regulator links temperature  \\\n",
       "1232  {'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...  37 celsius   \n",
       "1233                                                 {}  37 celsius   \n",
       "1234                                                 {}  37 celsius   \n",
       "1235                       {'LexA': ['rpsU-dnaG-rpoD']}  37 celsius   \n",
       "1236                                                 {}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "1232     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1233     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1234     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1235     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1236     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "     taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "1232      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1233      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1234      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1235      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1236      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "     sulfur-source calcium-source  \n",
       "1232   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1233   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1234   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1235   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1236   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assoc_muts_df = pd.read_pickle(\"./data/5_df.pkl\")\n",
    "display(assoc_muts_df.shape, assoc_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Significance study (level specific)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## permutation test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120002449</td>\n",
       "      <td>1179</td>\n",
       "      <td>2</td>\n",
       "      <td>xylR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125144791</td>\n",
       "      <td>113</td>\n",
       "      <td>4</td>\n",
       "      <td>rph-pyrE attenuator terminator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000854</td>\n",
       "      <td>687</td>\n",
       "      <td>4</td>\n",
       "      <td>rph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000886</td>\n",
       "      <td>4224</td>\n",
       "      <td>3</td>\n",
       "      <td>rpoC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000233</td>\n",
       "      <td>1746</td>\n",
       "      <td>1</td>\n",
       "      <td>dnaG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count                            name\n",
       "ECK120002449   1179                       2                            xylR\n",
       "ECK125144791    113                       4  rph-pyrE attenuator terminator\n",
       "ECK120000854    687                       4                             rph\n",
       "ECK120000886   4224                       3                            rpoC\n",
       "ECK120000233   1746                       1                            dnaG"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Genomic features can use the \"features\" column since there is no linking to them from a lower-level annotation (yet: sub-genetic components)\n",
    "\n",
    "comp_mut_df = pd.DataFrame(columns=[\"length\", \"observed mutation count\", \"name\"])\n",
    "for mut_feats_l in assoc_muts_df[\"genomic features\"]:\n",
    "    for mut_comp_d in mut_feats_l:\n",
    "        if mut_comp_d[\"RegulonDB ID\"] in comp_mut_df.index:\n",
    "            comp_mut_df.loc[mut_comp_d[\"RegulonDB ID\"], \"observed mutation count\"] += 1\n",
    "        else:\n",
    "            comp_len = mut_comp_d[\"range\"][1] - mut_comp_d[\"range\"][0] + 1\n",
    "            df = pd.DataFrame([[mut_comp_d[\"name\"], comp_len, 1]],\n",
    "                              columns=[\"name\", \"length\", \"observed mutation count\"],\n",
    "                              index=[mut_comp_d[\"RegulonDB ID\"]])  # \"name\" column just for visual inspection\n",
    "            comp_mut_df = comp_mut_df.append(df, sort=False)\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: This may be obsolete (no longer being used in the pipeline). Check and remove if so.\n",
    "for _, r in assoc_muts_df.iterrows():\n",
    "    for feat_d in r[\"genomic features\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"RegulonDB ID\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120002449</td>\n",
       "      <td>1179</td>\n",
       "      <td>2</td>\n",
       "      <td>xylR</td>\n",
       "      <td>0.000254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125144791</td>\n",
       "      <td>113</td>\n",
       "      <td>4</td>\n",
       "      <td>rph-pyrE attenuator terminator</td>\n",
       "      <td>0.000024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000854</td>\n",
       "      <td>687</td>\n",
       "      <td>4</td>\n",
       "      <td>rph</td>\n",
       "      <td>0.000148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000886</td>\n",
       "      <td>4224</td>\n",
       "      <td>3</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.000910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000233</td>\n",
       "      <td>1746</td>\n",
       "      <td>1</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>0.000376</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count                            name  \\\n",
       "ECK120002449   1179                       2                            xylR   \n",
       "ECK125144791    113                       4  rph-pyrE attenuator terminator   \n",
       "ECK120000854    687                       4                             rph   \n",
       "ECK120000886   4224                       3                            rpoC   \n",
       "ECK120000233   1746                       1                            dnaG   \n",
       "\n",
       "              proportion  \n",
       "ECK120002449    0.000254  \n",
       "ECK125144791    0.000024  \n",
       "ECK120000854    0.000148  \n",
       "ECK120000886    0.000910  \n",
       "ECK120000233    0.000376  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "\n",
    "TOTAL_GENOME_LEN = 4641652\n",
    "genome_remainder_len = TOTAL_GENOME_LEN - mutated_seq_len\n",
    "df = pd.DataFrame([[0, genome_remainder_len]], columns=[\"observed mutation count\", \"length\"], index=[\"genome remainder\"])\n",
    "comp_mut_df = comp_mut_df.append(df, sort=False)\n",
    "total_seq_len = TOTAL_GENOME_LEN\n",
    "\n",
    "\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120002449</td>\n",
       "      <td>1179</td>\n",
       "      <td>2</td>\n",
       "      <td>xylR</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0002</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125144791</td>\n",
       "      <td>113</td>\n",
       "      <td>4</td>\n",
       "      <td>rph-pyrE attenuator terminator</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000854</td>\n",
       "      <td>687</td>\n",
       "      <td>4</td>\n",
       "      <td>rph</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000886</td>\n",
       "      <td>4224</td>\n",
       "      <td>3</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.000910</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000233</td>\n",
       "      <td>1746</td>\n",
       "      <td>1</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>0.000376</td>\n",
       "      <td>206</td>\n",
       "      <td>0.0206</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count                            name  \\\n",
       "ECK120002449   1179                       2                            xylR   \n",
       "ECK125144791    113                       4  rph-pyrE attenuator terminator   \n",
       "ECK120000854    687                       4                             rph   \n",
       "ECK120000886   4224                       3                            rpoC   \n",
       "ECK120000233   1746                       1                            dnaG   \n",
       "\n",
       "              proportion  equal or larger count  p value  significant  \n",
       "ECK120002449    0.000254                      2   0.0002         True  \n",
       "ECK125144791    0.000024                      0   0.0000         True  \n",
       "ECK120000854    0.000148                      0   0.0000         True  \n",
       "ECK120000886    0.000910                      0   0.0000         True  \n",
       "ECK120000233    0.000376                    206   0.0206         True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=num_muts,\n",
    "        replace=True,\n",
    "        p=proportion_l\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120002449</td>\n",
       "      <td>1179</td>\n",
       "      <td>2</td>\n",
       "      <td>xylR</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0002</td>\n",
       "      <td>True</td>\n",
       "      <td>0.008</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125144791</td>\n",
       "      <td>113</td>\n",
       "      <td>4</td>\n",
       "      <td>rph-pyrE attenuator terminator</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000854</td>\n",
       "      <td>687</td>\n",
       "      <td>4</td>\n",
       "      <td>rph</td>\n",
       "      <td>0.000148</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000886</td>\n",
       "      <td>4224</td>\n",
       "      <td>3</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.000910</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000233</td>\n",
       "      <td>1746</td>\n",
       "      <td>1</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>0.000376</td>\n",
       "      <td>206</td>\n",
       "      <td>0.0206</td>\n",
       "      <td>True</td>\n",
       "      <td>0.824</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count                            name  \\\n",
       "ECK120002449   1179                       2                            xylR   \n",
       "ECK125144791    113                       4  rph-pyrE attenuator terminator   \n",
       "ECK120000854    687                       4                             rph   \n",
       "ECK120000886   4224                       3                            rpoC   \n",
       "ECK120000233   1746                       1                            dnaG   \n",
       "\n",
       "              proportion  equal or larger count  p value  significant  \\\n",
       "ECK120002449    0.000254                      2   0.0002         True   \n",
       "ECK125144791    0.000024                      0   0.0000         True   \n",
       "ECK120000854    0.000148                      0   0.0000         True   \n",
       "ECK120000886    0.000910                      0   0.0000         True   \n",
       "ECK120000233    0.000376                    206   0.0206         True   \n",
       "\n",
       "              corrected p value  corrected significance  \n",
       "ECK120002449              0.008                    True  \n",
       "ECK125144791              0.000                    True  \n",
       "ECK120000854              0.000                    True  \n",
       "ECK120000886              0.000                    True  \n",
       "ECK120000233              0.824                   False  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "# comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# writing significance status back into mutations DF.\n",
    "for i, r in assoc_muts_df.iterrows():\n",
    "    for d in r[\"genomic features\"]:\n",
    "        if d[\"RegulonDB ID\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"RegulonDB ID\"]][\"corrected significance\"]\n",
    "#             d[\"significant\"] = comp_mut_df.loc[d[\"RegulonDB ID\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputting dataframe for the next pipeline step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_muts_df.to_pickle(\"./data/6_df.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Currently commented out because some of the experiments have mutated features without names which will crash this cell & NB during whole pipeline execution. Uncomment when investigating individual experiments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 300\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # currently commented out because some of the experiments have mutated features without names\n",
    "# # which will crash this cell & NB during whole pipeline execution.\n",
    "# # Uncomment when investigating individual experiments.\n",
    "\n",
    "# TODO: found in multiple post-assoc NBs. Should extract into common script module.\n",
    "# def _get_name_signif_from_ID(RegulonDB_ID, json):\n",
    "#     name = \"\"\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"RegulonDB ID\"] == RegulonDB_ID:\n",
    "#             name = d[\"name\"]\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return (name, is_signif)\n",
    "\n",
    "\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     for feat_d in r[\"genomic features\"]:\n",
    "#         feat = feat_d[\"name\"]\n",
    "#         if \"attenuator terminator\" in feat:\n",
    "#             feat = feat.replace(\"attenuator terminator\", \"att term\")\n",
    "#         if \"terminator\" in feat:\n",
    "#             feat = feat.replace(\"terminator\", \"term\")\n",
    "        \n",
    "#         t = str(feat)\n",
    "#         if feat_d[\"significant\"]:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += 1\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-5:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,2),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.3, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 5 genomic mutated feature count\")\n",
    "# ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
