{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.metadata import get_condition_val_dict, get_condition_field_val_set\n",
    "from util.gene import get_coding_genetic_target_len_d, get_intergenic_len_d\n",
    "from util.genome import get_feature_hit_set, is_overlap, get_promoter_range_from_RegulonDB_df_row\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(114, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2527</td>\n",
       "      <td>28</td>\n",
       "      <td>None</td>\n",
       "      <td>[crl]</td>\n",
       "      <td>DEL</td>\n",
       "      <td>257908</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ776 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(257908, 258683)</td>\n",
       "      <td>{ECK125229291, ECK120001080, ECK125229290}</td>\n",
       "      <td>[{'name': 'crl', 'RegulonDB ID': 'ECK120001080...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'insA9', 'RegulonDB ID': 'ECK1252292...</td>\n",
       "      <td>{'ECK120001080': ['ECK120001080'], 'ECK1252292...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2528</td>\n",
       "      <td>3</td>\n",
       "      <td>coding (163/210 nt)</td>\n",
       "      <td>cspC</td>\n",
       "      <td>INS</td>\n",
       "      <td>1907273</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>(CGTCCTG)1→2</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1907273, 1907273)</td>\n",
       "      <td>{ECK120002092}</td>\n",
       "      <td>[{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...</td>\n",
       "      <td>{'ECK120002092': ['ECK120002092']}</td>\n",
       "      <td>[{'name': 'yobF-cspC', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028784': ['ECK120002092']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Transcription'}]</td>\n",
       "      <td>{'Transcription': ['ECK120028784']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2529</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>insB1, insA</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1978503</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ776 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1978503, 1979278)</td>\n",
       "      <td>{ECK120023915, ECK120003616, ECK120003617}</td>\n",
       "      <td>[{'name': 'insAB-5', 'RegulonDB ID': 'ECK12002...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120013472, ECK120017150}</td>\n",
       "      <td>{ECK125095446, ECK120034252}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'flhDC TFBS', 'RegulonDB ID': 'ECK12...</td>\n",
       "      <td>{'ECK120023915': ['ECK125095446', 'ECK12002391...</td>\n",
       "      <td>[{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...</td>\n",
       "      <td>{'ECK120014934': ['ECK120023915', 'ECK12002391...</td>\n",
       "      <td>[{'name': 'OmpR', 'significantly associated co...</td>\n",
       "      <td>{'MatA': ['flhDC', 'flhDC', 'flhDC'], 'OmpR': ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Mobilome: prophages, transposons'},...</td>\n",
       "      <td>{'Mobilome: prophages, transposons': ['ECK1250...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2530</td>\n",
       "      <td>32</td>\n",
       "      <td>intergenic (‑1/+1)</td>\n",
       "      <td>gatC/gatC</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2173363</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ2 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(2173363, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>{'ECK120002276': ['ECK120002276']}</td>\n",
       "      <td>[{'name': 'gatYZABCD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120015849': ['ECK120002276']}</td>\n",
       "      <td>[{'name': 'ArcA', 'significantly associated co...</td>\n",
       "      <td>{'ArcA': ['gatYZABCD'], 'GatR': ['gatYZABCD'],...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2531</td>\n",
       "      <td>4</td>\n",
       "      <td>N34K (AAT→AAG)</td>\n",
       "      <td>mreB</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3400986</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>TOL_putrescine</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3400986, 3400986)</td>\n",
       "      <td>{ECK120000601}</td>\n",
       "      <td>[{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>{'ECK120000601': ['ECK120000601']}</td>\n",
       "      <td>[{'name': 'mreBCD', 'RegulonDB ID': 'ECK120030...</td>\n",
       "      <td>{'ECK120030764': ['ECK120000601']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Cell cycle control, cell division, ...</td>\n",
       "      <td>{'Cell cycle control, cell division, chromosom...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>putrescine</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index              Details mutation target annotation Mutation Type  \\\n",
       "2527    28                 None                      [crl]           DEL   \n",
       "2528     3  coding (163/210 nt)                       cspC           INS   \n",
       "2529    29                 None                insB1, insA           DEL   \n",
       "2530    32   intergenic (‑1/+1)                  gatC/gatC           DEL   \n",
       "2531     4       N34K (AAT→AAG)                       mreB           SNP   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale             exp  flask  \\\n",
       "2527    257908     NC_000913         Δ776 bp    1  TOL_putrescine     50   \n",
       "2528   1907273     NC_000913    (CGTCCTG)1→2    1  TOL_putrescine     50   \n",
       "2529   1978503     NC_000913         Δ776 bp    1  TOL_putrescine     50   \n",
       "2530   2173363     NC_000913           Δ2 bp    1  TOL_putrescine     50   \n",
       "2531   3400986     NC_000913             A→C    1  TOL_putrescine     50   \n",
       "\n",
       "      isolate  presence  tech_rep  coding               range  \\\n",
       "2527        1       1.0         1    True    (257908, 258683)   \n",
       "2528        1       1.0         1    True  (1907273, 1907273)   \n",
       "2529        1       1.0         1    True  (1978503, 1979278)   \n",
       "2530        1       1.0         1   False  (2173363, 2173364)   \n",
       "2531        1       1.0         1    True  (3400986, 3400986)   \n",
       "\n",
       "                               gene RegulonDB ID  \\\n",
       "2527  {ECK125229291, ECK120001080, ECK125229290}   \n",
       "2528                              {ECK120002092}   \n",
       "2529  {ECK120023915, ECK120003616, ECK120003617}   \n",
       "2530                              {ECK120002276}   \n",
       "2531                              {ECK120000601}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "2527  [{'name': 'crl', 'RegulonDB ID': 'ECK120001080...  False       False   \n",
       "2528  [{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...  False       False   \n",
       "2529  [{'name': 'insAB-5', 'RegulonDB ID': 'ECK12002...  False       False   \n",
       "2530  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...  False       False   \n",
       "2531  [{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...  False       False   \n",
       "\n",
       "                              TFBS                      promoter RBS  \\\n",
       "2527                            {}                            {}  {}   \n",
       "2528                            {}                            {}  {}   \n",
       "2529  {ECK120013472, ECK120017150}  {ECK125095446, ECK120034252}  {}   \n",
       "2530                            {}                            {}  {}   \n",
       "2531                            {}                            {}  {}   \n",
       "\n",
       "     attenuator terminator terminator  genetic  \\\n",
       "2527                    {}         {}     True   \n",
       "2528                    {}         {}     True   \n",
       "2529                    {}         {}     True   \n",
       "2530                    {}         {}    False   \n",
       "2531                    {}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "2527  [{'name': 'insA9', 'RegulonDB ID': 'ECK1252292...   \n",
       "2528  [{'name': 'cspC', 'RegulonDB ID': 'ECK12000209...   \n",
       "2529  [{'name': 'flhDC TFBS', 'RegulonDB ID': 'ECK12...   \n",
       "2530  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...   \n",
       "2531  [{'name': 'mreB', 'RegulonDB ID': 'ECK12000060...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "2527  {'ECK120001080': ['ECK120001080'], 'ECK1252292...   \n",
       "2528                 {'ECK120002092': ['ECK120002092']}   \n",
       "2529  {'ECK120023915': ['ECK125095446', 'ECK12002391...   \n",
       "2530                 {'ECK120002276': ['ECK120002276']}   \n",
       "2531                 {'ECK120000601': ['ECK120000601']}   \n",
       "\n",
       "                                                operons  \\\n",
       "2527                                                 []   \n",
       "2528  [{'name': 'yobF-cspC', 'RegulonDB ID': 'ECK120...   \n",
       "2529  [{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...   \n",
       "2530  [{'name': 'gatYZABCD', 'RegulonDB ID': 'ECK120...   \n",
       "2531  [{'name': 'mreBCD', 'RegulonDB ID': 'ECK120030...   \n",
       "\n",
       "                                           operon links  \\\n",
       "2527                                                 {}   \n",
       "2528                 {'ECK120028784': ['ECK120002092']}   \n",
       "2529  {'ECK120014934': ['ECK120023915', 'ECK12002391...   \n",
       "2530                 {'ECK120015849': ['ECK120002276']}   \n",
       "2531                 {'ECK120030764': ['ECK120000601']}   \n",
       "\n",
       "                                             regulators  \\\n",
       "2527                                                 []   \n",
       "2528                                                 []   \n",
       "2529  [{'name': 'OmpR', 'significantly associated co...   \n",
       "2530  [{'name': 'ArcA', 'significantly associated co...   \n",
       "2531                                                 []   \n",
       "\n",
       "                                        regulator links pathways  \\\n",
       "2527                                                 {}       []   \n",
       "2528                                                 {}       []   \n",
       "2529  {'MatA': ['flhDC', 'flhDC', 'flhDC'], 'OmpR': ...       []   \n",
       "2530  {'ArcA': ['gatYZABCD'], 'GatR': ['gatYZABCD'],...       []   \n",
       "2531                                                 {}       []   \n",
       "\n",
       "     pathway links                                               COGs  \\\n",
       "2527            {}                                                 []   \n",
       "2528            {}                        [{'name': 'Transcription'}]   \n",
       "2529            {}  [{'name': 'Mobilome: prophages, transposons'},...   \n",
       "2530            {}                                                 []   \n",
       "2531            {}  [{'name': 'Cell cycle control, cell division, ...   \n",
       "\n",
       "                                              COG links temperature  \\\n",
       "2527                                                 {}  37 celsius   \n",
       "2528                {'Transcription': ['ECK120028784']}  37 celsius   \n",
       "2529  {'Mobilome: prophages, transposons': ['ECK1250...  37 celsius   \n",
       "2530                                                 {}  37 celsius   \n",
       "2531  {'Cell cycle control, cell division, chromosom...  37 celsius   \n",
       "\n",
       "     carbon-source  supplement strain-description taxonomy-id base-media  \\\n",
       "2527    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2528    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2529    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2530    glucose(2)  putrescine                 WT      511145         M9   \n",
       "2531    glucose(2)  putrescine                 WT      511145         M9   \n",
       "\n",
       "     nitrogen-source      phosphorous-source sulfur-source calcium-source  \n",
       "2527        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2528        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2529        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2530        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  \n",
       "2531        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assoc_muts_df = pd.read_pickle(\"./data/5_df.pkl\")\n",
    "display(assoc_muts_df.shape, assoc_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Significance study (level specific)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## permutation test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK125229291</td>\n",
       "      <td>276</td>\n",
       "      <td>7</td>\n",
       "      <td>insA9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125229290</td>\n",
       "      <td>504</td>\n",
       "      <td>7</td>\n",
       "      <td>insB9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120001080</td>\n",
       "      <td>1178</td>\n",
       "      <td>7</td>\n",
       "      <td>crl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120002092</td>\n",
       "      <td>210</td>\n",
       "      <td>2</td>\n",
       "      <td>cspC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120017150</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>flhDC TFBS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count        name\n",
       "ECK125229291    276                       7       insA9\n",
       "ECK125229290    504                       7       insB9\n",
       "ECK120001080   1178                       7         crl\n",
       "ECK120002092    210                       2        cspC\n",
       "ECK120017150     15                       7  flhDC TFBS"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Genomic features can use the \"features\" column since there is no linking to them from a lower-level annotation (yet: sub-genetic components)\n",
    "\n",
    "comp_mut_df = pd.DataFrame(columns=[\"length\", \"observed mutation count\", \"name\"])\n",
    "for mut_feats_l in assoc_muts_df[\"genomic features\"]:\n",
    "    for mut_comp_d in mut_feats_l:\n",
    "        if mut_comp_d[\"RegulonDB ID\"] in comp_mut_df.index:\n",
    "            comp_mut_df.loc[mut_comp_d[\"RegulonDB ID\"], \"observed mutation count\"] += 1\n",
    "        else:\n",
    "            comp_len = mut_comp_d[\"range\"][1] - mut_comp_d[\"range\"][0] + 1\n",
    "            df = pd.DataFrame([[mut_comp_d[\"name\"], comp_len, 1]],\n",
    "                              columns=[\"name\", \"length\", \"observed mutation count\"],\n",
    "                              index=[mut_comp_d[\"RegulonDB ID\"]])  # \"name\" column just for visual inspection\n",
    "            comp_mut_df = comp_mut_df.append(df, sort=False)\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: This may be obsolete (no longer being used in the pipeline). Check and remove if so.\n",
    "for _, r in assoc_muts_df.iterrows():\n",
    "    for feat_d in r[\"genomic features\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"RegulonDB ID\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK125229291</td>\n",
       "      <td>276</td>\n",
       "      <td>7</td>\n",
       "      <td>insA9</td>\n",
       "      <td>0.000059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125229290</td>\n",
       "      <td>504</td>\n",
       "      <td>7</td>\n",
       "      <td>insB9</td>\n",
       "      <td>0.000109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120001080</td>\n",
       "      <td>1178</td>\n",
       "      <td>7</td>\n",
       "      <td>crl</td>\n",
       "      <td>0.000254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120002092</td>\n",
       "      <td>210</td>\n",
       "      <td>2</td>\n",
       "      <td>cspC</td>\n",
       "      <td>0.000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120017150</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>flhDC TFBS</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count        name  proportion\n",
       "ECK125229291    276                       7       insA9    0.000059\n",
       "ECK125229290    504                       7       insB9    0.000109\n",
       "ECK120001080   1178                       7         crl    0.000254\n",
       "ECK120002092    210                       2        cspC    0.000045\n",
       "ECK120017150     15                       7  flhDC TFBS    0.000003"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "\n",
    "TOTAL_GENOME_LEN = 4641652\n",
    "genome_remainder_len = TOTAL_GENOME_LEN - mutated_seq_len\n",
    "df = pd.DataFrame([[0, genome_remainder_len]], columns=[\"observed mutation count\", \"length\"], index=[\"genome remainder\"])\n",
    "comp_mut_df = comp_mut_df.append(df, sort=False)\n",
    "total_seq_len = TOTAL_GENOME_LEN\n",
    "\n",
    "\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK125229291</td>\n",
       "      <td>276</td>\n",
       "      <td>7</td>\n",
       "      <td>insA9</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125229290</td>\n",
       "      <td>504</td>\n",
       "      <td>7</td>\n",
       "      <td>insB9</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120001080</td>\n",
       "      <td>1178</td>\n",
       "      <td>7</td>\n",
       "      <td>crl</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120002092</td>\n",
       "      <td>210</td>\n",
       "      <td>2</td>\n",
       "      <td>cspC</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120017150</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>flhDC TFBS</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count        name  proportion  \\\n",
       "ECK125229291    276                       7       insA9    0.000059   \n",
       "ECK125229290    504                       7       insB9    0.000109   \n",
       "ECK120001080   1178                       7         crl    0.000254   \n",
       "ECK120002092    210                       2        cspC    0.000045   \n",
       "ECK120017150     15                       7  flhDC TFBS    0.000003   \n",
       "\n",
       "              equal or larger count  p value  significant  \n",
       "ECK125229291                      0   0.0000         True  \n",
       "ECK125229290                      0   0.0000         True  \n",
       "ECK120001080                      0   0.0000         True  \n",
       "ECK120002092                      1   0.0001         True  \n",
       "ECK120017150                      0   0.0000         True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=num_muts,\n",
    "        replace=True,\n",
    "        p=proportion_l\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK125229291</td>\n",
       "      <td>276</td>\n",
       "      <td>7</td>\n",
       "      <td>insA9</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125229290</td>\n",
       "      <td>504</td>\n",
       "      <td>7</td>\n",
       "      <td>insB9</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120001080</td>\n",
       "      <td>1178</td>\n",
       "      <td>7</td>\n",
       "      <td>crl</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120002092</td>\n",
       "      <td>210</td>\n",
       "      <td>2</td>\n",
       "      <td>cspC</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0081</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120017150</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>flhDC TFBS</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count        name  proportion  \\\n",
       "ECK125229291    276                       7       insA9    0.000059   \n",
       "ECK125229290    504                       7       insB9    0.000109   \n",
       "ECK120001080   1178                       7         crl    0.000254   \n",
       "ECK120002092    210                       2        cspC    0.000045   \n",
       "ECK120017150     15                       7  flhDC TFBS    0.000003   \n",
       "\n",
       "              equal or larger count  p value  significant  corrected p value  \\\n",
       "ECK125229291                      0   0.0000         True             0.0000   \n",
       "ECK125229290                      0   0.0000         True             0.0000   \n",
       "ECK120001080                      0   0.0000         True             0.0000   \n",
       "ECK120002092                      1   0.0001         True             0.0081   \n",
       "ECK120017150                      0   0.0000         True             0.0000   \n",
       "\n",
       "              corrected significance  \n",
       "ECK125229291                    True  \n",
       "ECK125229290                    True  \n",
       "ECK120001080                    True  \n",
       "ECK120002092                    True  \n",
       "ECK120017150                    True  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "      <th>final significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK125229291</td>\n",
       "      <td>276</td>\n",
       "      <td>7</td>\n",
       "      <td>insA9</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125229290</td>\n",
       "      <td>504</td>\n",
       "      <td>7</td>\n",
       "      <td>insB9</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120001080</td>\n",
       "      <td>1178</td>\n",
       "      <td>7</td>\n",
       "      <td>crl</td>\n",
       "      <td>0.000254</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120002092</td>\n",
       "      <td>210</td>\n",
       "      <td>2</td>\n",
       "      <td>cspC</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0081</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120017150</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>flhDC TFBS</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120013472</td>\n",
       "      <td>20</td>\n",
       "      <td>7</td>\n",
       "      <td>flhDC TFBS</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125095446</td>\n",
       "      <td>81</td>\n",
       "      <td>7</td>\n",
       "      <td>flhDp2</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120034252</td>\n",
       "      <td>81</td>\n",
       "      <td>7</td>\n",
       "      <td>insA-5p</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120003617</td>\n",
       "      <td>276</td>\n",
       "      <td>7</td>\n",
       "      <td>insA-5</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120003616</td>\n",
       "      <td>504</td>\n",
       "      <td>7</td>\n",
       "      <td>insB-5</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120023915</td>\n",
       "      <td>698</td>\n",
       "      <td>7</td>\n",
       "      <td>insAB-5</td>\n",
       "      <td>0.000150</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120002276</td>\n",
       "      <td>1358</td>\n",
       "      <td>7</td>\n",
       "      <td>gatC</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000601</td>\n",
       "      <td>1044</td>\n",
       "      <td>5</td>\n",
       "      <td>mreB</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000393</td>\n",
       "      <td>758</td>\n",
       "      <td>7</td>\n",
       "      <td>glpR</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000886</td>\n",
       "      <td>4224</td>\n",
       "      <td>3</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>0.000910</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0405</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000398/ECK120001878</td>\n",
       "      <td>641</td>\n",
       "      <td>7</td>\n",
       "      <td>gltP/yjcO</td>\n",
       "      <td>0.000138</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000762</td>\n",
       "      <td>1203</td>\n",
       "      <td>3</td>\n",
       "      <td>proV</td>\n",
       "      <td>0.000259</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000955</td>\n",
       "      <td>2109</td>\n",
       "      <td>4</td>\n",
       "      <td>spoT</td>\n",
       "      <td>0.000454</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000891</td>\n",
       "      <td>1674</td>\n",
       "      <td>3</td>\n",
       "      <td>rpsA</td>\n",
       "      <td>0.000361</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0081</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120002480/ECK120002482</td>\n",
       "      <td>36</td>\n",
       "      <td>3</td>\n",
       "      <td>glyV/glyX</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125144791</td>\n",
       "      <td>113</td>\n",
       "      <td>2</td>\n",
       "      <td>rph-pyrE attenuator terminator</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          length observed mutation count  \\\n",
       "ECK125229291                 276                       7   \n",
       "ECK125229290                 504                       7   \n",
       "ECK120001080                1178                       7   \n",
       "ECK120002092                 210                       2   \n",
       "ECK120017150                  15                       7   \n",
       "ECK120013472                  20                       7   \n",
       "ECK125095446                  81                       7   \n",
       "ECK120034252                  81                       7   \n",
       "ECK120003617                 276                       7   \n",
       "ECK120003616                 504                       7   \n",
       "ECK120023915                 698                       7   \n",
       "ECK120002276                1358                       7   \n",
       "ECK120000601                1044                       5   \n",
       "ECK120000393                 758                       7   \n",
       "ECK120000886                4224                       3   \n",
       "ECK120000398/ECK120001878    641                       7   \n",
       "ECK120000762                1203                       3   \n",
       "ECK120000955                2109                       4   \n",
       "ECK120000891                1674                       3   \n",
       "ECK120002480/ECK120002482     36                       3   \n",
       "ECK125144791                 113                       2   \n",
       "\n",
       "                                                     name  proportion  \\\n",
       "ECK125229291                                        insA9    0.000059   \n",
       "ECK125229290                                        insB9    0.000109   \n",
       "ECK120001080                                          crl    0.000254   \n",
       "ECK120002092                                         cspC    0.000045   \n",
       "ECK120017150                                   flhDC TFBS    0.000003   \n",
       "ECK120013472                                   flhDC TFBS    0.000004   \n",
       "ECK125095446                                       flhDp2    0.000017   \n",
       "ECK120034252                                      insA-5p    0.000017   \n",
       "ECK120003617                                       insA-5    0.000059   \n",
       "ECK120003616                                       insB-5    0.000109   \n",
       "ECK120023915                                      insAB-5    0.000150   \n",
       "ECK120002276                                         gatC    0.000293   \n",
       "ECK120000601                                         mreB    0.000225   \n",
       "ECK120000393                                         glpR    0.000163   \n",
       "ECK120000886                                         rpoC    0.000910   \n",
       "ECK120000398/ECK120001878                       gltP/yjcO    0.000138   \n",
       "ECK120000762                                         proV    0.000259   \n",
       "ECK120000955                                         spoT    0.000454   \n",
       "ECK120000891                                         rpsA    0.000361   \n",
       "ECK120002480/ECK120002482                       glyV/glyX    0.000008   \n",
       "ECK125144791               rph-pyrE attenuator terminator    0.000024   \n",
       "\n",
       "                           equal or larger count  p value  significant  \\\n",
       "ECK125229291                                   0   0.0000         True   \n",
       "ECK125229290                                   0   0.0000         True   \n",
       "ECK120001080                                   0   0.0000         True   \n",
       "ECK120002092                                   1   0.0001         True   \n",
       "ECK120017150                                   0   0.0000         True   \n",
       "ECK120013472                                   0   0.0000         True   \n",
       "ECK125095446                                   0   0.0000         True   \n",
       "ECK120034252                                   0   0.0000         True   \n",
       "ECK120003617                                   0   0.0000         True   \n",
       "ECK120003616                                   0   0.0000         True   \n",
       "ECK120023915                                   0   0.0000         True   \n",
       "ECK120002276                                   0   0.0000         True   \n",
       "ECK120000601                                   0   0.0000         True   \n",
       "ECK120000393                                   0   0.0000         True   \n",
       "ECK120000886                                   5   0.0005         True   \n",
       "ECK120000398/ECK120001878                      0   0.0000         True   \n",
       "ECK120000762                                   0   0.0000         True   \n",
       "ECK120000955                                   0   0.0000         True   \n",
       "ECK120000891                                   1   0.0001         True   \n",
       "ECK120002480/ECK120002482                      0   0.0000         True   \n",
       "ECK125144791                                   0   0.0000         True   \n",
       "\n",
       "                           corrected p value  corrected significance  \\\n",
       "ECK125229291                          0.0000                    True   \n",
       "ECK125229290                          0.0000                    True   \n",
       "ECK120001080                          0.0000                    True   \n",
       "ECK120002092                          0.0081                    True   \n",
       "ECK120017150                          0.0000                    True   \n",
       "ECK120013472                          0.0000                    True   \n",
       "ECK125095446                          0.0000                    True   \n",
       "ECK120034252                          0.0000                    True   \n",
       "ECK120003617                          0.0000                    True   \n",
       "ECK120003616                          0.0000                    True   \n",
       "ECK120023915                          0.0000                    True   \n",
       "ECK120002276                          0.0000                    True   \n",
       "ECK120000601                          0.0000                    True   \n",
       "ECK120000393                          0.0000                    True   \n",
       "ECK120000886                          0.0405                    True   \n",
       "ECK120000398/ECK120001878             0.0000                    True   \n",
       "ECK120000762                          0.0000                    True   \n",
       "ECK120000955                          0.0000                    True   \n",
       "ECK120000891                          0.0081                    True   \n",
       "ECK120002480/ECK120002482             0.0000                    True   \n",
       "ECK125144791                          0.0000                    True   \n",
       "\n",
       "                           final significance  \n",
       "ECK125229291                             True  \n",
       "ECK125229290                             True  \n",
       "ECK120001080                             True  \n",
       "ECK120002092                             True  \n",
       "ECK120017150                             True  \n",
       "ECK120013472                             True  \n",
       "ECK125095446                             True  \n",
       "ECK120034252                             True  \n",
       "ECK120003617                             True  \n",
       "ECK120003616                             True  \n",
       "ECK120023915                             True  \n",
       "ECK120002276                             True  \n",
       "ECK120000601                             True  \n",
       "ECK120000393                             True  \n",
       "ECK120000886                             True  \n",
       "ECK120000398/ECK120001878                True  \n",
       "ECK120000762                             True  \n",
       "ECK120000955                             True  \n",
       "ECK120000891                             True  \n",
       "ECK120002480/ECK120002482                True  \n",
       "ECK125144791                             True  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp_mut_df[comp_mut_df[\"final significance\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# writing significance status back into mutations DF.\n",
    "for i, r in assoc_muts_df.iterrows():\n",
    "    for d in r[\"genomic features\"]:\n",
    "        if d[\"RegulonDB ID\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"RegulonDB ID\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputting dataframe for the next pipeline step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(assoc_muts_df.exp.unique())==1)\n",
    "exp = assoc_muts_df.exp.unique()[0]\n",
    "comp_mut_df.to_csv(\"./data/supp/permutation_test_results_per_experiment/\"+exp+\"_mut_geno_feat_signif.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_muts_df.to_pickle(\"./data/6_df.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Currently commented out because some of the experiments have mutated features without names which will crash this cell & NB during whole pipeline execution. Uncomment when investigating individual experiments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 300\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # currently commented out because some of the experiments have mutated features without names\n",
    "# # which will crash this cell & NB during whole pipeline execution.\n",
    "# # Uncomment when investigating individual experiments.\n",
    "\n",
    "# TODO: found in multiple post-assoc NBs. Should extract into common script module.\n",
    "# def _get_name_signif_from_ID(RegulonDB_ID, json):\n",
    "#     name = \"\"\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"RegulonDB ID\"] == RegulonDB_ID:\n",
    "#             name = d[\"name\"]\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return (name, is_signif)\n",
    "\n",
    "\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     for feat_d in r[\"genomic features\"]:\n",
    "#         feat = feat_d[\"name\"]\n",
    "#         if \"attenuator terminator\" in feat:\n",
    "#             feat = feat.replace(\"attenuator terminator\", \"att term\")\n",
    "#         if \"terminator\" in feat:\n",
    "#             feat = feat.replace(\"terminator\", \"term\")\n",
    "        \n",
    "#         t = str(feat)\n",
    "#         if feat_d[\"significant\"]:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += 1\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-5:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,2),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.3, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 5 genomic mutated feature count\")\n",
    "# ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
