{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.gene import get_coding_genetic_target_len_d, get_intergenic_len_d\n",
    "from util.genome import get_feature_hit_set\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(148, 45)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>630</td>\n",
       "      <td>66</td>\n",
       "      <td>T247M (ACG→ATG)</td>\n",
       "      <td>yafD</td>\n",
       "      <td>SNP</td>\n",
       "      <td>231861</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>0.72</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(231861, 231861)</td>\n",
       "      <td>{ECK120001594}</td>\n",
       "      <td>[{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...</td>\n",
       "      <td>{'ECK120001594': ['ECK120001594']}</td>\n",
       "      <td>[{'name': 'yafDE', 'RegulonDB ID': 'ECK1200291...</td>\n",
       "      <td>{'ECK120029169': ['ECK120001594']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'General function prediction only'}]</td>\n",
       "      <td>{'General function prediction only': ['ECK1200...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>631</td>\n",
       "      <td>126</td>\n",
       "      <td>R665S (CGT→AGT)</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3993145</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>C→A</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3993145, 3993145)</td>\n",
       "      <td>{ECK120000166}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>{'ECK120000166': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...</td>\n",
       "      <td>{'ECK120014835': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'CRP': ['cyaA']}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}]</td>\n",
       "      <td>{'Purine metabolism': ['ECK120014835']}</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism'}]</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>632</td>\n",
       "      <td>94</td>\n",
       "      <td>S59Y (TCC→TAC)</td>\n",
       "      <td>glpK</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4117047</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→T</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4117047, 4117047)</td>\n",
       "      <td>{ECK120000391}</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism'}]</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion'}]</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>633</td>\n",
       "      <td>72</td>\n",
       "      <td>T593P (ACC→CCC)</td>\n",
       "      <td>metL</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4131611</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>0.67</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4131611, 4131611)</td>\n",
       "      <td>{ECK120000583}</td>\n",
       "      <td>[{'name': 'metL', 'RegulonDB ID': 'ECK12000058...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'metL', 'RegulonDB ID': 'ECK12000058...</td>\n",
       "      <td>{'ECK120000583': ['ECK120000583']}</td>\n",
       "      <td>[{'name': 'metBL', 'RegulonDB ID': 'ECK1200150...</td>\n",
       "      <td>{'ECK120015057': ['ECK120000583']}</td>\n",
       "      <td>[{'name': 'MetJ', 'significantly associated co...</td>\n",
       "      <td>{'MetJ': ['metBL'], 'PhoP': ['metBL'], 'HypT':...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism'}]</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>634</td>\n",
       "      <td>107</td>\n",
       "      <td>E641K (GAA→AAA)</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4183165</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→A</td>\n",
       "      <td>1</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4183165, 4183165)</td>\n",
       "      <td>{ECK120000885}</td>\n",
       "      <td>[{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000885': ['ECK120000885']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000885']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Purine metabolism'}, {'name': 'Pyri...</td>\n",
       "      <td>{'Purine metabolism': ['ECK120016992'], 'Pyrim...</td>\n",
       "      <td>[{'name': 'Transcription'}]</td>\n",
       "      <td>{'Transcription': ['ECK120016992']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index          Details mutation target annotation Mutation Type  Position  \\\n",
       "630    66  T247M (ACG→ATG)                       yafD           SNP    231861   \n",
       "631   126  R665S (CGT→AGT)                       cyaA           SNP   3993145   \n",
       "632    94   S59Y (TCC→TAC)                       glpK           SNP   4117047   \n",
       "633    72  T593P (ACC→CCC)                       metL           SNP   4131611   \n",
       "634   107  E641K (GAA→AAA)                       rpoB           SNP   4183165   \n",
       "\n",
       "    Reference Seq Sequence Change  ale  exp  flask  isolate  presence  \\\n",
       "630     NC_000913             C→T    1  GYD    280        0      0.72   \n",
       "631     NC_000913             C→A    1  GYD    280        0      1.00   \n",
       "632     NC_000913             G→T    1  GYD    280        0      1.00   \n",
       "633     NC_000913             A→C    1  GYD    280        0      0.67   \n",
       "634     NC_000913             G→A    1  GYD    280        0      1.00   \n",
       "\n",
       "     tech_rep  coding               range gene RegulonDB ID  \\\n",
       "630         1    True    (231861, 231861)    {ECK120001594}   \n",
       "631         1    True  (3993145, 3993145)    {ECK120000166}   \n",
       "632         1    True  (4117047, 4117047)    {ECK120000391}   \n",
       "633         1    True  (4131611, 4131611)    {ECK120000583}   \n",
       "634         1    True  (4183165, 4183165)    {ECK120000885}   \n",
       "\n",
       "                                      genetic features   oriC  pseudogene  \\\n",
       "630  [{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...  False       False   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...  False       False   \n",
       "632  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...  False       False   \n",
       "633  [{'name': 'metL', 'RegulonDB ID': 'ECK12000058...  False       False   \n",
       "634  [{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "\n",
       "    TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "630   {}       {}  {}                    {}         {}     True   \n",
       "631   {}       {}  {}                    {}         {}     True   \n",
       "632   {}       {}  {}                    {}         {}     True   \n",
       "633   {}       {}  {}                    {}         {}     True   \n",
       "634   {}       {}  {}                    {}         {}     True   \n",
       "\n",
       "                                      genomic features  \\\n",
       "630  [{'name': 'yafD', 'RegulonDB ID': 'ECK12000159...   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...   \n",
       "632  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "633  [{'name': 'metL', 'RegulonDB ID': 'ECK12000058...   \n",
       "634  [{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...   \n",
       "\n",
       "                  genetic feature links  \\\n",
       "630  {'ECK120001594': ['ECK120001594']}   \n",
       "631  {'ECK120000166': ['ECK120000166']}   \n",
       "632  {'ECK120000391': ['ECK120000391']}   \n",
       "633  {'ECK120000583': ['ECK120000583']}   \n",
       "634  {'ECK120000885': ['ECK120000885']}   \n",
       "\n",
       "                                               operons  \\\n",
       "630  [{'name': 'yafDE', 'RegulonDB ID': 'ECK1200291...   \n",
       "631  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...   \n",
       "632  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "633  [{'name': 'metBL', 'RegulonDB ID': 'ECK1200150...   \n",
       "634  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "\n",
       "                           operon links  \\\n",
       "630  {'ECK120029169': ['ECK120001594']}   \n",
       "631  {'ECK120014835': ['ECK120000166']}   \n",
       "632  {'ECK120014723': ['ECK120000391']}   \n",
       "633  {'ECK120015057': ['ECK120000583']}   \n",
       "634  {'ECK120016992': ['ECK120000885']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "630                                                 []   \n",
       "631  [{'name': 'CRP', 'significantly associated con...   \n",
       "632  [{'name': 'GlpR', 'significantly associated co...   \n",
       "633  [{'name': 'MetJ', 'significantly associated co...   \n",
       "634                                                 []   \n",
       "\n",
       "                                       regulator links  \\\n",
       "630                                                 {}   \n",
       "631                                  {'CRP': ['cyaA']}   \n",
       "632            {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "633  {'MetJ': ['metBL'], 'PhoP': ['metBL'], 'HypT':...   \n",
       "634                                                 {}   \n",
       "\n",
       "                                              pathways  \\\n",
       "630                                                 []   \n",
       "631                    [{'name': 'Purine metabolism'}]   \n",
       "632              [{'name': 'Glycerolipid metabolism'}]   \n",
       "633  [{'name': 'Glycine, serine and threonine metab...   \n",
       "634  [{'name': 'Purine metabolism'}, {'name': 'Pyri...   \n",
       "\n",
       "                                         pathway links  \\\n",
       "630                                                 {}   \n",
       "631            {'Purine metabolism': ['ECK120014835']}   \n",
       "632      {'Glycerolipid metabolism': ['ECK120014723']}   \n",
       "633  {'Glycine, serine and threonine metabolism': [...   \n",
       "634  {'Purine metabolism': ['ECK120016992'], 'Pyrim...   \n",
       "\n",
       "                                                  COGs  \\\n",
       "630     [{'name': 'General function prediction only'}]   \n",
       "631  [{'name': 'Nucleotide transport and metabolism'}]   \n",
       "632     [{'name': 'Energy production and conversion'}]   \n",
       "633  [{'name': 'Amino acid transport and metabolism'}]   \n",
       "634                        [{'name': 'Transcription'}]   \n",
       "\n",
       "                                             COG links temperature  \\\n",
       "630  {'General function prediction only': ['ECK1200...  37 celsius   \n",
       "631  {'Nucleotide transport and metabolism': ['ECK1...  37 celsius   \n",
       "632  {'Energy production and conversion': ['ECK1200...  37 celsius   \n",
       "633  {'Amino acid transport and metabolism': ['ECK1...  37 celsius   \n",
       "634                {'Transcription': ['ECK120016992']}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "630  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "631  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "632  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "633  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "634  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "    taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "630      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "631      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "632      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "633      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "634      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "    sulfur-source calcium-source  \n",
       "630   MgSO4(0.24)     CaCl2(0.1)  \n",
       "631   MgSO4(0.24)     CaCl2(0.1)  \n",
       "632   MgSO4(0.24)     CaCl2(0.1)  \n",
       "633   MgSO4(0.24)     CaCl2(0.1)  \n",
       "634   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "assoc_muts_df = pd.read_pickle(\"./data/6_df.pkl\")\n",
    "display(assoc_muts_df.shape, assoc_muts_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build genetic feature links dictionary for mutflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Now in NB 2.2.1. Remove cell becuause obsolete\n",
    "# def get_genetic_links_d(mut_row):\n",
    "#     gen_trgt_d = dict()\n",
    "#     feat_df = pd.DataFrame()\n",
    "#     for feat_d in mut_row[\"genomic features\"]:\n",
    "#         feat_df = feat_df.append(feat_d, ignore_index=True)\n",
    "#     for gen_trgt in mut_row[\"genetic features\"]:\n",
    "#         gen_trgt_d[gen_trgt[\"RegulonDB ID\"]] = list(get_feature_hit_set(gen_trgt[\"range\"], feat_df, \"range\", \"RegulonDB ID\"))\n",
    "#     return gen_trgt_d\n",
    "\n",
    "\n",
    "# assoc_muts_df[\"genetic feature links\"] = assoc_muts_df.apply(get_genetic_links_d, axis=1)\n",
    "# assoc_muts_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Significance study (level specific)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120001594</td>\n",
       "      <td>801</td>\n",
       "      <td>1</td>\n",
       "      <td>yafD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000166</td>\n",
       "      <td>2547</td>\n",
       "      <td>21</td>\n",
       "      <td>cyaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000391</td>\n",
       "      <td>1509</td>\n",
       "      <td>30</td>\n",
       "      <td>glpK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000583</td>\n",
       "      <td>2433</td>\n",
       "      <td>2</td>\n",
       "      <td>metL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000885</td>\n",
       "      <td>4029</td>\n",
       "      <td>5</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120003261</td>\n",
       "      <td>1032</td>\n",
       "      <td>1</td>\n",
       "      <td>ynaI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120003260/ECK120003261</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>mppA/ynaI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120003259/ECK120003260</td>\n",
       "      <td>336</td>\n",
       "      <td>4</td>\n",
       "      <td>pgrR/mppA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120003261/ECK120003262</td>\n",
       "      <td>153</td>\n",
       "      <td>4</td>\n",
       "      <td>ynaI/insH-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120004249</td>\n",
       "      <td>1482</td>\n",
       "      <td>1</td>\n",
       "      <td>gspE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          length observed mutation count         name\n",
       "ECK120001594                 801                       1         yafD\n",
       "ECK120000166                2547                      21         cyaA\n",
       "ECK120000391                1509                      30         glpK\n",
       "ECK120000583                2433                       2         metL\n",
       "ECK120000885                4029                       5         rpoB\n",
       "...                          ...                     ...          ...\n",
       "ECK120003261                1032                       1         ynaI\n",
       "ECK120003260/ECK120003261     50                       0    mppA/ynaI\n",
       "ECK120003259/ECK120003260    336                       4    pgrR/mppA\n",
       "ECK120003261/ECK120003262    153                       4  ynaI/insH-4\n",
       "ECK120004249                1482                       1         gspE\n",
       "\n",
       "[74 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from util.feature import get_feat_mut_cnt_df_from_links\n",
    "comp_mut_df = get_feat_mut_cnt_df_from_links(assoc_muts_df, \"genetic features\", \"genetic feature links\")\n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in assoc_muts_df.iterrows():\n",
    "    for feat_d in r[\"genetic features\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"RegulonDB ID\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120001594</td>\n",
       "      <td>801</td>\n",
       "      <td>1</td>\n",
       "      <td>yafD</td>\n",
       "      <td>0.000173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000166</td>\n",
       "      <td>2547</td>\n",
       "      <td>21</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>0.000549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000391</td>\n",
       "      <td>1509</td>\n",
       "      <td>30</td>\n",
       "      <td>glpK</td>\n",
       "      <td>0.000325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000583</td>\n",
       "      <td>2433</td>\n",
       "      <td>2</td>\n",
       "      <td>metL</td>\n",
       "      <td>0.000524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000885</td>\n",
       "      <td>4029</td>\n",
       "      <td>5</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.000868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count  name  proportion\n",
       "ECK120001594    801                       1  yafD    0.000173\n",
       "ECK120000166   2547                      21  cyaA    0.000549\n",
       "ECK120000391   1509                      30  glpK    0.000325\n",
       "ECK120000583   2433                       2  metL    0.000524\n",
       "ECK120000885   4029                       5  rpoB    0.000868"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "\n",
    "TOTAL_GENOME_LEN = 4641652\n",
    "genome_remainder_len = TOTAL_GENOME_LEN - mutated_seq_len\n",
    "df = pd.DataFrame([[0, genome_remainder_len]], columns=[\"observed mutation count\", \"length\"], index=[\"genome remainder\"])\n",
    "comp_mut_df = comp_mut_df.append(df, sort=False)\n",
    "total_seq_len = TOTAL_GENOME_LEN\n",
    "\n",
    "\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120001594</td>\n",
       "      <td>801</td>\n",
       "      <td>1</td>\n",
       "      <td>yafD</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>306</td>\n",
       "      <td>0.0306</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000166</td>\n",
       "      <td>2547</td>\n",
       "      <td>21</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000391</td>\n",
       "      <td>1509</td>\n",
       "      <td>30</td>\n",
       "      <td>glpK</td>\n",
       "      <td>0.000325</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000583</td>\n",
       "      <td>2433</td>\n",
       "      <td>2</td>\n",
       "      <td>metL</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>43</td>\n",
       "      <td>0.0043</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000885</td>\n",
       "      <td>4029</td>\n",
       "      <td>5</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.000868</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count  name  proportion  \\\n",
       "ECK120001594    801                       1  yafD    0.000173   \n",
       "ECK120000166   2547                      21  cyaA    0.000549   \n",
       "ECK120000391   1509                      30  glpK    0.000325   \n",
       "ECK120000583   2433                       2  metL    0.000524   \n",
       "ECK120000885   4029                       5  rpoB    0.000868   \n",
       "\n",
       "              equal or larger count  p value  significant  \n",
       "ECK120001594                    306   0.0306         True  \n",
       "ECK120000166                      0   0.0000         True  \n",
       "ECK120000391                      0   0.0000         True  \n",
       "ECK120000583                     43   0.0043         True  \n",
       "ECK120000885                      0   0.0000         True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=num_muts,\n",
    "        replace=True,\n",
    "        p=proportion_l\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120001594</td>\n",
       "      <td>801</td>\n",
       "      <td>1</td>\n",
       "      <td>yafD</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>306</td>\n",
       "      <td>0.0306</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000166</td>\n",
       "      <td>2547</td>\n",
       "      <td>21</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000391</td>\n",
       "      <td>1509</td>\n",
       "      <td>30</td>\n",
       "      <td>glpK</td>\n",
       "      <td>0.000325</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000583</td>\n",
       "      <td>2433</td>\n",
       "      <td>2</td>\n",
       "      <td>metL</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>43</td>\n",
       "      <td>0.0043</td>\n",
       "      <td>True</td>\n",
       "      <td>0.3225</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120000885</td>\n",
       "      <td>4029</td>\n",
       "      <td>5</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>0.000868</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length observed mutation count  name  proportion  \\\n",
       "ECK120001594    801                       1  yafD    0.000173   \n",
       "ECK120000166   2547                      21  cyaA    0.000549   \n",
       "ECK120000391   1509                      30  glpK    0.000325   \n",
       "ECK120000583   2433                       2  metL    0.000524   \n",
       "ECK120000885   4029                       5  rpoB    0.000868   \n",
       "\n",
       "              equal or larger count  p value  significant  corrected p value  \\\n",
       "ECK120001594                    306   0.0306         True             1.0000   \n",
       "ECK120000166                      0   0.0000         True             0.0000   \n",
       "ECK120000391                      0   0.0000         True             0.0000   \n",
       "ECK120000583                     43   0.0043         True             0.3225   \n",
       "ECK120000885                      0   0.0000         True             0.0000   \n",
       "\n",
       "              corrected significance  \n",
       "ECK120001594                   False  \n",
       "ECK120000166                    True  \n",
       "ECK120000391                    True  \n",
       "ECK120000583                   False  \n",
       "ECK120000885                    True  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # not wanting to consider mutation events with only 1 features as being significant\n",
    "comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "# comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, r in assoc_muts_df.iterrows():\n",
    "    for d in r[\"genetic features\"]:\n",
    "        if d[\"RegulonDB ID\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"RegulonDB ID\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputting dataframe for the next step in the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(assoc_muts_df.exp.unique())==1)\n",
    "exp = assoc_muts_df.exp.unique()[0]\n",
    "comp_mut_df.to_csv(\"./data/supp/permutation_test_results_per_experiment/\"+exp+\"_mut_gen_feat_signif.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_muts_df.to_pickle(\"./data/7_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_ID(RegulonDB_ID, json):\n",
    "#     name = \"\"\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"RegulonDB ID\"] == RegulonDB_ID:\n",
    "#             name = d[\"name\"]\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return (name, is_signif)\n",
    "\n",
    "\n",
    "# # The reason we're using the links dictionary is because we want to count how genomic features were mutated for the genetic target.\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in assoc_muts_df.iterrows():\n",
    "#     link_d = r[\"genetic feature links\"]\n",
    "#     for targ_ID, hits in link_d.items():\n",
    "#         t, is_signif = _get_name_signif_from_ID(targ_ID, r[\"genetic features\"])\n",
    "#         t = str(t)\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += len(hits)\n",
    "        \n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\",\n",
    "#                   width=1\n",
    "#                  )\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.2, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 genetic mutated feature count\")\n",
    "# ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
