{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "import os, sys, itertools, csv\n",
    "pd.options.display.max_columns = 100\n",
    "\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from util.genome import get_feature_hit_set, is_overlap\n",
    "from util.operon import get_operon_name_set, get_operon_ID_set\n",
    "from util.feature import get_feat_d_from_ID\n",
    "from util.params import ENRICH_ALPHA, MULTI_HYP_CORR_METHOD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1232</td>\n",
       "      <td>4</td>\n",
       "      <td>R121C (CGC→TGC)</td>\n",
       "      <td>xylR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3735339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3735339, 3735339)</td>\n",
       "      <td>{ECK120002449}</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...</td>\n",
       "      <td>{'ECK120002449': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014482': ['ECK120002449']}</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1233</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1234</td>\n",
       "      <td>6</td>\n",
       "      <td>L770R (CTC→CGC)</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4187658</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→G</td>\n",
       "      <td>1</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>116</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(4187658, 4187658)</td>\n",
       "      <td>{ECK120000886}</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>{'ECK120000886': ['ECK120000886']}</td>\n",
       "      <td>[{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...</td>\n",
       "      <td>{'ECK120016992': ['ECK120000886']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1235</td>\n",
       "      <td>9</td>\n",
       "      <td>G164A (GGC→GCC)</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3211597</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→C</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3211597, 3211597)</td>\n",
       "      <td>{ECK120000233}</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...</td>\n",
       "      <td>{'ECK120000233': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120014515': ['ECK120000233']}</td>\n",
       "      <td>[{'name': 'LexA', 'significantly associated co...</td>\n",
       "      <td>{'LexA': ['rpsU-dnaG-rpoD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1236</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>rph</td>\n",
       "      <td>DEL</td>\n",
       "      <td>3815859</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ82 bp</td>\n",
       "      <td>2</td>\n",
       "      <td>SSW_XYL</td>\n",
       "      <td>109</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(3815859, 3815940)</td>\n",
       "      <td>{ECK120000854}</td>\n",
       "      <td>[{'name': 'rph', 'RegulonDB ID': 'ECK120000854...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125144791}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'rph-pyrE attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120000854': ['ECK125144791', 'ECK12000085...</td>\n",
       "      <td>[{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014627': ['ECK120000854', 'ECK12000085...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>xylose(4)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index          Details mutation target annotation Mutation Type  \\\n",
       "1232     4  R121C (CGC→TGC)                       xylR           SNP   \n",
       "1233     5             None                        rph           DEL   \n",
       "1234     6  L770R (CTC→CGC)                       rpoC           SNP   \n",
       "1235     9  G164A (GGC→GCC)                       dnaG           SNP   \n",
       "1236     5             None                        rph           DEL   \n",
       "\n",
       "      Position Reference Seq Sequence Change  ale      exp  flask  isolate  \\\n",
       "1232   3735339           NaN             C→T    1  SSW_XYL    116        1   \n",
       "1233   3815859           NaN          Δ82 bp    1  SSW_XYL    116        1   \n",
       "1234   4187658           NaN             T→G    1  SSW_XYL    116        1   \n",
       "1235   3211597           NaN             G→C    2  SSW_XYL    109        1   \n",
       "1236   3815859           NaN          Δ82 bp    2  SSW_XYL    109        1   \n",
       "\n",
       "      presence  tech_rep  coding               range gene RegulonDB ID  \\\n",
       "1232       1.0         1    True  (3735339, 3735339)    {ECK120002449}   \n",
       "1233       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "1234       1.0         1    True  (4187658, 4187658)    {ECK120000886}   \n",
       "1235       1.0         1    True  (3211597, 3211597)    {ECK120000233}   \n",
       "1236       1.0         1    True  (3815859, 3815940)    {ECK120000854}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...  False       False   \n",
       "1233  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...  False       False   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...  False       False   \n",
       "1236  [{'name': 'rph', 'RegulonDB ID': 'ECK120000854...  False       False   \n",
       "\n",
       "     TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "1232   {}       {}  {}                    {}         {}     True   \n",
       "1233   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "1234   {}       {}  {}                    {}         {}     True   \n",
       "1235   {}       {}  {}                    {}         {}     True   \n",
       "1236   {}       {}  {}        {ECK125144791}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1232  [{'name': 'xylR', 'RegulonDB ID': 'ECK12000244...   \n",
       "1233  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "1234  [{'name': 'rpoC', 'RegulonDB ID': 'ECK12000088...   \n",
       "1235  [{'name': 'dnaG', 'RegulonDB ID': 'ECK12000023...   \n",
       "1236  [{'name': 'rph-pyrE attenuator terminator', 'R...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1232                 {'ECK120002449': ['ECK120002449']}   \n",
       "1233  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "1234                 {'ECK120000886': ['ECK120000886']}   \n",
       "1235                 {'ECK120000233': ['ECK120000233']}   \n",
       "1236  {'ECK120000854': ['ECK125144791', 'ECK12000085...   \n",
       "\n",
       "                                                operons  \\\n",
       "1232  [{'name': 'xylFGHR', 'RegulonDB ID': 'ECK12001...   \n",
       "1233  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "1234  [{'name': 'rplKAJL-rpoBC', 'RegulonDB ID': 'EC...   \n",
       "1235  [{'name': 'rpsU-dnaG-rpoD', 'RegulonDB ID': 'E...   \n",
       "1236  [{'name': 'rph-pyrE', 'RegulonDB ID': 'ECK1200...   \n",
       "\n",
       "                                           operon links  \\\n",
       "1232                 {'ECK120014482': ['ECK120002449']}   \n",
       "1233  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "1234                 {'ECK120016992': ['ECK120000886']}   \n",
       "1235                 {'ECK120014515': ['ECK120000233']}   \n",
       "1236  {'ECK120014627': ['ECK120000854', 'ECK12000085...   \n",
       "\n",
       "                                             regulators  \\\n",
       "1232  [{'name': 'Fis', 'significantly associated con...   \n",
       "1233                                                 []   \n",
       "1234                                                 []   \n",
       "1235  [{'name': 'LexA', 'significantly associated co...   \n",
       "1236                                                 []   \n",
       "\n",
       "                                        regulator links temperature  \\\n",
       "1232  {'ArcA': ['xylFGHR'], 'Fis': ['xylFGHR'], 'CRP...  37 celsius   \n",
       "1233                                                 {}  37 celsius   \n",
       "1234                                                 {}  37 celsius   \n",
       "1235                       {'LexA': ['rpsU-dnaG-rpoD']}  37 celsius   \n",
       "1236                                                 {}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "1232     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1233     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1234     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1235     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "1236     xylose(4)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "     taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "1232      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1233      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1234      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1235      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "1236      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "     sulfur-source calcium-source  \n",
       "1232   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1233   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1234   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1235   MgSO4(0.24)     CaCl2(0.1)  \n",
       "1236   MgSO4(0.24)     CaCl2(0.1)  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "exp_muts_df = pd.read_pickle(\"./data/7_df.pkl\")\n",
    "\n",
    "# exp_muts_df = pd.read_pickle(\"./data/2_4_df.pkl\")  # debug\n",
    "# exp_muts_df = exp_muts_df[exp_muts_df.exp==\"SER\"]\n",
    "# exp_muts_df = exp_muts_df[exp_muts_df[\"mutation target annotation\"]==\"grxB/mdtH\"]\n",
    "\n",
    "display(len(exp_muts_df), exp_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OPERON_ID</th>\n",
       "      <th>OPERON_NAME</th>\n",
       "      <th>FIRSTGENEPOSLEFT</th>\n",
       "      <th>LASTGENEPOSRIGHT</th>\n",
       "      <th>REGULATIONPOSLEFT</th>\n",
       "      <th>REGULATIONPOSRIGHT</th>\n",
       "      <th>OPERON_STRAND</th>\n",
       "      <th>OPERON_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ECK120011191</td>\n",
       "      <td>rob</td>\n",
       "      <td>4634441</td>\n",
       "      <td>4635310</td>\n",
       "      <td>4634441</td>\n",
       "      <td>4635382</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(4634441, 4635382)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ECK120011203</td>\n",
       "      <td>argP</td>\n",
       "      <td>3059753</td>\n",
       "      <td>3060646</td>\n",
       "      <td>3059684</td>\n",
       "      <td>3060680</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3059684, 3060680)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ECK120011329</td>\n",
       "      <td>lrhA</td>\n",
       "      <td>2405703</td>\n",
       "      <td>2406641</td>\n",
       "      <td>2405703</td>\n",
       "      <td>2406866</td>\n",
       "      <td>reverse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(2405703, 2406866)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ECK120011490</td>\n",
       "      <td>rtcR</td>\n",
       "      <td>3558268</td>\n",
       "      <td>3559866</td>\n",
       "      <td>3558268</td>\n",
       "      <td>3559866</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(3558268, 3559866)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ECK120011629</td>\n",
       "      <td>lrp</td>\n",
       "      <td>932595</td>\n",
       "      <td>933089</td>\n",
       "      <td>932265</td>\n",
       "      <td>933089</td>\n",
       "      <td>forward</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>(932265, 933089)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      OPERON_ID OPERON_NAME  FIRSTGENEPOSLEFT  LASTGENEPOSRIGHT  \\\n",
       "0  ECK120011191         rob           4634441           4635310   \n",
       "1  ECK120011203        argP           3059753           3060646   \n",
       "2  ECK120011329        lrhA           2405703           2406641   \n",
       "3  ECK120011490        rtcR           3558268           3559866   \n",
       "4  ECK120011629         lrp            932595            933089   \n",
       "\n",
       "   REGULATIONPOSLEFT  REGULATIONPOSRIGHT OPERON_STRAND  \\\n",
       "0            4634441             4635382       reverse   \n",
       "1            3059684             3060680       forward   \n",
       "2            2405703             2406866       reverse   \n",
       "3            3558268             3559866       forward   \n",
       "4             932265              933089       forward   \n",
       "\n",
       "   OPERON_INTERNAL_COMMENT KEY_ID_ORG               range  \n",
       "0                      NaN      ECK12  (4634441, 4635382)  \n",
       "1                      NaN      ECK12  (3059684, 3060680)  \n",
       "2                      NaN      ECK12  (2405703, 2406866)  \n",
       "3                      NaN      ECK12  (3558268, 3559866)  \n",
       "4                      NaN      ECK12    (932265, 933089)  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "operon_df = pd.read_pickle(\"./data/operon_df.pkl\")\n",
    "operon_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRANSCRIPTION_UNIT_ID</th>\n",
       "      <th>NUMTU</th>\n",
       "      <th>TU_POSLEFT</th>\n",
       "      <th>TU_POSRIGHT</th>\n",
       "      <th>TU_TYPE</th>\n",
       "      <th>TU_OBJECT_CLASS</th>\n",
       "      <th>TU_OBJECT_ID</th>\n",
       "      <th>TU_OBJECT_NAME</th>\n",
       "      <th>TU_OBJECT_POSLEFT</th>\n",
       "      <th>TU_OBJECT_POSRIGHT</th>\n",
       "      <th>TU_OBJECT_STRAND</th>\n",
       "      <th>TU_OBJECT_COLORCLASS</th>\n",
       "      <th>TU_OBJECT_DESCRIPTION</th>\n",
       "      <th>TU_OBJECT_SIGMA</th>\n",
       "      <th>TU_OBJECT_EVIDENCE</th>\n",
       "      <th>TU_OBJECT_RI_TYPE</th>\n",
       "      <th>TU_OBJECT_TYPE</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>TU range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>PM</td>\n",
       "      <td>ECK120009851</td>\n",
       "      <td>astCp1</td>\n",
       "      <td>1832013</td>\n",
       "      <td>1832013</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sigma70</td>\n",
       "      <td>Human inference of promoter position</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1825955, 1832013)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003528</td>\n",
       "      <td>astE</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1826923</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1825955, 1832013)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003529</td>\n",
       "      <td>astB</td>\n",
       "      <td>1826916</td>\n",
       "      <td>1828259</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1825955, 1832013)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003532</td>\n",
       "      <td>astC</td>\n",
       "      <td>1830762</td>\n",
       "      <td>1831982</td>\n",
       "      <td>R</td>\n",
       "      <td>0,0,255</td>\n",
       "      <td>nitrogen metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1825955, 1832013)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ECK120008913</td>\n",
       "      <td>3</td>\n",
       "      <td>1825955</td>\n",
       "      <td>1832013</td>\n",
       "      <td>H</td>\n",
       "      <td>GN</td>\n",
       "      <td>ECK120003530</td>\n",
       "      <td>astD</td>\n",
       "      <td>1828256</td>\n",
       "      <td>1829734</td>\n",
       "      <td>R</td>\n",
       "      <td>51,153,255</td>\n",
       "      <td>amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(1825955, 1832013)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  TRANSCRIPTION_UNIT_ID  NUMTU  TU_POSLEFT  TU_POSRIGHT TU_TYPE  \\\n",
       "0          ECK120008913      3     1825955      1832013       H   \n",
       "1          ECK120008913      3     1825955      1832013       H   \n",
       "2          ECK120008913      3     1825955      1832013       H   \n",
       "3          ECK120008913      3     1825955      1832013       H   \n",
       "4          ECK120008913      3     1825955      1832013       H   \n",
       "\n",
       "  TU_OBJECT_CLASS  TU_OBJECT_ID TU_OBJECT_NAME  TU_OBJECT_POSLEFT  \\\n",
       "0              PM  ECK120009851         astCp1            1832013   \n",
       "1              GN  ECK120003528           astE            1825955   \n",
       "2              GN  ECK120003529           astB            1826916   \n",
       "3              GN  ECK120003532           astC            1830762   \n",
       "4              GN  ECK120003530           astD            1828256   \n",
       "\n",
       "   TU_OBJECT_POSRIGHT TU_OBJECT_STRAND TU_OBJECT_COLORCLASS  \\\n",
       "0             1832013                R                  NaN   \n",
       "1             1826923                R           51,153,255   \n",
       "2             1828259                R           51,153,255   \n",
       "3             1831982                R              0,0,255   \n",
       "4             1829734                R           51,153,255   \n",
       "\n",
       "  TU_OBJECT_DESCRIPTION TU_OBJECT_SIGMA                    TU_OBJECT_EVIDENCE  \\\n",
       "0                   NaN         Sigma70  Human inference of promoter position   \n",
       "1           amino acids             NaN                                   NaN   \n",
       "2           amino acids             NaN                                   NaN   \n",
       "3   nitrogen metabolism             NaN                                   NaN   \n",
       "4           amino acids             NaN                                   NaN   \n",
       "\n",
       "  TU_OBJECT_RI_TYPE TU_OBJECT_TYPE  EVIDENCE            TU range  \n",
       "0               NaN      predicted       NaN  (1825955, 1832013)  \n",
       "1               NaN      predicted       NaN  (1825955, 1832013)  \n",
       "2               NaN      predicted       NaN  (1825955, 1832013)  \n",
       "3               NaN      predicted       NaN  (1825955, 1832013)  \n",
       "4               NaN      predicted       NaN  (1825955, 1832013)  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TU_df = pd.read_csv(\"./data/RegulonDB10/transcription_unit.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "TU_df.columns = [\n",
    "    \"TRANSCRIPTION_UNIT_ID\",\n",
    "    \"PROMOTER_ID\",\n",
    "    \"TRANSCRIPTION_UNIT_NAME\",\n",
    "    \"OPERON_ID\",\n",
    "    \"TRANSCRIPTION_UNIT_NOTE\",\n",
    "    \"TU_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "\n",
    "\n",
    "TU_objects_df = pd.read_csv(\"./data/RegulonDB10/tu_objects_tmp.txt\",\n",
    "                            sep=\"\\t\",\n",
    "                            comment='#',\n",
    "                            header=None,\n",
    "                            )\n",
    "\n",
    "TU_objects_df.columns = [\n",
    "    \"TRANSCRIPTION_UNIT_ID\",\n",
    "    \"NUMTU\",\n",
    "    \"TU_POSLEFT\",\n",
    "    \"TU_POSRIGHT\",\n",
    "    \"TU_TYPE\",\n",
    "    \"TU_OBJECT_CLASS\",\n",
    "    \"TU_OBJECT_ID\",\n",
    "    \"TU_OBJECT_NAME\",\n",
    "    \"TU_OBJECT_POSLEFT\",\n",
    "    \"TU_OBJECT_POSRIGHT\",\n",
    "    \"TU_OBJECT_STRAND\",\n",
    "    \"TU_OBJECT_COLORCLASS\",\n",
    "    \"TU_OBJECT_DESCRIPTION\",\n",
    "    \"TU_OBJECT_SIGMA\",\n",
    "    \"TU_OBJECT_EVIDENCE\",\n",
    "    \"TU_OBJECT_RI_TYPE\",\n",
    "    \"TU_OBJECT_TYPE\",\n",
    "    \"EVIDENCE\"\n",
    "]\n",
    "\n",
    "TU_objects_df[\"TU range\"] = TU_objects_df.apply(\n",
    "    lambda row: (row[\"TU_POSLEFT\"], row[\"TU_POSRIGHT\"]), axis=1)\n",
    "\n",
    "TU_objects_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>operon description</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>unknown</td>\n",
       "      <td>265905</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  operon description  length\n",
       "0            unknown  265905"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "op_len_df = pd.read_pickle(\"./data/operon_length_df.pkl\")\n",
    "op_len_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from util.feature import get_feat_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120014482</td>\n",
       "      <td>2</td>\n",
       "      <td>xylFGHR</td>\n",
       "      <td>5280.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014627</td>\n",
       "      <td>12</td>\n",
       "      <td>rph-pyrE</td>\n",
       "      <td>1485.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120016992</td>\n",
       "      <td>3</td>\n",
       "      <td>rplKAJL-rpoBC</td>\n",
       "      <td>11376.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014515</td>\n",
       "      <td>1</td>\n",
       "      <td>rpsU-dnaG-rpoD</td>\n",
       "      <td>4317.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120020905</td>\n",
       "      <td>1</td>\n",
       "      <td>fimAICDFGH</td>\n",
       "      <td>7075.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028518</td>\n",
       "      <td>1</td>\n",
       "      <td>eutHA</td>\n",
       "      <td>2627.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014638</td>\n",
       "      <td>1</td>\n",
       "      <td>rpsMKD-rpoA-rplQ</td>\n",
       "      <td>3064.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120016920</td>\n",
       "      <td>2</td>\n",
       "      <td>rhoL-rho</td>\n",
       "      <td>1515.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120034085</td>\n",
       "      <td>1</td>\n",
       "      <td>ybeZYX-lnt</td>\n",
       "      <td>4245.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120029041</td>\n",
       "      <td>1</td>\n",
       "      <td>fabI</td>\n",
       "      <td>927.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028016</td>\n",
       "      <td>1</td>\n",
       "      <td>ydfUT</td>\n",
       "      <td>1816.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120023744</td>\n",
       "      <td>1</td>\n",
       "      <td>ydhQ</td>\n",
       "      <td>1332.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028746</td>\n",
       "      <td>1</td>\n",
       "      <td>ydiZ</td>\n",
       "      <td>291.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014934</td>\n",
       "      <td>1</td>\n",
       "      <td>flhDC</td>\n",
       "      <td>1312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028823</td>\n",
       "      <td>1</td>\n",
       "      <td>yeeJ</td>\n",
       "      <td>7125.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028632</td>\n",
       "      <td>1</td>\n",
       "      <td>yegX</td>\n",
       "      <td>866.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014834</td>\n",
       "      <td>1</td>\n",
       "      <td>cirA</td>\n",
       "      <td>2196.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014693</td>\n",
       "      <td>1</td>\n",
       "      <td>upp-uraA</td>\n",
       "      <td>2107.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014491</td>\n",
       "      <td>1</td>\n",
       "      <td>mutS</td>\n",
       "      <td>2636.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028479</td>\n",
       "      <td>1</td>\n",
       "      <td>ygjJK</td>\n",
       "      <td>3439.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125160702</td>\n",
       "      <td>1</td>\n",
       "      <td>yrbG-kdsDC-lptCAB-rpoN-hpf-ptsN-rapZ-npr</td>\n",
       "      <td>8186.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028315</td>\n",
       "      <td>1</td>\n",
       "      <td>gspCDEFGHIJKLMO</td>\n",
       "      <td>10749.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014704</td>\n",
       "      <td>1</td>\n",
       "      <td>livKHMGF</td>\n",
       "      <td>5032.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028344</td>\n",
       "      <td>3</td>\n",
       "      <td>rhsB</td>\n",
       "      <td>4236.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120026417</td>\n",
       "      <td>1</td>\n",
       "      <td>lldPRD</td>\n",
       "      <td>3837.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028201</td>\n",
       "      <td>1</td>\n",
       "      <td>ravA-viaA</td>\n",
       "      <td>3185.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028078</td>\n",
       "      <td>1</td>\n",
       "      <td>mhpR-lacI</td>\n",
       "      <td>2223.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028229</td>\n",
       "      <td>1</td>\n",
       "      <td>yihLM</td>\n",
       "      <td>1959.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125229275</td>\n",
       "      <td>1</td>\n",
       "      <td>pdeC</td>\n",
       "      <td>1640.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120028108</td>\n",
       "      <td>1</td>\n",
       "      <td>ytfQRT-yjfF</td>\n",
       "      <td>4617.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014757</td>\n",
       "      <td>1</td>\n",
       "      <td>treBC</td>\n",
       "      <td>3401.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK125165366</td>\n",
       "      <td>1</td>\n",
       "      <td>opgB</td>\n",
       "      <td>2320.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120029053</td>\n",
       "      <td>1</td>\n",
       "      <td>rhsD-ybbC-ylbH-ybbD</td>\n",
       "      <td>5640.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              observed mutation count  \\\n",
       "ECK120014482                        2   \n",
       "ECK120014627                       12   \n",
       "ECK120016992                        3   \n",
       "ECK120014515                        1   \n",
       "ECK120020905                        1   \n",
       "ECK120028518                        1   \n",
       "ECK120014638                        1   \n",
       "ECK120016920                        2   \n",
       "ECK120034085                        1   \n",
       "ECK120029041                        1   \n",
       "ECK120028016                        1   \n",
       "ECK120023744                        1   \n",
       "ECK120028746                        1   \n",
       "ECK120014934                        1   \n",
       "ECK120028823                        1   \n",
       "ECK120028632                        1   \n",
       "ECK120014834                        1   \n",
       "ECK120014693                        1   \n",
       "ECK120014491                        1   \n",
       "ECK120028479                        1   \n",
       "ECK125160702                        1   \n",
       "ECK120028315                        1   \n",
       "ECK120014704                        1   \n",
       "ECK120028344                        3   \n",
       "ECK120026417                        1   \n",
       "ECK120028201                        1   \n",
       "ECK120028078                        1   \n",
       "ECK120028229                        1   \n",
       "ECK125229275                        1   \n",
       "ECK120028108                        1   \n",
       "ECK120014757                        1   \n",
       "ECK125165366                        1   \n",
       "ECK120029053                        1   \n",
       "\n",
       "                                                  name   length  \n",
       "ECK120014482                                   xylFGHR   5280.0  \n",
       "ECK120014627                                  rph-pyrE   1485.0  \n",
       "ECK120016992                             rplKAJL-rpoBC  11376.0  \n",
       "ECK120014515                            rpsU-dnaG-rpoD   4317.0  \n",
       "ECK120020905                                fimAICDFGH   7075.0  \n",
       "ECK120028518                                     eutHA   2627.0  \n",
       "ECK120014638                          rpsMKD-rpoA-rplQ   3064.0  \n",
       "ECK120016920                                  rhoL-rho   1515.0  \n",
       "ECK120034085                                ybeZYX-lnt   4245.0  \n",
       "ECK120029041                                      fabI    927.0  \n",
       "ECK120028016                                     ydfUT   1816.0  \n",
       "ECK120023744                                      ydhQ   1332.0  \n",
       "ECK120028746                                      ydiZ    291.0  \n",
       "ECK120014934                                     flhDC   1312.0  \n",
       "ECK120028823                                      yeeJ   7125.0  \n",
       "ECK120028632                                      yegX    866.0  \n",
       "ECK120014834                                      cirA   2196.0  \n",
       "ECK120014693                                  upp-uraA   2107.0  \n",
       "ECK120014491                                      mutS   2636.0  \n",
       "ECK120028479                                     ygjJK   3439.0  \n",
       "ECK125160702  yrbG-kdsDC-lptCAB-rpoN-hpf-ptsN-rapZ-npr   8186.0  \n",
       "ECK120028315                           gspCDEFGHIJKLMO  10749.0  \n",
       "ECK120014704                                  livKHMGF   5032.0  \n",
       "ECK120028344                                      rhsB   4236.0  \n",
       "ECK120026417                                    lldPRD   3837.0  \n",
       "ECK120028201                                 ravA-viaA   3185.0  \n",
       "ECK120028078                                 mhpR-lacI   2223.0  \n",
       "ECK120028229                                     yihLM   1959.0  \n",
       "ECK125229275                                      pdeC   1640.0  \n",
       "ECK120028108                               ytfQRT-yjfF   4617.0  \n",
       "ECK120014757                                     treBC   3401.0  \n",
       "ECK125165366                                      opgB   2320.0  \n",
       "ECK120029053                       rhsD-ybbC-ylbH-ybbD   5640.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a bit behind annotations operons, etc.\n",
    "# Still uses same approach to counted, just not all data-types have been consolidated into the operon feat col\n",
    "\n",
    "feat_col_name = \"operons\"\n",
    "link_col_name = \"operon links\"\n",
    "comp_mut_df = pd.DataFrame(columns=[\"observed mutation count\"])\n",
    "for i, r in exp_muts_df.iterrows():\n",
    "    for f, links in r[link_col_name].items():\n",
    "        \n",
    "        f_d = dict()\n",
    "        if f == \"unknown\":\n",
    "            f_d[\"RegulonDB ID\"] = \"unknown\"\n",
    "            f_d[\"name\"] = \"unknown\"\n",
    "            f_d[\"range\"] = (0, op_len_df[op_len_df[\"operon description\"]==\"unknown\"].iloc[0][\"length\"])\n",
    "        else:\n",
    "            f_d = get_feat_d(RegulonDB_ID=f, json=r[feat_col_name])\n",
    "        \n",
    "        if f_d[\"RegulonDB ID\"] not in comp_mut_df.index:\n",
    "            f_len = f_d[\"range\"][1] - f_d[\"range\"][0] + 1\n",
    "            df = pd.DataFrame({\"name\": f_d[\"name\"], \"length\": f_len, \"observed mutation count\": len(links)},\n",
    "                              index=[f_d[\"RegulonDB ID\"]])  # \"name\" column just for visual inspection\n",
    "            comp_mut_df = comp_mut_df.append(df, sort=False)\n",
    "        else:\n",
    "            comp_mut_df.loc[f, \"observed mutation count\"] += len(links)\n",
    "            \n",
    "comp_mut_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, r in exp_muts_df.iterrows():\n",
    "    for feat_d in r[\"operons\"]:\n",
    "        feat_d[\"mutation set count\"] = comp_mut_df.loc[feat_d[\"RegulonDB ID\"]][\"observed mutation count\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120014482</td>\n",
       "      <td>2</td>\n",
       "      <td>xylFGHR</td>\n",
       "      <td>5280.0</td>\n",
       "      <td>0.001138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014627</td>\n",
       "      <td>12</td>\n",
       "      <td>rph-pyrE</td>\n",
       "      <td>1485.0</td>\n",
       "      <td>0.000320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120016992</td>\n",
       "      <td>3</td>\n",
       "      <td>rplKAJL-rpoBC</td>\n",
       "      <td>11376.0</td>\n",
       "      <td>0.002451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014515</td>\n",
       "      <td>1</td>\n",
       "      <td>rpsU-dnaG-rpoD</td>\n",
       "      <td>4317.0</td>\n",
       "      <td>0.000930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120020905</td>\n",
       "      <td>1</td>\n",
       "      <td>fimAICDFGH</td>\n",
       "      <td>7075.0</td>\n",
       "      <td>0.001524</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              observed mutation count            name   length  proportion\n",
       "ECK120014482                        2         xylFGHR   5280.0    0.001138\n",
       "ECK120014627                       12        rph-pyrE   1485.0    0.000320\n",
       "ECK120016992                        3   rplKAJL-rpoBC  11376.0    0.002451\n",
       "ECK120014515                        1  rpsU-dnaG-rpoD   4317.0    0.000930\n",
       "ECK120020905                        1      fimAICDFGH   7075.0    0.001524"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutated_seq_len = np.sum(comp_mut_df[\"length\"])\n",
    "\n",
    "# XXX Is this correct. What about overlapping operons?\n",
    "# They would make the genome remainder smaller than what it should be.\n",
    "TOTAL_GENOME_LEN = 4641652\n",
    "genome_remainder_len = TOTAL_GENOME_LEN - mutated_seq_len\n",
    "\n",
    "# Not account for the number of mutations that don't land in operons.\n",
    "df = pd.DataFrame([[0, genome_remainder_len]], columns=[\"observed mutation count\", \"length\"], index=[\"genome remainder\"])\n",
    "comp_mut_df = comp_mut_df.append(df, sort=False)\n",
    "total_seq_len = TOTAL_GENOME_LEN\n",
    "\n",
    "\n",
    "comp_mut_df[\"proportion\"] = comp_mut_df[\"length\"].apply(lambda comp_len: comp_len/total_seq_len)\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120014482</td>\n",
       "      <td>2</td>\n",
       "      <td>xylFGHR</td>\n",
       "      <td>5280.0</td>\n",
       "      <td>0.001138</td>\n",
       "      <td>17</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014627</td>\n",
       "      <td>12</td>\n",
       "      <td>rph-pyrE</td>\n",
       "      <td>1485.0</td>\n",
       "      <td>0.000320</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120016992</td>\n",
       "      <td>3</td>\n",
       "      <td>rplKAJL-rpoBC</td>\n",
       "      <td>11376.0</td>\n",
       "      <td>0.002451</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014515</td>\n",
       "      <td>1</td>\n",
       "      <td>rpsU-dnaG-rpoD</td>\n",
       "      <td>4317.0</td>\n",
       "      <td>0.000930</td>\n",
       "      <td>449</td>\n",
       "      <td>0.0449</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120020905</td>\n",
       "      <td>1</td>\n",
       "      <td>fimAICDFGH</td>\n",
       "      <td>7075.0</td>\n",
       "      <td>0.001524</td>\n",
       "      <td>780</td>\n",
       "      <td>0.0780</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              observed mutation count            name   length  proportion  \\\n",
       "ECK120014482                        2         xylFGHR   5280.0    0.001138   \n",
       "ECK120014627                       12        rph-pyrE   1485.0    0.000320   \n",
       "ECK120016992                        3   rplKAJL-rpoBC  11376.0    0.002451   \n",
       "ECK120014515                        1  rpsU-dnaG-rpoD   4317.0    0.000930   \n",
       "ECK120020905                        1      fimAICDFGH   7075.0    0.001524   \n",
       "\n",
       "              equal or larger count  p value  significant  \n",
       "ECK120014482                     17   0.0017         True  \n",
       "ECK120014627                      0   0.0000         True  \n",
       "ECK120016992                      4   0.0004         True  \n",
       "ECK120014515                    449   0.0449         True  \n",
       "ECK120020905                    780   0.0780        False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "comp_mut_df['equal or larger count'] = 0\n",
    "component_l = comp_mut_df.index.tolist()\n",
    "proportion_l = comp_mut_df[\"proportion\"].tolist()\n",
    "num_muts = np.sum(comp_mut_df[\"observed mutation count\"])\n",
    "\n",
    "# slides from UW speak about the relationship of permutations and precision.\n",
    "# should read through them to get a sense of what is the least iterations that should execute.\n",
    "num_itr = 10000\n",
    "for i in range(0, num_itr):\n",
    "    random_mut_comp_a = np.random.choice(\n",
    "        a=component_l,\n",
    "        size=num_muts,\n",
    "        replace=True,\n",
    "        p=proportion_l\n",
    "    )\n",
    "    random_mut_comp_count_d = Counter(random_mut_comp_a)\n",
    "    for component in random_mut_comp_count_d.keys():\n",
    "        if random_mut_comp_count_d[component] >= comp_mut_df.loc[component, 'observed mutation count']:\n",
    "            comp_mut_df.loc[component, 'equal or larger count'] += 1\n",
    "\n",
    "comp_mut_df[\"p value\"] = comp_mut_df[\"equal or larger count\"].apply(\n",
    "    lambda count: count/num_itr)\n",
    "\n",
    "comp_mut_df[\"significant\"] = comp_mut_df[\"p value\"].apply(lambda p_val: True if p_val < ENRICH_ALPHA else False)\n",
    "\n",
    "display(comp_mut_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120014482</td>\n",
       "      <td>2</td>\n",
       "      <td>xylFGHR</td>\n",
       "      <td>5280.0</td>\n",
       "      <td>0.001138</td>\n",
       "      <td>17</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0578</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014627</td>\n",
       "      <td>12</td>\n",
       "      <td>rph-pyrE</td>\n",
       "      <td>1485.0</td>\n",
       "      <td>0.000320</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120016992</td>\n",
       "      <td>3</td>\n",
       "      <td>rplKAJL-rpoBC</td>\n",
       "      <td>11376.0</td>\n",
       "      <td>0.002451</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0136</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014515</td>\n",
       "      <td>1</td>\n",
       "      <td>rpsU-dnaG-rpoD</td>\n",
       "      <td>4317.0</td>\n",
       "      <td>0.000930</td>\n",
       "      <td>449</td>\n",
       "      <td>0.0449</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120020905</td>\n",
       "      <td>1</td>\n",
       "      <td>fimAICDFGH</td>\n",
       "      <td>7075.0</td>\n",
       "      <td>0.001524</td>\n",
       "      <td>780</td>\n",
       "      <td>0.0780</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              observed mutation count            name   length  proportion  \\\n",
       "ECK120014482                        2         xylFGHR   5280.0    0.001138   \n",
       "ECK120014627                       12        rph-pyrE   1485.0    0.000320   \n",
       "ECK120016992                        3   rplKAJL-rpoBC  11376.0    0.002451   \n",
       "ECK120014515                        1  rpsU-dnaG-rpoD   4317.0    0.000930   \n",
       "ECK120020905                        1      fimAICDFGH   7075.0    0.001524   \n",
       "\n",
       "              equal or larger count  p value  significant  corrected p value  \\\n",
       "ECK120014482                     17   0.0017         True             0.0578   \n",
       "ECK120014627                      0   0.0000         True             0.0000   \n",
       "ECK120016992                      4   0.0004         True             0.0136   \n",
       "ECK120014515                    449   0.0449         True             1.0000   \n",
       "ECK120020905                    780   0.0780        False             1.0000   \n",
       "\n",
       "              corrected significance  \n",
       "ECK120014482                   False  \n",
       "ECK120014627                    True  \n",
       "ECK120016992                    True  \n",
       "ECK120014515                   False  \n",
       "ECK120020905                   False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from statsmodels.stats import multitest\n",
    "\n",
    "pvals = comp_mut_df[\"p value\"]\n",
    "rejects, pvals_corrected, alphacSidak, alphacBonf = multitest.multipletests(\n",
    "    pvals=pvals,\n",
    "    alpha=ENRICH_ALPHA,\n",
    "    method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "# pd.options.display.max_rows = 500\n",
    "comp_mut_df[\"corrected p value\"] = pvals_corrected\n",
    "comp_mut_df[\"corrected significance\"] = rejects\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>observed mutation count</th>\n",
       "      <th>name</th>\n",
       "      <th>length</th>\n",
       "      <th>proportion</th>\n",
       "      <th>equal or larger count</th>\n",
       "      <th>p value</th>\n",
       "      <th>significant</th>\n",
       "      <th>corrected p value</th>\n",
       "      <th>corrected significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ECK120014482</td>\n",
       "      <td>2</td>\n",
       "      <td>xylFGHR</td>\n",
       "      <td>5280.0</td>\n",
       "      <td>0.001138</td>\n",
       "      <td>17</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0578</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014627</td>\n",
       "      <td>12</td>\n",
       "      <td>rph-pyrE</td>\n",
       "      <td>1485.0</td>\n",
       "      <td>0.000320</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120016992</td>\n",
       "      <td>3</td>\n",
       "      <td>rplKAJL-rpoBC</td>\n",
       "      <td>11376.0</td>\n",
       "      <td>0.002451</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0004</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0136</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120014515</td>\n",
       "      <td>1</td>\n",
       "      <td>rpsU-dnaG-rpoD</td>\n",
       "      <td>4317.0</td>\n",
       "      <td>0.000930</td>\n",
       "      <td>449</td>\n",
       "      <td>0.0449</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ECK120020905</td>\n",
       "      <td>1</td>\n",
       "      <td>fimAICDFGH</td>\n",
       "      <td>7075.0</td>\n",
       "      <td>0.001524</td>\n",
       "      <td>780</td>\n",
       "      <td>0.0780</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              observed mutation count            name   length  proportion  \\\n",
       "ECK120014482                        2         xylFGHR   5280.0    0.001138   \n",
       "ECK120014627                       12        rph-pyrE   1485.0    0.000320   \n",
       "ECK120016992                        3   rplKAJL-rpoBC  11376.0    0.002451   \n",
       "ECK120014515                        1  rpsU-dnaG-rpoD   4317.0    0.000930   \n",
       "ECK120020905                        1      fimAICDFGH   7075.0    0.001524   \n",
       "\n",
       "              equal or larger count  p value  significant  corrected p value  \\\n",
       "ECK120014482                     17   0.0017         True             0.0578   \n",
       "ECK120014627                      0   0.0000         True             0.0000   \n",
       "ECK120016992                      4   0.0004         True             0.0136   \n",
       "ECK120014515                    449   0.0449         True             1.0000   \n",
       "ECK120020905                    780   0.0780        False             1.0000   \n",
       "\n",
       "              corrected significance  \n",
       "ECK120014482                   False  \n",
       "ECK120014627                    True  \n",
       "ECK120016992                    True  \n",
       "ECK120014515                   False  \n",
       "ECK120020905                   False  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# not wanting to consider mutation events with only 1 features as being significant\n",
    "# comp_mut_df[\"final significance\"] = comp_mut_df.apply(lambda r: r[\"corrected significance\"] if r[\"observed mutation count\"] > 1 else False, axis=1)\n",
    "comp_mut_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, r in exp_muts_df.iterrows():\n",
    "    for d in r[\"operons\"]:\n",
    "        if d[\"RegulonDB ID\"] in comp_mut_df.index:\n",
    "            d[\"significantly enriched\"] = comp_mut_df.loc[d[\"RegulonDB ID\"]][\"corrected significance\"]\n",
    "#             d[\"significant\"] = comp_mut_df.loc[d[\"RegulonDB ID\"]][\"final significance\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outputting dataframe for the next step in the pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_muts_df.to_pickle(\"./data/8_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# import seaborn as sns\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# %matplotlib inline\n",
    "# plt.rcParams[\"figure.dpi\"] = 100\n",
    "# sns.set_context(\"talk\")\n",
    "# sns.set_style(\"ticks\")\n",
    "\n",
    "\n",
    "# def _get_name_signif_from_ID(RegulonDB_ID, json):\n",
    "#     name = \"\"\n",
    "#     is_signif = False\n",
    "#     for d in json:\n",
    "#         if d[\"RegulonDB ID\"] == RegulonDB_ID:\n",
    "#             name = d[\"name\"]\n",
    "#             is_signif = d[\"significant\"]\n",
    "#             break\n",
    "#     return (name, is_signif)\n",
    "\n",
    "\n",
    "# targ_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     link_d = r[\"operon links\"]\n",
    "#     for targ_ID, hits in link_d.items():\n",
    "#         t, is_signif = _get_name_signif_from_ID(targ_ID, r[\"operons\"])\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#         if t not in targ_cnt_d.keys():\n",
    "#             targ_cnt_d[t] = 0\n",
    "#         targ_cnt_d[t] += len(hits)\n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# df = df[-20:]\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,10),\n",
    "#                   color=\"#4C72B0\", width=1)\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.1, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Top 20 operon mutated feature count\")\n",
    "# ax.get_legend().remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# targ_cnt_d = dict()\n",
    "# for _, r in exp_muts_df.iterrows():\n",
    "#     link_d = r[\"operon links\"]\n",
    "#     for targ_ID, hits in link_d.items():\n",
    "#         t, is_signif = _get_name_signif_from_ID(targ_ID, r[\"operons\"])\n",
    "#         if is_signif:\n",
    "#             t = '*' + t\n",
    "#             if t not in targ_cnt_d.keys():\n",
    "#                 targ_cnt_d[t] = 0\n",
    "#             targ_cnt_d[t] += len(hits)\n",
    "# df = pd.DataFrame.from_dict(targ_cnt_d, orient='index', columns=[\"mutated feature count\"])\n",
    "# df = df.sort_values(\"mutated feature count\")\n",
    "# ax = df.plot.barh(y='mutated feature count',\n",
    "# #                   figsize=(10,25),\n",
    "#                   figsize=(5,13),\n",
    "#                   color=\"#4C72B0\", width=1)\n",
    "# for i, v in enumerate(df[\"mutated feature count\"]):\n",
    "#     ax.text(v, i - 0.35, str(v))\n",
    "# sns.despine(ax=ax, top=True, right=True)\n",
    "# ax.set_title(\"Significant operons\")\n",
    "# ax.get_legend().remove()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
