{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "# FUNC_ANNOT = \"regulon\"\n",
    "FUNC_ANNOT = \"imodulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "67"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/EEPubiC_df.pkl\")\n",
    "display(len(muts_df),\n",
    "#         muts_df.head()\n",
    "       )\n",
    "\n",
    "\n",
    "# # Below are the COGs that got multiple hits with GYD\n",
    "# '''\n",
    "# Lipid transport and metabolism\n",
    "# Cell cycle control, cell division, chromosome partitioning\n",
    "# Transcription\n",
    "# Amino acid transport and metabolism\n",
    "# Secondary metabolites biosynthesis, transport and catabolism\n",
    "# Cell wall/membrane/envelope biogenesis\n",
    "# Posttranslational modification, protein turnover, chaperones\n",
    "# Replication, recombination and repair\n",
    "# Signal transduction mechanisms\n",
    "# Cell motility\n",
    "# Defense mechanisms\n",
    "# Inorganic ion transport and metabolism\n",
    "# Intracellular trafficking, secretion, and vesicular transport\n",
    "# Coenzyme transport and metabolism\n",
    "# Energy production and conversion\n",
    "# Translation, ribosomal structure and biogenesis\n",
    "# Carbohydrate transport and metabolism\n",
    "# Nucleotide transport and metabolism\n",
    "# '''\n",
    "\n",
    "\n",
    "# out_df = pd.DataFrame()\n",
    "# df = muts_df.copy()\n",
    "# for _, r in df.iterrows():\n",
    "#     for c in r[\"operons\"]:\n",
    "#         try:\n",
    "#             # GYD\n",
    "#             if \"rpoB\" in c[\"name\"] or \"rpoC\" in c[\"name\"] or \"rph\" in c[\"name\"]:\n",
    "#                 out_df = out_df.append(r, ignore_index=True)\n",
    "#                 break\n",
    "# #             if \"ptsHI-crr\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#             # PGI\n",
    "# #             if \"pnt\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#         except:\n",
    "#             print(\"feature annotation bug with\", c)\n",
    "\n",
    "# muts_df = out_df\n",
    "# display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>gene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>TSS</th>\n",
       "      <th>RBS</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>gene count</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>base-media</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>imodulons</th>\n",
       "      <th>imodulon links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>24</td>\n",
       "      <td>intergenic (‑495/‑46)</td>\n",
       "      <td>aroP/pdhR</td>\n",
       "      <td>834737.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>122046</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>2</td>\n",
       "      <td>EEPubiC</td>\n",
       "      <td>54</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(122046, 122046)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK120017175}</td>\n",
       "      <td>{ECK120010280}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000080/ECK120001076',...</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...</td>\n",
       "      <td>aroP/pdhR</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27_ubiC</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>Glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>Na2HPO4-KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td></td>\n",
       "      <td>{'ECK120000080/ECK120001076': ['ECK120017175',...</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120000080/ECK120001076',...</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...</td>\n",
       "      <td>[{'name': 'Glycolysis / Gluconeogenesis', 'mut...</td>\n",
       "      <td>{'Glycolysis / Gluconeogenesis': ['ECK12002395...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>24</td>\n",
       "      <td>intergenic (‑495/‑46)</td>\n",
       "      <td>aroP/pdhR</td>\n",
       "      <td>834737.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>122046</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→C</td>\n",
       "      <td>1</td>\n",
       "      <td>EEPubiC</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>0.22</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>(122046, 122046)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK120017175}</td>\n",
       "      <td>{ECK120010280}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000080/ECK120001076',...</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...</td>\n",
       "      <td>aroP/pdhR</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27_ubiC</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>Glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>Na2HPO4-KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td></td>\n",
       "      <td>{'ECK120000080/ECK120001076': ['ECK120017175',...</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120000080/ECK120001076',...</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...</td>\n",
       "      <td>[{'name': 'Glycolysis / Gluconeogenesis', 'mut...</td>\n",
       "      <td>{'Glycolysis / Gluconeogenesis': ['ECK12002395...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>0</td>\n",
       "      <td>E37* (GAA→TAA)</td>\n",
       "      <td>pdhR</td>\n",
       "      <td>834944.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>122200</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→T</td>\n",
       "      <td>1</td>\n",
       "      <td>EEPubiC</td>\n",
       "      <td>52</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(122200, 122200)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120001076}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>pdhR</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27_ubiC</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>Glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>Na2HPO4-KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td></td>\n",
       "      <td>{'ECK120001076': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120023950']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>30</td>\n",
       "      <td>G62C (GGC→TGC)</td>\n",
       "      <td>pdhR</td>\n",
       "      <td>834691.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>122275</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→T</td>\n",
       "      <td>4</td>\n",
       "      <td>EEPubiC</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(122275, 122275)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120001076}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>pdhR</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27_ubiC</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>Glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>Na2HPO4-KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td></td>\n",
       "      <td>{'ECK120001076': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120023950']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>34</td>\n",
       "      <td>L64* (TTG→TAG)</td>\n",
       "      <td>pdhR</td>\n",
       "      <td>834795.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>122282</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>T→A</td>\n",
       "      <td>3</td>\n",
       "      <td>EEPubiC</td>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(122282, 122282)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120001076}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>pdhR</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27_ubiC</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>Glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>Na2HPO4-KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td></td>\n",
       "      <td>{'ECK120001076': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120023950']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index                Details mutation target annotation    Mut ID  \\\n",
       "571    24  intergenic (‑495/‑46)                  aroP/pdhR  834737.0   \n",
       "555    24  intergenic (‑495/‑46)                  aroP/pdhR  834737.0   \n",
       "556     0         E37* (GAA→TAA)                       pdhR  834944.0   \n",
       "604    30         G62C (GGC→TGC)                       pdhR  834691.0   \n",
       "595    34         L64* (TTG→TAG)                       pdhR  834795.0   \n",
       "\n",
       "    Mutation Type  Position Reference Seq Sequence Change  ale      exp  \\\n",
       "571           SNP    122046     NC_000913             T→C    2  EEPubiC   \n",
       "555           SNP    122046     NC_000913             T→C    1  EEPubiC   \n",
       "556           SNP    122200     NC_000913             G→T    1  EEPubiC   \n",
       "604           SNP    122275     NC_000913             G→T    4  EEPubiC   \n",
       "595           SNP    122282     NC_000913             T→A    3  EEPubiC   \n",
       "\n",
       "     flask  isolate  presence  tech_rep  coding             range   oriC  \\\n",
       "571     54        1      1.00         1   False  (122046, 122046)  False   \n",
       "555     52        0      0.22         2   False  (122046, 122046)  False   \n",
       "556     52        1      1.00         1    True  (122200, 122200)  False   \n",
       "604     52        0      1.00         1    True  (122275, 122275)  False   \n",
       "595     53        1      1.00         1    True  (122282, 122282)  False   \n",
       "\n",
       "     pseudogene            gene            TFBS        promoter TSS RBS  \\\n",
       "571       False              {}  {ECK120017175}  {ECK120010280}  {}  {}   \n",
       "555       False              {}  {ECK120017175}  {ECK120010280}  {}  {}   \n",
       "556       False  {ECK120001076}              {}              {}  {}  {}   \n",
       "604       False  {ECK120001076}              {}              {}  {}  {}   \n",
       "595       False  {ECK120001076}              {}              {}  {}  {}   \n",
       "\n",
       "    cis-regulatory RNA attenuator terminator terminator  genetic  \\\n",
       "571                 {}                    {}         {}    False   \n",
       "555                 {}                    {}         {}    False   \n",
       "556                 {}                    {}         {}     True   \n",
       "604                 {}                    {}         {}     True   \n",
       "595                 {}                    {}         {}     True   \n",
       "\n",
       "                                      genetic features  \\\n",
       "571  [{'RegulonDB ID': 'ECK120000080/ECK120001076',...   \n",
       "555  [{'RegulonDB ID': 'ECK120000080/ECK120001076',...   \n",
       "556  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "604  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "595  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "\n",
       "                                      genomic features RegulonDB Gene  \\\n",
       "571  [{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...      aroP/pdhR   \n",
       "555  [{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...      aroP/pdhR   \n",
       "556  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...           pdhR   \n",
       "604  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...           pdhR   \n",
       "595  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...           pdhR   \n",
       "\n",
       "     mutation size  gene count  disrupt  repaired taxonomy-id  \\\n",
       "571              1           1    False     False      511145   \n",
       "555              1           1    False     False      511145   \n",
       "556              1           1     True     False      511145   \n",
       "604              1           1    False     False      511145   \n",
       "595              1           1     True     False      511145   \n",
       "\n",
       "    strain-description base-media temperature carbon-source nitrogen-source  \\\n",
       "571         BOP27_ubiC         M9  37 celsius    Glucose(4)        NH4Cl(1)   \n",
       "555         BOP27_ubiC         M9  37 celsius    Glucose(4)        NH4Cl(1)   \n",
       "556         BOP27_ubiC         M9  37 celsius    Glucose(4)        NH4Cl(1)   \n",
       "604         BOP27_ubiC         M9  37 celsius    Glucose(4)        NH4Cl(1)   \n",
       "595         BOP27_ubiC         M9  37 celsius    Glucose(4)        NH4Cl(1)   \n",
       "\n",
       "    phosphorous-source sulfur-source supplement  \\\n",
       "571  Na2HPO4-KH2PO4(3)   MgSO4(0.24)              \n",
       "555  Na2HPO4-KH2PO4(3)   MgSO4(0.24)              \n",
       "556  Na2HPO4-KH2PO4(3)   MgSO4(0.24)              \n",
       "604  Na2HPO4-KH2PO4(3)   MgSO4(0.24)              \n",
       "595  Na2HPO4-KH2PO4(3)   MgSO4(0.24)              \n",
       "\n",
       "                                 genetic feature links  \\\n",
       "571  {'ECK120000080/ECK120001076': ['ECK120017175',...   \n",
       "555  {'ECK120000080/ECK120001076': ['ECK120017175',...   \n",
       "556                 {'ECK120001076': ['ECK120001076']}   \n",
       "604                 {'ECK120001076': ['ECK120001076']}   \n",
       "595                 {'ECK120001076': ['ECK120001076']}   \n",
       "\n",
       "                                               operons  \\\n",
       "571  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "555  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "556  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "604  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "595  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "\n",
       "                                          operon links  \\\n",
       "571  {'ECK120023950': ['ECK120000080/ECK120001076',...   \n",
       "555  {'ECK120023950': ['ECK120000080/ECK120001076',...   \n",
       "556                 {'ECK120023950': ['ECK120001076']}   \n",
       "604                 {'ECK120023950': ['ECK120001076']}   \n",
       "595                 {'ECK120023950': ['ECK120001076']}   \n",
       "\n",
       "                                             imodulons  \\\n",
       "571  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "555  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "556  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "604  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "595  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "\n",
       "                                        imodulon links  \\\n",
       "571  {'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...   \n",
       "555  {'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...   \n",
       "556  {'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...   \n",
       "604  {'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...   \n",
       "595  {'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...   \n",
       "\n",
       "                                                  COGs  \\\n",
       "571  [{'name': 'Energy production and conversion', ...   \n",
       "555  [{'name': 'Energy production and conversion', ...   \n",
       "556  [{'name': 'Transcription', 'mutation set count...   \n",
       "604  [{'name': 'Transcription', 'mutation set count...   \n",
       "595  [{'name': 'Transcription', 'mutation set count...   \n",
       "\n",
       "                                             COG links  \\\n",
       "571  {'Energy production and conversion': ['ECK1200...   \n",
       "555  {'Energy production and conversion': ['ECK1200...   \n",
       "556                {'Transcription': ['ECK120023950']}   \n",
       "604                {'Transcription': ['ECK120023950']}   \n",
       "595                {'Transcription': ['ECK120023950']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "571  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "555  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "556  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "604  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "595  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "\n",
       "                                       regulator links  \\\n",
       "571  {'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...   \n",
       "555  {'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...   \n",
       "556  {'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...   \n",
       "604  {'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...   \n",
       "595  {'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...   \n",
       "\n",
       "                                              pathways  \\\n",
       "571  [{'name': 'Glycolysis / Gluconeogenesis', 'mut...   \n",
       "555  [{'name': 'Glycolysis / Gluconeogenesis', 'mut...   \n",
       "556                                                 []   \n",
       "604                                                 []   \n",
       "595                                                 []   \n",
       "\n",
       "                                         pathway links  \n",
       "571  {'Glycolysis / Gluconeogenesis': ['ECK12002395...  \n",
       "555  {'Glycolysis / Gluconeogenesis': ['ECK12002395...  \n",
       "556                                                 {}  \n",
       "604                                                 {}  \n",
       "595                                                 {}  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# muts_df = muts_df[muts_df[\"mutation target annotation\"].str.contains(\"pnt\")]\n",
    "# muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"CRP\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"GlpR\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Mlc\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "# #     if \"Nucleotide transport and metabolism\" in r[\"COG links\"].keys() \\\n",
    "#     if \"cyaA\" in r[\"RegulonDB Gene\"]:\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Glycerolipid metabolism\" in r[\"pathway links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = muts_df.copy()\n",
    "# df[\"significantly associated condition set\"] = df[\"significantly associated condition set\"].apply(\n",
    "#     lambda x: str(x))\n",
    "# display(df[\"significantly associated condition set\"].unique())\n",
    "\n",
    "\n",
    "# # GYD\n",
    "# muts_df = muts_df[\n",
    "#     (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)'})\n",
    "#     | (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)', \"BOP27\"})\n",
    "# ]\n",
    "\n",
    "# # SER\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"] == {'ΔsdaA ΔsdaB ΔtdcG ΔglyA', 'L-Serine'}]\n",
    "\n",
    "# # # GLU\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'glucose(2)'}]\n",
    "\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]==set()]\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'log', '511145', 'M9'}]\n",
    "\n",
    "\n",
    "# display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(67, 17)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>imodulons</th>\n",
       "      <th>imodulon links</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000080/ECK120001076',...</td>\n",
       "      <td>{'ECK120000080/ECK120001076': ['ECK120017175',...</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120000080/ECK120001076',...</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Glycolysis / Gluconeogenesis', 'mut...</td>\n",
       "      <td>{'Glycolysis / Gluconeogenesis': ['ECK12002395...</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...</td>\n",
       "      <td>aroP/pdhR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000080/ECK120001076',...</td>\n",
       "      <td>{'ECK120000080/ECK120001076': ['ECK120017175',...</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120000080/ECK120001076',...</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Glycolysis / Gluconeogenesis', 'mut...</td>\n",
       "      <td>{'Glycolysis / Gluconeogenesis': ['ECK12002395...</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...</td>\n",
       "      <td>aroP/pdhR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>{'ECK120001076': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120023950']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...</td>\n",
       "      <td>pdhR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>{'ECK120001076': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120023950']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...</td>\n",
       "      <td>pdhR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>[{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...</td>\n",
       "      <td>{'ECK120001076': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK120023950': ['ECK120001076']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 8.0, 's...</td>\n",
       "      <td>{'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120023950']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'ArcA-1', 'mutation set count': 7.0,...</td>\n",
       "      <td>{'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...</td>\n",
       "      <td>pdhR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutation Type  disrupt  repaired  \\\n",
       "571           SNP    False     False   \n",
       "555           SNP    False     False   \n",
       "556           SNP     True     False   \n",
       "604           SNP    False     False   \n",
       "595           SNP     True     False   \n",
       "\n",
       "                                      genomic features  \\\n",
       "571  [{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...   \n",
       "555  [{'name': 'pdhR-aceEF-lpd TFBS', 'RegulonDB ID...   \n",
       "556  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "604  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "595  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "\n",
       "                                      genetic features  \\\n",
       "571  [{'RegulonDB ID': 'ECK120000080/ECK120001076',...   \n",
       "555  [{'RegulonDB ID': 'ECK120000080/ECK120001076',...   \n",
       "556  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "604  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "595  [{'name': 'pdhR', 'RegulonDB ID': 'ECK12000107...   \n",
       "\n",
       "                                 genetic feature links  \\\n",
       "571  {'ECK120000080/ECK120001076': ['ECK120017175',...   \n",
       "555  {'ECK120000080/ECK120001076': ['ECK120017175',...   \n",
       "556                 {'ECK120001076': ['ECK120001076']}   \n",
       "604                 {'ECK120001076': ['ECK120001076']}   \n",
       "595                 {'ECK120001076': ['ECK120001076']}   \n",
       "\n",
       "                                               operons  \\\n",
       "571  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "555  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "556  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "604  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "595  [{'name': 'pdhR-aceEF-lpd', 'RegulonDB ID': 'E...   \n",
       "\n",
       "                                          operon links  \\\n",
       "571  {'ECK120023950': ['ECK120000080/ECK120001076',...   \n",
       "555  {'ECK120023950': ['ECK120000080/ECK120001076',...   \n",
       "556                 {'ECK120023950': ['ECK120001076']}   \n",
       "604                 {'ECK120023950': ['ECK120001076']}   \n",
       "595                 {'ECK120023950': ['ECK120001076']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "571  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "555  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "556  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "604  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "595  [{'name': 'CRP', 'mutation set count': 8.0, 's...   \n",
       "\n",
       "                                       regulator links  \\\n",
       "571  {'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...   \n",
       "555  {'CRP': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'], ...   \n",
       "556  {'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...   \n",
       "604  {'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...   \n",
       "595  {'CRP': ['pdhR-aceEF-lpd'], 'FNR': ['pdhR-aceE...   \n",
       "\n",
       "                                                  COGs  \\\n",
       "571  [{'name': 'Energy production and conversion', ...   \n",
       "555  [{'name': 'Energy production and conversion', ...   \n",
       "556  [{'name': 'Transcription', 'mutation set count...   \n",
       "604  [{'name': 'Transcription', 'mutation set count...   \n",
       "595  [{'name': 'Transcription', 'mutation set count...   \n",
       "\n",
       "                                             COG links  \\\n",
       "571  {'Energy production and conversion': ['ECK1200...   \n",
       "555  {'Energy production and conversion': ['ECK1200...   \n",
       "556                {'Transcription': ['ECK120023950']}   \n",
       "604                {'Transcription': ['ECK120023950']}   \n",
       "595                {'Transcription': ['ECK120023950']}   \n",
       "\n",
       "                                              pathways  \\\n",
       "571  [{'name': 'Glycolysis / Gluconeogenesis', 'mut...   \n",
       "555  [{'name': 'Glycolysis / Gluconeogenesis', 'mut...   \n",
       "556                                                 []   \n",
       "604                                                 []   \n",
       "595                                                 []   \n",
       "\n",
       "                                         pathway links  \\\n",
       "571  {'Glycolysis / Gluconeogenesis': ['ECK12002395...   \n",
       "555  {'Glycolysis / Gluconeogenesis': ['ECK12002395...   \n",
       "556                                                 {}   \n",
       "604                                                 {}   \n",
       "595                                                 {}   \n",
       "\n",
       "                                             imodulons  \\\n",
       "571  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "555  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "556  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "604  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "595  [{'name': 'ArcA-1', 'mutation set count': 7.0,...   \n",
       "\n",
       "                                        imodulon links RegulonDB Gene  \n",
       "571  {'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...      aroP/pdhR  \n",
       "555  {'ArcA-1': ['pdhR-aceEF-lpd', 'pdhR-aceEF-lpd'...      aroP/pdhR  \n",
       "556  {'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...           pdhR  \n",
       "604  {'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...           pdhR  \n",
       "595  {'ArcA-1': ['pdhR-aceEF-lpd'], 'Pyruvate': ['p...           pdhR  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "    \"disrupt\",\n",
    "    \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "    \"imodulons\",\n",
    "    \"imodulon links\",\n",
    "    \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "imodulon_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)\n",
    "        \n",
    "    for imodulon, links in r[\"imodulon links\"].items():\n",
    "        if imodulon not in imodulon_counts.keys():\n",
    "            imodulon_counts[imodulon] = 0\n",
    "        imodulon_counts[imodulon] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # For development. Delete when NB is stable.\n",
    "# for _, r in annot_df.iterrows():\n",
    "#     for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "#         effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "#         mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "#         mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "            \n",
    "#         genomic_feat_annot = \"\"\n",
    "#         if genomic_feat_d[\"significant\"]: genomic_feat_annot += \"*\"\n",
    "#         genomic_feat_annot += str(genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]) + ' '\n",
    "#         genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "    \n",
    "            \n",
    "#         display(genomic_feat_annot + ' #### ' + mut_type_feat_effect)\n",
    "# #         display(mut_type_feat_effect, \"################################\", r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*2 pdhR-aceEF-lpd TFBS</td>\n",
       "      <td>2 pdhR-aceEF-lpd TFBS SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*2 pdhRp</td>\n",
       "      <td>2 pdhRp SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>*2 pdhR-aceEF-lpd TFBS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>*2 pdhRp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7 ArcA-1</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7 ArcA-1</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>*7 Pyruvate</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>*7 Pyruvate</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>*2 pdhR-aceEF-lpd TFBS</td>\n",
       "      <td>2 pdhR-aceEF-lpd TFBS SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>*2 pdhRp</td>\n",
       "      <td>2 pdhRp SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>*2 pdhR-aceEF-lpd TFBS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>*2 pdhRp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>*4 aroP/pdhR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>7 ArcA-1</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>7 ArcA-1</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>*7 Pyruvate</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>*7 Pyruvate</td>\n",
       "      <td>*7 pdhR-aceEF-lpd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    source                      target     type  value\n",
       "0   *2 pdhR-aceEF-lpd TFBS  2 pdhR-aceEF-lpd TFBS SNP   unknown    1.0\n",
       "1                 *2 pdhRp                2 pdhRp SNP   unknown    1.0\n",
       "2             *4 aroP/pdhR      *2 pdhR-aceEF-lpd TFBS  unknown    1.0\n",
       "3             *4 aroP/pdhR                    *2 pdhRp  unknown    1.0\n",
       "4       *7 pdhR-aceEF-lpd                 *4 aroP/pdhR  unknown    1.0\n",
       "5       *7 pdhR-aceEF-lpd                 *4 aroP/pdhR  unknown    1.0\n",
       "6                 7 ArcA-1          *7 pdhR-aceEF-lpd   unknown    1.0\n",
       "7                 7 ArcA-1          *7 pdhR-aceEF-lpd   unknown    1.0\n",
       "8              *7 Pyruvate          *7 pdhR-aceEF-lpd   unknown    1.0\n",
       "9              *7 Pyruvate          *7 pdhR-aceEF-lpd   unknown    1.0\n",
       "10  *2 pdhR-aceEF-lpd TFBS  2 pdhR-aceEF-lpd TFBS SNP   unknown    1.0\n",
       "11                *2 pdhRp                2 pdhRp SNP   unknown    1.0\n",
       "12            *4 aroP/pdhR      *2 pdhR-aceEF-lpd TFBS  unknown    1.0\n",
       "13            *4 aroP/pdhR                    *2 pdhRp  unknown    1.0\n",
       "14      *7 pdhR-aceEF-lpd                 *4 aroP/pdhR  unknown    1.0\n",
       "15      *7 pdhR-aceEF-lpd                 *4 aroP/pdhR  unknown    1.0\n",
       "16                7 ArcA-1          *7 pdhR-aceEF-lpd   unknown    1.0\n",
       "17                7 ArcA-1          *7 pdhR-aceEF-lpd   unknown    1.0\n",
       "18             *7 Pyruvate          *7 pdhR-aceEF-lpd   unknown    1.0\n",
       "19             *7 Pyruvate          *7 pdhR-aceEF-lpd   unknown    1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "imodulons = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"pathway\":\n",
    "        for pathway, ops in r[\"pathway links\"].items():\n",
    "#             if pathway == \"Glycerolipid metabolism\":  # for GYD Glycerolipid metabolism.\n",
    "            pathway_feat_d = _get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "            pathway_annot = \"\"\n",
    "            mut_count = pathway_counts[pathway]\n",
    "            if pathway_feat_d[\"significant\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "            pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "            pathways.append(pathway_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": pathway_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "#             if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "            reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "    if FUNC_ANNOT == \"COG\":\n",
    "        for COG, ops in r[\"COG links\"].items():\n",
    "#             if COG == \"Nucleotide transport and metabolism\":\n",
    "            COG_feat_d = _get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "            COG_annot = \"\"\n",
    "            mut_count = COG_counts[COG]\n",
    "            if COG_feat_d[\"significant\"] and mut_count > 1: COG_annot += \"*\"\n",
    "            COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "            COGs.append(COG_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": COG_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "                \n",
    "                \n",
    "    if FUNC_ANNOT == \"imodulon\":\n",
    "        for imodulon, ops in r[\"imodulon links\"].items():\n",
    "            reg_feat_d = _get_feat_d(name=imodulon, json=r[\"imodulons\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = imodulon_counts[imodulon]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            imodulons.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "                \n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)\n",
    "imodulons = remove_duplicates(imodulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dbc6250830ba453cac5d7a0bd15f404f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'imodulon', 'type': 'process', 'title': 'imodulon', 'nodes': ['imodulon^7 ArcA-1',…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "    \"imodulon\": ProcessGroup(imodulons, title=\"imodulon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "if FUNC_ANNOT == \"imodulon\":\n",
    "    l = imodulons\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "#                        flow_partition=mut_effect_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# p_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "# size = dict(width=3000, height=4000)\n",
    "# size = dict(width=2000, height=2500)\n",
    "# size = dict(width=1600, height=1000)\n",
    "size = dict(width=1500, height=800)\n",
    "# size = dict(width=1600, height=500)\n",
    "# size = dict(width=1100, height=300)\n",
    "# size = dict(width=1350, height=300)\n",
    "# size = dict(width=1000, height=200)\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
