{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "FUNC_ANNOT = \"regulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/glycerol_2_5_df.pkl\")\n",
    "display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GYD', 'SSW_GLY'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(muts_df.exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>TSS</th>\n",
       "      <th>ale</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>base-media</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>coding</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene count</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>oriC</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>repaired</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>temperature</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>317489.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>A→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.59</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 3.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>317489.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>A→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>14.0</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>68.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.50</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 3.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1738</th>\n",
       "      <td>intergenic (‑61/+32)</td>\n",
       "      <td>317490.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83590.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.59</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(83590, 83590)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 3.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1815</th>\n",
       "      <td>intergenic (+174/+54)</td>\n",
       "      <td>573114.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>543207.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>glxK/allE</td>\n",
       "      <td>A→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>25.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>36.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>100</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>glxK/allE</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.50</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(543207, 543207)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'ECK120002891/ECK120002892': ['ECK120002891/E...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1816</th>\n",
       "      <td>intergenic (+177/+51)</td>\n",
       "      <td>573115.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>543210.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>glxK/allE</td>\n",
       "      <td>A→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>25.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>36.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>101</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>glxK/allE</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.50</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(543210, 543210)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'ECK120002891/ECK120002892': ['ECK120002891/E...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Details    Mut ID Mutation Type  Position RBS  \\\n",
       "1737   intergenic (‑58/+35)  317489.0           SNP   83587.0  {}   \n",
       "1763   intergenic (‑58/+35)  317489.0           SNP   83587.0  {}   \n",
       "1738   intergenic (‑61/+32)  317490.0           SNP   83590.0  {}   \n",
       "1815  intergenic (+174/+54)  573114.0           SNP  543207.0  {}   \n",
       "1816  intergenic (+177/+51)  573115.0           SNP  543210.0  {}   \n",
       "\n",
       "     Reference Seq RegulonDB Gene Sequence Change TFBS TSS   ale  \\\n",
       "1737     NC_000913      leuA/leuL             A→C   {}  {}  10.0   \n",
       "1763     NC_000913      leuA/leuL             A→C   {}  {}  14.0   \n",
       "1738     NC_000913      leuA/leuL             G→T   {}  {}  10.0   \n",
       "1815     NC_000913      glxK/allE             A→T   {}  {}  25.0   \n",
       "1816     NC_000913      glxK/allE             A→T   {}  {}  25.0   \n",
       "\n",
       "     attenuator terminator base-media  carbon-source cis-regulatory RNA  \\\n",
       "1737        {ECK125143560}         M9  glycerol(2.5)     {ECK120001253}   \n",
       "1763        {ECK125143560}         M9  glycerol(2.5)     {ECK120001253}   \n",
       "1738        {ECK125143560}         M9  glycerol(2.5)     {ECK120001253}   \n",
       "1815                    {}         M9  glycerol(2.5)                 {}   \n",
       "1816                    {}         M9  glycerol(2.5)                 {}   \n",
       "\n",
       "      coding  disrupt  exp  flask gene  gene count  genetic  \\\n",
       "1737     0.0      0.0  GYD  106.0   {}         1.0      0.0   \n",
       "1763     0.0      0.0  GYD   68.0   {}         1.0      0.0   \n",
       "1738     0.0      0.0  GYD  106.0   {}         1.0      0.0   \n",
       "1815     0.0      0.0  GYD   36.0   {}         1.0      0.0   \n",
       "1816     0.0      0.0  GYD   36.0   {}         1.0      0.0   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1737  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1763  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1738  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1815  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   \n",
       "1816  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   \n",
       "\n",
       "                                       genomic features index  isolate  \\\n",
       "1737  [{'name': 'leuLABCD attenuator terminator', 'R...    27      0.0   \n",
       "1763  [{'name': 'leuLABCD attenuator terminator', 'R...    27      0.0   \n",
       "1738  [{'name': 'leuLABCD attenuator terminator', 'R...    28      0.0   \n",
       "1815  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   100      0.0   \n",
       "1816  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   101      0.0   \n",
       "\n",
       "      mutation size mutation target annotation nitrogen-source  oriC  \\\n",
       "1737            1.0                  leuA/leuL        NH4Cl(1)   0.0   \n",
       "1763            1.0                  leuA/leuL        NH4Cl(1)   0.0   \n",
       "1738            1.0                  leuA/leuL        NH4Cl(1)   0.0   \n",
       "1815            1.0                  glxK/allE        NH4Cl(1)   0.0   \n",
       "1816            1.0                  glxK/allE        NH4Cl(1)   0.0   \n",
       "\n",
       "     phosphorous-source  presence promoter  pseudogene             range  \\\n",
       "1737          KH2PO4(3)      0.59       {}         0.0    (83587, 83587)   \n",
       "1763          KH2PO4(3)      0.50       {}         0.0    (83587, 83587)   \n",
       "1738          KH2PO4(3)      0.59       {}         0.0    (83590, 83590)   \n",
       "1815          KH2PO4(3)      0.50       {}         0.0  (543207, 543207)   \n",
       "1816          KH2PO4(3)      0.50       {}         0.0  (543210, 543210)   \n",
       "\n",
       "      repaired strain-description sulfur-source      supplement taxonomy-id  \\\n",
       "1737       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "1763       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "1738       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "1815       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "1816       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "\n",
       "      tech_rep temperature      terminator  \\\n",
       "1737       1.0  37 celsius  {ECK120034959}   \n",
       "1763       1.0  37 celsius  {ECK120034959}   \n",
       "1738       1.0  37 celsius  {ECK120034959}   \n",
       "1815       2.0  37 celsius              {}   \n",
       "1816       2.0  37 celsius              {}   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1737    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1763    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1738    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1815  {'ECK120002891/ECK120002892': ['ECK120002891/E...   \n",
       "1816  {'ECK120002891/ECK120002892': ['ECK120002891/E...   \n",
       "\n",
       "                                                operons  \\\n",
       "1737  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1763  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1738  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "                                         operon links  \\\n",
       "1737  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1763  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1738  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1815                                               {}   \n",
       "1816                                               {}   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1737  [{'name': 'Amino acid transport and metabolism...   \n",
       "1763  [{'name': 'Amino acid transport and metabolism...   \n",
       "1738  [{'name': 'Amino acid transport and metabolism...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "                                              COG links  \\\n",
       "1737  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1763  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1738  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1815                                                 {}   \n",
       "1816                                                 {}   \n",
       "\n",
       "                                             regulators  \\\n",
       "1737  [{'name': 'LeuO', 'mutation set count': 3.0, '...   \n",
       "1763  [{'name': 'LeuO', 'mutation set count': 3.0, '...   \n",
       "1738  [{'name': 'LeuO', 'mutation set count': 3.0, '...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "             regulator links  \\\n",
       "1737  {'LeuO': ['leuLABCD']}   \n",
       "1763  {'LeuO': ['leuLABCD']}   \n",
       "1738  {'LeuO': ['leuLABCD']}   \n",
       "1815                      {}   \n",
       "1816                      {}   \n",
       "\n",
       "                                               pathways  \\\n",
       "1737  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1763  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1738  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "                                          pathway links  \n",
       "1737  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "1763  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "1738  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "1815                                                 {}  \n",
       "1816                                                 {}  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(107, 16)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 3.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>GYD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 3.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>GYD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1738</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 3.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>GYD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1815</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>{'ECK120002891/ECK120002892': ['ECK120002891/E...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>glxK/allE</td>\n",
       "      <td>GYD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1816</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002891/ECK120002892',...</td>\n",
       "      <td>{'ECK120002891/ECK120002892': ['ECK120002891/E...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>glxK/allE</td>\n",
       "      <td>GYD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type  disrupt  repaired  \\\n",
       "1737           SNP      0.0       0.0   \n",
       "1763           SNP      0.0       0.0   \n",
       "1738           SNP      0.0       0.0   \n",
       "1815           SNP      0.0       0.0   \n",
       "1816           SNP      0.0       0.0   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1737  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1763  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1738  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1815  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   \n",
       "1816  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1737  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1763  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1738  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1815  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   \n",
       "1816  [{'RegulonDB ID': 'ECK120002891/ECK120002892',...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1737    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1763    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1738    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1815  {'ECK120002891/ECK120002892': ['ECK120002891/E...   \n",
       "1816  {'ECK120002891/ECK120002892': ['ECK120002891/E...   \n",
       "\n",
       "                                                operons  \\\n",
       "1737  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1763  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1738  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "                                         operon links  \\\n",
       "1737  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1763  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1738  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1815                                               {}   \n",
       "1816                                               {}   \n",
       "\n",
       "                                             regulators  \\\n",
       "1737  [{'name': 'LeuO', 'mutation set count': 3.0, '...   \n",
       "1763  [{'name': 'LeuO', 'mutation set count': 3.0, '...   \n",
       "1738  [{'name': 'LeuO', 'mutation set count': 3.0, '...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "             regulator links  \\\n",
       "1737  {'LeuO': ['leuLABCD']}   \n",
       "1763  {'LeuO': ['leuLABCD']}   \n",
       "1738  {'LeuO': ['leuLABCD']}   \n",
       "1815                      {}   \n",
       "1816                      {}   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1737  [{'name': 'Amino acid transport and metabolism...   \n",
       "1763  [{'name': 'Amino acid transport and metabolism...   \n",
       "1738  [{'name': 'Amino acid transport and metabolism...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "                                              COG links  \\\n",
       "1737  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1763  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1738  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1815                                                 {}   \n",
       "1816                                                 {}   \n",
       "\n",
       "                                               pathways  \\\n",
       "1737  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1763  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1738  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1815                                                 []   \n",
       "1816                                                 []   \n",
       "\n",
       "                                          pathway links RegulonDB Gene  exp  \n",
       "1737  {'Valine, leucine and isoleucine biosynthesis'...      leuA/leuL  GYD  \n",
       "1763  {'Valine, leucine and isoleucine biosynthesis'...      leuA/leuL  GYD  \n",
       "1738  {'Valine, leucine and isoleucine biosynthesis'...      leuA/leuL  GYD  \n",
       "1815                                                 {}      glxK/allE  GYD  \n",
       "1816                                                 {}      glxK/allE  GYD  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "    \"disrupt\",\n",
    "    \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "    \"RegulonDB Gene\",  # for debugging\n",
    "    \"exp\"\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>3 leuLABCD attenuator terminator SNP</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*3 leuLABCD</td>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3 LeuO</td>\n",
       "      <td>*3 leuLABCD</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>3 leuLABCD attenuator terminator SNP</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>*3 leuLABCD</td>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3 LeuO</td>\n",
       "      <td>*3 leuLABCD</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>3 leuLABCD attenuator terminator SNP</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>*3 leuLABCD</td>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3 LeuO</td>\n",
       "      <td>*3 leuLABCD</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>*2 glxK/allE</td>\n",
       "      <td>2 glxK/allE SNP</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>*2 glxK/allE</td>\n",
       "      <td>*2 glxK/allE</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>*2 glxK/allE</td>\n",
       "      <td>2 glxK/allE SNP</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>*2 glxK/allE</td>\n",
       "      <td>*2 glxK/allE</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>*2 fimZ</td>\n",
       "      <td>2 fimZ MOB disrupt</td>\n",
       "      <td>SSW_GLY</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>*2 fimZ</td>\n",
       "      <td>*2 fimZ</td>\n",
       "      <td>SSW_GLY</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>*2 fimZ</td>\n",
       "      <td>*2 fimZ</td>\n",
       "      <td>SSW_GLY</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>*2 fimZ</td>\n",
       "      <td>2 fimZ MOB disrupt</td>\n",
       "      <td>GYD</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               source                                 target  \\\n",
       "0   *3 leuLABCD attenuator terminator  3 leuLABCD attenuator terminator SNP    \n",
       "1                        *3 leuA/leuL      *3 leuLABCD attenuator terminator   \n",
       "2                        *3 leuLABCD                            *3 leuA/leuL   \n",
       "3                              3 LeuO                           *3 leuLABCD    \n",
       "4   *3 leuLABCD attenuator terminator  3 leuLABCD attenuator terminator SNP    \n",
       "5                        *3 leuA/leuL      *3 leuLABCD attenuator terminator   \n",
       "6                        *3 leuLABCD                            *3 leuA/leuL   \n",
       "7                              3 LeuO                           *3 leuLABCD    \n",
       "8   *3 leuLABCD attenuator terminator  3 leuLABCD attenuator terminator SNP    \n",
       "9                        *3 leuA/leuL      *3 leuLABCD attenuator terminator   \n",
       "10                       *3 leuLABCD                            *3 leuA/leuL   \n",
       "11                             3 LeuO                           *3 leuLABCD    \n",
       "12                       *2 glxK/allE                       2 glxK/allE SNP    \n",
       "13                       *2 glxK/allE                           *2 glxK/allE   \n",
       "14                       *2 glxK/allE                       2 glxK/allE SNP    \n",
       "15                       *2 glxK/allE                           *2 glxK/allE   \n",
       "16                            *2 fimZ                     2 fimZ MOB disrupt   \n",
       "17                            *2 fimZ                                *2 fimZ   \n",
       "18                           *2 fimZ                                 *2 fimZ   \n",
       "19                            *2 fimZ                     2 fimZ MOB disrupt   \n",
       "\n",
       "       type  value  \n",
       "0       GYD    1.0  \n",
       "1       GYD    1.0  \n",
       "2       GYD    1.0  \n",
       "3       GYD    1.0  \n",
       "4       GYD    1.0  \n",
       "5       GYD    1.0  \n",
       "6       GYD    1.0  \n",
       "7       GYD    1.0  \n",
       "8       GYD    1.0  \n",
       "9       GYD    1.0  \n",
       "10      GYD    1.0  \n",
       "11      GYD    1.0  \n",
       "12      GYD    1.0  \n",
       "13      GYD    1.0  \n",
       "14      GYD    1.0  \n",
       "15      GYD    1.0  \n",
       "16  SSW_GLY    1.0  \n",
       "17  SSW_GLY    1.0  \n",
       "18  SSW_GLY    1.0  \n",
       "19      GYD    1.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "            \"type\": r[\"exp\"],\n",
    "#             \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": r[\"exp\"],\n",
    "#                 \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": r[\"exp\"],\n",
    "#                 \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"pathway\":\n",
    "        for pathway, ops in r[\"pathway links\"].items():\n",
    "#             if pathway == \"Glycerolipid metabolism\":  # for GYD Glycerolipid metabolism.\n",
    "            pathway_feat_d = _get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "            pathway_annot = \"\"\n",
    "            mut_count = pathway_counts[pathway]\n",
    "            if pathway_feat_d[\"significant\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "            pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "            pathways.append(pathway_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": pathway_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": r[\"exp\"],\n",
    "#                     \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "#             if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "            reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": r[\"exp\"],\n",
    "#                     \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "    if FUNC_ANNOT == \"COG\":\n",
    "        for COG, ops in r[\"COG links\"].items():\n",
    "#             if COG == \"Nucleotide transport and metabolism\":\n",
    "            COG_feat_d = _get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "            COG_annot = \"\"\n",
    "            mut_count = COG_counts[COG]\n",
    "            if COG_feat_d[\"significant\"] and mut_count > 1: COG_annot += \"*\"\n",
    "            COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "            COGs.append(COG_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": COG_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": r[\"exp\"],\n",
    "#                     \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e6822e8ae1a844bdaeba0f9170e7c5e2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'regulon', 'type': 'process', 'title': 'regulon', 'nodes': ['regulon^3 LeuO', 'reg…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "exp_flow_partition = Partition.Simple(\n",
    "    \"type\", [\"GYD\", \"SSW_GLY\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "                       flow_partition=exp_flow_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# palette_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "palette_d = {\"GYD\": \"#6BAAD3\", \"SSW_GLY\": \"#DD8452\"}\n",
    "size = dict(width=1400, height=900)\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=palette_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
