{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "# FUNC_ANNOT = \"regulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/SER_df.pkl\")\n",
    "display(len(muts_df),\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>gene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>TSS</th>\n",
       "      <th>RBS</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>gene count</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>base-media</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>imodulons</th>\n",
       "      <th>imodulon links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1363</th>\n",
       "      <td>27</td>\n",
       "      <td>Y356C (TAC→TGC)</td>\n",
       "      <td>thrA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→G</td>\n",
       "      <td>5</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1403, 1403)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>thrA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>{'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1345</th>\n",
       "      <td>0</td>\n",
       "      <td>S357R (AGC→CGC)</td>\n",
       "      <td>thrA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→C</td>\n",
       "      <td>3</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1405, 1405)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>thrA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>{'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>19</td>\n",
       "      <td>S359R (AGT→AGA)</td>\n",
       "      <td>thrA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→A</td>\n",
       "      <td>4</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1413, 1413)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>thrA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>{'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1347</th>\n",
       "      <td>1</td>\n",
       "      <td>coding (31‑34/1374 nt)</td>\n",
       "      <td>aroP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MOB</td>\n",
       "      <td>121518</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IS5 (+) +4 bp</td>\n",
       "      <td>3</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(121518, 121518)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120000080}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...</td>\n",
       "      <td>aroP</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>{'ECK120000080': ['ECK120000080']}</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12001447...</td>\n",
       "      <td>{'ECK120014477': ['ECK120000080']}</td>\n",
       "      <td>[{'name': 'GntR/TyrR', 'mutation set count': 1...</td>\n",
       "      <td>{'GntR/TyrR': ['aroP'], 'NagC/TyrR': ['aroP']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Cra', 'mutation set count': 3.0, 's...</td>\n",
       "      <td>{'Cra': ['aroP'], 'TyrR': ['aroP']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1370</th>\n",
       "      <td>28</td>\n",
       "      <td>L26L (TTG→TTA)</td>\n",
       "      <td>ybhN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>822421</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>5</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(822421, 822421)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120003005}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...</td>\n",
       "      <td>[{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...</td>\n",
       "      <td>ybhN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>{'ECK120003005': ['ECK120003005']}</td>\n",
       "      <td>[{'name': 'ybhP-clsB-ybhN', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK125160624': ['ECK120003005']}</td>\n",
       "      <td>[{'name': 'RpoS', 'mutation set count': 2.0, '...</td>\n",
       "      <td>{'RpoS': ['ybhP-clsB-ybhN']}</td>\n",
       "      <td>[{'name': 'Function unknown', 'mutation set co...</td>\n",
       "      <td>{'Function unknown': ['ECK125160624']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index                 Details mutation target annotation  Mut ID  \\\n",
       "1363    27         Y356C (TAC→TGC)                       thrA     NaN   \n",
       "1345     0         S357R (AGC→CGC)                       thrA     NaN   \n",
       "1355    19         S359R (AGT→AGA)                       thrA     NaN   \n",
       "1347     1  coding (31‑34/1374 nt)                       aroP     NaN   \n",
       "1370    28          L26L (TTG→TTA)                       ybhN     NaN   \n",
       "\n",
       "     Mutation Type  Position Reference Seq Sequence Change  ale  exp  flask  \\\n",
       "1363           SNP      1403           NaN             A→G    5  SER     50   \n",
       "1345           SNP      1405           NaN             A→C    3  SER     50   \n",
       "1355           SNP      1413           NaN             T→A    4  SER     50   \n",
       "1347           MOB    121518           NaN   IS5 (+) +4 bp    3  SER     50   \n",
       "1370           SNP    822421           NaN             C→T    5  SER     50   \n",
       "\n",
       "      isolate  presence  tech_rep  coding             range   oriC  \\\n",
       "1363        1       1.0         1    True      (1403, 1403)  False   \n",
       "1345        1       1.0         1    True      (1405, 1405)  False   \n",
       "1355        1       1.0         1    True      (1413, 1413)  False   \n",
       "1347        1       1.0         1    True  (121518, 121518)  False   \n",
       "1370        1       1.0         1    True  (822421, 822421)  False   \n",
       "\n",
       "      pseudogene            gene TFBS promoter TSS RBS cis-regulatory RNA  \\\n",
       "1363       False  {ECK120000987}   {}       {}  {}  {}                 {}   \n",
       "1345       False  {ECK120000987}   {}       {}  {}  {}                 {}   \n",
       "1355       False  {ECK120000987}   {}       {}  {}  {}                 {}   \n",
       "1347       False  {ECK120000080}   {}       {}  {}  {}                 {}   \n",
       "1370       False  {ECK120003005}   {}       {}  {}  {}                 {}   \n",
       "\n",
       "     attenuator terminator terminator  genetic  \\\n",
       "1363                    {}         {}     True   \n",
       "1345                    {}         {}     True   \n",
       "1355                    {}         {}     True   \n",
       "1347                    {}         {}     True   \n",
       "1370                    {}         {}     True   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1363  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1345  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1355  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1347  [{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...   \n",
       "1370  [{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...   \n",
       "\n",
       "                                       genomic features RegulonDB Gene  \\\n",
       "1363  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...           thrA   \n",
       "1345  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...           thrA   \n",
       "1355  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...           thrA   \n",
       "1347  [{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...           aroP   \n",
       "1370  [{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...           ybhN   \n",
       "\n",
       "      mutation size  gene count  disrupt  repaired taxonomy-id  \\\n",
       "1363              1           1    False     False      511145   \n",
       "1345              1           1    False     False      511145   \n",
       "1355              1           1    False     False      511145   \n",
       "1347              0           1     True     False      511145   \n",
       "1370              1           1    False     False      511145   \n",
       "\n",
       "           strain-description base-media temperature carbon-source  \\\n",
       "1363  ΔsdaA ΔsdaB ΔtdcG ΔglyA         M9  37 celsius    glucose(4)   \n",
       "1345  ΔsdaA ΔsdaB ΔtdcG ΔglyA         M9  37 celsius    glucose(4)   \n",
       "1355  ΔsdaA ΔsdaB ΔtdcG ΔglyA         M9  37 celsius    glucose(4)   \n",
       "1347  ΔsdaA ΔsdaB ΔtdcG ΔglyA         M9  37 celsius    glucose(4)   \n",
       "1370  ΔsdaA ΔsdaB ΔtdcG ΔglyA         M9  37 celsius    glucose(4)   \n",
       "\n",
       "     nitrogen-source phosphorous-source sulfur-source supplement  \\\n",
       "1363        NH4Cl(1)          KH2PO4(3)   MgSO4(0.24)   L-Serine   \n",
       "1345        NH4Cl(1)          KH2PO4(3)   MgSO4(0.24)   L-Serine   \n",
       "1355        NH4Cl(1)          KH2PO4(3)   MgSO4(0.24)   L-Serine   \n",
       "1347        NH4Cl(1)          KH2PO4(3)   MgSO4(0.24)   L-Serine   \n",
       "1370        NH4Cl(1)          KH2PO4(3)   MgSO4(0.24)   L-Serine   \n",
       "\n",
       "                   genetic feature links  \\\n",
       "1363  {'ECK120000987': ['ECK120000987']}   \n",
       "1345  {'ECK120000987': ['ECK120000987']}   \n",
       "1355  {'ECK120000987': ['ECK120000987']}   \n",
       "1347  {'ECK120000080': ['ECK120000080']}   \n",
       "1370  {'ECK120003005': ['ECK120003005']}   \n",
       "\n",
       "                                                operons  \\\n",
       "1363  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "1345  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "1355  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "1347  [{'name': 'aroP', 'RegulonDB ID': 'ECK12001447...   \n",
       "1370  [{'name': 'ybhP-clsB-ybhN', 'RegulonDB ID': 'E...   \n",
       "\n",
       "                            operon links  \\\n",
       "1363  {'ECK120014725': ['ECK120000987']}   \n",
       "1345  {'ECK120014725': ['ECK120000987']}   \n",
       "1355  {'ECK120014725': ['ECK120000987']}   \n",
       "1347  {'ECK120014477': ['ECK120000080']}   \n",
       "1370  {'ECK125160624': ['ECK120003005']}   \n",
       "\n",
       "                                              imodulons  \\\n",
       "1363  [{'name': 'Leu/Ile', 'mutation set count': 3.0...   \n",
       "1345  [{'name': 'Leu/Ile', 'mutation set count': 3.0...   \n",
       "1355  [{'name': 'Leu/Ile', 'mutation set count': 3.0...   \n",
       "1347  [{'name': 'GntR/TyrR', 'mutation set count': 1...   \n",
       "1370  [{'name': 'RpoS', 'mutation set count': 2.0, '...   \n",
       "\n",
       "                                         imodulon links  \\\n",
       "1363  {'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...   \n",
       "1345  {'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...   \n",
       "1355  {'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...   \n",
       "1347     {'GntR/TyrR': ['aroP'], 'NagC/TyrR': ['aroP']}   \n",
       "1370                       {'RpoS': ['ybhP-clsB-ybhN']}   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1363  [{'name': 'Amino acid transport and metabolism...   \n",
       "1345  [{'name': 'Amino acid transport and metabolism...   \n",
       "1355  [{'name': 'Amino acid transport and metabolism...   \n",
       "1347  [{'name': 'Amino acid transport and metabolism...   \n",
       "1370  [{'name': 'Function unknown', 'mutation set co...   \n",
       "\n",
       "                                              COG links  \\\n",
       "1363  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1345  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1355  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1347  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1370             {'Function unknown': ['ECK125160624']}   \n",
       "\n",
       "                                             regulators  \\\n",
       "1363                                                 []   \n",
       "1345                                                 []   \n",
       "1355                                                 []   \n",
       "1347  [{'name': 'Cra', 'mutation set count': 3.0, 's...   \n",
       "1370                                                 []   \n",
       "\n",
       "                          regulator links  \\\n",
       "1363                                   {}   \n",
       "1345                                   {}   \n",
       "1355                                   {}   \n",
       "1347  {'Cra': ['aroP'], 'TyrR': ['aroP']}   \n",
       "1370                                   {}   \n",
       "\n",
       "                                               pathways  \\\n",
       "1363  [{'name': 'Glycine, serine and threonine metab...   \n",
       "1345  [{'name': 'Glycine, serine and threonine metab...   \n",
       "1355  [{'name': 'Glycine, serine and threonine metab...   \n",
       "1347                                                 []   \n",
       "1370                                                 []   \n",
       "\n",
       "                                          pathway links  \n",
       "1363  {'Glycine, serine and threonine metabolism': [...  \n",
       "1345  {'Glycine, serine and threonine metabolism': [...  \n",
       "1355  {'Glycine, serine and threonine metabolism': [...  \n",
       "1347                                                 {}  \n",
       "1370                                                 {}  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specific to the AVA publication\n",
    "temp_df = pd.DataFrame()\n",
    "for i, r in muts_df.iterrows():\n",
    "    if \"thrA\" in r[\"RegulonDB Gene\"]:\n",
    "        temp_df = temp_df.append(r)\n",
    "muts_df = temp_df\n",
    "len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3, 16)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1363</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>thrA</td>\n",
       "      <td>thrA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1345</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>thrA</td>\n",
       "      <td>thrA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>thrA</td>\n",
       "      <td>thrA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type  disrupt  repaired  \\\n",
       "1363           SNP      0.0       0.0   \n",
       "1345           SNP      0.0       0.0   \n",
       "1355           SNP      0.0       0.0   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1363  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1345  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1355  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1363  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1345  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1355  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "\n",
       "                   genetic feature links  \\\n",
       "1363  {'ECK120000987': ['ECK120000987']}   \n",
       "1345  {'ECK120000987': ['ECK120000987']}   \n",
       "1355  {'ECK120000987': ['ECK120000987']}   \n",
       "\n",
       "                                                operons  \\\n",
       "1363  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "1345  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "1355  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "\n",
       "                            operon links regulators regulator links  \\\n",
       "1363  {'ECK120014725': ['ECK120000987']}         []              {}   \n",
       "1345  {'ECK120014725': ['ECK120000987']}         []              {}   \n",
       "1355  {'ECK120014725': ['ECK120000987']}         []              {}   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1363  [{'name': 'Amino acid transport and metabolism...   \n",
       "1345  [{'name': 'Amino acid transport and metabolism...   \n",
       "1355  [{'name': 'Amino acid transport and metabolism...   \n",
       "\n",
       "                                              COG links  \\\n",
       "1363  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1345  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1355  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "\n",
       "                                               pathways  \\\n",
       "1363  [{'name': 'Glycine, serine and threonine metab...   \n",
       "1345  [{'name': 'Glycine, serine and threonine metab...   \n",
       "1355  [{'name': 'Glycine, serine and threonine metab...   \n",
       "\n",
       "                                          pathway links  \\\n",
       "1363  {'Glycine, serine and threonine metabolism': [...   \n",
       "1345  {'Glycine, serine and threonine metabolism': [...   \n",
       "1355  {'Glycine, serine and threonine metabolism': [...   \n",
       "\n",
       "     mutation target annotation RegulonDB Gene  \n",
       "1363                       thrA           thrA  \n",
       "1345                       thrA           thrA  \n",
       "1355                       thrA           thrA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "    \"disrupt\",\n",
    "    \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "    \"mutation target annotation\",  # for assigning significant flow\n",
    "    \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>Details</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>TSS</th>\n",
       "      <th>ale</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>base-media</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>coding</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene count</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>imodulon links</th>\n",
       "      <th>imodulons</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>operon links</th>\n",
       "      <th>operons</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>repaired</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>temperature</th>\n",
       "      <th>terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1363</th>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>Y356C (TAC→TGC)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1403.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SER</td>\n",
       "      <td>50.0</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>27</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thrA</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(1403, 1403)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1345</th>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>S357R (AGC→CGC)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1405.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>3.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SER</td>\n",
       "      <td>50.0</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thrA</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(1405, 1405)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>S359R (AGT→AGA)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1413.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T→A</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>4.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SER</td>\n",
       "      <td>50.0</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>19</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>thrA</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(1413, 1413)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              COG links  \\\n",
       "1363  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1345  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1355  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "\n",
       "                                                   COGs          Details  \\\n",
       "1363  [{'name': 'Amino acid transport and metabolism...  Y356C (TAC→TGC)   \n",
       "1345  [{'name': 'Amino acid transport and metabolism...  S357R (AGC→CGC)   \n",
       "1355  [{'name': 'Amino acid transport and metabolism...  S359R (AGT→AGA)   \n",
       "\n",
       "      Mut ID Mutation Type  Position RBS  Reference Seq RegulonDB Gene  \\\n",
       "1363     NaN           SNP    1403.0  {}            NaN           thrA   \n",
       "1345     NaN           SNP    1405.0  {}            NaN           thrA   \n",
       "1355     NaN           SNP    1413.0  {}            NaN           thrA   \n",
       "\n",
       "     Sequence Change TFBS TSS  ale attenuator terminator base-media  \\\n",
       "1363             A→G   {}  {}  5.0                    {}         M9   \n",
       "1345             A→C   {}  {}  3.0                    {}         M9   \n",
       "1355             T→A   {}  {}  4.0                    {}         M9   \n",
       "\n",
       "     carbon-source cis-regulatory RNA  coding  disrupt  exp  flask  \\\n",
       "1363    glucose(4)                 {}     1.0      0.0  SER   50.0   \n",
       "1345    glucose(4)                 {}     1.0      0.0  SER   50.0   \n",
       "1355    glucose(4)                 {}     1.0      0.0  SER   50.0   \n",
       "\n",
       "                gene  gene count  genetic               genetic feature links  \\\n",
       "1363  {ECK120000987}         1.0      1.0  {'ECK120000987': ['ECK120000987']}   \n",
       "1345  {ECK120000987}         1.0      1.0  {'ECK120000987': ['ECK120000987']}   \n",
       "1355  {ECK120000987}         1.0      1.0  {'ECK120000987': ['ECK120000987']}   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1363  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1345  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1355  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1363  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1345  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "1355  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "\n",
       "                                         imodulon links  \\\n",
       "1363  {'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...   \n",
       "1345  {'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...   \n",
       "1355  {'Leu/Ile': ['thrLABC'], 'thrA-KO': ['thrLABC'...   \n",
       "\n",
       "                                              imodulons index  isolate  \\\n",
       "1363  [{'name': 'Leu/Ile', 'mutation set count': 3.0...    27      1.0   \n",
       "1345  [{'name': 'Leu/Ile', 'mutation set count': 3.0...     0      1.0   \n",
       "1355  [{'name': 'Leu/Ile', 'mutation set count': 3.0...    19      1.0   \n",
       "\n",
       "      mutation size mutation target annotation nitrogen-source  \\\n",
       "1363            1.0                       thrA        NH4Cl(1)   \n",
       "1345            1.0                       thrA        NH4Cl(1)   \n",
       "1355            1.0                       thrA        NH4Cl(1)   \n",
       "\n",
       "                            operon links  \\\n",
       "1363  {'ECK120014725': ['ECK120000987']}   \n",
       "1345  {'ECK120014725': ['ECK120000987']}   \n",
       "1355  {'ECK120014725': ['ECK120000987']}   \n",
       "\n",
       "                                                operons  oriC  \\\n",
       "1363  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   0.0   \n",
       "1345  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   0.0   \n",
       "1355  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   0.0   \n",
       "\n",
       "                                          pathway links  \\\n",
       "1363  {'Glycine, serine and threonine metabolism': [...   \n",
       "1345  {'Glycine, serine and threonine metabolism': [...   \n",
       "1355  {'Glycine, serine and threonine metabolism': [...   \n",
       "\n",
       "                                               pathways phosphorous-source  \\\n",
       "1363  [{'name': 'Glycine, serine and threonine metab...          KH2PO4(3)   \n",
       "1345  [{'name': 'Glycine, serine and threonine metab...          KH2PO4(3)   \n",
       "1355  [{'name': 'Glycine, serine and threonine metab...          KH2PO4(3)   \n",
       "\n",
       "      presence promoter  pseudogene         range regulator links regulators  \\\n",
       "1363       1.0       {}         0.0  (1403, 1403)              {}         []   \n",
       "1345       1.0       {}         0.0  (1405, 1405)              {}         []   \n",
       "1355       1.0       {}         0.0  (1413, 1413)              {}         []   \n",
       "\n",
       "      repaired       strain-description sulfur-source supplement taxonomy-id  \\\n",
       "1363       0.0  ΔsdaA ΔsdaB ΔtdcG ΔglyA   MgSO4(0.24)   L-Serine      511145   \n",
       "1345       0.0  ΔsdaA ΔsdaB ΔtdcG ΔglyA   MgSO4(0.24)   L-Serine      511145   \n",
       "1355       0.0  ΔsdaA ΔsdaB ΔtdcG ΔglyA   MgSO4(0.24)   L-Serine      511145   \n",
       "\n",
       "      tech_rep temperature terminator  \n",
       "1363       1.0  37 celsius         {}  \n",
       "1345       1.0  37 celsius         {}  \n",
       "1355       1.0  37 celsius         {}  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>3 thrA SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3 Glycine, serine and threonine metabolism</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3 Cysteine and methionine metabolism</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3 Lysine biosynthesis</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>3 thrA SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3 Glycine, serine and threonine metabolism</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3 Cysteine and methionine metabolism</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3 Lysine biosynthesis</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>3 thrA SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3 Glycine, serine and threonine metabolism</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3 Cysteine and methionine metabolism</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>3 Lysine biosynthesis</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        source       target type  value\n",
       "0                                      *3 thrA  3 thrA SNP     T    1.0\n",
       "1                                      *3 thrA      *3 thrA    T    1.0\n",
       "2                                  *3 thrLABC       *3 thrA    T    1.0\n",
       "3   3 Glycine, serine and threonine metabolism  *3 thrLABC     T    1.0\n",
       "4         3 Cysteine and methionine metabolism  *3 thrLABC     T    1.0\n",
       "5                        3 Lysine biosynthesis  *3 thrLABC     T    1.0\n",
       "6                                      *3 thrA  3 thrA SNP     T    1.0\n",
       "7                                      *3 thrA      *3 thrA    T    1.0\n",
       "8                                  *3 thrLABC       *3 thrA    T    1.0\n",
       "9   3 Glycine, serine and threonine metabolism  *3 thrLABC     T    1.0\n",
       "10        3 Cysteine and methionine metabolism  *3 thrLABC     T    1.0\n",
       "11                       3 Lysine biosynthesis  *3 thrLABC     T    1.0\n",
       "12                                     *3 thrA  3 thrA SNP     T    1.0\n",
       "13                                     *3 thrA      *3 thrA    T    1.0\n",
       "14                                 *3 thrLABC       *3 thrA    T    1.0\n",
       "15  3 Glycine, serine and threonine metabolism  *3 thrLABC     T    1.0\n",
       "16        3 Cysteine and methionine metabolism  *3 thrLABC     T    1.0\n",
       "17                       3 Lysine biosynthesis  *3 thrLABC     T    1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "\n",
    "\n",
    "KEY_MUT_PREFIX = [\"thr\", 'gcv', \"rho\", \"lrp\"]\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "\n",
    "    signif_flow = \"F\"\n",
    "    for k in KEY_MUT_PREFIX:\n",
    "        if k in r[\"mutation target annotation\"]:\n",
    "            signif_flow = \"T\"\n",
    "            break\n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "            \"type\": signif_flow,\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "                \"type\": signif_flow,\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "                \"type\": signif_flow,\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"pathway\":\n",
    "        for pathway, ops in r[\"pathway links\"].items():\n",
    "            pathway_feat_d = _get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "            pathway_annot = \"\"\n",
    "            mut_count = pathway_counts[pathway]\n",
    "            if pathway_feat_d[\"significant\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "            pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "            pathways.append(pathway_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": pathway_annot,\n",
    "                    \"target\": operon_annot,\n",
    "                    \"type\": signif_flow,\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "            reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "                    \"type\": signif_flow,\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "    if FUNC_ANNOT == \"COG\":\n",
    "        for COG, ops in r[\"COG links\"].items():\n",
    "#             if COG == \"Nucleotide transport and metabolism\":\n",
    "            COG_feat_d = _get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "            COG_annot = \"\"\n",
    "            mut_count = COG_counts[COG]\n",
    "            if COG_feat_d[\"significant\"] and mut_count > 1: COG_annot += \"*\"\n",
    "            COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "            COGs.append(COG_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": COG_annot,\n",
    "                    \"target\": operon_annot,\n",
    "                    \"type\": signif_flow,\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7dcb6485dd204f14a41a7f5a0d9a8834",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'pathway', 'type': 'process', 'title': 'pathway', 'nodes': ['pathway^3 Glycine, se…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "\n",
    "mut_signif_partition = Partition.Simple(\n",
    "    \"type\", ['T', 'F'])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "                       flow_partition=mut_signif_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "p_d = {'T': \"#6BAAD3\", 'F': \"#6BAAD3\"}\n",
    "size = dict(width=1200, height=200)\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
