{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "FUNC_ANNOT = \"regulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "163"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./data/GYD_df.pkl\")\n",
    "display(len(all_muts_df),\n",
    "#         muts_df.head()\n",
    "       )\n",
    "\n",
    "\n",
    "# # Below are the COGs that got multiple hits with GYD\n",
    "# '''\n",
    "# Lipid transport and metabolism\n",
    "# Cell cycle control, cell division, chromosome partitioning\n",
    "# Transcription\n",
    "# Amino acid transport and metabolism\n",
    "# Secondary metabolites biosynthesis, transport and catabolism\n",
    "# Cell wall/membrane/envelope biogenesis\n",
    "# Posttranslational modification, protein turnover, chaperones\n",
    "# Replication, recombination and repair\n",
    "# Signal transduction mechanisms\n",
    "# Cell motility\n",
    "# Defense mechanisms\n",
    "# Inorganic ion transport and metabolism\n",
    "# Intracellular trafficking, secretion, and vesicular transport\n",
    "# Coenzyme transport and metabolism\n",
    "# Energy production and conversion\n",
    "# Translation, ribosomal structure and biogenesis\n",
    "# Carbohydrate transport and metabolism\n",
    "# Nucleotide transport and metabolism\n",
    "# '''\n",
    "\n",
    "\n",
    "# out_df = pd.DataFrame()\n",
    "# df = muts_df.copy()\n",
    "# for _, r in df.iterrows():\n",
    "#     for c in r[\"operons\"]:\n",
    "#         try:\n",
    "#             # GYD\n",
    "#             if \"rpoB\" in c[\"name\"] or \"rpoC\" in c[\"name\"] or \"rph\" in c[\"name\"]:\n",
    "#                 out_df = out_df.append(r, ignore_index=True)\n",
    "#                 break\n",
    "# #             if \"ptsHI-crr\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#             # PGI\n",
    "# #             if \"pnt\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#         except:\n",
    "#             print(\"feature annotation bug with\", c)\n",
    "\n",
    "# muts_df = out_df\n",
    "# display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>Details</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>TSS</th>\n",
       "      <th>ale</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>base-media</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>coding</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene count</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>operon links</th>\n",
       "      <th>operons</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>repaired</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>temperature</th>\n",
       "      <th>terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1682</th>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>R665S (CGT→AGT)</td>\n",
       "      <td>1103591.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3993145.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>C→A</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280.0</td>\n",
       "      <td>{ECK120000166}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'ECK120000166': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...</td>\n",
       "      <td>126</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>cyaA</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014835': ['ECK120000166']}</td>\n",
       "      <td>[{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Purine metabolism': ['ECK120014835']}</td>\n",
       "      <td>[{'name': 'Purine metabolism', 'mutation set c...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(3993145, 3993145)</td>\n",
       "      <td>{'CRP': ['cyaA']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 72.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1683</th>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>S59Y (TCC→TAC)</td>\n",
       "      <td>1103607.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4117047.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>glpK</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>280.0</td>\n",
       "      <td>{ECK120000391}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>94</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>glpK</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism', 'mutation...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4117047, 4117047)</td>\n",
       "      <td>{'CRP': ['glpFKX'], 'GlpR': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 72.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1686</th>\n",
       "      <td>{'Transcription': ['ECK120014937']}</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>intergenic (‑93/‑505)</td>\n",
       "      <td>1105393.0</td>\n",
       "      <td>MOB</td>\n",
       "      <td>1293015.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>hns/tdk</td>\n",
       "      <td>IS1 (+) +8 bp</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>2.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>182.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ECK120000450/ECK120000983': ['ECK120010201']}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000450/ECK120000983',...</td>\n",
       "      <td>[{'name': 'hnsp', 'RegulonDB ID': 'ECK12001020...</td>\n",
       "      <td>130</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>hns/tdk</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014937': ['ECK120000450/ECK120000983']}</td>\n",
       "      <td>[{'name': 'hns', 'RegulonDB ID': 'ECK120014937...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{ECK120010201}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(1293015, 1293015)</td>\n",
       "      <td>{'GadX': ['hns'], 'H-NS': ['hns'], 'CspA': ['h...</td>\n",
       "      <td>[{'name': 'GadX', 'mutation set count': 6.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1688</th>\n",
       "      <td>{'Carbohydrate transport and metabolism': ['EC...</td>\n",
       "      <td>[{'name': 'Carbohydrate transport and metaboli...</td>\n",
       "      <td>I34T (ATC→ACC)</td>\n",
       "      <td>1105329.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2535934.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>crr</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>2.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>182.0</td>\n",
       "      <td>{ECK120000161}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'ECK120000161': ['ECK120000161']}</td>\n",
       "      <td>[{'name': 'crr', 'RegulonDB ID': 'ECK120000161...</td>\n",
       "      <td>[{'name': 'crr', 'RegulonDB ID': 'ECK120000161...</td>\n",
       "      <td>95</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>crr</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014520': ['ECK120000161']}</td>\n",
       "      <td>[{'name': 'ptsHI-crr', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(2535934, 2535934)</td>\n",
       "      <td>{'CRP': ['ptsHI-crr'], 'Mlc': ['ptsHI-crr'], '...</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 72.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1691</th>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>N229H (AAC→CAC)</td>\n",
       "      <td>1104408.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4116538.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>glpK</td>\n",
       "      <td>T→G</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>2.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>182.0</td>\n",
       "      <td>{ECK120000391}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>glpK</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism', 'mutation...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(4116538, 4116538)</td>\n",
       "      <td>{'CRP': ['glpFKX'], 'GlpR': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'CRP', 'mutation set count': 72.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              COG links  \\\n",
       "1682  {'Nucleotide transport and metabolism': ['ECK1...   \n",
       "1683  {'Energy production and conversion': ['ECK1200...   \n",
       "1686                {'Transcription': ['ECK120014937']}   \n",
       "1688  {'Carbohydrate transport and metabolism': ['EC...   \n",
       "1691  {'Energy production and conversion': ['ECK1200...   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1682  [{'name': 'Nucleotide transport and metabolism...   \n",
       "1683  [{'name': 'Energy production and conversion', ...   \n",
       "1686  [{'name': 'Transcription', 'mutation set count...   \n",
       "1688  [{'name': 'Carbohydrate transport and metaboli...   \n",
       "1691  [{'name': 'Energy production and conversion', ...   \n",
       "\n",
       "                    Details     Mut ID Mutation Type   Position RBS  \\\n",
       "1682        R665S (CGT→AGT)  1103591.0           SNP  3993145.0  {}   \n",
       "1683         S59Y (TCC→TAC)  1103607.0           SNP  4117047.0  {}   \n",
       "1686  intergenic (‑93/‑505)  1105393.0           MOB  1293015.0  {}   \n",
       "1688         I34T (ATC→ACC)  1105329.0           SNP  2535934.0  {}   \n",
       "1691        N229H (AAC→CAC)  1104408.0           SNP  4116538.0  {}   \n",
       "\n",
       "     Reference Seq RegulonDB Gene Sequence Change TFBS TSS  ale  \\\n",
       "1682     NC_000913           cyaA             C→A   {}  {}  1.0   \n",
       "1683     NC_000913           glpK             G→T   {}  {}  1.0   \n",
       "1686     NC_000913        hns/tdk   IS1 (+) +8 bp   {}  {}  2.0   \n",
       "1688     NC_000913            crr             T→C   {}  {}  2.0   \n",
       "1691     NC_000913           glpK             T→G   {}  {}  2.0   \n",
       "\n",
       "     attenuator terminator base-media  carbon-source cis-regulatory RNA  \\\n",
       "1682                    {}         M9  glycerol(2.5)                 {}   \n",
       "1683                    {}         M9  glycerol(2.5)                 {}   \n",
       "1686                    {}         M9  glycerol(2.5)                 {}   \n",
       "1688                    {}         M9  glycerol(2.5)                 {}   \n",
       "1691                    {}         M9  glycerol(2.5)                 {}   \n",
       "\n",
       "      coding  disrupt  exp  flask            gene  gene count  genetic  \\\n",
       "1682     1.0      0.0  GYD  280.0  {ECK120000166}         1.0      1.0   \n",
       "1683     1.0      0.0  GYD  280.0  {ECK120000391}         1.0      1.0   \n",
       "1686     0.0      1.0  GYD  182.0              {}         1.0      0.0   \n",
       "1688     1.0      0.0  GYD  182.0  {ECK120000161}         1.0      1.0   \n",
       "1691     1.0      0.0  GYD  182.0  {ECK120000391}         1.0      1.0   \n",
       "\n",
       "                                genetic feature links  \\\n",
       "1682               {'ECK120000166': ['ECK120000166']}   \n",
       "1683               {'ECK120000391': ['ECK120000391']}   \n",
       "1686  {'ECK120000450/ECK120000983': ['ECK120010201']}   \n",
       "1688               {'ECK120000161': ['ECK120000161']}   \n",
       "1691               {'ECK120000391': ['ECK120000391']}   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1682  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...   \n",
       "1683  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "1686  [{'RegulonDB ID': 'ECK120000450/ECK120000983',...   \n",
       "1688  [{'name': 'crr', 'RegulonDB ID': 'ECK120000161...   \n",
       "1691  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "\n",
       "                                       genomic features index  isolate  \\\n",
       "1682  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12000016...   126      0.0   \n",
       "1683  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...    94      0.0   \n",
       "1686  [{'name': 'hnsp', 'RegulonDB ID': 'ECK12001020...   130      1.0   \n",
       "1688  [{'name': 'crr', 'RegulonDB ID': 'ECK120000161...    95      0.0   \n",
       "1691  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...    16      1.0   \n",
       "\n",
       "      mutation size mutation target annotation nitrogen-source  \\\n",
       "1682            1.0                       cyaA        NH4Cl(1)   \n",
       "1683            1.0                       glpK        NH4Cl(1)   \n",
       "1686            0.0                    hns/tdk        NH4Cl(1)   \n",
       "1688            1.0                        crr        NH4Cl(1)   \n",
       "1691            1.0                       glpK        NH4Cl(1)   \n",
       "\n",
       "                                         operon links  \\\n",
       "1682               {'ECK120014835': ['ECK120000166']}   \n",
       "1683               {'ECK120014723': ['ECK120000391']}   \n",
       "1686  {'ECK120014937': ['ECK120000450/ECK120000983']}   \n",
       "1688               {'ECK120014520': ['ECK120000161']}   \n",
       "1691               {'ECK120014723': ['ECK120000391']}   \n",
       "\n",
       "                                                operons  oriC  \\\n",
       "1682  [{'name': 'cyaA', 'RegulonDB ID': 'ECK12001483...   0.0   \n",
       "1683  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   0.0   \n",
       "1686  [{'name': 'hns', 'RegulonDB ID': 'ECK120014937...   0.0   \n",
       "1688  [{'name': 'ptsHI-crr', 'RegulonDB ID': 'ECK120...   0.0   \n",
       "1691  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   0.0   \n",
       "\n",
       "                                      pathway links  \\\n",
       "1682        {'Purine metabolism': ['ECK120014835']}   \n",
       "1683  {'Glycerolipid metabolism': ['ECK120014723']}   \n",
       "1686                                             {}   \n",
       "1688                                             {}   \n",
       "1691  {'Glycerolipid metabolism': ['ECK120014723']}   \n",
       "\n",
       "                                               pathways phosphorous-source  \\\n",
       "1682  [{'name': 'Purine metabolism', 'mutation set c...          KH2PO4(3)   \n",
       "1683  [{'name': 'Glycerolipid metabolism', 'mutation...          KH2PO4(3)   \n",
       "1686                                                 []          KH2PO4(3)   \n",
       "1688                                                 []          KH2PO4(3)   \n",
       "1691  [{'name': 'Glycerolipid metabolism', 'mutation...          KH2PO4(3)   \n",
       "\n",
       "      presence        promoter  pseudogene               range  \\\n",
       "1682       1.0              {}         0.0  (3993145, 3993145)   \n",
       "1683       1.0              {}         0.0  (4117047, 4117047)   \n",
       "1686       1.0  {ECK120010201}         0.0  (1293015, 1293015)   \n",
       "1688       1.0              {}         0.0  (2535934, 2535934)   \n",
       "1691       1.0              {}         0.0  (4116538, 4116538)   \n",
       "\n",
       "                                        regulator links  \\\n",
       "1682                                  {'CRP': ['cyaA']}   \n",
       "1683            {'CRP': ['glpFKX'], 'GlpR': ['glpFKX']}   \n",
       "1686  {'GadX': ['hns'], 'H-NS': ['hns'], 'CspA': ['h...   \n",
       "1688  {'CRP': ['ptsHI-crr'], 'Mlc': ['ptsHI-crr'], '...   \n",
       "1691            {'CRP': ['glpFKX'], 'GlpR': ['glpFKX']}   \n",
       "\n",
       "                                             regulators  repaired  \\\n",
       "1682  [{'name': 'CRP', 'mutation set count': 72.0, '...       0.0   \n",
       "1683  [{'name': 'CRP', 'mutation set count': 72.0, '...       0.0   \n",
       "1686  [{'name': 'GadX', 'mutation set count': 6.0, '...       0.0   \n",
       "1688  [{'name': 'CRP', 'mutation set count': 72.0, '...       0.0   \n",
       "1691  [{'name': 'CRP', 'mutation set count': 72.0, '...       0.0   \n",
       "\n",
       "     strain-description sulfur-source      supplement taxonomy-id  tech_rep  \\\n",
       "1682              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1683              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1686              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1688              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1691              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "\n",
       "     temperature terminator  \n",
       "1682  37 celsius         {}  \n",
       "1683  37 celsius         {}  \n",
       "1686  37 celsius         {}  \n",
       "1688  37 celsius         {}  \n",
       "1691  37 celsius         {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# This is for filtering out mutations with genomic features statistically associated to specific conditions\n",
    "muts_df = pd.DataFrame()\n",
    "for _, mut in all_muts_df.iterrows():\n",
    "    for feat in mut[\"genomic features\"]:\n",
    "        if 'glycerol(2.5)' in feat['significantly associated conditions']:\n",
    "            muts_df = muts_df.append(mut)\n",
    "            # need to break since a mutation could have more than one feature\n",
    "            # which are associated with a condition,\n",
    "            # therefore adding a duplicate entry.\n",
    "            break\n",
    "display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COG links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>Details</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>TSS</th>\n",
       "      <th>ale</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>base-media</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>coding</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene count</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>operon links</th>\n",
       "      <th>operons</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>repaired</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>temperature</th>\n",
       "      <th>terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>317489.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>A→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>14.0</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>68.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.50</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 4.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>317489.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>A→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.59</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 4.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1738</th>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>intergenic (‑61/+32)</td>\n",
       "      <td>317490.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83590.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>G→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>10.0</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.59</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(83590, 83590)</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 4.0, '...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1782</th>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>intergenic (+40/+36)</td>\n",
       "      <td>1105347.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1018977.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>matP/ompA</td>\n",
       "      <td>C→A</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>20.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>33.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ECK120002403/ECK120000662': ['ECK120002403/E...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>81</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>matP/ompA</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.67</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(1018977, 1018977)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1832</th>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>intergenic (+46/+30)</td>\n",
       "      <td>570538.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1018983.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>matP/ompA</td>\n",
       "      <td>A→T</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>35.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GYD</td>\n",
       "      <td>9.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ECK120002403/ECK120000662': ['ECK120002403/E...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>135</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>matP/ompA</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>0.56</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>(1018983, 1018983)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              COG links  \\\n",
       "1763  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1737  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1738  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1782                                                 {}   \n",
       "1832                                                 {}   \n",
       "\n",
       "                                                   COGs               Details  \\\n",
       "1763  [{'name': 'Amino acid transport and metabolism...  intergenic (‑58/+35)   \n",
       "1737  [{'name': 'Amino acid transport and metabolism...  intergenic (‑58/+35)   \n",
       "1738  [{'name': 'Amino acid transport and metabolism...  intergenic (‑61/+32)   \n",
       "1782                                                 []  intergenic (+40/+36)   \n",
       "1832                                                 []  intergenic (+46/+30)   \n",
       "\n",
       "         Mut ID Mutation Type   Position RBS Reference Seq RegulonDB Gene  \\\n",
       "1763   317489.0           SNP    83587.0  {}     NC_000913      leuA/leuL   \n",
       "1737   317489.0           SNP    83587.0  {}     NC_000913      leuA/leuL   \n",
       "1738   317490.0           SNP    83590.0  {}     NC_000913      leuA/leuL   \n",
       "1782  1105347.0           SNP  1018977.0  {}     NC_000913      matP/ompA   \n",
       "1832   570538.0           SNP  1018983.0  {}     NC_000913      matP/ompA   \n",
       "\n",
       "     Sequence Change TFBS TSS   ale attenuator terminator base-media  \\\n",
       "1763             A→C   {}  {}  14.0        {ECK125143560}         M9   \n",
       "1737             A→C   {}  {}  10.0        {ECK125143560}         M9   \n",
       "1738             G→T   {}  {}  10.0        {ECK125143560}         M9   \n",
       "1782             C→A   {}  {}  20.0                    {}         M9   \n",
       "1832             A→T   {}  {}  35.0                    {}         M9   \n",
       "\n",
       "      carbon-source cis-regulatory RNA  coding  disrupt  exp  flask gene  \\\n",
       "1763  glycerol(2.5)     {ECK120001253}     0.0      0.0  GYD   68.0   {}   \n",
       "1737  glycerol(2.5)     {ECK120001253}     0.0      0.0  GYD  106.0   {}   \n",
       "1738  glycerol(2.5)     {ECK120001253}     0.0      0.0  GYD  106.0   {}   \n",
       "1782  glycerol(2.5)                 {}     0.0      0.0  GYD   33.0   {}   \n",
       "1832  glycerol(2.5)                 {}     0.0      0.0  GYD    9.0   {}   \n",
       "\n",
       "      gene count  genetic                              genetic feature links  \\\n",
       "1763         1.0      0.0    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1737         1.0      0.0    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1738         1.0      0.0    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1782         1.0      0.0  {'ECK120002403/ECK120000662': ['ECK120002403/E...   \n",
       "1832         1.0      0.0  {'ECK120002403/ECK120000662': ['ECK120002403/E...   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1763  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1737  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1738  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1782  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...   \n",
       "1832  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...   \n",
       "\n",
       "                                       genomic features index  isolate  \\\n",
       "1763  [{'name': 'leuLABCD attenuator terminator', 'R...    27      0.0   \n",
       "1737  [{'name': 'leuLABCD attenuator terminator', 'R...    27      0.0   \n",
       "1738  [{'name': 'leuLABCD attenuator terminator', 'R...    28      0.0   \n",
       "1782  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...    81      0.0   \n",
       "1832  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...   135      0.0   \n",
       "\n",
       "      mutation size mutation target annotation nitrogen-source  \\\n",
       "1763            1.0                  leuA/leuL        NH4Cl(1)   \n",
       "1737            1.0                  leuA/leuL        NH4Cl(1)   \n",
       "1738            1.0                  leuA/leuL        NH4Cl(1)   \n",
       "1782            1.0                  matP/ompA        NH4Cl(1)   \n",
       "1832            1.0                  matP/ompA        NH4Cl(1)   \n",
       "\n",
       "                                         operon links  \\\n",
       "1763  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1737  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1738  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1782                                               {}   \n",
       "1832                                               {}   \n",
       "\n",
       "                                                operons  oriC  \\\n",
       "1763  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   0.0   \n",
       "1737  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   0.0   \n",
       "1738  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   0.0   \n",
       "1782                                                 []   0.0   \n",
       "1832                                                 []   0.0   \n",
       "\n",
       "                                          pathway links  \\\n",
       "1763  {'Valine, leucine and isoleucine biosynthesis'...   \n",
       "1737  {'Valine, leucine and isoleucine biosynthesis'...   \n",
       "1738  {'Valine, leucine and isoleucine biosynthesis'...   \n",
       "1782                                                 {}   \n",
       "1832                                                 {}   \n",
       "\n",
       "                                               pathways phosphorous-source  \\\n",
       "1763  [{'name': 'Valine, leucine and isoleucine bios...          KH2PO4(3)   \n",
       "1737  [{'name': 'Valine, leucine and isoleucine bios...          KH2PO4(3)   \n",
       "1738  [{'name': 'Valine, leucine and isoleucine bios...          KH2PO4(3)   \n",
       "1782                                                 []          KH2PO4(3)   \n",
       "1832                                                 []          KH2PO4(3)   \n",
       "\n",
       "      presence promoter  pseudogene               range  \\\n",
       "1763      0.50       {}         0.0      (83587, 83587)   \n",
       "1737      0.59       {}         0.0      (83587, 83587)   \n",
       "1738      0.59       {}         0.0      (83590, 83590)   \n",
       "1782      0.67       {}         0.0  (1018977, 1018977)   \n",
       "1832      0.56       {}         0.0  (1018983, 1018983)   \n",
       "\n",
       "             regulator links  \\\n",
       "1763  {'LeuO': ['leuLABCD']}   \n",
       "1737  {'LeuO': ['leuLABCD']}   \n",
       "1738  {'LeuO': ['leuLABCD']}   \n",
       "1782                      {}   \n",
       "1832                      {}   \n",
       "\n",
       "                                             regulators  repaired  \\\n",
       "1763  [{'name': 'LeuO', 'mutation set count': 4.0, '...       0.0   \n",
       "1737  [{'name': 'LeuO', 'mutation set count': 4.0, '...       0.0   \n",
       "1738  [{'name': 'LeuO', 'mutation set count': 4.0, '...       0.0   \n",
       "1782                                                 []       0.0   \n",
       "1832                                                 []       0.0   \n",
       "\n",
       "     strain-description sulfur-source      supplement taxonomy-id  tech_rep  \\\n",
       "1763              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1737              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1738              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1782              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "1832              BOP27   MgSO4(0.24)  trace elements      511145       1.0   \n",
       "\n",
       "     temperature      terminator  \n",
       "1763  37 celsius  {ECK120034959}  \n",
       "1737  37 celsius  {ECK120034959}  \n",
       "1738  37 celsius  {ECK120034959}  \n",
       "1782  37 celsius              {}  \n",
       "1832  37 celsius              {}  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# muts_df = muts_df[muts_df[\"mutation target annotation\"].str.contains(\"pnt\")]\n",
    "# muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"CRP\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"GlpR\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Mlc\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "# #     if \"Nucleotide transport and metabolism\" in r[\"COG links\"].keys() \\\n",
    "#     if \"cyaA\" in r[\"RegulonDB Gene\"]:\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Glycerolipid metabolism\" in r[\"pathway links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = muts_df.copy()\n",
    "# df[\"significantly associated condition set\"] = df[\"significantly associated condition set\"].apply(\n",
    "#     lambda x: str(x))\n",
    "# display(df[\"significantly associated condition set\"].unique())\n",
    "\n",
    "\n",
    "# # GYD\n",
    "# muts_df = muts_df[\n",
    "#     (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)'})\n",
    "#     | (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)', \"BOP27\"})\n",
    "# ]\n",
    "\n",
    "# # SER\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"] == {'ΔsdaA ΔsdaB ΔtdcG ΔglyA', 'L-Serine'}]\n",
    "\n",
    "# # # GLU\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'glucose(2)'}]\n",
    "\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]==set()]\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'log', '511145', 'M9'}]\n",
    "\n",
    "\n",
    "# display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(91, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 4.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1737</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 4.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1738</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'mutation set count': 4.0, '...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1782</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>{'ECK120002403/ECK120000662': ['ECK120002403/E...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>matP/ompA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1832</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120002403/ECK120000662',...</td>\n",
       "      <td>{'ECK120002403/ECK120000662': ['ECK120002403/E...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>matP/ompA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type  disrupt  repaired  \\\n",
       "1763           SNP      0.0       0.0   \n",
       "1737           SNP      0.0       0.0   \n",
       "1738           SNP      0.0       0.0   \n",
       "1782           SNP      0.0       0.0   \n",
       "1832           SNP      0.0       0.0   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1763  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1737  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1738  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1782  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...   \n",
       "1832  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1763  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1737  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1738  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...   \n",
       "1782  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...   \n",
       "1832  [{'RegulonDB ID': 'ECK120002403/ECK120000662',...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1763    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1737    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1738    {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1782  {'ECK120002403/ECK120000662': ['ECK120002403/E...   \n",
       "1832  {'ECK120002403/ECK120000662': ['ECK120002403/E...   \n",
       "\n",
       "                                                operons  \\\n",
       "1763  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1737  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1738  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1782                                                 []   \n",
       "1832                                                 []   \n",
       "\n",
       "                                         operon links  \\\n",
       "1763  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1737  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1738  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1782                                               {}   \n",
       "1832                                               {}   \n",
       "\n",
       "                                             regulators  \\\n",
       "1763  [{'name': 'LeuO', 'mutation set count': 4.0, '...   \n",
       "1737  [{'name': 'LeuO', 'mutation set count': 4.0, '...   \n",
       "1738  [{'name': 'LeuO', 'mutation set count': 4.0, '...   \n",
       "1782                                                 []   \n",
       "1832                                                 []   \n",
       "\n",
       "             regulator links  \\\n",
       "1763  {'LeuO': ['leuLABCD']}   \n",
       "1737  {'LeuO': ['leuLABCD']}   \n",
       "1738  {'LeuO': ['leuLABCD']}   \n",
       "1782                      {}   \n",
       "1832                      {}   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1763  [{'name': 'Amino acid transport and metabolism...   \n",
       "1737  [{'name': 'Amino acid transport and metabolism...   \n",
       "1738  [{'name': 'Amino acid transport and metabolism...   \n",
       "1782                                                 []   \n",
       "1832                                                 []   \n",
       "\n",
       "                                              COG links  \\\n",
       "1763  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1737  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1738  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1782                                                 {}   \n",
       "1832                                                 {}   \n",
       "\n",
       "                                               pathways  \\\n",
       "1763  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1737  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1738  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1782                                                 []   \n",
       "1832                                                 []   \n",
       "\n",
       "                                          pathway links RegulonDB Gene  \n",
       "1763  {'Valine, leucine and isoleucine biosynthesis'...      leuA/leuL  \n",
       "1737  {'Valine, leucine and isoleucine biosynthesis'...      leuA/leuL  \n",
       "1738  {'Valine, leucine and isoleucine biosynthesis'...      leuA/leuL  \n",
       "1782                                                 {}      matP/ompA  \n",
       "1832                                                 {}      matP/ompA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "    \"disrupt\",\n",
    "    \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "    \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # For development. Delete when NB is stable.\n",
    "# for _, r in annot_df.iterrows():\n",
    "#     for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "#         effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "#         mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "#         mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "            \n",
    "#         genomic_feat_annot = \"\"\n",
    "#         if genomic_feat_d[\"significant\"]: genomic_feat_annot += \"*\"\n",
    "#         genomic_feat_annot += str(genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]) + ' '\n",
    "#         genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "    \n",
    "            \n",
    "#         display(genomic_feat_annot + ' #### ' + mut_type_feat_effect)\n",
    "# #         display(mut_type_feat_effect, \"################################\", r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>3 leuLABCD attenuator terminator SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3 leuLABCD</td>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3 LeuO</td>\n",
       "      <td>3 leuLABCD</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>3 leuLABCD attenuator terminator SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3 leuLABCD</td>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3 LeuO</td>\n",
       "      <td>3 leuLABCD</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>3 leuLABCD attenuator terminator SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>*3 leuLABCD attenuator terminator</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3 leuLABCD</td>\n",
       "      <td>*3 leuA/leuL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>3 LeuO</td>\n",
       "      <td>3 leuLABCD</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>3 matP/ompA SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>3 matP/ompA SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>3 matP/ompA SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>*3 matP/ompA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>*2 opgGH terminator</td>\n",
       "      <td>2 opgGH terminator SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>*2 yceK attenuator terminator</td>\n",
       "      <td>2 yceK attenuator terminator SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               source                                 target  \\\n",
       "0   *3 leuLABCD attenuator terminator  3 leuLABCD attenuator terminator SNP    \n",
       "1                        *3 leuA/leuL      *3 leuLABCD attenuator terminator   \n",
       "2                         3 leuLABCD                            *3 leuA/leuL   \n",
       "3                              3 LeuO                            3 leuLABCD    \n",
       "4   *3 leuLABCD attenuator terminator  3 leuLABCD attenuator terminator SNP    \n",
       "5                        *3 leuA/leuL      *3 leuLABCD attenuator terminator   \n",
       "6                         3 leuLABCD                            *3 leuA/leuL   \n",
       "7                              3 LeuO                            3 leuLABCD    \n",
       "8   *3 leuLABCD attenuator terminator  3 leuLABCD attenuator terminator SNP    \n",
       "9                        *3 leuA/leuL      *3 leuLABCD attenuator terminator   \n",
       "10                        3 leuLABCD                            *3 leuA/leuL   \n",
       "11                             3 LeuO                            3 leuLABCD    \n",
       "12                       *3 matP/ompA                       3 matP/ompA SNP    \n",
       "13                       *3 matP/ompA                           *3 matP/ompA   \n",
       "14                       *3 matP/ompA                       3 matP/ompA SNP    \n",
       "15                       *3 matP/ompA                           *3 matP/ompA   \n",
       "16                       *3 matP/ompA                       3 matP/ompA SNP    \n",
       "17                       *3 matP/ompA                           *3 matP/ompA   \n",
       "18                *2 opgGH terminator                2 opgGH terminator SNP    \n",
       "19      *2 yceK attenuator terminator      2 yceK attenuator terminator SNP    \n",
       "\n",
       "       type  value  \n",
       "0   unknown    1.0  \n",
       "1   unknown    1.0  \n",
       "2   unknown    1.0  \n",
       "3   unknown    1.0  \n",
       "4   unknown    1.0  \n",
       "5   unknown    1.0  \n",
       "6   unknown    1.0  \n",
       "7   unknown    1.0  \n",
       "8   unknown    1.0  \n",
       "9   unknown    1.0  \n",
       "10  unknown    1.0  \n",
       "11  unknown    1.0  \n",
       "12  unknown    1.0  \n",
       "13  unknown    1.0  \n",
       "14  unknown    1.0  \n",
       "15  unknown    1.0  \n",
       "16  unknown    1.0  \n",
       "17  unknown    1.0  \n",
       "18  unknown    1.0  \n",
       "19  unknown    1.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"pathway\":\n",
    "        for pathway, ops in r[\"pathway links\"].items():\n",
    "#             if pathway == \"Glycerolipid metabolism\":  # for GYD Glycerolipid metabolism.\n",
    "            pathway_feat_d = _get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "            pathway_annot = \"\"\n",
    "            mut_count = pathway_counts[pathway]\n",
    "            if pathway_feat_d[\"significant\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "            pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "            pathways.append(pathway_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": pathway_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "#             if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "            reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "    if FUNC_ANNOT == \"COG\":\n",
    "        for COG, ops in r[\"COG links\"].items():\n",
    "#             if COG == \"Nucleotide transport and metabolism\":\n",
    "            COG_feat_d = _get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "            COG_annot = \"\"\n",
    "            mut_count = COG_counts[COG]\n",
    "            if COG_feat_d[\"significant\"] and mut_count > 1: COG_annot += \"*\"\n",
    "            COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "            COGs.append(COG_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": COG_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specific to the GYD fig of the AVA paper.\n",
    "flow_df = flow_df[~(flow_df[\"target\"].str.contains(\"1 pdxK\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "44c38b57cfb8485097602401602fcced",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'regulon', 'type': 'process', 'title': 'regulon', 'nodes': ['regulon^3 LeuO', 'reg…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "#                        flow_partition=mut_effect_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# p_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "# size = dict(width=3000, height=4000)\n",
    "size = dict(width=1500, height=1000)\n",
    "# size = dict(width=1600, height=1000)\n",
    "# size = dict(width=1350, height=800)\n",
    "# size = dict(width=1600, height=500)\n",
    "# size = dict(width=1100, height=300)\n",
    "# size = dict(width=1350, height=300)\n",
    "# size = dict(width=1000, height=200)\n",
    "# size = dict(width=1200, height=300)  # for GYD GlpR mutflow fig\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
