{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "# FUNC_ANNOT = \"regulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "148"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/GYD_df.pkl\")\n",
    "display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>base-media</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>calcium-source</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>694</td>\n",
       "      <td>7</td>\n",
       "      <td>intergenic (+74/‑188)</td>\n",
       "      <td>carB/caiF</td>\n",
       "      <td>DEL</td>\n",
       "      <td>34112</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>12</td>\n",
       "      <td>GYD</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(34112, 34112)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000131/ECK120002708',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125257016}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'carAB terminator', 'RegulonDB ID': ...</td>\n",
       "      <td>{'ECK120000131/ECK120002708': ['ECK125257016']}</td>\n",
       "      <td>[{'name': 'carAB', 'RegulonDB ID': 'ECK1200149...</td>\n",
       "      <td>{'ECK120014911': ['ECK120000131/ECK120002708']}</td>\n",
       "      <td>[{'name': 'ArgR', 'significantly associated co...</td>\n",
       "      <td>{'Fis': ['carAB'], 'RutR': ['carAB'], 'IHF': [...</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism', 'mutation s...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014911', 'EC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>712</td>\n",
       "      <td>27</td>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>14</td>\n",
       "      <td>GYD</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'significantly associated co...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>686</td>\n",
       "      <td>27</td>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>10</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'significantly associated co...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>687</td>\n",
       "      <td>28</td>\n",
       "      <td>intergenic (‑61/+32)</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83590</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→T</td>\n",
       "      <td>10</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(83590, 83590)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'LeuO', 'significantly associated co...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>675</td>\n",
       "      <td>44</td>\n",
       "      <td>coding (314‑317/597 nt)</td>\n",
       "      <td>yadK</td>\n",
       "      <td>MOB</td>\n",
       "      <td>151283</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>IS5 (–) +4 bp</td>\n",
       "      <td>7</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(151283, 151283)</td>\n",
       "      <td>{ECK120002200}</td>\n",
       "      <td>[{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...</td>\n",
       "      <td>{'ECK120002200': ['ECK120002200']}</td>\n",
       "      <td>[{'name': 'yadMLKC', 'RegulonDB ID': 'ECK12003...</td>\n",
       "      <td>{'ECK120030520': ['ECK120002200']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NaCl(0.5g/L) trace elements</td>\n",
       "      <td>WT</td>\n",
       "      <td>511145</td>\n",
       "      <td>M9</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3) Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>[{'name': 'Cell motility', 'mutation set count...</td>\n",
       "      <td>{'Cell motility': ['ECK120030520']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index                  Details mutation target annotation Mutation Type  \\\n",
       "694     7    intergenic (+74/‑188)                  carB/caiF           DEL   \n",
       "712    27     intergenic (‑58/+35)                  leuA/leuL           SNP   \n",
       "686    27     intergenic (‑58/+35)                  leuA/leuL           SNP   \n",
       "687    28     intergenic (‑61/+32)                  leuA/leuL           SNP   \n",
       "675    44  coding (314‑317/597 nt)                       yadK           MOB   \n",
       "\n",
       "     Position Reference Seq Sequence Change  ale  exp  flask  isolate  \\\n",
       "694     34112     NC_000913           Δ1 bp   12  GYD    114        0   \n",
       "712     83587     NC_000913             A→C   14  GYD     68        0   \n",
       "686     83587     NC_000913             A→C   10  GYD    106        0   \n",
       "687     83590     NC_000913             G→T   10  GYD    106        0   \n",
       "675    151283     NC_000913   IS5 (–) +4 bp    7  GYD    106        1   \n",
       "\n",
       "     presence  tech_rep  coding             range gene RegulonDB ID  \\\n",
       "694      1.00         1   False    (34112, 34112)                {}   \n",
       "712      0.50         1   False    (83587, 83587)                {}   \n",
       "686      0.59         1   False    (83587, 83587)                {}   \n",
       "687      0.59         1   False    (83590, 83590)                {}   \n",
       "675      1.00         1    True  (151283, 151283)    {ECK120002200}   \n",
       "\n",
       "                                      genetic features   oriC  pseudogene  \\\n",
       "694  [{'RegulonDB ID': 'ECK120000131/ECK120002708',...  False       False   \n",
       "712  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...  False       False   \n",
       "686  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...  False       False   \n",
       "687  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...  False       False   \n",
       "675  [{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...  False       False   \n",
       "\n",
       "    TFBS promoter RBS attenuator terminator      terminator  genetic  \\\n",
       "694   {}       {}  {}                    {}  {ECK125257016}    False   \n",
       "712   {}       {}  {}        {ECK125143560}  {ECK120034959}    False   \n",
       "686   {}       {}  {}        {ECK125143560}  {ECK120034959}    False   \n",
       "687   {}       {}  {}        {ECK125143560}  {ECK120034959}    False   \n",
       "675   {}       {}  {}                    {}              {}     True   \n",
       "\n",
       "                                      genomic features  \\\n",
       "694  [{'name': 'carAB terminator', 'RegulonDB ID': ...   \n",
       "712  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "686  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "687  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "675  [{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...   \n",
       "\n",
       "                               genetic feature links  \\\n",
       "694  {'ECK120000131/ECK120002708': ['ECK125257016']}   \n",
       "712  {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "686  {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "687  {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "675               {'ECK120002200': ['ECK120002200']}   \n",
       "\n",
       "                                               operons  \\\n",
       "694  [{'name': 'carAB', 'RegulonDB ID': 'ECK1200149...   \n",
       "712  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "686  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "687  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "675  [{'name': 'yadMLKC', 'RegulonDB ID': 'ECK12003...   \n",
       "\n",
       "                                        operon links  \\\n",
       "694  {'ECK120014911': ['ECK120000131/ECK120002708']}   \n",
       "712  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "686  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "687  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "675               {'ECK120030520': ['ECK120002200']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "694  [{'name': 'ArgR', 'significantly associated co...   \n",
       "712  [{'name': 'LeuO', 'significantly associated co...   \n",
       "686  [{'name': 'LeuO', 'significantly associated co...   \n",
       "687  [{'name': 'LeuO', 'significantly associated co...   \n",
       "675                                                 []   \n",
       "\n",
       "                                       regulator links temperature  \\\n",
       "694  {'Fis': ['carAB'], 'RutR': ['carAB'], 'IHF': [...  37 celsius   \n",
       "712                             {'LeuO': ['leuLABCD']}  37 celsius   \n",
       "686                             {'LeuO': ['leuLABCD']}  37 celsius   \n",
       "687                             {'LeuO': ['leuLABCD']}  37 celsius   \n",
       "675                                                 {}  37 celsius   \n",
       "\n",
       "     carbon-source                   supplement strain-description  \\\n",
       "694  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "712  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "686  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "687  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "675  glycerol(2.5)  NaCl(0.5g/L) trace elements                 WT   \n",
       "\n",
       "    taxonomy-id base-media nitrogen-source      phosphorous-source  \\\n",
       "694      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "712      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "686      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "687      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "675      511145         M9        NH4Cl(1)  KH2PO4(3) Na2HPO4(6.8)   \n",
       "\n",
       "    sulfur-source calcium-source  \\\n",
       "694   MgSO4(0.24)     CaCl2(0.1)   \n",
       "712   MgSO4(0.24)     CaCl2(0.1)   \n",
       "686   MgSO4(0.24)     CaCl2(0.1)   \n",
       "687   MgSO4(0.24)     CaCl2(0.1)   \n",
       "675   MgSO4(0.24)     CaCl2(0.1)   \n",
       "\n",
       "                                                  COGs  \\\n",
       "694  [{'name': 'Amino acid transport and metabolism...   \n",
       "712  [{'name': 'Amino acid transport and metabolism...   \n",
       "686  [{'name': 'Amino acid transport and metabolism...   \n",
       "687  [{'name': 'Amino acid transport and metabolism...   \n",
       "675  [{'name': 'Cell motility', 'mutation set count...   \n",
       "\n",
       "                                             COG links  \\\n",
       "694  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "712  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "686  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "687  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "675                {'Cell motility': ['ECK120030520']}   \n",
       "\n",
       "                                              pathways  \\\n",
       "694  [{'name': 'Pyrimidine metabolism', 'mutation s...   \n",
       "712  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "686  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "687  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "675                                                 []   \n",
       "\n",
       "                                         pathway links  \n",
       "694  {'Pyrimidine metabolism': ['ECK120014911', 'EC...  \n",
       "712  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "686  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "687  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "675                                                 {}  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specific to the AVA publication\n",
    "temp_df = pd.DataFrame()\n",
    "for i, r in muts_df.iterrows():\n",
    "#     if \"Nucleotide transport and metabolism\" in r[\"COG links\"].keys() \\\n",
    "    if \"glpK\" in r[\"mutation target annotation\"]:\n",
    "        temp_df = temp_df.append(r)\n",
    "muts_df = temp_df\n",
    "len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28, 12)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>751</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism', 'mutation...</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>677</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism', 'mutation...</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>744</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism', 'mutation...</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>678</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism', 'mutation...</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>640</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>[{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...</td>\n",
       "      <td>{'ECK120000391': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014723': ['ECK120000391']}</td>\n",
       "      <td>[{'name': 'GlpR', 'significantly associated co...</td>\n",
       "      <td>{'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Glycerolipid metabolism', 'mutation...</td>\n",
       "      <td>{'Glycerolipid metabolism': ['ECK120014723']}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutation Type                                   genomic features  \\\n",
       "751           SNP  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "677           SNP  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "744           SNP  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "678           SNP  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "640           SNP  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "\n",
       "                                      genetic features  \\\n",
       "751  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "677  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "744  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "678  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "640  [{'name': 'glpK', 'RegulonDB ID': 'ECK12000039...   \n",
       "\n",
       "                  genetic feature links  \\\n",
       "751  {'ECK120000391': ['ECK120000391']}   \n",
       "677  {'ECK120000391': ['ECK120000391']}   \n",
       "744  {'ECK120000391': ['ECK120000391']}   \n",
       "678  {'ECK120000391': ['ECK120000391']}   \n",
       "640  {'ECK120000391': ['ECK120000391']}   \n",
       "\n",
       "                                               operons  \\\n",
       "751  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "677  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "744  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "678  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "640  [{'name': 'glpFKX', 'RegulonDB ID': 'ECK120014...   \n",
       "\n",
       "                           operon links  \\\n",
       "751  {'ECK120014723': ['ECK120000391']}   \n",
       "677  {'ECK120014723': ['ECK120000391']}   \n",
       "744  {'ECK120014723': ['ECK120000391']}   \n",
       "678  {'ECK120014723': ['ECK120000391']}   \n",
       "640  {'ECK120014723': ['ECK120000391']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "751  [{'name': 'GlpR', 'significantly associated co...   \n",
       "677  [{'name': 'GlpR', 'significantly associated co...   \n",
       "744  [{'name': 'GlpR', 'significantly associated co...   \n",
       "678  [{'name': 'GlpR', 'significantly associated co...   \n",
       "640  [{'name': 'GlpR', 'significantly associated co...   \n",
       "\n",
       "                             regulator links  \\\n",
       "751  {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "677  {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "744  {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "678  {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "640  {'GlpR': ['glpFKX'], 'CRP': ['glpFKX']}   \n",
       "\n",
       "                                                  COGs  \\\n",
       "751  [{'name': 'Energy production and conversion', ...   \n",
       "677  [{'name': 'Energy production and conversion', ...   \n",
       "744  [{'name': 'Energy production and conversion', ...   \n",
       "678  [{'name': 'Energy production and conversion', ...   \n",
       "640  [{'name': 'Energy production and conversion', ...   \n",
       "\n",
       "                                             COG links  \\\n",
       "751  {'Energy production and conversion': ['ECK1200...   \n",
       "677  {'Energy production and conversion': ['ECK1200...   \n",
       "744  {'Energy production and conversion': ['ECK1200...   \n",
       "678  {'Energy production and conversion': ['ECK1200...   \n",
       "640  {'Energy production and conversion': ['ECK1200...   \n",
       "\n",
       "                                              pathways  \\\n",
       "751  [{'name': 'Glycerolipid metabolism', 'mutation...   \n",
       "677  [{'name': 'Glycerolipid metabolism', 'mutation...   \n",
       "744  [{'name': 'Glycerolipid metabolism', 'mutation...   \n",
       "678  [{'name': 'Glycerolipid metabolism', 'mutation...   \n",
       "640  [{'name': 'Glycerolipid metabolism', 'mutation...   \n",
       "\n",
       "                                     pathway links  \n",
       "751  {'Glycerolipid metabolism': ['ECK120014723']}  \n",
       "677  {'Glycerolipid metabolism': ['ECK120014723']}  \n",
       "744  {'Glycerolipid metabolism': ['ECK120014723']}  \n",
       "678  {'Glycerolipid metabolism': ['ECK120014723']}  \n",
       "640  {'Glycerolipid metabolism': ['ECK120014723']}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "#     \"disrupt\",\n",
    "#     \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "#     \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>28 glpK SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>*30 Glycerolipid metabolism</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>28 glpK SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>*30 Glycerolipid metabolism</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>28 glpK SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>*30 Glycerolipid metabolism</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>28 glpK SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>*30 Glycerolipid metabolism</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>28 glpK SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>*28 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>*30 glpK</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>*30 Glycerolipid metabolism</td>\n",
       "      <td>*30 glpFKX</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         source        target     type  value\n",
       "0                      *28 glpK  28 glpK SNP   unknown    1.0\n",
       "1                      *30 glpK      *28 glpK  unknown    1.0\n",
       "2                   *30 glpFKX       *30 glpK  unknown    1.0\n",
       "3   *30 Glycerolipid metabolism   *30 glpFKX   unknown    1.0\n",
       "4                      *28 glpK  28 glpK SNP   unknown    1.0\n",
       "5                      *30 glpK      *28 glpK  unknown    1.0\n",
       "6                   *30 glpFKX       *30 glpK  unknown    1.0\n",
       "7   *30 Glycerolipid metabolism   *30 glpFKX   unknown    1.0\n",
       "8                      *28 glpK  28 glpK SNP   unknown    1.0\n",
       "9                      *30 glpK      *28 glpK  unknown    1.0\n",
       "10                  *30 glpFKX       *30 glpK  unknown    1.0\n",
       "11  *30 Glycerolipid metabolism   *30 glpFKX   unknown    1.0\n",
       "12                     *28 glpK  28 glpK SNP   unknown    1.0\n",
       "13                     *30 glpK      *28 glpK  unknown    1.0\n",
       "14                  *30 glpFKX       *30 glpK  unknown    1.0\n",
       "15  *30 Glycerolipid metabolism   *30 glpFKX   unknown    1.0\n",
       "16                     *28 glpK  28 glpK SNP   unknown    1.0\n",
       "17                     *30 glpK      *28 glpK  unknown    1.0\n",
       "18                  *30 glpFKX       *30 glpK  unknown    1.0\n",
       "19  *30 Glycerolipid metabolism   *30 glpFKX   unknown    1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significantly enriched\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significantly enriched\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significantly enriched\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significantly enriched\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significantly enriched\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"pathway\":\n",
    "        for pathway, ops in r[\"pathway links\"].items():\n",
    "            if pathway == \"Glycerolipid metabolism\":  # for GYD Glycerolipid metabolism.\n",
    "                pathway_feat_d = _get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "                pathway_annot = \"\"\n",
    "                mut_count = pathway_counts[pathway]\n",
    "                if pathway_feat_d[\"significantly enriched\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "                pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "                pathways.append(pathway_annot)\n",
    "                for operon_id in ops:\n",
    "                    op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                    operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                    operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                    operon_annot = \"\"\n",
    "                    mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                    if op_feat_d[\"significantly enriched\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                    operon_annot += str(mut_count) + ' '\n",
    "                    operon_annot += operon_name\n",
    "                    operons.append(operon_annot)\n",
    "                    flow_row = {\n",
    "                        \"source\": pathway_annot,\n",
    "                        \"target\": operon_annot,\n",
    "    #                     \"type\": r[\"mutation effect\"],\n",
    "                        \"type\": \"unknown\",\n",
    "                        \"value\": 1\n",
    "                    }\n",
    "                    flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "#             if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "            reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significantly enriched\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significantly enriched\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "    if FUNC_ANNOT == \"COG\":\n",
    "        for COG, ops in r[\"COG links\"].items():\n",
    "#             if COG == \"Nucleotide transport and metabolism\":\n",
    "            COG_feat_d = _get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "            COG_annot = \"\"\n",
    "            mut_count = COG_counts[COG]\n",
    "            if COG_feat_d[\"significantly enriched\"] and mut_count > 1: COG_annot += \"*\"\n",
    "            COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "            COGs.append(COG_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significantly enriched\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": COG_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f361b40e830644858dfd1001e3c31be6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'pathway', 'type': 'process', 'title': 'pathway', 'nodes': ['pathway^*30 Glyceroli…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "\n",
    "sdd = SankeyDefinition(\n",
    "    nodes,\n",
    "    bundles,\n",
    "    ordering,\n",
    ")\n",
    "\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "size = dict(width=1250, height=150)\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
