{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "FUNC_ANNOT = \"regulon\"\n",
    "# FUNC_ANNOT = \"imodulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/SER_df.pkl\")\n",
    "display(len(muts_df),\n",
    "       )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for _, r in muts_df.iterrows():\n",
    "#     for f in r[\"regulators\"]:\n",
    "#         if \"GcvA\" in f[\"name\"]:\n",
    "#             display(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>base-media</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>860</td>\n",
       "      <td>27</td>\n",
       "      <td>Y356C (TAC→TGC)</td>\n",
       "      <td>thrA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→G</td>\n",
       "      <td>5</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1403, 1403)</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(2)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>842</td>\n",
       "      <td>0</td>\n",
       "      <td>S357R (AGC→CGC)</td>\n",
       "      <td>thrA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A→C</td>\n",
       "      <td>3</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1405, 1405)</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(2)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>852</td>\n",
       "      <td>19</td>\n",
       "      <td>S359R (AGT→AGA)</td>\n",
       "      <td>thrA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1413</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→A</td>\n",
       "      <td>4</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(1413, 1413)</td>\n",
       "      <td>{ECK120000987}</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(2)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Glycine, serine and threonine metab...</td>\n",
       "      <td>{'Glycine, serine and threonine metabolism': [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>844</td>\n",
       "      <td>1</td>\n",
       "      <td>coding (31‑34/1374 nt)</td>\n",
       "      <td>aroP</td>\n",
       "      <td>MOB</td>\n",
       "      <td>121518</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IS5 (+) +4 bp</td>\n",
       "      <td>3</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(121518, 121518)</td>\n",
       "      <td>{ECK120000080}</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...</td>\n",
       "      <td>{'ECK120000080': ['ECK120000080']}</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12001447...</td>\n",
       "      <td>{'ECK120014477': ['ECK120000080']}</td>\n",
       "      <td>[{'name': 'TyrR', 'significantly associated co...</td>\n",
       "      <td>{'TyrR': ['aroP'], 'Cra': ['aroP']}</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(2)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>867</td>\n",
       "      <td>28</td>\n",
       "      <td>L26L (TTG→TTA)</td>\n",
       "      <td>ybhN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>822421</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>5</td>\n",
       "      <td>SER</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(822421, 822421)</td>\n",
       "      <td>{ECK120003005}</td>\n",
       "      <td>[{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...</td>\n",
       "      <td>{'ECK120003005': ['ECK120003005']}</td>\n",
       "      <td>[{'name': 'ybhP-clsB-ybhN', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK125160624': ['ECK120003005']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>511145</td>\n",
       "      <td>ΔsdaA ΔsdaB ΔtdcG ΔglyA</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(2)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index                 Details mutation target annotation Mutation Type  \\\n",
       "860    27         Y356C (TAC→TGC)                       thrA           SNP   \n",
       "842     0         S357R (AGC→CGC)                       thrA           SNP   \n",
       "852    19         S359R (AGT→AGA)                       thrA           SNP   \n",
       "844     1  coding (31‑34/1374 nt)                       aroP           MOB   \n",
       "867    28          L26L (TTG→TTA)                       ybhN           SNP   \n",
       "\n",
       "     Position Reference Seq Sequence Change  ale  exp  flask  isolate  \\\n",
       "860      1403           NaN             A→G    5  SER     50        1   \n",
       "842      1405           NaN             A→C    3  SER     50        1   \n",
       "852      1413           NaN             T→A    4  SER     50        1   \n",
       "844    121518           NaN   IS5 (+) +4 bp    3  SER     50        1   \n",
       "867    822421           NaN             C→T    5  SER     50        1   \n",
       "\n",
       "     presence  tech_rep  coding             range gene RegulonDB ID  \\\n",
       "860       1.0         1    True      (1403, 1403)    {ECK120000987}   \n",
       "842       1.0         1    True      (1405, 1405)    {ECK120000987}   \n",
       "852       1.0         1    True      (1413, 1413)    {ECK120000987}   \n",
       "844       1.0         1    True  (121518, 121518)    {ECK120000080}   \n",
       "867       1.0         1    True  (822421, 822421)    {ECK120003005}   \n",
       "\n",
       "                                      genetic features   oriC  pseudogene  \\\n",
       "860  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...  False       False   \n",
       "842  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...  False       False   \n",
       "852  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...  False       False   \n",
       "844  [{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...  False       False   \n",
       "867  [{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...  False       False   \n",
       "\n",
       "    TFBS promoter RBS attenuator terminator terminator  genetic  \\\n",
       "860   {}       {}  {}                    {}         {}     True   \n",
       "842   {}       {}  {}                    {}         {}     True   \n",
       "852   {}       {}  {}                    {}         {}     True   \n",
       "844   {}       {}  {}                    {}         {}     True   \n",
       "867   {}       {}  {}                    {}         {}     True   \n",
       "\n",
       "                                      genomic features  \\\n",
       "860  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "842  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "852  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "844  [{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...   \n",
       "867  [{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...   \n",
       "\n",
       "                  genetic feature links  \\\n",
       "860  {'ECK120000987': ['ECK120000987']}   \n",
       "842  {'ECK120000987': ['ECK120000987']}   \n",
       "852  {'ECK120000987': ['ECK120000987']}   \n",
       "844  {'ECK120000080': ['ECK120000080']}   \n",
       "867  {'ECK120003005': ['ECK120003005']}   \n",
       "\n",
       "                                               operons  \\\n",
       "860  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "842  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "852  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "844  [{'name': 'aroP', 'RegulonDB ID': 'ECK12001447...   \n",
       "867  [{'name': 'ybhP-clsB-ybhN', 'RegulonDB ID': 'E...   \n",
       "\n",
       "                           operon links  \\\n",
       "860  {'ECK120014725': ['ECK120000987']}   \n",
       "842  {'ECK120014725': ['ECK120000987']}   \n",
       "852  {'ECK120014725': ['ECK120000987']}   \n",
       "844  {'ECK120014477': ['ECK120000080']}   \n",
       "867  {'ECK125160624': ['ECK120003005']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "860                                                 []   \n",
       "842                                                 []   \n",
       "852                                                 []   \n",
       "844  [{'name': 'TyrR', 'significantly associated co...   \n",
       "867                                                 []   \n",
       "\n",
       "                         regulator links taxonomy-id       strain-description  \\\n",
       "860                                   {}      511145  ΔsdaA ΔsdaB ΔtdcG ΔglyA   \n",
       "842                                   {}      511145  ΔsdaA ΔsdaB ΔtdcG ΔglyA   \n",
       "852                                   {}      511145  ΔsdaA ΔsdaB ΔtdcG ΔglyA   \n",
       "844  {'TyrR': ['aroP'], 'Cra': ['aroP']}      511145  ΔsdaA ΔsdaB ΔtdcG ΔglyA   \n",
       "867                                   {}      511145  ΔsdaA ΔsdaB ΔtdcG ΔglyA   \n",
       "\n",
       "    base-media temperature carbon-source nitrogen-source phosphorous-source  \\\n",
       "860         M9  37 celsius    glucose(4)        NH4Cl(1)          KH2PO4(2)   \n",
       "842         M9  37 celsius    glucose(4)        NH4Cl(1)          KH2PO4(2)   \n",
       "852         M9  37 celsius    glucose(4)        NH4Cl(1)          KH2PO4(2)   \n",
       "844         M9  37 celsius    glucose(4)        NH4Cl(1)          KH2PO4(2)   \n",
       "867         M9  37 celsius    glucose(4)        NH4Cl(1)          KH2PO4(2)   \n",
       "\n",
       "    sulfur-source supplement  \\\n",
       "860   MgSO4(0.24)   L-Serine   \n",
       "842   MgSO4(0.24)   L-Serine   \n",
       "852   MgSO4(0.24)   L-Serine   \n",
       "844   MgSO4(0.24)   L-Serine   \n",
       "867   MgSO4(0.24)   L-Serine   \n",
       "\n",
       "                                                  COGs  \\\n",
       "860  [{'name': 'Amino acid transport and metabolism...   \n",
       "842  [{'name': 'Amino acid transport and metabolism...   \n",
       "852  [{'name': 'Amino acid transport and metabolism...   \n",
       "844  [{'name': 'Amino acid transport and metabolism...   \n",
       "867                                                 []   \n",
       "\n",
       "                                             COG links  \\\n",
       "860  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "842  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "852  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "844  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "867                                                 {}   \n",
       "\n",
       "                                              pathways  \\\n",
       "860  [{'name': 'Glycine, serine and threonine metab...   \n",
       "842  [{'name': 'Glycine, serine and threonine metab...   \n",
       "852  [{'name': 'Glycine, serine and threonine metab...   \n",
       "844                                                 []   \n",
       "867                                                 []   \n",
       "\n",
       "                                         pathway links  \n",
       "860  {'Glycine, serine and threonine metabolism': [...  \n",
       "842  {'Glycine, serine and threonine metabolism': [...  \n",
       "852  {'Glycine, serine and threonine metabolism': [...  \n",
       "844                                                 {}  \n",
       "867                                                 {}  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# muts_df = muts_df[muts_df[\"mutation target annotation\"].str.contains(\"pnt\")]\n",
    "# muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"CRP\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"GlpR\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Mlc\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "# #     if \"Nucleotide transport and metabolism\" in r[\"COG links\"].keys() \\\n",
    "#     if \"cyaA\" in r[\"RegulonDB Gene\"]:\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Glycerolipid metabolism\" in r[\"pathway links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = muts_df.copy()\n",
    "# df[\"significantly associated condition set\"] = df[\"significantly associated condition set\"].apply(\n",
    "#     lambda x: str(x))\n",
    "# display(df[\"significantly associated condition set\"].unique())\n",
    "\n",
    "\n",
    "# # GYD\n",
    "# muts_df = muts_df[\n",
    "#     (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)'})\n",
    "#     | (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)', \"BOP27\"})\n",
    "# ]\n",
    "\n",
    "# # SER\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"] == {'ΔsdaA ΔsdaB ΔtdcG ΔglyA', 'L-Serine'}]\n",
    "\n",
    "# # # GLU\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'glucose(2)'}]\n",
    "\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]==set()]\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'log', '511145', 'M9'}]\n",
    "\n",
    "\n",
    "# display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27, 8)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>860</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>842</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>852</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>[{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...</td>\n",
       "      <td>{'ECK120000987': ['ECK120000987']}</td>\n",
       "      <td>[{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...</td>\n",
       "      <td>{'ECK120014725': ['ECK120000987']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>844</td>\n",
       "      <td>MOB</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...</td>\n",
       "      <td>{'ECK120000080': ['ECK120000080']}</td>\n",
       "      <td>[{'name': 'aroP', 'RegulonDB ID': 'ECK12001447...</td>\n",
       "      <td>{'ECK120014477': ['ECK120000080']}</td>\n",
       "      <td>[{'name': 'TyrR', 'significantly associated co...</td>\n",
       "      <td>{'TyrR': ['aroP'], 'Cra': ['aroP']}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>867</td>\n",
       "      <td>SNP</td>\n",
       "      <td>[{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...</td>\n",
       "      <td>[{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...</td>\n",
       "      <td>{'ECK120003005': ['ECK120003005']}</td>\n",
       "      <td>[{'name': 'ybhP-clsB-ybhN', 'RegulonDB ID': 'E...</td>\n",
       "      <td>{'ECK125160624': ['ECK120003005']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutation Type                                   genomic features  \\\n",
       "860           SNP  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "842           SNP  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "852           SNP  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "844           MOB  [{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...   \n",
       "867           SNP  [{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...   \n",
       "\n",
       "                                      genetic features  \\\n",
       "860  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "842  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "852  [{'name': 'thrA', 'RegulonDB ID': 'ECK12000098...   \n",
       "844  [{'name': 'aroP', 'RegulonDB ID': 'ECK12000008...   \n",
       "867  [{'name': 'ybhN', 'RegulonDB ID': 'ECK12000300...   \n",
       "\n",
       "                  genetic feature links  \\\n",
       "860  {'ECK120000987': ['ECK120000987']}   \n",
       "842  {'ECK120000987': ['ECK120000987']}   \n",
       "852  {'ECK120000987': ['ECK120000987']}   \n",
       "844  {'ECK120000080': ['ECK120000080']}   \n",
       "867  {'ECK120003005': ['ECK120003005']}   \n",
       "\n",
       "                                               operons  \\\n",
       "860  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "842  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "852  [{'name': 'thrLABC', 'RegulonDB ID': 'ECK12001...   \n",
       "844  [{'name': 'aroP', 'RegulonDB ID': 'ECK12001447...   \n",
       "867  [{'name': 'ybhP-clsB-ybhN', 'RegulonDB ID': 'E...   \n",
       "\n",
       "                           operon links  \\\n",
       "860  {'ECK120014725': ['ECK120000987']}   \n",
       "842  {'ECK120014725': ['ECK120000987']}   \n",
       "852  {'ECK120014725': ['ECK120000987']}   \n",
       "844  {'ECK120014477': ['ECK120000080']}   \n",
       "867  {'ECK125160624': ['ECK120003005']}   \n",
       "\n",
       "                                            regulators  \\\n",
       "860                                                 []   \n",
       "842                                                 []   \n",
       "852                                                 []   \n",
       "844  [{'name': 'TyrR', 'significantly associated co...   \n",
       "867                                                 []   \n",
       "\n",
       "                         regulator links  \n",
       "860                                   {}  \n",
       "842                                   {}  \n",
       "852                                   {}  \n",
       "844  {'TyrR': ['aroP'], 'Cra': ['aroP']}  \n",
       "867                                   {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "#     \"disrupt\",\n",
    "#     \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "#     \"COGs\",\n",
    "#     \"COG links\",\n",
    "#     \"pathways\",\n",
    "#     \"pathway links\",\n",
    "#     \"imodulons\",\n",
    "#     \"imodulon links\",\n",
    "#     \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "# for i, r in annot_df.iterrows():\n",
    "#     if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "#         del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "imodulon_counts = dict()\n",
    "\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another.\n",
    "        # Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "#     for COG, links in r[\"COG links\"].items():\n",
    "#         if COG not in COG_counts.keys():\n",
    "#             COG_counts[COG] = 0\n",
    "#         COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "#     for pathway, links in r[\"pathway links\"].items():\n",
    "#         if pathway not in pathway_counts.keys():\n",
    "#             pathway_counts[pathway] = 0\n",
    "#         pathway_counts[pathway] += len(links)\n",
    "        \n",
    "#     for imodulon, links in r[\"imodulon links\"].items():\n",
    "#         if imodulon not in imodulon_counts.keys():\n",
    "#             imodulon_counts[imodulon] = 0\n",
    "#         imodulon_counts[imodulon] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # For development. Delete when NB is stable.\n",
    "# for _, r in annot_df.iterrows():\n",
    "#     for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "#         effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "#         mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "#         mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "            \n",
    "#         genomic_feat_annot = \"\"\n",
    "#         if genomic_feat_d[\"significant\"]: genomic_feat_annot += \"*\"\n",
    "#         genomic_feat_annot += str(genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]) + ' '\n",
    "#         genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "    \n",
    "            \n",
    "#         display(genomic_feat_annot + ' #### ' + mut_type_feat_effect)\n",
    "# #         display(mut_type_feat_effect, \"################################\", r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from util.feature import get_feat_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>3 thrA SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>3 thrA SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>3 thrA SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>*3 thrLABC</td>\n",
       "      <td>*3 thrA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>1 aroP</td>\n",
       "      <td>1 aroP MOB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>1 aroP</td>\n",
       "      <td>1 aroP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>1 aroP</td>\n",
       "      <td>1 aroP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>1 TyrR</td>\n",
       "      <td>1 aroP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>3 Cra</td>\n",
       "      <td>1 aroP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>1 ybhN</td>\n",
       "      <td>1 ybhN SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>1 ybhN</td>\n",
       "      <td>1 ybhN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>1 ybhP-clsB-ybhN</td>\n",
       "      <td>1 ybhN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>1 lrpp</td>\n",
       "      <td>1 lrpp SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>1 trxB/lrp</td>\n",
       "      <td>1 lrpp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>*2 lrp</td>\n",
       "      <td>1 trxB/lrp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               source       target     type  value\n",
       "0             *3 thrA  3 thrA SNP   unknown    1.0\n",
       "1             *3 thrA      *3 thrA  unknown    1.0\n",
       "2         *3 thrLABC       *3 thrA  unknown    1.0\n",
       "3             *3 thrA  3 thrA SNP   unknown    1.0\n",
       "4             *3 thrA      *3 thrA  unknown    1.0\n",
       "5         *3 thrLABC       *3 thrA  unknown    1.0\n",
       "6             *3 thrA  3 thrA SNP   unknown    1.0\n",
       "7             *3 thrA      *3 thrA  unknown    1.0\n",
       "8         *3 thrLABC       *3 thrA  unknown    1.0\n",
       "9              1 aroP  1 aroP MOB   unknown    1.0\n",
       "10             1 aroP       1 aroP  unknown    1.0\n",
       "11            1 aroP        1 aroP  unknown    1.0\n",
       "12             1 TyrR      1 aroP   unknown    1.0\n",
       "13              3 Cra      1 aroP   unknown    1.0\n",
       "14             1 ybhN  1 ybhN SNP   unknown    1.0\n",
       "15             1 ybhN       1 ybhN  unknown    1.0\n",
       "16  1 ybhP-clsB-ybhN        1 ybhN  unknown    1.0\n",
       "17             1 lrpp  1 lrpp SNP   unknown    1.0\n",
       "18         1 trxB/lrp       1 lrpp  unknown    1.0\n",
       "19            *2 lrp    1 trxB/lrp  unknown    1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "# COGs = list()\n",
    "regulators = list()\n",
    "# pathways = list()\n",
    "# imodulons = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "#     if FUNC_ANNOT == \"pathway\":\n",
    "#         for pathway, ops in r[\"pathway links\"].items():\n",
    "# #             if pathway == \"Glycerolipid metabolism\":  # for GYD Glycerolipid metabolism.\n",
    "#             pathway_feat_d = get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "#             pathway_annot = \"\"\n",
    "#             mut_count = pathway_counts[pathway]\n",
    "#             if pathway_feat_d[\"significant\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "#             pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "#             pathways.append(pathway_annot)\n",
    "#             for operon_id in ops:\n",
    "#                 op_feat_d = get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "#                 operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "#                 operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "#                 operon_annot = \"\"\n",
    "#                 mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "#                 if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "#                 operon_annot += str(mut_count) + ' '\n",
    "#                 operon_annot += operon_name\n",
    "#                 operons.append(operon_annot)\n",
    "#                 flow_row = {\n",
    "#                     \"source\": pathway_annot,\n",
    "#                     \"target\": operon_annot,\n",
    "# #                     \"type\": r[\"mutation effect\"],\n",
    "#                     \"type\": \"unknown\",\n",
    "#                     \"value\": 1\n",
    "#                 }\n",
    "#                 flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "#             if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "            reg_feat_d = get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "#     if FUNC_ANNOT == \"COG\":\n",
    "#         for COG, ops in r[\"COG links\"].items():\n",
    "# #             if COG == \"Nucleotide transport and metabolism\":\n",
    "#             COG_feat_d = get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "#             COG_annot = \"\"\n",
    "#             mut_count = COG_counts[COG]\n",
    "#             if COG_feat_d[\"significant\"] and mut_count > 1: COG_annot += \"*\"\n",
    "#             COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "#             COGs.append(COG_annot)\n",
    "#             for operon_id in ops:\n",
    "#                 op_feat_d = get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "#                 operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "#                 operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "#                 operon_annot = \"\"\n",
    "#                 mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "#                 if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "#                 operon_annot += str(mut_count) + ' '\n",
    "#                 operon_annot += operon_name\n",
    "#                 operons.append(operon_annot)\n",
    "#                 flow_row = {\n",
    "#                     \"source\": COG_annot,\n",
    "#                     \"target\": operon_annot,\n",
    "# #                     \"type\": r[\"mutation effect\"],\n",
    "#                     \"type\": \"unknown\",\n",
    "#                     \"value\": 1\n",
    "#                 }\n",
    "#                 flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "                \n",
    "                \n",
    "#     if FUNC_ANNOT == \"imodulon\":\n",
    "#         for imodulon, ops in r[\"imodulon links\"].items():\n",
    "#             reg_feat_d = get_feat_d(name=imodulon, json=r[\"imodulons\"])\n",
    "#             reg_annot = \"\"\n",
    "#             mut_count = imodulon_counts[imodulon]\n",
    "#             if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "#             reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "#             imodulons.append(reg_annot)\n",
    "#             for operon_name in ops:\n",
    "#                 op_feat_d = get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "#                 operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "#                 operon_annot = \"\"\n",
    "#                 mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "#                 if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "#                 operon_annot += str(mut_count) + ' '\n",
    "#                 operon_annot += operon_name\n",
    "#                 operons.append(operon_annot)\n",
    "#                 flow_row = {\n",
    "#                     \"source\": reg_annot,\n",
    "#                     \"target\": operon_annot,\n",
    "#                     \"type\": \"unknown\",\n",
    "#                     \"value\": 1\n",
    "#                 }\n",
    "#                 flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "                \n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "# pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "# COGs = remove_duplicates(COGs)\n",
    "# imodulons = remove_duplicates(imodulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add italics to genomic features\n",
    "# feats_to_italicize = []\n",
    "# italicize_ignore_annots = [\" TFBS\"]\n",
    "# for i in range(0, len(genomic_feats)):\n",
    "#     f = genomic_feats[i]\n",
    "#     f = f[f.find(' ')+1:]\n",
    "#     for x in italicize_ignore_annots:\n",
    "#         if x in f:\n",
    "#             f = f.replace(x, \"\")\n",
    "#     feats_to_italicize.append(f)\n",
    "# feats_to_italicize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes = []\n",
    "# for i in range(0, len(genomic_feats)):\n",
    "#     f = genomic_feats[i]\n",
    "#     f = f[f.find(' ')+1:]\n",
    "#     if (f[-1] != 'p') and (\"TFBS\" not in f):\n",
    "#         genes.append(f)\n",
    "# genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# promoters = []\n",
    "# for i in range(0, len(genomic_feats)):\n",
    "#     f = genomic_feats[i]\n",
    "#     f = f[f.find(' ')+1:]\n",
    "#     if f[-1] == 'p':\n",
    "#         promoters.append(f)\n",
    "# promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tfbs = []\n",
    "# for i in range(0, len(genomic_feats)):\n",
    "#     f = genomic_feats[i]\n",
    "#     f = f[f.find(' ')+1:]\n",
    "#     if (\"TFBS\" in f):\n",
    "#         tfbs.append(f)\n",
    "# tfbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in range(0, len(mut_feat_effects)):\n",
    "#     for i_f in feats_to_italicize:\n",
    "#         if i_f in mut_feat_effects[i]:\n",
    "#             mut_feat_effects[i] = mut_feat_effects[i].replace(i_f, '$\\it{' + i_f + '}$')\n",
    "    \n",
    "# #     # very specific to SER\n",
    "# #     i_f = 'gcvB gcvA'\n",
    "# #     if i_f in mut_feat_effects[i]:\n",
    "# #         mut_feat_effects[i] = mut_feat_effects[i].replace(i_f, '$\\it{' + i_f + '}$')\n",
    "\n",
    "# mut_feat_effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'3 thrA SNP ': '$\\\\it{thrA}$',\n",
       " '1 aroP MOB ': '$\\\\it{aroP}$',\n",
       " '1 ybhN SNP ': '$\\\\it{ybhN}$',\n",
       " '1 lrpp SNP ': '$\\\\it{lrpp}$',\n",
       " '1 lrp SNP ': '$\\\\it{lrp}$',\n",
       " '1 pykF SNP ': '$\\\\it{pykF}$',\n",
       " '1 yodD SNP ': '$\\\\it{yodD}$',\n",
       " '1 yojI SNP ': '$\\\\it{yojI}$',\n",
       " '1 frc MOB ': '$\\\\it{frc}$',\n",
       " '2 gcvRp DEL ': '$\\\\it{gcvRp}$',\n",
       " '1 gshA SNP ': '$\\\\it{gshA}$',\n",
       " '1 eno SNP ': '$\\\\it{eno}$',\n",
       " '1 gcvB gcvA TFBS DEL ': '$\\\\it{gcvB gcvA}$ TFBS',\n",
       " '1 gcvAp DEL ': '$\\\\it{gcvAp}$',\n",
       " '1 gcvB MOB ': '$\\\\it{gcvB}$',\n",
       " '1 argP SNP ': '$\\\\it{argP}$',\n",
       " '1 tufA SNP ': '$\\\\it{tufA}$',\n",
       " '1 rpe SNP ': '$\\\\it{rpe}$',\n",
       " '1 malT SNP ': '$\\\\it{malT}$',\n",
       " '1 rhoLp MOB ': '$\\\\it{rhoLp}$',\n",
       " '1 rho SNP ': '$\\\\it{rho}$',\n",
       " '1 rpoB SNP ': '$\\\\it{rpoB}$',\n",
       " '1 lamB SNP ': '$\\\\it{lamB}$',\n",
       " '1 ytfBp SNP ': '$\\\\it{ytfBp}$',\n",
       " '1 cycA SNP ': '$\\\\it{cycA}$'}"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# to_replace_d = {}\n",
    "# for i in range(0, len(mut_feat_effects)):\n",
    "#     f = mut_feat_effects[i]\n",
    "#     f = f[:-1]  # removing final space\n",
    "#     f = f[:f.rfind(' ')]\n",
    "#     f = f[f.find(' ') + 1:]\n",
    "#     if \"TFBS\" in f:\n",
    "#         to_replace = f.replace(\" TFBS\", \"\")\n",
    "#         f = f.replace(to_replace, '$\\it{' + to_replace + '}$')\n",
    "# #     if (f[-1] == 'p') and (f != \"lrp\"):\n",
    "# #         f = f.replace(f, '$\\it{' + f + '}$')\n",
    "#     else:\n",
    "#         f = f.replace(f, '$\\it{' + f + '}$')\n",
    "        \n",
    "# #         f = '$\\it{' + f.replace(\" TFBS\", \"\") + '}$'\n",
    "        \n",
    "# #     if \" TFBS\" in f:\n",
    "# #         f = f.replace(\" TFBS\", \"\")\n",
    "#     to_replace_d[mut_feat_effects[i]] = f\n",
    "# #     for i_f in feats_to_italicize:\n",
    "# #         if i_f in mut_feat_effects[i]:\n",
    "# #             tl[i] = mut_feat_effects[i].replace(i_f, '$\\it{' + i_f + '}$')\n",
    "# to_replace_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "439ed4cd9d024800a9315713fdb5de49",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'regulon', 'type': 'process', 'title': 'regulon', 'nodes': ['regulon^1 TyrR', 'reg…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "# if FUNC_ANNOT == \"pathway\": l = pathways\n",
    "if FUNC_ANNOT == \"regulon\": l = regulators\n",
    "# if FUNC_ANNOT == \"COG\": l = COGs\n",
    "# if FUNC_ANNOT == \"imodulon\": l = imodulons\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "#                        flow_partition=mut_effect_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# p_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "# size = dict(width=3000, height=4000)\n",
    "# size = dict(width=2000, height=2500)\n",
    "# size = dict(width=1600, height=1000)\n",
    "size = dict(width=1250, height=400)\n",
    "# size = dict(width=1600, height=500)\n",
    "# size = dict(width=1100, height=300)\n",
    "# size = dict(width=1350, height=300)\n",
    "# size = dict(width=1000, height=200)\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
