{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "flow_df = pd.DataFrame()\n",
    "mutations = []\n",
    "genomic_feats = []\n",
    "genetic_feats = []\n",
    "operons = []\n",
    "regulons = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"*3 regulon C\",\n",
    "        \"target\": \"*3 operon C\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 1.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"*3 operon C\",\n",
    "        \"target\": \"1 gene A\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 1.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"1 gene A\",\n",
    "        \"target\": \"1 feature A\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 1.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"1 feature A\",\n",
    "        \"target\": \"1 feature A DEL\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 1.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "\n",
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"*3 regulon C\",\n",
    "        \"target\": \"*3 operon C\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 2.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"*3 operon C\",\n",
    "        \"target\": \"2 gene B\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 2.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"2 gene B\",\n",
    "        \"target\": \"*2 feature B\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 2.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "flow_df = flow_df.append(\n",
    "    {\n",
    "        \"source\": \"*2 feature B\",\n",
    "        \"target\": \"*2 feature B SNP\",\n",
    "        \"type\": \"unknown\",\n",
    "        \"value\": 2.0\n",
    "    },\n",
    "    ignore_index=True)\n",
    "\n",
    "mutations.append(\"1 feature A DEL\")\n",
    "genomic_feats.append(\"1 feature A\",)\n",
    "genetic_feats.append(\"1 gene A\",)\n",
    "regulons.append(\"1 regulon A\")\n",
    "\n",
    "mutations.append(\"*2 feature B SNP\")\n",
    "genomic_feats.append(\"*2 feature B\",)\n",
    "genetic_feats.append(\"2 gene B\",)\n",
    "operons.append(\"*3 operon C\")\n",
    "regulons.append(\"*3 regulon C\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f82ff77dc3ce408ca70ee9248c1277b4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'regulon', 'type': 'process', 'title': 'regulon', 'nodes': ['regulon^*3 regulon C'â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mutation_partition = Partition.Simple(\"process\", mutations)\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "regulon_partition = Partition.Simple(\"process\", regulons)\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mutations, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "    \"regulon\": ProcessGroup(regulons, title=\"regulon\"),\n",
    "}\n",
    "\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "nodes[\"regulon\"].partition = regulon_partition\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"regulon\",\n",
    "        \"operon\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "ordering = [\n",
    "    [\"regulon\"],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "                      )\n",
    "\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "\n",
    "size = dict(width=1200, height=300)\n",
    "\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size\n",
    ")\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
