{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "FUNC_ANNOT = \"regulon\"\n",
    "# FUNC_ANNOT = \"imodulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "102"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/Central_carbon_knockout_pgi_df.pkl\")\n",
    "display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['index', 'Details', 'mutation target annotation', 'Mutation Type',\n",
       "       'Position', 'Reference Seq', 'Sequence Change', 'ale', 'exp',\n",
       "       'flask', 'isolate', 'presence', 'tech_rep', 'coding', 'range',\n",
       "       'gene RegulonDB ID', 'genetic features', 'oriC', 'pseudogene',\n",
       "       'TFBS', 'promoter', 'TSS', 'RBS', 'attenuator terminator',\n",
       "       'terminator', 'genetic', 'genomic features',\n",
       "       'genetic feature links', 'operons', 'operon links', 'regulators',\n",
       "       'regulator links', 'taxonomy-id', 'strain-description',\n",
       "       'base-media', 'temperature', 'carbon-source', 'nitrogen-source',\n",
       "       'phosphorous-source', 'sulfur-source', 'supplement', 'COGs',\n",
       "       'COG links', 'pathways', 'pathway links'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df.columns.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>TSS</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>base-media</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>325</td>\n",
       "      <td>159</td>\n",
       "      <td>intergenic (‑172/‑6)</td>\n",
       "      <td>marC/marR</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1619080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ35 bp</td>\n",
       "      <td>3</td>\n",
       "      <td>Central_carbon_knockout_pgi</td>\n",
       "      <td>108</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(1619080, 1619114)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001584/ECK120001399',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120017169, ECK120030315}</td>\n",
       "      <td>{ECK120010245}</td>\n",
       "      <td>{ECK120010245}</td>\n",
       "      <td>{ECK120014209}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'marRAB TFBS', 'RegulonDB ID': 'ECK1...</td>\n",
       "      <td>{'ECK120001584/ECK120001399': ['ECK120014209',...</td>\n",
       "      <td>[{'name': 'marRAB', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014736': ['ECK120001584/ECK120001399',...</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'Fis': ['marRAB', 'marRAB', 'marRAB'], 'CRP':...</td>\n",
       "      <td>511145</td>\n",
       "      <td>Δpgi</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(2)</td>\n",
       "      <td>MgSO4(3)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120014736', 'ECK1200147...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index               Details mutation target annotation Mutation Type  \\\n",
       "325   159  intergenic (‑172/‑6)                  marC/marR           DEL   \n",
       "\n",
       "     Position Reference Seq Sequence Change  ale                          exp  \\\n",
       "325   1619080           NaN          Δ35 bp    3  Central_carbon_knockout_pgi   \n",
       "\n",
       "     flask  isolate  presence  tech_rep  coding               range  \\\n",
       "325    108        1       1.0         1   False  (1619080, 1619114)   \n",
       "\n",
       "    gene RegulonDB ID                                   genetic features  \\\n",
       "325                {}  [{'RegulonDB ID': 'ECK120001584/ECK120001399',...   \n",
       "\n",
       "      oriC  pseudogene                          TFBS        promoter  \\\n",
       "325  False       False  {ECK120017169, ECK120030315}  {ECK120010245}   \n",
       "\n",
       "                TSS             RBS attenuator terminator terminator  genetic  \\\n",
       "325  {ECK120010245}  {ECK120014209}                    {}         {}    False   \n",
       "\n",
       "                                      genomic features  \\\n",
       "325  [{'name': 'marRAB TFBS', 'RegulonDB ID': 'ECK1...   \n",
       "\n",
       "                                 genetic feature links  \\\n",
       "325  {'ECK120001584/ECK120001399': ['ECK120014209',...   \n",
       "\n",
       "                                               operons  \\\n",
       "325  [{'name': 'marRAB', 'RegulonDB ID': 'ECK120014...   \n",
       "\n",
       "                                          operon links  \\\n",
       "325  {'ECK120014736': ['ECK120001584/ECK120001399',...   \n",
       "\n",
       "                                            regulators  \\\n",
       "325  [{'name': 'Fis', 'significantly associated con...   \n",
       "\n",
       "                                       regulator links taxonomy-id  \\\n",
       "325  {'Fis': ['marRAB', 'marRAB', 'marRAB'], 'CRP':...      511145   \n",
       "\n",
       "    strain-description base-media temperature carbon-source nitrogen-source  \\\n",
       "325               Δpgi         M9  37 celsius    glucose(4)        NH4Cl(1)   \n",
       "\n",
       "    phosphorous-source sulfur-source      supplement  \\\n",
       "325          KH2PO4(2)      MgSO4(3)  trace elements   \n",
       "\n",
       "                                                  COGs  \\\n",
       "325  [{'name': 'Transcription', 'mutation set count...   \n",
       "\n",
       "                                             COG links pathways pathway links  \n",
       "325  {'Transcription': ['ECK120014736', 'ECK1200147...       []            {}  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_df = muts_df[\n",
    "    muts_df[\"mutation target annotation\"].str.contains(\"marC/marR\")\n",
    "]\n",
    "muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>TSS</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>base-media</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>325</td>\n",
       "      <td>159</td>\n",
       "      <td>intergenic (‑172/‑6)</td>\n",
       "      <td>marC/marR</td>\n",
       "      <td>DEL</td>\n",
       "      <td>1619080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Δ35 bp</td>\n",
       "      <td>3</td>\n",
       "      <td>Central_carbon_knockout_pgi</td>\n",
       "      <td>108</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(1619080, 1619114)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001584/ECK120001399',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{ECK120017169, ECK120030315}</td>\n",
       "      <td>{ECK120010245}</td>\n",
       "      <td>{ECK120010245}</td>\n",
       "      <td>{ECK120014209}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'marRAB TFBS', 'RegulonDB ID': 'ECK1...</td>\n",
       "      <td>{'ECK120001584/ECK120001399': ['ECK120014209',...</td>\n",
       "      <td>[{'name': 'marRAB', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014736': ['ECK120001584/ECK120001399',...</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'Fis': ['marRAB', 'marRAB', 'marRAB'], 'CRP':...</td>\n",
       "      <td>511145</td>\n",
       "      <td>Δpgi</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(2)</td>\n",
       "      <td>MgSO4(3)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK120014736', 'ECK1200147...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index               Details mutation target annotation Mutation Type  \\\n",
       "325   159  intergenic (‑172/‑6)                  marC/marR           DEL   \n",
       "\n",
       "     Position Reference Seq Sequence Change  ale                          exp  \\\n",
       "325   1619080           NaN          Δ35 bp    3  Central_carbon_knockout_pgi   \n",
       "\n",
       "     flask  isolate  presence  tech_rep  coding               range  \\\n",
       "325    108        1       1.0         1   False  (1619080, 1619114)   \n",
       "\n",
       "    gene RegulonDB ID                                   genetic features  \\\n",
       "325                {}  [{'RegulonDB ID': 'ECK120001584/ECK120001399',...   \n",
       "\n",
       "      oriC  pseudogene                          TFBS        promoter  \\\n",
       "325  False       False  {ECK120017169, ECK120030315}  {ECK120010245}   \n",
       "\n",
       "                TSS             RBS attenuator terminator terminator  genetic  \\\n",
       "325  {ECK120010245}  {ECK120014209}                    {}         {}    False   \n",
       "\n",
       "                                      genomic features  \\\n",
       "325  [{'name': 'marRAB TFBS', 'RegulonDB ID': 'ECK1...   \n",
       "\n",
       "                                 genetic feature links  \\\n",
       "325  {'ECK120001584/ECK120001399': ['ECK120014209',...   \n",
       "\n",
       "                                               operons  \\\n",
       "325  [{'name': 'marRAB', 'RegulonDB ID': 'ECK120014...   \n",
       "\n",
       "                                          operon links  \\\n",
       "325  {'ECK120014736': ['ECK120001584/ECK120001399',...   \n",
       "\n",
       "                                            regulators  \\\n",
       "325  [{'name': 'Fis', 'significantly associated con...   \n",
       "\n",
       "                                       regulator links taxonomy-id  \\\n",
       "325  {'Fis': ['marRAB', 'marRAB', 'marRAB'], 'CRP':...      511145   \n",
       "\n",
       "    strain-description base-media temperature carbon-source nitrogen-source  \\\n",
       "325               Δpgi         M9  37 celsius    glucose(4)        NH4Cl(1)   \n",
       "\n",
       "    phosphorous-source sulfur-source      supplement  \\\n",
       "325          KH2PO4(2)      MgSO4(3)  trace elements   \n",
       "\n",
       "                                                  COGs  \\\n",
       "325  [{'name': 'Transcription', 'mutation set count...   \n",
       "\n",
       "                                             COG links pathways pathway links  \n",
       "325  {'Transcription': ['ECK120014736', 'ECK1200147...       []            {}  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 8)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>325</td>\n",
       "      <td>DEL</td>\n",
       "      <td>[{'name': 'marRAB TFBS', 'RegulonDB ID': 'ECK1...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001584/ECK120001399',...</td>\n",
       "      <td>{'ECK120001584/ECK120001399': ['ECK120014209',...</td>\n",
       "      <td>[{'name': 'marRAB', 'RegulonDB ID': 'ECK120014...</td>\n",
       "      <td>{'ECK120014736': ['ECK120001584/ECK120001399',...</td>\n",
       "      <td>[{'name': 'Fis', 'significantly associated con...</td>\n",
       "      <td>{'Fis': ['marRAB', 'marRAB', 'marRAB'], 'CRP':...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutation Type                                   genomic features  \\\n",
       "325           DEL  [{'name': 'marRAB TFBS', 'RegulonDB ID': 'ECK1...   \n",
       "\n",
       "                                      genetic features  \\\n",
       "325  [{'RegulonDB ID': 'ECK120001584/ECK120001399',...   \n",
       "\n",
       "                                 genetic feature links  \\\n",
       "325  {'ECK120001584/ECK120001399': ['ECK120014209',...   \n",
       "\n",
       "                                               operons  \\\n",
       "325  [{'name': 'marRAB', 'RegulonDB ID': 'ECK120014...   \n",
       "\n",
       "                                          operon links  \\\n",
       "325  {'ECK120014736': ['ECK120001584/ECK120001399',...   \n",
       "\n",
       "                                            regulators  \\\n",
       "325  [{'name': 'Fis', 'significantly associated con...   \n",
       "\n",
       "                                       regulator links  \n",
       "325  {'Fis': ['marRAB', 'marRAB', 'marRAB'], 'CRP':...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "#     \"disrupt\",\n",
    "#     \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "#     \"COGs\",\n",
    "#     \"COG links\",\n",
    "#     \"pathways\",\n",
    "#     \"pathway links\",\n",
    "#     \"imodulons\",\n",
    "#     \"imodulon links\",\n",
    "#     \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "# for i, r in annot_df.iterrows():\n",
    "#     if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "#         del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'marRAB TFBS DEL ': 2,\n",
       " 'marRp DEL ': 1,\n",
       " 'marRp TSS DEL ': 1,\n",
       " 'marR RBS DEL ': 1}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{'ECK120030315': 1, 'ECK120017169': 1, 'ECK120010245': 2, 'ECK120014209': 1}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "imodulon_counts = dict()\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another.\n",
    "        # Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "\n",
    "        mut_type_feat_effect = str(genomic_feat[\"name\"]) + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "#     for COG, links in r[\"COG links\"].items():\n",
    "#         if COG not in COG_counts.keys():\n",
    "#             COG_counts[COG] = 0\n",
    "#         COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "#     for pathway, links in r[\"pathway links\"].items():\n",
    "#         if pathway not in pathway_counts.keys():\n",
    "#             pathway_counts[pathway] = 0\n",
    "#         pathway_counts[pathway] += len(links)\n",
    "        \n",
    "#     for imodulon, links in r[\"imodulon links\"].items():\n",
    "#         if imodulon not in imodulon_counts.keys():\n",
    "#             imodulon_counts[imodulon] = 0\n",
    "#         imodulon_counts[imodulon] += len(links)\n",
    "\n",
    "\n",
    "display(mut_type_feat_effect_counts, genomic_feat_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from util.feature import get_feat_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'source': '1 marRAB TFBS', 'target': '2 marRAB TFBS DEL ', 'type': 'unknown', 'value': 1}\n",
      "{'source': '1 marRAB TFBS', 'target': '2 marRAB TFBS DEL ', 'type': 'unknown', 'value': 1}\n",
      "{'source': '*2 marRp', 'target': '1 marRp DEL ', 'type': 'unknown', 'value': 1}\n",
      "{'source': '*2 marRp TSS', 'target': '1 marRp TSS DEL ', 'type': 'unknown', 'value': 1}\n",
      "{'source': '1 marR RBS', 'target': '1 marR RBS DEL ', 'type': 'unknown', 'value': 1}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>1 marRAB TFBS</td>\n",
       "      <td>2 marRAB TFBS DEL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1 marRAB TFBS</td>\n",
       "      <td>2 marRAB TFBS DEL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>*2 marRp</td>\n",
       "      <td>1 marRp DEL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>*2 marRp TSS</td>\n",
       "      <td>1 marRp TSS DEL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>1 marR RBS</td>\n",
       "      <td>1 marR RBS DEL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>1 marR RBS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>1 marRAB TFBS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>1 marRAB TFBS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>*2 marRp</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>*4 marC/marR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>3 Fis</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>3 Fis</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>3 Fis</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>3 CRP</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>3 CRP</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>3 CRP</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>3 MarA</td>\n",
       "      <td>*4 marRAB</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           source              target     type  value\n",
       "0   1 marRAB TFBS  2 marRAB TFBS DEL   unknown    1.0\n",
       "1   1 marRAB TFBS  2 marRAB TFBS DEL   unknown    1.0\n",
       "2        *2 marRp        1 marRp DEL   unknown    1.0\n",
       "3    *2 marRp TSS    1 marRp TSS DEL   unknown    1.0\n",
       "4      1 marR RBS     1 marR RBS DEL   unknown    1.0\n",
       "5    *4 marC/marR          1 marR RBS  unknown    1.0\n",
       "6    *4 marC/marR       1 marRAB TFBS  unknown    1.0\n",
       "7    *4 marC/marR       1 marRAB TFBS  unknown    1.0\n",
       "8    *4 marC/marR            *2 marRp  unknown    1.0\n",
       "9      *4 marRAB         *4 marC/marR  unknown    1.0\n",
       "10     *4 marRAB         *4 marC/marR  unknown    1.0\n",
       "11     *4 marRAB         *4 marC/marR  unknown    1.0\n",
       "12     *4 marRAB         *4 marC/marR  unknown    1.0\n",
       "13          3 Fis          *4 marRAB   unknown    1.0\n",
       "14          3 Fis          *4 marRAB   unknown    1.0\n",
       "15          3 Fis          *4 marRAB   unknown    1.0\n",
       "16          3 CRP          *4 marRAB   unknown    1.0\n",
       "17          3 CRP          *4 marRAB   unknown    1.0\n",
       "18          3 CRP          *4 marRAB   unknown    1.0\n",
       "19         3 MarA          *4 marRAB   unknown    1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "# COGs = list()\n",
    "regulators = list()\n",
    "# pathways = list()\n",
    "# imodulons = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = str(genomic_feat_d[\"name\"]) + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        print(flow_row)\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "#     if FUNC_ANNOT == \"pathway\":\n",
    "#         for pathway, ops in r[\"pathway links\"].items():\n",
    "# #             if pathway == \"Glycerolipid metabolism\":  # for GYD Glycerolipid metabolism.\n",
    "#             pathway_feat_d = get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "#             pathway_annot = \"\"\n",
    "#             mut_count = pathway_counts[pathway]\n",
    "#             if pathway_feat_d[\"significant\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "#             pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "#             pathways.append(pathway_annot)\n",
    "#             for operon_id in ops:\n",
    "#                 op_feat_d = get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "#                 operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "#                 operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "#                 operon_annot = \"\"\n",
    "#                 mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "#                 if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "#                 operon_annot += str(mut_count) + ' '\n",
    "#                 operon_annot += operon_name\n",
    "#                 operons.append(operon_annot)\n",
    "#                 flow_row = {\n",
    "#                     \"source\": pathway_annot,\n",
    "#                     \"target\": operon_annot,\n",
    "# #                     \"type\": r[\"mutation effect\"],\n",
    "#                     \"type\": \"unknown\",\n",
    "#                     \"value\": 1\n",
    "#                 }\n",
    "#                 flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "#             if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "            reg_feat_d = get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "#     if FUNC_ANNOT == \"COG\":\n",
    "#         for COG, ops in r[\"COG links\"].items():\n",
    "# #             if COG == \"Nucleotide transport and metabolism\":\n",
    "#             COG_feat_d = get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "#             COG_annot = \"\"\n",
    "#             mut_count = COG_counts[COG]\n",
    "#             if COG_feat_d[\"significant\"] and mut_count > 1: COG_annot += \"*\"\n",
    "#             COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "#             COGs.append(COG_annot)\n",
    "#             for operon_id in ops:\n",
    "#                 op_feat_d = get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "#                 operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "#                 operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "#                 operon_annot = \"\"\n",
    "#                 mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "#                 if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "#                 operon_annot += str(mut_count) + ' '\n",
    "#                 operon_annot += operon_name\n",
    "#                 operons.append(operon_annot)\n",
    "#                 flow_row = {\n",
    "#                     \"source\": COG_annot,\n",
    "#                     \"target\": operon_annot,\n",
    "# #                     \"type\": r[\"mutation effect\"],\n",
    "#                     \"type\": \"unknown\",\n",
    "#                     \"value\": 1\n",
    "#                 }\n",
    "#                 flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "                \n",
    "                \n",
    "#     if FUNC_ANNOT == \"imodulon\":\n",
    "#         for imodulon, ops in r[\"imodulon links\"].items():\n",
    "#             reg_feat_d = get_feat_d(name=imodulon, json=r[\"imodulons\"])\n",
    "#             reg_annot = \"\"\n",
    "#             mut_count = imodulon_counts[imodulon]\n",
    "#             if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "#             reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "#             imodulons.append(reg_annot)\n",
    "#             for operon_name in ops:\n",
    "#                 op_feat_d = get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "#                 operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "#                 operon_annot = \"\"\n",
    "#                 mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "#                 if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "#                 operon_annot += str(mut_count) + ' '\n",
    "#                 operon_annot += operon_name\n",
    "#                 operons.append(operon_annot)\n",
    "#                 flow_row = {\n",
    "#                     \"source\": reg_annot,\n",
    "#                     \"target\": operon_annot,\n",
    "#                     \"type\": \"unknown\",\n",
    "#                     \"value\": 1\n",
    "#                 }\n",
    "#                 flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "                \n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "# pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "# COGs = remove_duplicates(COGs)\n",
    "# imodulons = remove_duplicates(imodulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "00e832caaf2c45e79f8d9aefd337f5b7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'regulon', 'type': 'process', 'title': 'regulon', 'nodes': ['regulon^3 Fis', 'regu…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "# if FUNC_ANNOT == \"pathway\": l = pathways\n",
    "if FUNC_ANNOT == \"regulon\": l = regulators\n",
    "# if FUNC_ANNOT == \"COG\": l = COGs\n",
    "# if FUNC_ANNOT == \"imodulon\": l = imodulons\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "#                        flow_partition=mut_effect_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# p_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "\n",
    "size = dict(width=1200, height=400)\n",
    "\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
