{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "FUNC_ANNOT = \"regulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "209"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/BOP27_df.pkl\")\n",
    "display(len(muts_df),\n",
    "#         muts_df.head()\n",
    "       )\n",
    "\n",
    "\n",
    "# # Below are the COGs that got multiple hits with GYD\n",
    "# '''\n",
    "# Lipid transport and metabolism\n",
    "# Cell cycle control, cell division, chromosome partitioning\n",
    "# Transcription\n",
    "# Amino acid transport and metabolism\n",
    "# Secondary metabolites biosynthesis, transport and catabolism\n",
    "# Cell wall/membrane/envelope biogenesis\n",
    "# Posttranslational modification, protein turnover, chaperones\n",
    "# Replication, recombination and repair\n",
    "# Signal transduction mechanisms\n",
    "# Cell motility\n",
    "# Defense mechanisms\n",
    "# Inorganic ion transport and metabolism\n",
    "# Intracellular trafficking, secretion, and vesicular transport\n",
    "# Coenzyme transport and metabolism\n",
    "# Energy production and conversion\n",
    "# Translation, ribosomal structure and biogenesis\n",
    "# Carbohydrate transport and metabolism\n",
    "# Nucleotide transport and metabolism\n",
    "# '''\n",
    "\n",
    "\n",
    "# out_df = pd.DataFrame()\n",
    "# df = muts_df.copy()\n",
    "# for _, r in df.iterrows():\n",
    "#     for c in r[\"operons\"]:\n",
    "#         try:\n",
    "#             # GYD\n",
    "#             if \"rpoB\" in c[\"name\"] or \"rpoC\" in c[\"name\"] or \"rph\" in c[\"name\"]:\n",
    "#                 out_df = out_df.append(r, ignore_index=True)\n",
    "#                 break\n",
    "# #             if \"ptsHI-crr\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#             # PGI\n",
    "# #             if \"pnt\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#         except:\n",
    "#             print(\"feature annotation bug with\", c)\n",
    "\n",
    "# muts_df = out_df\n",
    "# display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'42C',\n",
       " 'C13',\n",
       " 'GLU',\n",
       " 'GYD',\n",
       " 'SSW_GLU_AC',\n",
       " 'SSW_GLU_GLY',\n",
       " 'SSW_GLU_XYL',\n",
       " 'SSW_GLY',\n",
       " 'SSW_XYL'}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(muts_df.exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cyaA', 'fumC', 'glpK', 'rcsC', 'rhsB', 'rhsE', 'rpoC', 'yegE'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(muts_df[muts_df.exp!=\"GYD\"][\"RegulonDB Gene\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Details</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>RBS</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>TSS</th>\n",
       "      <th>ale</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>base-media</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>coding</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene count</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>index</th>\n",
       "      <th>isolate</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>oriC</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>presence</th>\n",
       "      <th>promoter</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>range</th>\n",
       "      <th>repaired</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>temperature</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2861</th>\n",
       "      <td>pseudogene (1274/2037 nt)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1529175.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>7.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4) or acetate(4)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SSW_GLU_AC</td>\n",
       "      <td>90.0</td>\n",
       "      <td>{ECK120000008}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(1529175, 1529175)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2920</th>\n",
       "      <td>pseudogene (1274/2037 nt)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1529175.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>9.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4) or acetate(4)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SSW_GLU_AC</td>\n",
       "      <td>92.0</td>\n",
       "      <td>{ECK120000008}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(1529175, 1529175)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>pseudogene (1274/2037 nt)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1529175.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>U-13C glucose(2)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C13</td>\n",
       "      <td>134.0</td>\n",
       "      <td>{ECK120000008}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(1529175, 1529175)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>pseudogene (1274/2037 nt)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1529175.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>6.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>U-13C glucose(2)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C13</td>\n",
       "      <td>133.0</td>\n",
       "      <td>{ECK120000008}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(1529175, 1529175)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2862</th>\n",
       "      <td>pseudogene (1279/2037 nt)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1529180.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>T→C</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>7.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4) or acetate(4)</td>\n",
       "      <td>{}</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SSW_GLU_AC</td>\n",
       "      <td>90.0</td>\n",
       "      <td>{ECK120000008}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>rhsE</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>{}</td>\n",
       "      <td>1.0</td>\n",
       "      <td>(1529180, 1529180)</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>511145</td>\n",
       "      <td>2.0</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Details  Mut ID Mutation Type   Position RBS  \\\n",
       "2861  pseudogene (1274/2037 nt)     NaN           SNP  1529175.0  {}   \n",
       "2920  pseudogene (1274/2037 nt)     NaN           SNP  1529175.0  {}   \n",
       "205   pseudogene (1274/2037 nt)     NaN           SNP  1529175.0  {}   \n",
       "233   pseudogene (1274/2037 nt)     NaN           SNP  1529175.0  {}   \n",
       "2862  pseudogene (1279/2037 nt)     NaN           SNP  1529180.0  {}   \n",
       "\n",
       "     Reference Seq RegulonDB Gene Sequence Change TFBS TSS  ale  \\\n",
       "2861           NaN           rhsE             T→C   {}  {}  7.0   \n",
       "2920           NaN           rhsE             T→C   {}  {}  9.0   \n",
       "205            NaN           rhsE             T→C   {}  {}  1.0   \n",
       "233            NaN           rhsE             T→C   {}  {}  6.0   \n",
       "2862           NaN           rhsE             T→C   {}  {}  7.0   \n",
       "\n",
       "     attenuator terminator base-media             carbon-source  \\\n",
       "2861                    {}         M9  glucose(4) or acetate(4)   \n",
       "2920                    {}         M9  glucose(4) or acetate(4)   \n",
       "205                     {}         M9          U-13C glucose(2)   \n",
       "233                     {}         M9          U-13C glucose(2)   \n",
       "2862                    {}         M9  glucose(4) or acetate(4)   \n",
       "\n",
       "     cis-regulatory RNA  coding  disrupt         exp  flask            gene  \\\n",
       "2861                 {}     0.0      0.0  SSW_GLU_AC   90.0  {ECK120000008}   \n",
       "2920                 {}     0.0      0.0  SSW_GLU_AC   92.0  {ECK120000008}   \n",
       "205                  {}     0.0      0.0         C13  134.0  {ECK120000008}   \n",
       "233                  {}     0.0      0.0         C13  133.0  {ECK120000008}   \n",
       "2862                 {}     0.0      0.0  SSW_GLU_AC   90.0  {ECK120000008}   \n",
       "\n",
       "      gene count  genetic                                   genetic features  \\\n",
       "2861         1.0      1.0  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "2920         1.0      1.0  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "205          1.0      1.0  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "233          1.0      1.0  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "2862         1.0      1.0  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "\n",
       "                                       genomic features index  isolate  \\\n",
       "2861  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...     8      0.0   \n",
       "2920  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...     8      0.0   \n",
       "205   [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...     7      1.0   \n",
       "233   [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...     7      1.0   \n",
       "2862  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...     9      0.0   \n",
       "\n",
       "      mutation size mutation target annotation nitrogen-source  oriC  \\\n",
       "2861            1.0                       rhsE        NH4Cl(1)   0.0   \n",
       "2920            1.0                       rhsE        NH4Cl(1)   0.0   \n",
       "205             1.0                       rhsE        NH4Cl(1)   0.0   \n",
       "233             1.0                       rhsE        NH4Cl(1)   0.0   \n",
       "2862            1.0                       rhsE        NH4Cl(1)   0.0   \n",
       "\n",
       "     phosphorous-source  presence promoter  pseudogene               range  \\\n",
       "2861          KH2PO4(3)       1.0       {}         1.0  (1529175, 1529175)   \n",
       "2920          KH2PO4(3)       1.0       {}         1.0  (1529175, 1529175)   \n",
       "205           KH2PO4(3)       1.0       {}         1.0  (1529175, 1529175)   \n",
       "233           KH2PO4(3)       1.0       {}         1.0  (1529175, 1529175)   \n",
       "2862          KH2PO4(3)       1.0       {}         1.0  (1529180, 1529180)   \n",
       "\n",
       "      repaired strain-description sulfur-source      supplement taxonomy-id  \\\n",
       "2861       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "2920       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "205        0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "233        0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "2862       0.0              BOP27   MgSO4(0.24)  trace elements      511145   \n",
       "\n",
       "      tech_rep temperature terminator               genetic feature links  \\\n",
       "2861       2.0  37 celsius         {}  {'ECK120000008': ['ECK120000008']}   \n",
       "2920       1.0  37 celsius         {}  {'ECK120000008': ['ECK120000008']}   \n",
       "205        1.0  37 celsius         {}  {'ECK120000008': ['ECK120000008']}   \n",
       "233        1.0  37 celsius         {}  {'ECK120000008': ['ECK120000008']}   \n",
       "2862       2.0  37 celsius         {}  {'ECK120000008': ['ECK120000008']}   \n",
       "\n",
       "                                                operons  \\\n",
       "2861  [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "2920  [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "205   [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "233   [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "2862  [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "\n",
       "                            operon links COGs COG links regulators  \\\n",
       "2861  {'ECK120028885': ['ECK120000008']}   []        {}         []   \n",
       "2920  {'ECK120028885': ['ECK120000008']}   []        {}         []   \n",
       "205   {'ECK120028885': ['ECK120000008']}   []        {}         []   \n",
       "233   {'ECK120028885': ['ECK120000008']}   []        {}         []   \n",
       "2862  {'ECK120028885': ['ECK120000008']}   []        {}         []   \n",
       "\n",
       "     regulator links pathways pathway links  \n",
       "2861              {}       []            {}  \n",
       "2920              {}       []            {}  \n",
       "205               {}       []            {}  \n",
       "233               {}       []            {}  \n",
       "2862              {}       []            {}  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# muts_df = muts_df[muts_df[\"mutation target annotation\"].str.contains(\"pnt\")]\n",
    "# muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"CRP\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"GlpR\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Mlc\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "# #     if \"Nucleotide transport and metabolism\" in r[\"COG links\"].keys() \\\n",
    "#     if \"cyaA\" in r[\"RegulonDB Gene\"]:\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Glycerolipid metabolism\" in r[\"pathway links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = muts_df.copy()\n",
    "# df[\"significantly associated condition set\"] = df[\"significantly associated condition set\"].apply(\n",
    "#     lambda x: str(x))\n",
    "# display(df[\"significantly associated condition set\"].unique())\n",
    "\n",
    "\n",
    "# # GYD\n",
    "# muts_df = muts_df[\n",
    "#     (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)'})\n",
    "#     | (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)', \"BOP27\"})\n",
    "# ]\n",
    "\n",
    "# # SER\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"] == {'ΔsdaA ΔsdaB ΔtdcG ΔglyA', 'L-Serine'}]\n",
    "\n",
    "# # # GLU\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'glucose(2)'}]\n",
    "\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]==set()]\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'log', '511145', 'M9'}]\n",
    "\n",
    "\n",
    "# display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(209, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2861</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>rhsE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2920</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>rhsE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>rhsE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>rhsE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2862</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>[{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...</td>\n",
       "      <td>{'ECK120000008': ['ECK120000008']}</td>\n",
       "      <td>[{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120028885': ['ECK120000008']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>rhsE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type  disrupt  repaired  \\\n",
       "2861           SNP      0.0       0.0   \n",
       "2920           SNP      0.0       0.0   \n",
       "205            SNP      0.0       0.0   \n",
       "233            SNP      0.0       0.0   \n",
       "2862           SNP      0.0       0.0   \n",
       "\n",
       "                                       genomic features  \\\n",
       "2861  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "2920  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "205   [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "233   [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "2862  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "\n",
       "                                       genetic features  \\\n",
       "2861  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "2920  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "205   [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "233   [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "2862  [{'name': 'rhsE', 'RegulonDB ID': 'ECK12000000...   \n",
       "\n",
       "                   genetic feature links  \\\n",
       "2861  {'ECK120000008': ['ECK120000008']}   \n",
       "2920  {'ECK120000008': ['ECK120000008']}   \n",
       "205   {'ECK120000008': ['ECK120000008']}   \n",
       "233   {'ECK120000008': ['ECK120000008']}   \n",
       "2862  {'ECK120000008': ['ECK120000008']}   \n",
       "\n",
       "                                                operons  \\\n",
       "2861  [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "2920  [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "205   [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "233   [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "2862  [{'name': 'rhsE-ydcD', 'RegulonDB ID': 'ECK120...   \n",
       "\n",
       "                            operon links regulators regulator links COGs  \\\n",
       "2861  {'ECK120028885': ['ECK120000008']}         []              {}   []   \n",
       "2920  {'ECK120028885': ['ECK120000008']}         []              {}   []   \n",
       "205   {'ECK120028885': ['ECK120000008']}         []              {}   []   \n",
       "233   {'ECK120028885': ['ECK120000008']}         []              {}   []   \n",
       "2862  {'ECK120028885': ['ECK120000008']}         []              {}   []   \n",
       "\n",
       "     COG links pathways pathway links RegulonDB Gene  \n",
       "2861        {}       []            {}           rhsE  \n",
       "2920        {}       []            {}           rhsE  \n",
       "205         {}       []            {}           rhsE  \n",
       "233         {}       []            {}           rhsE  \n",
       "2862        {}       []            {}           rhsE  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "    \"disrupt\",\n",
    "    \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "    \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # For development. Delete when NB is stable.\n",
    "# for _, r in annot_df.iterrows():\n",
    "#     for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "#         effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "#         mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "#         mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "            \n",
    "#         genomic_feat_annot = \"\"\n",
    "#         if genomic_feat_d[\"significant\"]: genomic_feat_annot += \"*\"\n",
    "#         genomic_feat_annot += str(genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]) + ' '\n",
    "#         genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "    \n",
    "            \n",
    "#         display(genomic_feat_annot + ' #### ' + mut_type_feat_effect)\n",
    "# #         display(mut_type_feat_effect, \"################################\", r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>28 rhsE SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>*28 rhsE-ydcD</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>28 rhsE SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>*28 rhsE-ydcD</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>28 rhsE SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>*28 rhsE-ydcD</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>28 rhsE SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>*28 rhsE-ydcD</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>28 rhsE SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>*28 rhsE-ydcD</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>28 rhsE SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>*28 rhsE-ydcD</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>28 rhsE SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>*28 rhsE</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source        target     type  value\n",
       "0         *28 rhsE  28 rhsE SNP   unknown    1.0\n",
       "1         *28 rhsE      *28 rhsE  unknown    1.0\n",
       "2   *28 rhsE-ydcD       *28 rhsE  unknown    1.0\n",
       "3         *28 rhsE  28 rhsE SNP   unknown    1.0\n",
       "4         *28 rhsE      *28 rhsE  unknown    1.0\n",
       "5   *28 rhsE-ydcD       *28 rhsE  unknown    1.0\n",
       "6         *28 rhsE  28 rhsE SNP   unknown    1.0\n",
       "7         *28 rhsE      *28 rhsE  unknown    1.0\n",
       "8   *28 rhsE-ydcD       *28 rhsE  unknown    1.0\n",
       "9         *28 rhsE  28 rhsE SNP   unknown    1.0\n",
       "10        *28 rhsE      *28 rhsE  unknown    1.0\n",
       "11  *28 rhsE-ydcD       *28 rhsE  unknown    1.0\n",
       "12        *28 rhsE  28 rhsE SNP   unknown    1.0\n",
       "13        *28 rhsE      *28 rhsE  unknown    1.0\n",
       "14  *28 rhsE-ydcD       *28 rhsE  unknown    1.0\n",
       "15        *28 rhsE  28 rhsE SNP   unknown    1.0\n",
       "16        *28 rhsE      *28 rhsE  unknown    1.0\n",
       "17  *28 rhsE-ydcD       *28 rhsE  unknown    1.0\n",
       "18        *28 rhsE  28 rhsE SNP   unknown    1.0\n",
       "19        *28 rhsE      *28 rhsE  unknown    1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "        if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "        if r[\"repaired\"]: effect = \"repair\"    \n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "#             if r[\"disrupt\"]: genomic_feat_annot += \" disrupt\"\n",
    "#             if r[\"repaired\"]: genomic_feat_annot += \" repair\"\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "#                 \"type\": r[\"mutation effect\"],\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"pathway\":\n",
    "        for pathway, ops in r[\"pathway links\"].items():\n",
    "#             if pathway == \"Glycerolipid metabolism\":  # for GYD Glycerolipid metabolism.\n",
    "            pathway_feat_d = _get_feat_d(name=pathway, json=r[\"pathways\"])\n",
    "            pathway_annot = \"\"\n",
    "            mut_count = pathway_counts[pathway]\n",
    "            if pathway_feat_d[\"significant\"] and mut_count > 1: pathway_annot += \"*\"\n",
    "            pathway_annot += str(mut_count) + ' ' + pathway_feat_d[\"name\"]\n",
    "            pathways.append(pathway_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": pathway_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "#             if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "            reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "            reg_annot = \"\"\n",
    "            mut_count = regulator_counts[regulator]\n",
    "            if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "            reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "            regulators.append(reg_annot)\n",
    "            for operon_name in ops:\n",
    "                op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": reg_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "                \n",
    "    if FUNC_ANNOT == \"COG\":\n",
    "        for COG, ops in r[\"COG links\"].items():\n",
    "#             if COG == \"Nucleotide transport and metabolism\":\n",
    "            COG_feat_d = _get_feat_d(name=COG, json=r[\"COGs\"])\n",
    "            COG_annot = \"\"\n",
    "            mut_count = COG_counts[COG]\n",
    "            if COG_feat_d[\"significant\"] and mut_count > 1: COG_annot += \"*\"\n",
    "            COG_annot += str(mut_count) + ' ' + COG_feat_d[\"name\"]\n",
    "            COGs.append(COG_annot)\n",
    "            for operon_id in ops:\n",
    "                op_feat_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "                operon_name = operon_df[operon_df[\"OPERON_ID\"]==operon].iloc[0][\"OPERON_NAME\"]\n",
    "                operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                operon_annot = \"\"\n",
    "                mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                operon_annot += str(mut_count) + ' '\n",
    "                operon_annot += operon_name\n",
    "                operons.append(operon_annot)\n",
    "                flow_row = {\n",
    "                    \"source\": COG_annot,\n",
    "                    \"target\": operon_annot,\n",
    "#                     \"type\": r[\"mutation effect\"],\n",
    "                    \"type\": \"unknown\",\n",
    "                    \"value\": 1\n",
    "                }\n",
    "                flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "    \n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specific to the GYD fig of the AVA paper.\n",
    "# flow_df = flow_df[~(flow_df[\"target\"].str.contains(\"1 pdxK\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5d349eea048f4871b429dc4fe4dad5c4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'regulon', 'type': 'process', 'title': 'regulon', 'nodes': ['regulon^*31 MarA', 'r…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "#                        flow_partition=mut_effect_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# p_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "size = dict(width=1200, height=400)\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
