{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "FUNC_ANNOT = \"regulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "163"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/GYD_df.pkl\")\n",
    "display(len(muts_df),\n",
    "#         muts_df.head()\n",
    "       )\n",
    "\n",
    "\n",
    "# out_df = pd.DataFrame()\n",
    "# df = muts_df.copy()\n",
    "# for _, r in df.iterrows():\n",
    "#     for c in r[\"operons\"]:\n",
    "#         try:\n",
    "#             # GYD\n",
    "#             if \"rpoB\" in c[\"name\"] or \"rpoC\" in c[\"name\"] or \"rph\" in c[\"name\"]:\n",
    "#                 out_df = out_df.append(r, ignore_index=True)\n",
    "#                 break\n",
    "# #             if \"ptsHI-crr\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#             # PGI\n",
    "# #             if \"pnt\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#         except:\n",
    "#             print(\"feature annotation bug with\", c)\n",
    "\n",
    "# muts_df = out_df\n",
    "# display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mut ID</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>TSS</th>\n",
       "      <th>RBS</th>\n",
       "      <th>cis-regulatory RNA</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>regulators</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>gene count</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>base-media</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>imodulons</th>\n",
       "      <th>imodulon links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>7</td>\n",
       "      <td>intergenic (+74/‑188)</td>\n",
       "      <td>carB/caiF</td>\n",
       "      <td>1105159.0</td>\n",
       "      <td>DEL</td>\n",
       "      <td>34112</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>12</td>\n",
       "      <td>GYD</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(34112, 34112)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000131/ECK120002708',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125257016}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'carAB terminator', 'RegulonDB ID': ...</td>\n",
       "      <td>[{'name': 'IHF', 'significantly associated con...</td>\n",
       "      <td>carB/caiF</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>{'ECK120000131/ECK120002708': ['ECK125257016']}</td>\n",
       "      <td>[{'name': 'carAB', 'RegulonDB ID': 'ECK1200149...</td>\n",
       "      <td>{'ECK120014911': ['ECK120000131/ECK120002708']}</td>\n",
       "      <td>[{'name': 'PurR-2', 'mutation set count': 2.0,...</td>\n",
       "      <td>{'PurR-2': ['carAB'], 'ArgR': ['carAB']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>{'PurR': ['carAB'], 'PepA': ['carAB'], 'Fis': ...</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism', 'mutation s...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014911', 'EC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1151</th>\n",
       "      <td>27</td>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>317489.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>14</td>\n",
       "      <td>GYD</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'name': 'LeuO', 'significantly associated co...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>{'Leu/Ile': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1125</th>\n",
       "      <td>27</td>\n",
       "      <td>intergenic (‑58/+35)</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>317489.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83587</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>A→C</td>\n",
       "      <td>10</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(83587, 83587)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'name': 'LeuO', 'significantly associated co...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>{'Leu/Ile': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1126</th>\n",
       "      <td>28</td>\n",
       "      <td>intergenic (‑61/+32)</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>317490.0</td>\n",
       "      <td>SNP</td>\n",
       "      <td>83590</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>G→T</td>\n",
       "      <td>10</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>0.59</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>(83590, 83590)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001204/ECK120001253',...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK120001253}</td>\n",
       "      <td>{ECK125143560}</td>\n",
       "      <td>{ECK120034959}</td>\n",
       "      <td>False</td>\n",
       "      <td>[{'name': 'leuLABCD attenuator terminator', 'R...</td>\n",
       "      <td>[{'name': 'LeuO', 'significantly associated co...</td>\n",
       "      <td>leuA/leuL</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>{'ECK120001204/ECK120001253': ['ECK125143560']}</td>\n",
       "      <td>[{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...</td>\n",
       "      <td>{'ECK120014726': ['ECK120001204/ECK120001253']}</td>\n",
       "      <td>[{'name': 'Leu/Ile', 'mutation set count': 3.0...</td>\n",
       "      <td>{'Leu/Ile': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Amino acid transport and metabolism...</td>\n",
       "      <td>{'Amino acid transport and metabolism': ['ECK1...</td>\n",
       "      <td>{'LeuO': ['leuLABCD']}</td>\n",
       "      <td>[{'name': 'Valine, leucine and isoleucine bios...</td>\n",
       "      <td>{'Valine, leucine and isoleucine biosynthesis'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1114</th>\n",
       "      <td>44</td>\n",
       "      <td>coding (314‑317/597 nt)</td>\n",
       "      <td>yadK</td>\n",
       "      <td>1108569.0</td>\n",
       "      <td>MOB</td>\n",
       "      <td>151283</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>IS5 (–) +4 bp</td>\n",
       "      <td>7</td>\n",
       "      <td>GYD</td>\n",
       "      <td>106</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(151283, 151283)</td>\n",
       "      <td>{ECK120002200}</td>\n",
       "      <td>[{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...</td>\n",
       "      <td>[]</td>\n",
       "      <td>yadK</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>511145</td>\n",
       "      <td>BOP27</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glycerol(2.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>{'ECK120002200': ['ECK120002200']}</td>\n",
       "      <td>[{'name': 'yadMLKC', 'RegulonDB ID': 'ECK12003...</td>\n",
       "      <td>{'ECK120030520': ['ECK120002200']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'name': 'Cell motility', 'mutation set count...</td>\n",
       "      <td>{'Cell motility': ['ECK120030520']}</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index                  Details mutation target annotation     Mut ID  \\\n",
       "1133     7    intergenic (+74/‑188)                  carB/caiF  1105159.0   \n",
       "1151    27     intergenic (‑58/+35)                  leuA/leuL   317489.0   \n",
       "1125    27     intergenic (‑58/+35)                  leuA/leuL   317489.0   \n",
       "1126    28     intergenic (‑61/+32)                  leuA/leuL   317490.0   \n",
       "1114    44  coding (314‑317/597 nt)                       yadK  1108569.0   \n",
       "\n",
       "     Mutation Type  Position Reference Seq Sequence Change  ale  exp  flask  \\\n",
       "1133           DEL     34112     NC_000913           Δ1 bp   12  GYD    114   \n",
       "1151           SNP     83587     NC_000913             A→C   14  GYD     68   \n",
       "1125           SNP     83587     NC_000913             A→C   10  GYD    106   \n",
       "1126           SNP     83590     NC_000913             G→T   10  GYD    106   \n",
       "1114           MOB    151283     NC_000913   IS5 (–) +4 bp    7  GYD    106   \n",
       "\n",
       "      isolate  presence  tech_rep  coding             range gene RegulonDB ID  \\\n",
       "1133        0      1.00         1   False    (34112, 34112)                {}   \n",
       "1151        0      0.50         1   False    (83587, 83587)                {}   \n",
       "1125        0      0.59         1   False    (83587, 83587)                {}   \n",
       "1126        0      0.59         1   False    (83590, 83590)                {}   \n",
       "1114        1      1.00         1    True  (151283, 151283)    {ECK120002200}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1133  [{'RegulonDB ID': 'ECK120000131/ECK120002708',...  False       False   \n",
       "1151  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...  False       False   \n",
       "1125  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...  False       False   \n",
       "1126  [{'RegulonDB ID': 'ECK120001204/ECK120001253',...  False       False   \n",
       "1114  [{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...  False       False   \n",
       "\n",
       "     TFBS promoter TSS RBS cis-regulatory RNA attenuator terminator  \\\n",
       "1133   {}       {}  {}  {}                 {}                    {}   \n",
       "1151   {}       {}  {}  {}     {ECK120001253}        {ECK125143560}   \n",
       "1125   {}       {}  {}  {}     {ECK120001253}        {ECK125143560}   \n",
       "1126   {}       {}  {}  {}     {ECK120001253}        {ECK125143560}   \n",
       "1114   {}       {}  {}  {}                 {}                    {}   \n",
       "\n",
       "          terminator  genetic  \\\n",
       "1133  {ECK125257016}    False   \n",
       "1151  {ECK120034959}    False   \n",
       "1125  {ECK120034959}    False   \n",
       "1126  {ECK120034959}    False   \n",
       "1114              {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1133  [{'name': 'carAB terminator', 'RegulonDB ID': ...   \n",
       "1151  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1125  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1126  [{'name': 'leuLABCD attenuator terminator', 'R...   \n",
       "1114  [{'name': 'yadK', 'RegulonDB ID': 'ECK12000220...   \n",
       "\n",
       "                                             regulators RegulonDB Gene  \\\n",
       "1133  [{'name': 'IHF', 'significantly associated con...      carB/caiF   \n",
       "1151  [{'name': 'LeuO', 'significantly associated co...      leuA/leuL   \n",
       "1125  [{'name': 'LeuO', 'significantly associated co...      leuA/leuL   \n",
       "1126  [{'name': 'LeuO', 'significantly associated co...      leuA/leuL   \n",
       "1114                                                 []           yadK   \n",
       "\n",
       "      mutation size  gene count  disrupt  repaired taxonomy-id  \\\n",
       "1133              1           1    False     False      511145   \n",
       "1151              1           1    False     False      511145   \n",
       "1125              1           1    False     False      511145   \n",
       "1126              1           1    False     False      511145   \n",
       "1114              0           1     True     False      511145   \n",
       "\n",
       "     strain-description base-media temperature  carbon-source nitrogen-source  \\\n",
       "1133              BOP27         M9  37 celsius  glycerol(2.5)        NH4Cl(1)   \n",
       "1151              BOP27         M9  37 celsius  glycerol(2.5)        NH4Cl(1)   \n",
       "1125              BOP27         M9  37 celsius  glycerol(2.5)        NH4Cl(1)   \n",
       "1126              BOP27         M9  37 celsius  glycerol(2.5)        NH4Cl(1)   \n",
       "1114              BOP27         M9  37 celsius  glycerol(2.5)        NH4Cl(1)   \n",
       "\n",
       "     phosphorous-source sulfur-source      supplement  \\\n",
       "1133          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1151          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1125          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1126          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1114          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "\n",
       "                                genetic feature links  \\\n",
       "1133  {'ECK120000131/ECK120002708': ['ECK125257016']}   \n",
       "1151  {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1125  {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1126  {'ECK120001204/ECK120001253': ['ECK125143560']}   \n",
       "1114               {'ECK120002200': ['ECK120002200']}   \n",
       "\n",
       "                                                operons  \\\n",
       "1133  [{'name': 'carAB', 'RegulonDB ID': 'ECK1200149...   \n",
       "1151  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1125  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1126  [{'name': 'leuLABCD', 'RegulonDB ID': 'ECK1200...   \n",
       "1114  [{'name': 'yadMLKC', 'RegulonDB ID': 'ECK12003...   \n",
       "\n",
       "                                         operon links  \\\n",
       "1133  {'ECK120014911': ['ECK120000131/ECK120002708']}   \n",
       "1151  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1125  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1126  {'ECK120014726': ['ECK120001204/ECK120001253']}   \n",
       "1114               {'ECK120030520': ['ECK120002200']}   \n",
       "\n",
       "                                              imodulons  \\\n",
       "1133  [{'name': 'PurR-2', 'mutation set count': 2.0,...   \n",
       "1151  [{'name': 'Leu/Ile', 'mutation set count': 3.0...   \n",
       "1125  [{'name': 'Leu/Ile', 'mutation set count': 3.0...   \n",
       "1126  [{'name': 'Leu/Ile', 'mutation set count': 3.0...   \n",
       "1114                                                 []   \n",
       "\n",
       "                                imodulon links  \\\n",
       "1133  {'PurR-2': ['carAB'], 'ArgR': ['carAB']}   \n",
       "1151                 {'Leu/Ile': ['leuLABCD']}   \n",
       "1125                 {'Leu/Ile': ['leuLABCD']}   \n",
       "1126                 {'Leu/Ile': ['leuLABCD']}   \n",
       "1114                                        {}   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1133  [{'name': 'Amino acid transport and metabolism...   \n",
       "1151  [{'name': 'Amino acid transport and metabolism...   \n",
       "1125  [{'name': 'Amino acid transport and metabolism...   \n",
       "1126  [{'name': 'Amino acid transport and metabolism...   \n",
       "1114  [{'name': 'Cell motility', 'mutation set count...   \n",
       "\n",
       "                                              COG links  \\\n",
       "1133  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1151  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1125  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1126  {'Amino acid transport and metabolism': ['ECK1...   \n",
       "1114                {'Cell motility': ['ECK120030520']}   \n",
       "\n",
       "                                        regulator links  \\\n",
       "1133  {'PurR': ['carAB'], 'PepA': ['carAB'], 'Fis': ...   \n",
       "1151                             {'LeuO': ['leuLABCD']}   \n",
       "1125                             {'LeuO': ['leuLABCD']}   \n",
       "1126                             {'LeuO': ['leuLABCD']}   \n",
       "1114                                                 {}   \n",
       "\n",
       "                                               pathways  \\\n",
       "1133  [{'name': 'Pyrimidine metabolism', 'mutation s...   \n",
       "1151  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1125  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1126  [{'name': 'Valine, leucine and isoleucine bios...   \n",
       "1114                                                 []   \n",
       "\n",
       "                                          pathway links  \n",
       "1133  {'Pyrimidine metabolism': ['ECK120014911', 'EC...  \n",
       "1151  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "1125  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "1126  {'Valine, leucine and isoleucine biosynthesis'...  \n",
       "1114                                                 {}  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# muts_df = muts_df[muts_df[\"mutation target annotation\"].str.contains(\"pnt\")]\n",
    "# muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specific to the AVA publication\n",
    "temp_df = pd.DataFrame()\n",
    "for i, r in muts_df.iterrows():\n",
    "    if \"CRP\" in r[\"regulator links\"].keys():\n",
    "        temp_df = temp_df.append(r)\n",
    "muts_df = temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"GlpR\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Mlc\" in r[\"regulator links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "# #     if \"Nucleotide transport and metabolism\" in r[\"COG links\"].keys() \\\n",
    "#     if \"cyaA\" in r[\"RegulonDB Gene\"]:\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Specific to the AVA publication\n",
    "# temp_df = pd.DataFrame()\n",
    "# for i, r in muts_df.iterrows():\n",
    "#     if \"Glycerolipid metabolism\" in r[\"pathway links\"].keys():\n",
    "#         temp_df = temp_df.append(r)\n",
    "# muts_df = temp_df\n",
    "# len(muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = muts_df.copy()\n",
    "# df[\"significantly associated condition set\"] = df[\"significantly associated condition set\"].apply(\n",
    "#     lambda x: str(x))\n",
    "# display(df[\"significantly associated condition set\"].unique())\n",
    "\n",
    "\n",
    "# # GYD\n",
    "# muts_df = muts_df[\n",
    "#     (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)'})\n",
    "#     | (muts_df[\"significantly associated condition set\"]=={'glycerol(2.5)', \"BOP27\"})\n",
    "# ]\n",
    "\n",
    "# # SER\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"] == {'ΔsdaA ΔsdaB ΔtdcG ΔglyA', 'L-Serine'}]\n",
    "\n",
    "# # # GLU\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'glucose(2)'}]\n",
    "\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]==set()]\n",
    "\n",
    "\n",
    "# # muts_df = muts_df[\n",
    "# #     muts_df[\"significantly associated condition set\"]=={'log', '511145', 'M9'}]\n",
    "\n",
    "\n",
    "# display(len(muts_df), muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(70, 15)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>disrupt</th>\n",
       "      <th>repaired</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "      <th>RegulonDB Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1075</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001044/ECK120003456',...</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120001044/ECK120003456',...</td>\n",
       "      <td>{'ECK120001044/ECK120003456': ['ECK120001044/E...</td>\n",
       "      <td>[{'name': 'uidABC', 'RegulonDB ID': 'ECK120026...</td>\n",
       "      <td>{'ECK120026438': ['ECK120001044/ECK120003456']}</td>\n",
       "      <td>[{'name': 'UidR', 'significantly associated co...</td>\n",
       "      <td>{'UxuR': ['uidABC'], 'CRP': ['uidABC'], 'UidR'...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>uidA/uidR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1083</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'flhCp3', 'RegulonDB ID': 'ECK125137...</td>\n",
       "      <td>[{'name': 'flhD', 'RegulonDB ID': 'ECK12000031...</td>\n",
       "      <td>{'ECK120000314': ['ECK125137156', 'ECK12000031...</td>\n",
       "      <td>[{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...</td>\n",
       "      <td>{'ECK120014934': ['ECK120000314']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'RcsAB': ['flhDC'], 'FliZ': ['flhDC'], 'MatA'...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>flhD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1085</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'cdd', 'RegulonDB ID': 'ECK120000133...</td>\n",
       "      <td>[{'name': 'cdd', 'RegulonDB ID': 'ECK120000133...</td>\n",
       "      <td>{'ECK120000133': ['ECK120000133']}</td>\n",
       "      <td>[{'name': 'cdd', 'RegulonDB ID': 'ECK120014505...</td>\n",
       "      <td>{'ECK120014505': ['ECK120000133']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'CytR': ['cdd'], 'CRP': ['cdd'], 'NsrR': ['cd...</td>\n",
       "      <td>[{'name': 'Nucleotide transport and metabolism...</td>\n",
       "      <td>{'Nucleotide transport and metabolism': ['ECK1...</td>\n",
       "      <td>[{'name': 'Pyrimidine metabolism', 'mutation s...</td>\n",
       "      <td>{'Pyrimidine metabolism': ['ECK120014505'], 'D...</td>\n",
       "      <td>cdd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1076</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'crr', 'RegulonDB ID': 'ECK120000161...</td>\n",
       "      <td>[{'name': 'crr', 'RegulonDB ID': 'ECK120000161...</td>\n",
       "      <td>{'ECK120000161': ['ECK120000161']}</td>\n",
       "      <td>[{'name': 'ptsHI-crr', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120014520': ['ECK120000161']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'Mlc': ['ptsHI-crr'], 'CRP': ['ptsHI-crr'], '...</td>\n",
       "      <td>[{'name': 'Carbohydrate transport and metaboli...</td>\n",
       "      <td>{'Carbohydrate transport and metabolism': ['EC...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>crr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189</th>\n",
       "      <td>SNP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'crr', 'RegulonDB ID': 'ECK120000161...</td>\n",
       "      <td>[{'name': 'crr', 'RegulonDB ID': 'ECK120000161...</td>\n",
       "      <td>{'ECK120000161': ['ECK120000161']}</td>\n",
       "      <td>[{'name': 'ptsHI-crr', 'RegulonDB ID': 'ECK120...</td>\n",
       "      <td>{'ECK120014520': ['ECK120000161']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'Mlc': ['ptsHI-crr'], 'CRP': ['ptsHI-crr'], '...</td>\n",
       "      <td>[{'name': 'Carbohydrate transport and metaboli...</td>\n",
       "      <td>{'Carbohydrate transport and metabolism': ['EC...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>crr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type  disrupt  repaired  \\\n",
       "1075           SNP      0.0       0.0   \n",
       "1083           SNP      0.0       0.0   \n",
       "1085           SNP      0.0       0.0   \n",
       "1076           SNP      0.0       0.0   \n",
       "1189           SNP      0.0       0.0   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1075  [{'RegulonDB ID': 'ECK120001044/ECK120003456',...   \n",
       "1083  [{'name': 'flhCp3', 'RegulonDB ID': 'ECK125137...   \n",
       "1085  [{'name': 'cdd', 'RegulonDB ID': 'ECK120000133...   \n",
       "1076  [{'name': 'crr', 'RegulonDB ID': 'ECK120000161...   \n",
       "1189  [{'name': 'crr', 'RegulonDB ID': 'ECK120000161...   \n",
       "\n",
       "                                       genetic features  \\\n",
       "1075  [{'RegulonDB ID': 'ECK120001044/ECK120003456',...   \n",
       "1083  [{'name': 'flhD', 'RegulonDB ID': 'ECK12000031...   \n",
       "1085  [{'name': 'cdd', 'RegulonDB ID': 'ECK120000133...   \n",
       "1076  [{'name': 'crr', 'RegulonDB ID': 'ECK120000161...   \n",
       "1189  [{'name': 'crr', 'RegulonDB ID': 'ECK120000161...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1075  {'ECK120001044/ECK120003456': ['ECK120001044/E...   \n",
       "1083  {'ECK120000314': ['ECK125137156', 'ECK12000031...   \n",
       "1085                 {'ECK120000133': ['ECK120000133']}   \n",
       "1076                 {'ECK120000161': ['ECK120000161']}   \n",
       "1189                 {'ECK120000161': ['ECK120000161']}   \n",
       "\n",
       "                                                operons  \\\n",
       "1075  [{'name': 'uidABC', 'RegulonDB ID': 'ECK120026...   \n",
       "1083  [{'name': 'flhDC', 'RegulonDB ID': 'ECK1200149...   \n",
       "1085  [{'name': 'cdd', 'RegulonDB ID': 'ECK120014505...   \n",
       "1076  [{'name': 'ptsHI-crr', 'RegulonDB ID': 'ECK120...   \n",
       "1189  [{'name': 'ptsHI-crr', 'RegulonDB ID': 'ECK120...   \n",
       "\n",
       "                                         operon links  \\\n",
       "1075  {'ECK120026438': ['ECK120001044/ECK120003456']}   \n",
       "1083               {'ECK120014934': ['ECK120000314']}   \n",
       "1085               {'ECK120014505': ['ECK120000133']}   \n",
       "1076               {'ECK120014520': ['ECK120000161']}   \n",
       "1189               {'ECK120014520': ['ECK120000161']}   \n",
       "\n",
       "                                             regulators  \\\n",
       "1075  [{'name': 'UidR', 'significantly associated co...   \n",
       "1083  [{'name': 'CRP', 'significantly associated con...   \n",
       "1085  [{'name': 'CRP', 'significantly associated con...   \n",
       "1076  [{'name': 'CRP', 'significantly associated con...   \n",
       "1189  [{'name': 'CRP', 'significantly associated con...   \n",
       "\n",
       "                                        regulator links  \\\n",
       "1075  {'UxuR': ['uidABC'], 'CRP': ['uidABC'], 'UidR'...   \n",
       "1083  {'RcsAB': ['flhDC'], 'FliZ': ['flhDC'], 'MatA'...   \n",
       "1085  {'CytR': ['cdd'], 'CRP': ['cdd'], 'NsrR': ['cd...   \n",
       "1076  {'Mlc': ['ptsHI-crr'], 'CRP': ['ptsHI-crr'], '...   \n",
       "1189  {'Mlc': ['ptsHI-crr'], 'CRP': ['ptsHI-crr'], '...   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1075                                                 []   \n",
       "1083                                                 []   \n",
       "1085  [{'name': 'Nucleotide transport and metabolism...   \n",
       "1076  [{'name': 'Carbohydrate transport and metaboli...   \n",
       "1189  [{'name': 'Carbohydrate transport and metaboli...   \n",
       "\n",
       "                                              COG links  \\\n",
       "1075                                                 {}   \n",
       "1083                                                 {}   \n",
       "1085  {'Nucleotide transport and metabolism': ['ECK1...   \n",
       "1076  {'Carbohydrate transport and metabolism': ['EC...   \n",
       "1189  {'Carbohydrate transport and metabolism': ['EC...   \n",
       "\n",
       "                                               pathways  \\\n",
       "1075                                                 []   \n",
       "1083                                                 []   \n",
       "1085  [{'name': 'Pyrimidine metabolism', 'mutation s...   \n",
       "1076                                                 []   \n",
       "1189                                                 []   \n",
       "\n",
       "                                          pathway links RegulonDB Gene  \n",
       "1075                                                 {}      uidA/uidR  \n",
       "1083                                                 {}           flhD  \n",
       "1085  {'Pyrimidine metabolism': ['ECK120014505'], 'D...            cdd  \n",
       "1076                                                 {}            crr  \n",
       "1189                                                 {}            crr  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "    \"disrupt\",\n",
    "    \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "    \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # For development. Delete when NB is stable.\n",
    "# for _, r in annot_df.iterrows():\n",
    "#     for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "#         effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "#         mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "#         mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "            \n",
    "#         genomic_feat_annot = \"\"\n",
    "#         if genomic_feat_d[\"significant\"]: genomic_feat_annot += \"*\"\n",
    "#         genomic_feat_annot += str(genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]) + ' '\n",
    "#         genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "    \n",
    "            \n",
    "#         display(genomic_feat_annot + ' #### ' + mut_type_feat_effect)\n",
    "# #         display(mut_type_feat_effect, \"################################\", r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>type</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1 uidA/uidR</td>\n",
       "      <td>1 uidA/uidR SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1 uidA/uidR</td>\n",
       "      <td>1 uidA/uidR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1 uidABC</td>\n",
       "      <td>1 uidA/uidR</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>*72 CRP</td>\n",
       "      <td>1 uidABC</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1 flhCp3</td>\n",
       "      <td>1 flhCp3 SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1 flhD</td>\n",
       "      <td>1 flhD SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2 flhD</td>\n",
       "      <td>1 flhCp3</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2 flhD</td>\n",
       "      <td>1 flhD</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1 flhDC</td>\n",
       "      <td>2 flhD</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>*72 CRP</td>\n",
       "      <td>1 flhDC</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1 cdd</td>\n",
       "      <td>1 cdd SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1 cdd</td>\n",
       "      <td>1 cdd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1 cdd</td>\n",
       "      <td>1 cdd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>*72 CRP</td>\n",
       "      <td>1 cdd</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>*10 crr</td>\n",
       "      <td>10 crr SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>*10 crr</td>\n",
       "      <td>*10 crr</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>*11 ptsHI-crr</td>\n",
       "      <td>*10 crr</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>*72 CRP</td>\n",
       "      <td>*11 ptsHI-crr</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>*10 crr</td>\n",
       "      <td>10 crr SNP</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>*10 crr</td>\n",
       "      <td>*10 crr</td>\n",
       "      <td>unknown</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source            target     type  value\n",
       "0      1 uidA/uidR  1 uidA/uidR SNP   unknown    1.0\n",
       "1      1 uidA/uidR       1 uidA/uidR  unknown    1.0\n",
       "2        1 uidABC        1 uidA/uidR  unknown    1.0\n",
       "3          *72 CRP         1 uidABC   unknown    1.0\n",
       "4         1 flhCp3     1 flhCp3 SNP   unknown    1.0\n",
       "5           1 flhD       1 flhD SNP   unknown    1.0\n",
       "6           2 flhD          1 flhCp3  unknown    1.0\n",
       "7           2 flhD            1 flhD  unknown    1.0\n",
       "8         1 flhDC             2 flhD  unknown    1.0\n",
       "9          *72 CRP          1 flhDC   unknown    1.0\n",
       "10           1 cdd        1 cdd SNP   unknown    1.0\n",
       "11           1 cdd             1 cdd  unknown    1.0\n",
       "12          1 cdd              1 cdd  unknown    1.0\n",
       "13         *72 CRP            1 cdd   unknown    1.0\n",
       "14         *10 crr       10 crr SNP   unknown    1.0\n",
       "15         *10 crr           *10 crr  unknown    1.0\n",
       "16  *11 ptsHI-crr            *10 crr  unknown    1.0\n",
       "17         *72 CRP    *11 ptsHI-crr   unknown    1.0\n",
       "18         *10 crr       10 crr SNP   unknown    1.0\n",
       "19         *10 crr           *10 crr  unknown    1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _adjust_name_len_for_floweaver(feat_name):\n",
    "    if len(feat_name) > 7: \n",
    "        feat_name = feat_name[:7] + \"...\"\n",
    "    return feat_name\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "        if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "        if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "            if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "        if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "            if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "            if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "                reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "                reg_annot = \"\"\n",
    "                mut_count = regulator_counts[regulator]\n",
    "                if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "                reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "                regulators.append(reg_annot)\n",
    "                for operon_name in ops:\n",
    "                    op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                    operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                    operon_annot = \"\"\n",
    "                    mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "                    if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                    operon_annot += str(mut_count) + ' '\n",
    "                    operon_annot += operon_name\n",
    "                    operons.append(operon_annot)\n",
    "                    flow_row = {\n",
    "                        \"source\": reg_annot,\n",
    "                        \"target\": operon_annot,\n",
    "                        \"type\": \"unknown\",\n",
    "                        \"value\": 1\n",
    "                    }\n",
    "                    flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "\n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specific to the GYD fig of the AVA paper.\n",
    "flow_df = flow_df[~(flow_df[\"target\"].str.contains(\"1 pdxK\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b21e79a1f9394fd3b690a0e13b53be1d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "SankeyWidget(groups=[{'id': 'regulon', 'type': 'process', 'title': 'regulon', 'nodes': ['regulon^*72 CRP']}, {…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "#                        flow_partition=mut_effect_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# p_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "size = dict(width=3000, height=4000)\n",
    "# size = dict(width=2000, height=2500)\n",
    "size = dict(width=1500, height=800)\n",
    "\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
