{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNC_ANNOT = \"pathway\"\n",
    "# FUNC_ANNOT = \"COG\"\n",
    "FUNC_ANNOT = \"regulon\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "muts_df = pd.read_pickle(\"./data/PGI_df.pkl\")\n",
    "display(len(muts_df),\n",
    "#         muts_df.head()\n",
    "       )\n",
    "\n",
    "\n",
    "# out_df = pd.DataFrame()\n",
    "# df = muts_df.copy()\n",
    "# for _, r in df.iterrows():\n",
    "#     for c in r[\"operons\"]:\n",
    "#         try:\n",
    "#             # GYD\n",
    "#             if \"rpoB\" in c[\"name\"] or \"rpoC\" in c[\"name\"] or \"rph\" in c[\"name\"]:\n",
    "#                 out_df = out_df.append(r, ignore_index=True)\n",
    "#                 break\n",
    "# #             if \"ptsHI-crr\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#             # PGI\n",
    "# #             if \"pnt\" in c[\"name\"]:\n",
    "# #                 out_df = out_df.append(r, ignore_index=True)\n",
    "# #                 break\n",
    "#         except:\n",
    "#             print(\"feature annotation bug with\", c)\n",
    "\n",
    "# muts_df = out_df\n",
    "# display(len(muts_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>ale</th>\n",
       "      <th>exp</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>presence</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>coding</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>oriC</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>TFBS</th>\n",
       "      <th>promoter</th>\n",
       "      <th>TSS</th>\n",
       "      <th>RBS</th>\n",
       "      <th>attenuator terminator</th>\n",
       "      <th>terminator</th>\n",
       "      <th>genetic</th>\n",
       "      <th>genomic features</th>\n",
       "      <th>genetic feature links</th>\n",
       "      <th>operons</th>\n",
       "      <th>operon links</th>\n",
       "      <th>regulators</th>\n",
       "      <th>regulator links</th>\n",
       "      <th>taxonomy-id</th>\n",
       "      <th>strain-description</th>\n",
       "      <th>base-media</th>\n",
       "      <th>temperature</th>\n",
       "      <th>carbon-source</th>\n",
       "      <th>nitrogen-source</th>\n",
       "      <th>phosphorous-source</th>\n",
       "      <th>sulfur-source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>COGs</th>\n",
       "      <th>COG links</th>\n",
       "      <th>pathways</th>\n",
       "      <th>pathway links</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1628</th>\n",
       "      <td>27</td>\n",
       "      <td>coding (2443‑2444/2598 nt)</td>\n",
       "      <td>acnB</td>\n",
       "      <td>SUB</td>\n",
       "      <td>134057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2 bp→CT</td>\n",
       "      <td>5</td>\n",
       "      <td>PGI</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(134057, 134058)</td>\n",
       "      <td>{ECK120002193}</td>\n",
       "      <td>[{'name': 'acnB', 'RegulonDB ID': 'ECK12000219...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'acnB', 'RegulonDB ID': 'ECK12000219...</td>\n",
       "      <td>{'ECK120002193': ['ECK120002193']}</td>\n",
       "      <td>[{'name': 'acnB', 'RegulonDB ID': 'ECK12001450...</td>\n",
       "      <td>{'ECK120014501': ['ECK120002193']}</td>\n",
       "      <td>[{'name': 'ArcA', 'significantly associated co...</td>\n",
       "      <td>{'ArcA': ['acnB'], 'Fis': ['acnB'], 'CRP': ['a...</td>\n",
       "      <td>511145</td>\n",
       "      <td>Δpgi</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>[{'name': 'Energy production and conversion', ...</td>\n",
       "      <td>{'Energy production and conversion': ['ECK1200...</td>\n",
       "      <td>[{'name': 'Citrate cycle (TCA cycle)', 'mutati...</td>\n",
       "      <td>{'Citrate cycle (TCA cycle)': ['ECK120014501']...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1658</th>\n",
       "      <td>49</td>\n",
       "      <td>L220R (CTC→CGC)</td>\n",
       "      <td>ispU</td>\n",
       "      <td>SNP</td>\n",
       "      <td>195561</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T→G</td>\n",
       "      <td>10</td>\n",
       "      <td>PGI</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(195561, 195561)</td>\n",
       "      <td>{ECK120002712}</td>\n",
       "      <td>[{'name': 'ispU', 'RegulonDB ID': 'ECK12000271...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125136392, ECK125136393}</td>\n",
       "      <td>{ECK125136392}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'cdsAp2', 'RegulonDB ID': 'ECK125136...</td>\n",
       "      <td>{'ECK120002712': ['ECK120002712', 'ECK12513639...</td>\n",
       "      <td>[{'name': 'ispU', 'RegulonDB ID': 'ECK12003056...</td>\n",
       "      <td>{'ECK120030566': ['ECK120002712']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>511145</td>\n",
       "      <td>Δpgi</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>[{'name': 'Lipid transport and metabolism', 'm...</td>\n",
       "      <td>{'Lipid transport and metabolism': ['ECK120030...</td>\n",
       "      <td>[{'name': 'Terpenoid backbone biosynthesis', '...</td>\n",
       "      <td>{'Terpenoid backbone biosynthesis': ['ECK12003...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1599</th>\n",
       "      <td>0</td>\n",
       "      <td>P56L (CCT→CTT)</td>\n",
       "      <td>fabZ</td>\n",
       "      <td>SNP</td>\n",
       "      <td>202267</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→T</td>\n",
       "      <td>1</td>\n",
       "      <td>PGI</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(202267, 202267)</td>\n",
       "      <td>{ECK120001257}</td>\n",
       "      <td>[{'name': 'fabZ', 'RegulonDB ID': 'ECK12000125...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'fabZ', 'RegulonDB ID': 'ECK12000125...</td>\n",
       "      <td>{'ECK120001257': ['ECK120001257']}</td>\n",
       "      <td>[{'name': 'bamA-skp-lpxD-fabZ-lpxAB-rnhB-dnaE'...</td>\n",
       "      <td>{'ECK125166138': ['ECK120001257']}</td>\n",
       "      <td>[{'name': 'CpxR', 'significantly associated co...</td>\n",
       "      <td>{'CpxR': ['bamA-skp-lpxD-fabZ-lpxAB-rnhB-dnaE'...</td>\n",
       "      <td>511145</td>\n",
       "      <td>Δpgi</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>[{'name': 'Lipid transport and metabolism', 'm...</td>\n",
       "      <td>{'Lipid transport and metabolism': ['ECK125166...</td>\n",
       "      <td>[{'name': 'Fatty acid biosynthesis', 'mutation...</td>\n",
       "      <td>{'Fatty acid biosynthesis': ['ECK125166138'], ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1644</th>\n",
       "      <td>38</td>\n",
       "      <td>A88E (GCG→GAG)</td>\n",
       "      <td>mrdB</td>\n",
       "      <td>SNP</td>\n",
       "      <td>666051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G→T</td>\n",
       "      <td>7</td>\n",
       "      <td>PGI</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(666051, 666051)</td>\n",
       "      <td>{ECK120000600}</td>\n",
       "      <td>[{'name': 'mrdB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'mrdB', 'RegulonDB ID': 'ECK12000060...</td>\n",
       "      <td>{'ECK120000600': ['ECK120000600']}</td>\n",
       "      <td>[{'name': 'rsfS-rlmH-mrdAB-rlpA', 'RegulonDB I...</td>\n",
       "      <td>{'ECK125147414': ['ECK120000600']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>511145</td>\n",
       "      <td>Δpgi</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>[{'name': 'Cell cycle control, cell division, ...</td>\n",
       "      <td>{'Cell cycle control, cell division, chromosom...</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1656</th>\n",
       "      <td>41</td>\n",
       "      <td>G73C (GGC→TGC)</td>\n",
       "      <td>nagC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>701378</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C→A</td>\n",
       "      <td>9</td>\n",
       "      <td>PGI</td>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>(701378, 701378)</td>\n",
       "      <td>{ECK120000629}</td>\n",
       "      <td>[{'name': 'nagC', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ECK125236293, ECK125236296, ECK125236295, ECK...</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>True</td>\n",
       "      <td>[{'name': 'nagC', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>{'ECK120000629': ['ECK120000629']}</td>\n",
       "      <td>[{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...</td>\n",
       "      <td>{'ECK125162049': ['ECK120000629']}</td>\n",
       "      <td>[{'name': 'CRP', 'significantly associated con...</td>\n",
       "      <td>{'CRP': ['nagBAC-umpH'], 'NagC': ['nagBAC-umpH']}</td>\n",
       "      <td>511145</td>\n",
       "      <td>Δpgi</td>\n",
       "      <td>M9</td>\n",
       "      <td>37 celsius</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>trace elements</td>\n",
       "      <td>[{'name': 'Transcription', 'mutation set count...</td>\n",
       "      <td>{'Transcription': ['ECK125162049']}</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index                     Details mutation target annotation  \\\n",
       "1628    27  coding (2443‑2444/2598 nt)                       acnB   \n",
       "1658    49             L220R (CTC→CGC)                       ispU   \n",
       "1599     0              P56L (CCT→CTT)                       fabZ   \n",
       "1644    38              A88E (GCG→GAG)                       mrdB   \n",
       "1656    41              G73C (GGC→TGC)                       nagC   \n",
       "\n",
       "     Mutation Type  Position Reference Seq Sequence Change  ale  exp  flask  \\\n",
       "1628           SUB    134057           NaN         2 bp→CT    5  PGI     50   \n",
       "1658           SNP    195561           NaN             T→G   10  PGI     50   \n",
       "1599           SNP    202267           NaN             C→T    1  PGI     50   \n",
       "1644           SNP    666051           NaN             G→T    7  PGI     50   \n",
       "1656           SNP    701378           NaN             C→A    9  PGI     50   \n",
       "\n",
       "      isolate  presence  tech_rep  coding             range gene RegulonDB ID  \\\n",
       "1628        1       1.0         1    True  (134057, 134058)    {ECK120002193}   \n",
       "1658        1       1.0         1    True  (195561, 195561)    {ECK120002712}   \n",
       "1599        1       1.0         1    True  (202267, 202267)    {ECK120001257}   \n",
       "1644        1       1.0         1    True  (666051, 666051)    {ECK120000600}   \n",
       "1656        1       1.0         1    True  (701378, 701378)    {ECK120000629}   \n",
       "\n",
       "                                       genetic features   oriC  pseudogene  \\\n",
       "1628  [{'name': 'acnB', 'RegulonDB ID': 'ECK12000219...  False       False   \n",
       "1658  [{'name': 'ispU', 'RegulonDB ID': 'ECK12000271...  False       False   \n",
       "1599  [{'name': 'fabZ', 'RegulonDB ID': 'ECK12000125...  False       False   \n",
       "1644  [{'name': 'mrdB', 'RegulonDB ID': 'ECK12000060...  False       False   \n",
       "1656  [{'name': 'nagC', 'RegulonDB ID': 'ECK12000062...  False       False   \n",
       "\n",
       "     TFBS                                           promoter             TSS  \\\n",
       "1628   {}                                                 {}              {}   \n",
       "1658   {}                       {ECK125136392, ECK125136393}  {ECK125136392}   \n",
       "1599   {}                                                 {}              {}   \n",
       "1644   {}                                                 {}              {}   \n",
       "1656   {}  {ECK125236293, ECK125236296, ECK125236295, ECK...              {}   \n",
       "\n",
       "     RBS attenuator terminator terminator  genetic  \\\n",
       "1628  {}                    {}         {}     True   \n",
       "1658  {}                    {}         {}     True   \n",
       "1599  {}                    {}         {}     True   \n",
       "1644  {}                    {}         {}     True   \n",
       "1656  {}                    {}         {}     True   \n",
       "\n",
       "                                       genomic features  \\\n",
       "1628  [{'name': 'acnB', 'RegulonDB ID': 'ECK12000219...   \n",
       "1658  [{'name': 'cdsAp2', 'RegulonDB ID': 'ECK125136...   \n",
       "1599  [{'name': 'fabZ', 'RegulonDB ID': 'ECK12000125...   \n",
       "1644  [{'name': 'mrdB', 'RegulonDB ID': 'ECK12000060...   \n",
       "1656  [{'name': 'nagC', 'RegulonDB ID': 'ECK12000062...   \n",
       "\n",
       "                                  genetic feature links  \\\n",
       "1628                 {'ECK120002193': ['ECK120002193']}   \n",
       "1658  {'ECK120002712': ['ECK120002712', 'ECK12513639...   \n",
       "1599                 {'ECK120001257': ['ECK120001257']}   \n",
       "1644                 {'ECK120000600': ['ECK120000600']}   \n",
       "1656                 {'ECK120000629': ['ECK120000629']}   \n",
       "\n",
       "                                                operons  \\\n",
       "1628  [{'name': 'acnB', 'RegulonDB ID': 'ECK12001450...   \n",
       "1658  [{'name': 'ispU', 'RegulonDB ID': 'ECK12003056...   \n",
       "1599  [{'name': 'bamA-skp-lpxD-fabZ-lpxAB-rnhB-dnaE'...   \n",
       "1644  [{'name': 'rsfS-rlmH-mrdAB-rlpA', 'RegulonDB I...   \n",
       "1656  [{'name': 'nagBAC-umpH', 'RegulonDB ID': 'ECK1...   \n",
       "\n",
       "                            operon links  \\\n",
       "1628  {'ECK120014501': ['ECK120002193']}   \n",
       "1658  {'ECK120030566': ['ECK120002712']}   \n",
       "1599  {'ECK125166138': ['ECK120001257']}   \n",
       "1644  {'ECK125147414': ['ECK120000600']}   \n",
       "1656  {'ECK125162049': ['ECK120000629']}   \n",
       "\n",
       "                                             regulators  \\\n",
       "1628  [{'name': 'ArcA', 'significantly associated co...   \n",
       "1658                                                 []   \n",
       "1599  [{'name': 'CpxR', 'significantly associated co...   \n",
       "1644                                                 []   \n",
       "1656  [{'name': 'CRP', 'significantly associated con...   \n",
       "\n",
       "                                        regulator links taxonomy-id  \\\n",
       "1628  {'ArcA': ['acnB'], 'Fis': ['acnB'], 'CRP': ['a...      511145   \n",
       "1658                                                 {}      511145   \n",
       "1599  {'CpxR': ['bamA-skp-lpxD-fabZ-lpxAB-rnhB-dnaE'...      511145   \n",
       "1644                                                 {}      511145   \n",
       "1656  {'CRP': ['nagBAC-umpH'], 'NagC': ['nagBAC-umpH']}      511145   \n",
       "\n",
       "     strain-description base-media temperature carbon-source nitrogen-source  \\\n",
       "1628               Δpgi         M9  37 celsius    glucose(4)        NH4Cl(1)   \n",
       "1658               Δpgi         M9  37 celsius    glucose(4)        NH4Cl(1)   \n",
       "1599               Δpgi         M9  37 celsius    glucose(4)        NH4Cl(1)   \n",
       "1644               Δpgi         M9  37 celsius    glucose(4)        NH4Cl(1)   \n",
       "1656               Δpgi         M9  37 celsius    glucose(4)        NH4Cl(1)   \n",
       "\n",
       "     phosphorous-source sulfur-source      supplement  \\\n",
       "1628          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1658          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1599          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1644          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "1656          KH2PO4(3)   MgSO4(0.24)  trace elements   \n",
       "\n",
       "                                                   COGs  \\\n",
       "1628  [{'name': 'Energy production and conversion', ...   \n",
       "1658  [{'name': 'Lipid transport and metabolism', 'm...   \n",
       "1599  [{'name': 'Lipid transport and metabolism', 'm...   \n",
       "1644  [{'name': 'Cell cycle control, cell division, ...   \n",
       "1656  [{'name': 'Transcription', 'mutation set count...   \n",
       "\n",
       "                                              COG links  \\\n",
       "1628  {'Energy production and conversion': ['ECK1200...   \n",
       "1658  {'Lipid transport and metabolism': ['ECK120030...   \n",
       "1599  {'Lipid transport and metabolism': ['ECK125166...   \n",
       "1644  {'Cell cycle control, cell division, chromosom...   \n",
       "1656                {'Transcription': ['ECK125162049']}   \n",
       "\n",
       "                                               pathways  \\\n",
       "1628  [{'name': 'Citrate cycle (TCA cycle)', 'mutati...   \n",
       "1658  [{'name': 'Terpenoid backbone biosynthesis', '...   \n",
       "1599  [{'name': 'Fatty acid biosynthesis', 'mutation...   \n",
       "1644                                                 []   \n",
       "1656                                                 []   \n",
       "\n",
       "                                          pathway links  \n",
       "1628  {'Citrate cycle (TCA cycle)': ['ECK120014501']...  \n",
       "1658  {'Terpenoid backbone biosynthesis': ['ECK12003...  \n",
       "1599  {'Fatty acid biosynthesis': ['ECK125166138'], ...  \n",
       "1644                                                 {}  \n",
       "1656                                                 {}  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# sorts mutations by position so that mutations to proximal features will be inserted into flow_df next to each other.\n",
    "muts_df = muts_df.sort_values(by=\"Position\")\n",
    "muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specific to the AVA publication\n",
    "temp_df = pd.DataFrame()\n",
    "for i, r in muts_df.iterrows():\n",
    "    if \"ECK120014614\" in r[\"operon links\"].keys():  # ECK120014614 is the regulonDB ID for the pntAB operon.\n",
    "        temp_df = temp_df.append(r)\n",
    "muts_df = temp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['repaired', 'RegulonDB Gene', 'disrupt'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-93ce06d4618b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mannot_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmuts_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0mannot_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mannot_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol_l\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m \u001b[0;31m# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2932\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2933\u001b[0m             indexer = self.loc._convert_to_indexer(key, axis=1,\n\u001b[0;32m-> 2934\u001b[0;31m                                                    raise_missing=True)\n\u001b[0m\u001b[1;32m   2935\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2936\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_convert_to_indexer\u001b[0;34m(self, obj, axis, is_setter, raise_missing)\u001b[0m\n\u001b[1;32m   1352\u001b[0m                 kwargs = {'raise_missing': True if is_setter else\n\u001b[1;32m   1353\u001b[0m                           raise_missing}\n\u001b[0;32m-> 1354\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1355\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1356\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[0;34m(self, key, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1159\u001b[0m         self._validate_read_indexer(keyarr, indexer,\n\u001b[1;32m   1160\u001b[0m                                     \u001b[0mo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis_number\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1161\u001b[0;31m                                     raise_missing=raise_missing)\n\u001b[0m\u001b[1;32m   1162\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1163\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[0;34m(self, key, indexer, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1250\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'loc'\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1251\u001b[0m                 \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1252\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"{} not in index\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnot_found\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1254\u001b[0m             \u001b[0;31m# we skip the warning on Categorical/Interval\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['repaired', 'RegulonDB Gene', 'disrupt'] not in index\""
     ]
    }
   ],
   "source": [
    "col_l = [\n",
    "    \"Mutation Type\",\n",
    "    \"disrupt\",\n",
    "    \"repaired\",\n",
    "    \"genomic features\",\n",
    "    \"genetic features\",\n",
    "    \"genetic feature links\",\n",
    "    \"operons\",\n",
    "    \"operon links\",\n",
    "    \"regulators\",\n",
    "    \"regulator links\",\n",
    "    \"COGs\",\n",
    "    \"COG links\",\n",
    "    \"pathways\",\n",
    "    \"pathway links\",\n",
    "    \"RegulonDB Gene\"  # for debugging\n",
    "]\n",
    "\n",
    "annot_df = muts_df.copy()\n",
    "annot_df = annot_df[col_l]\n",
    "# annot_df[\"mutation effect\"] = annot_df.apply(lambda mut_row:\"repaired\" if mut_row[\"repaired\"] else (\"disrupt\" if mut_row[\"disrupt\"] else \"unknown\"), axis=1)\n",
    "\n",
    "\n",
    "display(annot_df.shape, annot_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just removing all \"Function unknown\" links since connections to this COG isn't useful.\n",
    "for i, r in annot_df.iterrows():\n",
    "    if \"Function unknown\" in r[\"COG links\"].keys():\n",
    "        del r[\"COG links\"][\"Function unknown\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_feat_effect_counts = dict()\n",
    "genomic_feat_counts = dict()\n",
    "genetic_feat_counts = dict()\n",
    "operon_counts = dict()\n",
    "COG_counts = dict()\n",
    "regulator_counts = dict()\n",
    "pathway_counts = dict()\n",
    "# Using the individual feature columns rather than link columns because this guarantees considering all linked and non-linked features\n",
    "# for each of these levels, except genomic features.\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        \n",
    "        # These are usually exclusive of each other per feature.\n",
    "        # There is a design problem here: A multi-feature mutation could disrupt feature (gene) and repair another. Need to measure the predicted effect per feature rather than mutation to enable this (considering a design bug fix).\n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "            \n",
    "        mut_type_feat_effect = genomic_feat[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        if mut_type_feat_effect not in mut_type_feat_effect_counts.keys():\n",
    "            mut_type_feat_effect_counts[mut_type_feat_effect] = 0\n",
    "        mut_type_feat_effect_counts[mut_type_feat_effect] += 1\n",
    "    \n",
    "    for genomic_feat in r[\"genomic features\"]:\n",
    "        genomic_feat_annot = genomic_feat[\"RegulonDB ID\"] #+ ' ' + r[\"Mutation Type\"]\n",
    "        if genomic_feat_annot not in genomic_feat_counts.keys():\n",
    "            genomic_feat_counts[genomic_feat_annot] = 0\n",
    "        genomic_feat_counts[genomic_feat_annot] += 1\n",
    "\n",
    "    for genetic_feat, links in r[\"genetic feature links\"].items():\n",
    "        if genetic_feat not in genetic_feat_counts.keys():\n",
    "            genetic_feat_counts[genetic_feat] = 0\n",
    "        genetic_feat_counts[genetic_feat] += len(links)\n",
    "\n",
    "    # count the operons according to the amount of links from genes since an operon is only included if its gene is hit.\n",
    "    for operon, links in r[\"operon links\"].items():\n",
    "        if operon not in operon_counts.keys():\n",
    "            operon_counts[operon] = 0\n",
    "        operon_counts[operon] += len(links)\n",
    "\n",
    "    # count the COGs according to the amount of links from operons since a COG is only included if linked to a mutated operon (its genes)\n",
    "    for COG, links in r[\"COG links\"].items():\n",
    "        if COG not in COG_counts.keys():\n",
    "            COG_counts[COG] = 0\n",
    "        COG_counts[COG] += len(links)\n",
    "        \n",
    "    for reg, links in r[\"regulator links\"].items():\n",
    "        if reg not in regulator_counts.keys():\n",
    "            regulator_counts[reg] = 0\n",
    "        regulator_counts[reg] += len(links)\n",
    "        \n",
    "    for pathway, links in r[\"pathway links\"].items():\n",
    "        if pathway not in pathway_counts.keys():\n",
    "            pathway_counts[pathway] = 0\n",
    "        pathway_counts[pathway] += len(links)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "operon_df = pd.read_csv(\"./data/RegulonDB10/operon.txt\", sep=\"\\t\", comment='#', header=None)\n",
    "operon_df.columns = [\n",
    "    \"OPERON_ID\",\n",
    "    \"OPERON_NAME\",\n",
    "    \"FIRSTGENEPOSLEFT\",\n",
    "    \"LASTGENEPOSRIGHT\",\n",
    "    \"REGULATIONPOSLEFT\",\n",
    "    \"REGULATIONPOSRIGHT\",\n",
    "    \"OPERON_STRAND\",\n",
    "    \"OPERON_INTERNAL_COMMENT\",\n",
    "    \"KEY_ID_ORG\"]\n",
    "operon_df[\"range\"] = operon_df.apply(lambda row: (row[\"REGULATIONPOSLEFT\"], row[\"REGULATIONPOSRIGHT\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# operon names can be the same as a gene, which is problemmatic for floweaver\n",
    "def _adjust_name_for_floweaver(operon):\n",
    "    return operon + ' '\n",
    "\n",
    "\n",
    "def _adjust_name_len_for_floweaver(feat_name):\n",
    "    if len(feat_name) > 7: \n",
    "        feat_name = feat_name[:7] + \"...\"\n",
    "    return feat_name\n",
    "\n",
    "\n",
    "def _get_feat_d(json, RegulonDB_ID=None, name=None):\n",
    "    feat_d = dict()\n",
    "    for d in json:\n",
    "        if (RegulonDB_ID and d[\"RegulonDB ID\"] == RegulonDB_ID) or (name and d[\"name\"] == name):\n",
    "            feat_d = d\n",
    "            break\n",
    "    return feat_d\n",
    "\n",
    "\n",
    "# Will probably have to put elsewhere at some point due to annotation work (count, significance)\n",
    "mut_feat_effects = list()\n",
    "genomic_feats = list()\n",
    "genetic_feats = list()\n",
    "operons = list()\n",
    "COGs = list()\n",
    "regulators = list()\n",
    "pathways = list()\n",
    "\n",
    "\n",
    "flow_df = pd.DataFrame()\n",
    "for _, r in annot_df.iterrows():\n",
    "    \n",
    "\n",
    "    for genomic_feat_d in r[\"genomic features\"]:\n",
    "        \n",
    "        effect = \"\"\n",
    "#         if r[\"disrupt\"]: effect = \"disrupt\"\n",
    "#         if r[\"repaired\"]: effect = \"repair\"\n",
    "        mut_type_feat_effect = genomic_feat_d[\"name\"] + ' ' + r[\"Mutation Type\"] + ' ' + effect\n",
    "        mut_type_feat_effect = str(mut_type_feat_effect_counts[mut_type_feat_effect]) + ' ' + mut_type_feat_effect\n",
    "        mut_feat_effects.append(mut_type_feat_effect)\n",
    "            \n",
    "        genomic_feat_annot = \"\"\n",
    "        mut_count = genomic_feat_counts[genomic_feat_d[\"RegulonDB ID\"]]\n",
    "#         if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "        genomic_feat_annot += str(mut_count) + ' '\n",
    "        genomic_feat_annot += str(genomic_feat_d[\"name\"])\n",
    "        flow_row = {\n",
    "            \"source\": genomic_feat_annot,\n",
    "            \"target\": mut_type_feat_effect,\n",
    "            \"type\": \"unknown\",\n",
    "            \"value\": 1\n",
    "        }\n",
    "        flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "        \n",
    "    \n",
    "    \n",
    "    for genetic_feat_ID, genomic_feat_set in r[\"genetic feature links\"].items():\n",
    "        genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "        genetic_feat_annot = \"\"\n",
    "        mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "#         if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "        genetic_feat_annot += str(mut_count) + ' '\n",
    "        genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "        genetic_feats.append(genetic_feat_annot)\n",
    "        for genomic_feat_ID in genomic_feat_set:\n",
    "            genomic_feat_d = _get_feat_d(RegulonDB_ID=genomic_feat_ID, json=r[\"genomic features\"])\n",
    "            k = genomic_feat_ID #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feat_annot = \"\"\n",
    "            mut_count = genomic_feat_counts[k]\n",
    "#             if genomic_feat_d[\"significant\"] and mut_count > 1: genomic_feat_annot += \"*\"\n",
    "            genomic_feat_annot += str(mut_count) + ' '\n",
    "            genomic_feat_annot += str(genomic_feat_d[\"name\"]) #+ ' ' + r[\"Mutation Type\"]\n",
    "            genomic_feats.append(genomic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": genetic_feat_annot,\n",
    "                \"target\": genomic_feat_annot,\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "    \n",
    "    \n",
    "    for operon_id, gen_targs in r[\"operon links\"].items():\n",
    "        operon_d = _get_feat_d(RegulonDB_ID=operon_id, json=r[\"operons\"])\n",
    "        operon_name = _adjust_name_for_floweaver(operon_d[\"name\"])\n",
    "        operon_annot = \"\"\n",
    "        mut_count = operon_counts[operon_id]\n",
    "#         if operon_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "        operon_annot += str(mut_count) + ' '\n",
    "        operon_annot += operon_name\n",
    "        operons.append(operon_annot)\n",
    "        for genetic_feat_ID in gen_targs:\n",
    "            genetic_feat_d = _get_feat_d(RegulonDB_ID=genetic_feat_ID, json=r[\"genetic features\"])\n",
    "            genetic_feat_annot = \"\"\n",
    "            mut_count = genetic_feat_counts[genetic_feat_ID]\n",
    "#             if genetic_feat_d[\"significant\"] and mut_count > 1: genetic_feat_annot += \"*\"\n",
    "            genetic_feat_annot += str(mut_count) + ' '\n",
    "            genetic_feat_annot += str(genetic_feat_d[\"name\"])\n",
    "            genetic_feats.append(genetic_feat_annot)\n",
    "            flow_row = {\n",
    "                \"source\": operon_annot,\n",
    "                \"target\": genetic_feat_annot,\n",
    "                \"type\": \"unknown\",\n",
    "                \"value\": 1\n",
    "            }\n",
    "            flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "            \n",
    "            \n",
    "    if FUNC_ANNOT == \"regulon\":\n",
    "        for regulator, ops in r[\"regulator links\"].items():\n",
    "            if regulator == \"CRP\":  # for GYD CRP fig\n",
    "#             if regulator == \"GlpR\":  # for GYD GlpR fig\n",
    "#             if regulator == \"Mlc\":  # for GYD Mlc fig\n",
    "                reg_feat_d = _get_feat_d(name=regulator, json=r[\"regulators\"])\n",
    "                reg_annot = \"\"\n",
    "                mut_count = regulator_counts[regulator]\n",
    "#                 if reg_feat_d[\"significant\"] and mut_count > 1: reg_annot += \"*\"\n",
    "                reg_annot += str(mut_count) + ' ' + reg_feat_d[\"name\"]\n",
    "                regulators.append(reg_annot)\n",
    "                for operon_name in ops:\n",
    "                    op_feat_d = _get_feat_d(name=operon_name, json=r[\"operons\"])\n",
    "                    operon_name = _adjust_name_for_floweaver(op_feat_d[\"name\"])\n",
    "                    operon_annot = \"\"\n",
    "                    mut_count = operon_counts[op_feat_d[\"RegulonDB ID\"]]\n",
    "#                     if op_feat_d[\"significant\"] and mut_count > 1: operon_annot += \"*\"\n",
    "                    operon_annot += str(mut_count) + ' '\n",
    "                    operon_annot += operon_name\n",
    "                    operons.append(operon_annot)\n",
    "                    flow_row = {\n",
    "                        \"source\": reg_annot,\n",
    "                        \"target\": operon_annot,\n",
    "                        \"type\": \"unknown\",\n",
    "                        \"value\": 1\n",
    "                    }\n",
    "                    flow_df = flow_df.append(flow_row, ignore_index=True)\n",
    "\n",
    "\n",
    "display(flow_df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specific to the GYD fig of the AVA paper.\n",
    "flow_df = flow_df[~(flow_df[\"target\"].str.contains(\"1 pdxK\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# processing to help a cleaner order of features.\n",
    "# remove duplicates but maintains the order established above\n",
    "def remove_duplicates(feats):\n",
    "    l = list()\n",
    "    for x in feats:\n",
    "        if x not in l:\n",
    "            l.append(x)\n",
    "    return l\n",
    "\n",
    "mut_feat_effects = remove_duplicates(mut_feat_effects)\n",
    "genomic_feats = remove_duplicates(genomic_feats)\n",
    "genetic_feats = remove_duplicates(genetic_feats)\n",
    "operons = remove_duplicates(operons)\n",
    "pathways = remove_duplicates(pathways)\n",
    "regulators = remove_duplicates(regulators)\n",
    "COGs = remove_duplicates(COGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from ipysankeywidget import SankeyWidget\n",
    "from floweaver import *\n",
    "\n",
    "\n",
    "nodes = {\n",
    "    \"mutation\": ProcessGroup(mut_feat_effects, title=\"mutation\"),\n",
    "    \"genomic feature\": ProcessGroup(genomic_feats, title=\"genomic feature\"),\n",
    "    \"genetic feature\": ProcessGroup(genetic_feats, title=\"genetic feature\"),\n",
    "    \"operon\": ProcessGroup(operons, title=\"operon\"),\n",
    "}\n",
    "\n",
    "l = []\n",
    "if FUNC_ANNOT == \"pathway\":\n",
    "    l = pathways\n",
    "if FUNC_ANNOT == \"regulon\":\n",
    "    l = regulators\n",
    "if FUNC_ANNOT == \"COG\":\n",
    "    l = COGs\n",
    "nodes[FUNC_ANNOT] = ProcessGroup(l, title=FUNC_ANNOT)\n",
    "\n",
    "\n",
    "# !!! These partition DFs need to have accessible variables (ex: mut_type_partition) accessible external to the nodes\n",
    "# otherwise floweaver will crash during its rendering step.\n",
    "mutation_partition = Partition.Simple(\"process\", mut_feat_effects)\n",
    "nodes[\"mutation\"].partition = mutation_partition\n",
    "\n",
    "genomic_feat_partition = Partition.Simple(\"process\", genomic_feats)\n",
    "nodes[\"genomic feature\"].partition = genomic_feat_partition\n",
    "\n",
    "genetic_feat_partition = Partition.Simple(\"process\", genetic_feats)\n",
    "nodes[\"genetic feature\"].partition = genetic_feat_partition\n",
    "\n",
    "operon_partition = Partition.Simple(\"process\", operons)\n",
    "nodes[\"operon\"].partition = operon_partition\n",
    "\n",
    "nodes[FUNC_ANNOT].partition = Partition.Simple(\"process\", l)\n",
    "\n",
    "\n",
    "# The order that the labels is entered into the bundles is explicit. Don't arbitrarily give an order.\n",
    "bundles = [\n",
    "    Bundle(\n",
    "        \"genomic feature\",\n",
    "        \"mutation\"\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"genetic feature\",\n",
    "        \"genomic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        \"operon\",\n",
    "        \"genetic feature\",\n",
    "    ),\n",
    "    Bundle(\n",
    "        FUNC_ANNOT,\n",
    "        \"operon\",\n",
    "    ),\n",
    "]\n",
    "\n",
    "\n",
    "ordering = [\n",
    "    [FUNC_ANNOT],\n",
    "    [\"operon\"],\n",
    "    [\"genetic feature\"],\n",
    "    [\"genomic feature\"],\n",
    "    [\"mutation\"]\n",
    "]\n",
    "\n",
    "#\n",
    "# mut_effect_partition = Partition.Simple(\n",
    "#     \"type\", [\"unknown\", \"disrupt\", \"repaired\"])\n",
    "sdd = SankeyDefinition(nodes,\n",
    "                       bundles,\n",
    "                       ordering,\n",
    "#                        flow_partition=mut_effect_partition\n",
    "                      )\n",
    "\n",
    "\n",
    "# p_d = {\"disrupt\": \"#C44E52\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#55A868\"}\n",
    "p_d = {\"disrupt\": \"#6BAAD3\", \"unknown\": \"#6BAAD3\", \"repaired\": \"#6BAAD3\"}\n",
    "# size = dict(width=3000, height=4000)\n",
    "# size = dict(width=2000, height=2500)\n",
    "size = dict(width=1000, height=200)\n",
    "\n",
    "results = weave(sdd,\n",
    "                flow_df,\n",
    "                palette=p_d,\n",
    "                ).to_widget(\n",
    "    margins=dict(left=400, right=350),\n",
    "    **size)\n",
    "results.auto_save_svg(\"mutflow.svg\")\n",
    "results.auto_save_png(\"mutflow.png\")\n",
    "results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
